---
title: "SW0100_P239_TotalSeq_IL27RAKO_Johannes_v1.2.Rmd"
output: 
  html_document: 
    toc: yes
    toc_float: yes
    collapsed: yes
    code_folding: hide
    fig_height: 8
    number_sections: yes
    toc_depth: 4
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# About data
- This analysis is based on a TotalSeq single cell RNA (GEX) plus Antibody derived tag (ADT; protein) sequencing experiment by Dr. Johannes Platten and Dr. Sarah Walachowski. <br>
- The experiment involved BALF from three wild type (WT; samples BM_WS_5, BM_WS_6, BM_WS_8) and three IL27RA knockout mice (IL27RAko; samples BM_WS_9, BM_WS_10, BM_WS_15). <br>
- Mice were infected for 48h with JR32 strain of Legionella pneumophila and then BALF cells were collected by Johannes and, frozen and shipped for further staining, FACSorting and sequencing by Sarah. Sequencing was performed at the Boston University Single Cell Sequencing Core. <br>
- Neutrophils were sorted out before Totalseq antibodies staining, so no neutrophil clusters are expected in the analysis. <br>
- The sequencing experiment included two libraries per sample: 1 ADT and 1 GEX. <br>
- A total of 198 BioLegend TotalSeq A ADTs were used. <br>
- The targeted number of cells for sequencing was 4000  <br>
- One lane was used for sequencing using P3 flow cell. <br>
- While demultiplexing the binary base call files to generate fastq files (Cell Ranger mkfastq tool) prior to quantification, --barcode-mismatches=0 flag was set instead of the default --barcode-mismatches=1 flag. The 0 flag was set as there was an error message from Cell Ranger that sample indices : RPI-1 (ATCACGAT) and SI-GA-D4 (ATTCCGAT) only differ by two bases. This leads to barcode collision while demultiplexing.  This parameter was confirmed by the 10X Genomics support team. <br>
- CellRanger QC reports indicate that the estimated number of cells are within the targeted cell range for all samples except BM_WS_6. <br>
- The reads/cell for GEX are in the good quality range, while for ADT they are below the good quality range. <br>
- The analysis is based on the filtered counts matrices generated after performing alignment and quantification using CellRanger count. Filtered matrices are used as the --expect-cells parameter was set to 4000 while performing Cell Ranger quantification.<br>
- This analysis is based on excluding doublets prior to clustering, keeping k.anchor = 5, and not scaling ADT (not scaling ADT as recommended by Seurat developers). <br>
- This version (v1.2) includes an update in the cell type annotation of NK cell clusters. <br>

# Loading libraries
```{r}
options(RENV_CONFIG_SANDBOX_ENABLED = FALSE)
suppressPackageStartupMessages({library(Seurat)
  library(SeuratWrappers)
library(ggplot2)
library(patchwork)
library(dplyr)
library(clustree)
library(RColorBrewer)
library(future)
  library(magrittr)
  library(ggcharts)
  library(tidyverse)
  library(DT)
  library(scDblFinder)
  library(DoubletFinder)
  library(openxlsx)
  library(SingleCellExperiment)
  library(writexl)
  library(ggrepel)
  options(future.globals.maxSize = 80000 * 1024^2)
  options(ggrepel.max.overlaps = Inf)
  #plan("multiprocess", workers = 8)
})

#specifying data directories
input_dir <- file.path("~/RNA_ADT_counts/")
output_dir <- file.path("~/analysis/R_analysis/")
fig_path <- file.path(paste0(output_dir, "SW0100_P239_figures_v1.2/")) #changed to v1.2 for figures after re-annotation of NK clusters
tab_path <- file.path(paste0(output_dir, "SW0100_P239_tables_v1.2/")) #changed to v1.2 for tables after re-annotation of NK clusters
data_path <- file.path(paste0(output_dir, "SW0100_P239_Rdata_v1.2/")) #changed to v1.2 for data after re-annotation of NK clusters

```

# Load data<br>
```{r}
#Reference Source code: https://satijalab.org/seurat/articles/multimodal_vignette.html#loading-data-from-10x-multi-modal-experiments-1
#load in data by providing path to folder with raw counts matrices as .gz files
#Create a Seurat object for each sample
for (sample in c("BM_WS_5", "BM_WS_6", "BM_WS_8", "BM_WS_9", "BM_WS_10", "BM_WS_15")){
        sample_file <- Read10X(file.path(paste0(input_dir, sample, "/outs/filtered_feature_bc_matrix/")))
        sample_obj <- CreateSeuratObject(counts = sample_file[["Gene Expression"]], min.cells = 3, min.features = 200, project = sample)
        sample_obj[["ADT"]] <- CreateAssayObject(counts = sample_file[["Antibody Capture"]][, colnames(sample_file[["Antibody Capture"]]) %in% colnames(sample_obj)])
        assign(sample, sample_obj)
}

#Check that both RNA and ADT assay data is available
for (obj in c(BM_WS_5, BM_WS_6, BM_WS_8, BM_WS_9, BM_WS_10, BM_WS_15)){
  print(Seurat::Assays(obj))
}


#add metadata (#Source code: https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/seurat/seurat_01_qc.html)
BM_WS_5$group <- "wt"; BM_WS_5$sample_id <- "BM_WS_5"
BM_WS_6$group <- "wt"; BM_WS_6$sample_id <- "BM_WS_6"
BM_WS_8$group <- "wt"; BM_WS_8$sample_id <- "BM_WS_8"
BM_WS_9$group <- "il27raKO"; BM_WS_9$sample_id <- "BM_WS_9"
BM_WS_10$group <- "il27raKO"; BM_WS_10$sample_id <- "BM_WS_10"
BM_WS_15$group <- "il27raKO"; BM_WS_15$sample_id <- "BM_WS_15"

```

# Check features measured in the ADT assay
```{r}
# Extract a list of features measured in the ADT assay
for (obj in list(BM_WS_5, BM_WS_6, BM_WS_8, BM_WS_9, BM_WS_10, BM_WS_15)){
  print(rownames(obj[["ADT"]]))
}
```


## Initial QC{.tabset .tabset-fade .tabset-pills}<br>
### BM_WS_5
(tab content)#<br>
Notes (Source: 1. Seurat tutorial pbmc3k tutorial. Available from: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html ;<br> 2. Ilicic T, et al. Classification of low quality cells from single-cell RNA-seq data. Genome Biol. 2016;17:29. Available from: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4758103/; <br> 3. Harvard Chan Bioinformatics core tutorials. Available from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/04_SC_quality_control.md):<br>
1.The QC metrics evaluated include: <br>
a. nFeature_RNA (number of genes)<br>
b. nCount_RNA (number of unique molecular identifiers [UMIs])<br>
c. percent.mt (percentage of percentage of cell reads originating from mitochondrial genes)<br>
d. nFeature_ADT (number of antibdiy derived tags)<br>
2. These QC metrics are evaluated as they help in assessment of the following:<br>
a. The number of unique genes detected in each cell (nFeature_RNA).<br>
    -Low-quality cells or empty droplets will often have very few genes<br>
    -Cell doublets or multiplets may exhibit an aberrantly high gene count<br>
b. The total number of molecules detected within a cell (correlates strongly with unique genes; nCount_RNA)<br>
c. Mitochondrial QC metrics, which calculates the percentage of counts originating from mitochondrial genes (percent.mt)<br>
     -The percentage of reads that map to the mitochondrial genome<br>
     -Low-quality / dying cells often exhibit extensive mitochondrial contamination<br>
     -Higher percentage of counts from mitocondrial genes may indicate these cells may be dead or dying, as in these cells  cell membrane might be broken, cytoplasmic RNA lost, but RNAs enclosed in the mitochondria will be retained driving the high mitochondrial count<br>
d. The number of unique antibody derived tags (ADTs) detected in each cell (nFeature_ADT).<br>

```{r}
#Reference: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
#Assign mitochondrial reads to metadata
BM_WS_5[["percent.mt"]] <- Seurat::PercentageFeatureSet(BM_WS_5, pattern = "^mt-") #searching for genes with this pattern will elicit mitochondrial genes which are then counted


#UMI counts per gene
summary(apply(GetAssayData(BM_WS_5, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_5, slot = "counts"), 2, sum))


#Visualize QC metrics
#pdf(paste0(fig_path, "BM_WS_5_initial_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_5, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_5, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_5, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_5, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_5, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p1 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "BM_WS_5_initial_QC_plot2.pdf") , width = 15, height = 15)
p1
##dev.off()
```



### BM_WS_6
(tab content)#<br>
```{r}
#Reference: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
#Assign mitochondrial reads to metadata
BM_WS_6[["percent.mt"]] <- Seurat::PercentageFeatureSet(BM_WS_6, pattern = "^mt-") #searching for genes with this pattern will elicit mitochondrial genes which are then counted


#UMI counts per gene
summary(apply(GetAssayData(BM_WS_6, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_6, slot = "counts"), 2, sum))


#Visualize QC metrics
#pdf(paste0(fig_path, "BM_WS_6_initial_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_6, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_6, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_6, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_6, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_6, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p1 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "BM_WS_6_initial_QC_plot2.pdf") , width = 15, height = 15)
p1
##dev.off()
```



### BM_WS_8
(tab content)#<br>
```{r}
#Reference: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
#Assign mitochondrial reads to metadata
BM_WS_8[["percent.mt"]] <- Seurat::PercentageFeatureSet(BM_WS_8, pattern = "^mt-") #searching for genes with this pattern will elicit mitochondrial genes which are then counted


#UMI counts per gene
summary(apply(GetAssayData(BM_WS_8, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_8, slot = "counts"), 2, sum))


#Visualize QC metrics
#pdf(paste0(fig_path, "BM_WS_8_initial_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_8, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_8, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_8, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_8, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_8, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p1 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "BM_WS_8_initial_QC_plot2.pdf") , width = 15, height = 15)
p1
##dev.off()
```


### BM_WS_9
(tab content)#<br>
```{r}
#Reference: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
#Assign mitochondrial reads to metadata
BM_WS_9[["percent.mt"]] <- Seurat::PercentageFeatureSet(BM_WS_9, pattern = "^mt-") #searching for genes with this pattern will elicit mitochondrial genes which are then counted


#UMI counts per gene
summary(apply(GetAssayData(BM_WS_9, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_9, slot = "counts"), 2, sum))


#Visualize QC metrics
#pdf(paste0(fig_path, "BM_WS_9_initial_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_9, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_9, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_9, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_9, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_9, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p1 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "BM_WS_9_initial_QC_plot2.pdf") , width = 15, height = 15)
p1
##dev.off()
```


### BM_WS_10
(tab content)#<br>
```{r}
#Reference: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
#Assign mitochondrial reads to metadata
BM_WS_10[["percent.mt"]] <- Seurat::PercentageFeatureSet(BM_WS_10, pattern = "^mt-") #searching for genes with this pattern will elicit mitochondrial genes which are then counted


#UMI counts per gene
summary(apply(GetAssayData(BM_WS_10, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_10, slot = "counts"), 2, sum))


#Visualize QC metrics
#pdf(paste0(fig_path, "BM_WS_10_initial_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_10, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_10, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_10, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_10, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_10, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p1 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "BM_WS_10_initial_QC_plot2.pdf") , width = 15, height = 15)
p1
##dev.off()
```


### BM_WS_15
(tab content)#<br>
```{r}
#Reference: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html
#Assign mitochondrial reads to metadata
BM_WS_15[["percent.mt"]] <- Seurat::PercentageFeatureSet(BM_WS_15, pattern = "^mt-") #searching for genes with this pattern will elicit mitochondrial genes which are then counted


#UMI counts per gene
summary(apply(GetAssayData(BM_WS_15, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_15, slot = "counts"), 2, sum))


#Visualize QC metrics
#pdf(paste0(fig_path, "BM_WS_15_initial_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_15, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_15, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_15, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_15, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_15, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p1 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "BM_WS_15_initial_QC_plot2.pdf") , width = 15, height = 15)
p1
##dev.off()
```


# Check qc metrics including 1% and 99% quantiles
```{r}
# Check 1% and 99% quantiles
qc_metrics <- list()
obj_list <- list("BM_WS_5" = BM_WS_5, "BM_WS_6" = BM_WS_6, "BM_WS_8" = BM_WS_8, "BM_WS_9" = BM_WS_9, "BM_WS_10" = BM_WS_10, "BM_WS_15" = BM_WS_15)
for (i in 1:length(obj_list)){
  qc_metrics[[names(obj_list)[i]]]$min_nFeature_RNA <- min(obj_list[[i]]@meta.data$nFeature_RNA)
  qc_metrics[[names(obj_list)[i]]]$median_nFeature_RNA <- median(obj_list[[i]]@meta.data$nFeature_RNA)
  qc_metrics[[names(obj_list)[i]]]$mean_nFeature_RNA <- mean(obj_list[[i]]@meta.data$nFeature_RNA)
  qc_metrics[[names(obj_list)[i]]]$max_nFeature_RNA <- max(obj_list[[i]]@meta.data$nFeature_RNA) 
  qc_metrics[[names(obj_list)[i]]]$sd_nFeature_RNA <- sd(obj_list[[i]]@meta.data$nFeature_RNA)
  qc_metrics[[names(obj_list)[i]]]$quantile0.01_nFeature_RNA <- quantile(obj_list[[i]]@meta.data$nFeature_RNA, 0.01)
  qc_metrics[[names(obj_list)[i]]]$quantile0.99_nFeature_RNA <- quantile(obj_list[[i]]@meta.data$nFeature_RNA, 0.99)
  
  qc_metrics[[names(obj_list)[i]]]$min_nCount_RNA <- min(obj_list[[i]]@meta.data$nCount_RNA)
  qc_metrics[[names(obj_list)[i]]]$median_nCount_RNA <- median(obj_list[[i]]@meta.data$nCount_RNA)
  qc_metrics[[names(obj_list)[i]]]$max_nCount_RNA <- max(obj_list[[i]]@meta.data$nCount_RNA) 
  qc_metrics[[names(obj_list)[i]]]$sd_nCount_RNA <- sd(obj_list[[i]]@meta.data$nCount_RNA)
  qc_metrics[[names(obj_list)[i]]]$mean_nCount_RNA <- mean(obj_list[[i]]@meta.data$nCount_RNA)
  qc_metrics[[names(obj_list)[i]]]$quantile0.01_nCount_RNA <- quantile(obj_list[[i]]@meta.data$nCount_RNA, 0.01)
  qc_metrics[[names(obj_list)[i]]]$quantile0.99_nCount_RNA <- quantile(obj_list[[i]]@meta.data$nCount_RNA, 0.99)
  
  
  qc_metrics[[names(obj_list)[i]]]$min_nFeature_ADT <- min(obj_list[[i]]@meta.data$nFeature_ADT)
  qc_metrics[[names(obj_list)[i]]]$median_nFeature_ADT <- median(obj_list[[i]]@meta.data$nFeature_ADT)
  qc_metrics[[names(obj_list)[i]]]$mean_nFeature_ADT <- mean(obj_list[[i]]@meta.data$nFeature_ADT)
  qc_metrics[[names(obj_list)[i]]]$max_nFeature_ADT <- max(obj_list[[i]]@meta.data$nFeature_ADT) 
  qc_metrics[[names(obj_list)[i]]]$sd_nFeature_ADT <- sd(obj_list[[i]]@meta.data$nFeature_ADT)
  qc_metrics[[names(obj_list)[i]]]$quantile0.01_nFeature_ADT <- quantile(obj_list[[i]]@meta.data$nFeature_ADT, 0.01)
  qc_metrics[[names(obj_list)[i]]]$quantile0.99_nFeature_ADT <- quantile(obj_list[[i]]@meta.data$nFeature_ADT, 0.99)
  
  qc_metrics[[names(obj_list)[i]]]$min_nCount_ADT <- min(obj_list[[i]]@meta.data$nCount_ADT)
  qc_metrics[[names(obj_list)[i]]]$median_nCount_ADT <- median(obj_list[[i]]@meta.data$nCount_ADT)
  qc_metrics[[names(obj_list)[i]]]$max_nCount_ADT <- max(obj_list[[i]]@meta.data$nCount_ADT) 
  qc_metrics[[names(obj_list)[i]]]$sd_nCount_ADT <- sd(obj_list[[i]]@meta.data$nCount_ADT)
  qc_metrics[[names(obj_list)[i]]]$mean_nCount_ADT <- mean(obj_list[[i]]@meta.data$nCount_ADT)
  qc_metrics[[names(obj_list)[i]]]$quantile0.01_nCount_ADT <- quantile(obj_list[[i]]@meta.data$nCount_ADT, 0.01)
  qc_metrics[[names(obj_list)[i]]]$quantile0.99_nCount_ADT <- quantile(obj_list[[i]]@meta.data$nCount_ADT, 0.99)
  qc_metrics[[names(obj_list)[i]]]$sample_id <- names(obj_list)[i]
}

#source code to flatten the list: https://stackoverflow.com/questions/16300344/how-to-flatten-a-list-of-lists
#writexl::write_xlsx(tibble(qc_metrics)  %>% unnest_wider(qc_metrics), paste0(tab_path, "R_QC_metrics_TotalSeq_IL27RAKO_Johannes.xlsx") )

```


## Check viability based on only filtering out cells with mitochondrial reads >10%
```{r}
#unfiltered samples
tmp_cell_unfiltered <- list()

for (i in 1:length(obj_list)){
  tmp <- obj_list[[i]]
  tmp_cell_unfiltered[[names(obj_list[i])]] <- as.data.frame(table(tmp$orig.ident))
}

#filtered samples
tmp_cell_filtered <- list()

for (i in 1:length(obj_list)){
  tmp <- obj_list[[i]]
  tmp <- subset(tmp, subset = percent.mt < 10)
  tmp_cell_filtered[[names(obj_list[i])]] <- as.data.frame(table(tmp$orig.ident))
}

tmp_all <- cbind(unlist(tmp_cell_unfiltered), unlist(tmp_cell_filtered))
colnames(tmp_all) <- c("Unfiltered", "Mito_filtered")
tmp_all <- tmp_all[c(2, 4, 6, 8, 10, 12), ] #exclude rows with other variables
tmp_all <- as.data.frame(tmp_all)
tmp_all$sample_id <- row.names(tmp_all)
tmp_all$sample_id <- gsub(".Freq", "", tmp_all$sample_id)
tmp_all
tmp_all <- tmp_all[c(3, 1, 2)] # change order so that sample_id is the first column

#writexl::write_xlsx(tmp_all, paste0(tab_path, "Viable_cells_Mitochondrial_TotalSeq_IL27RAKO_Johannes.xlsx") )
rm(tmp, tmp_cell_unfiltered, tmp_cell_filtered, obj_list, tmp_all)

```


```{r, eval=FALSE, echo=FALSE}
# Merge datasets into one single seurat object
alldata <- merge(BM_WS_5, c(BM_WS_6, BM_WS_8, BM_WS_9, BM_WS_10, BM_WS_15), add.cell.ids = c("BM_WS_5", "BM_WS_6", "BM_WS_8", "BM_WS_9", "BM_WS_10", "BM_WS_15"))

alldata$sample_id <- factor(alldata$sample_id, levels = c("BM_WS_5", "BM_WS_6", "BM_WS_8", "BM_WS_9", "BM_WS_10", "BM_WS_15"))
Idents(alldata) <- alldata$sample_id

#pdf(paste0(fig_path, "AllSamples_QC_plot1.pdf") , width = 15, height = 15)
VlnPlot(alldata, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nFeature_ADT", "nCount_ADT"), ncol = 3, pt.size = 0.0)
##dev.off()

rm(alldata)

```


### Filtering dataset based on QC metrics
(tab content)
Notes (Reference: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/04_SC_quality_control.md):<br>
1. Based on QC analysis, filtering out cells that have very low or high nFeature_RNA (number of genes) , very high nCount_RNA and nCount_ADT (nCount values chosen based on the highest point of the peak of the violin plot with minimal/no cells)  and with >10% mitochondrial ratio, as these may represent dead/dying cells or low complexity cells (Reference: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/04_SC_quality_control.md) <br>
```{r}
#Based on QC analysis, filtering out cells that have very low or high nFeature_RNA (number of genes) , very high nCount_RNA and nCount_ADT (nCount values chosen based on the highest point of the peak of the violin plot with minimal/no cells)  and with >10% mitochondrial ratio, as these may represent dead/dying cells or low complexity cells (Reference: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/04_SC_quality_control.md)

BM_WS_5_filt <- subset(BM_WS_5, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & nCount_RNA < 40000 & nCount_ADT < 5000 & percent.mt < 10)
BM_WS_5_filt

BM_WS_6_filt <- subset(BM_WS_6, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & nCount_RNA < 50000 & nCount_ADT < 3000 & percent.mt < 10)
BM_WS_6_filt


BM_WS_8_filt <- subset(BM_WS_8, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & nCount_RNA < 50000 & nCount_ADT < 5000 & percent.mt < 10)
BM_WS_8_filt


BM_WS_9_filt <- subset(BM_WS_9, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & nCount_RNA < 40000 & nCount_ADT < 4000 & percent.mt < 10)
BM_WS_9_filt

BM_WS_10_filt <- subset(BM_WS_10, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & nCount_RNA < 40000 & nCount_ADT < 5000 & percent.mt < 10)
BM_WS_10_filt


BM_WS_15_filt <- subset(BM_WS_15, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & nCount_RNA < 40000 & nCount_ADT < 5000 & percent.mt < 10)
BM_WS_15_filt

```


## Visualizing QC metrics after filtering
### BM_WS_5
(tab content)
```{r}
#Visualize QC metrics after filtering

#UMI counts per gene
summary(apply(GetAssayData(BM_WS_5_filt, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_5_filt, slot = "counts"), 2, sum))

#pdf(paste0(fig_path, "/BM_WS_5_filtered_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_5_filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_5_filt, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_5_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_5_filt, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_5_filt, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p2 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "/BM_WS_5_filtered_QC_plot2.pdf") , width = 15, height = 15)
p2
##dev.off()
```


### BM_WS_6
(tab content)
```{r}
#Visualize QC metrics after filtering

#UMI counts per gene
summary(apply(GetAssayData(BM_WS_6_filt, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_6_filt, slot = "counts"), 2, sum))

#pdf(paste0(fig_path, "/BM_WS_6_filtered_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_6_filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_6_filt, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_6_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_6_filt, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_6_filt, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p2 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "/BM_WS_6_filtered_QC_plot2.pdf") , width = 15, height = 15)
p2
##dev.off()
```


### BM_WS_8
(tab content)
```{r}
#Visualize QC metrics after filtering

#UMI counts per gene
summary(apply(GetAssayData(BM_WS_8_filt, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_8_filt, slot = "counts"), 2, sum))

#pdf(paste0(fig_path, "/BM_WS_8_filtered_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_8_filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_8_filt, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_8_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_8_filt, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_8_filt, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p2 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "/BM_WS_8_filtered_QC_plot2.pdf") , width = 15, height = 15)
p2
##dev.off()
```


### BM_WS_9
(tab content)
```{r}
#Visualize QC metrics after filtering

#UMI counts per gene
summary(apply(GetAssayData(BM_WS_9_filt, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_9_filt, slot = "counts"), 2, sum))

#pdf(paste0(fig_path, "/BM_WS_9_filtered_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_9_filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_9_filt, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_9_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_9_filt, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_9_filt, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p2 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "/BM_WS_9_filtered_QC_plot2.pdf") , width = 15, height = 15)
p2
##dev.off()
```


### BM_WS_10
(tab content)
```{r}
#Visualize QC metrics after filtering

#UMI counts per gene
summary(apply(GetAssayData(BM_WS_10_filt, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_10_filt, slot = "counts"), 2, sum))

#pdf(paste0(fig_path, "/BM_WS_10_filtered_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_10_filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_10_filt, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_10_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_10_filt, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_10_filt, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p2 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "/BM_WS_10_filtered_QC_plot2.pdf") , width = 15, height = 15)
p2
##dev.off()
```


### BM_WS_15
(tab content)
```{r}
#Visualize QC metrics after filtering

#UMI counts per gene
summary(apply(GetAssayData(BM_WS_15_filt, slot = "counts"), 1, sum))


#UMI counts per cell
summary(apply(GetAssayData(BM_WS_15_filt, slot = "counts"), 2, sum))

#pdf(paste0(fig_path, "/BM_WS_15_filtered_QC_plot1.pdf") , width = 10, height = 10)
VlnPlot(BM_WS_15_filt, features = c("nFeature_RNA", "nCount_RNA", "percent.mt", "nCount_ADT", "nFeature_ADT"), ncol = 3)
##dev.off()

plot1 <- FeatureScatter(BM_WS_15_filt, feature1 = "nCount_RNA", feature2 = "percent.mt") + NoLegend()
plot2 <- FeatureScatter(BM_WS_15_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") + NoLegend()
plot3 <- FeatureScatter(BM_WS_15_filt, feature1 = "nCount_RNA", feature2 = "nCount_ADT") + NoLegend()
plot4 <- FeatureScatter(BM_WS_15_filt, feature1 = "nCount_ADT", feature2 = "nFeature_ADT") + NoLegend()
p2 <- plot1 + plot2 + plot3 + plot4

#pdf(paste0(fig_path, "/BM_WS_15_filtered_QC_plot2.pdf") , width = 15, height = 15)
p2
##dev.off()
```


##Detect and exclude doublets/multiplets <br>
Notes (Reference: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md ):<br>
dbr, the doublet rate was calculated as 3.22%  from Seurat: https://satijalab.org/costpercell/ based on the parameters: number of cells desired = 4000 (approximating cell ranger results), Number of 10x lanes = 1, Number of multiplexed samples = 1
This also aligns with the rates estimated by 10X Genomics (Reference: https://assets.ctfassets.net/an68im79xiti/4UZmzpmHGn09BRRIIyagG2/56766ac8e3488f9d66b0275c907751ac/CG000206_ChromiumNextGEMSingleCell3_v3.1_CellSurfaceProtein_Rev_D.pdf)
###  BM_WS_5
```{r}
#Source code: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md

DefaultAssay(BM_WS_5_filt) <- "RNA"

sce_lung_dtSet <- as.SingleCellExperiment(BM_WS_5_filt)
set.seed(100)
sce_lung_dtSet <- scDblFinder(sce_lung_dtSet, dbr=0.0322) #dbr, the doublet rate was calculated as 3.22%  from Seurat: https://satijalab.org/costpercell/ based on the parameters: number of cells desired = 4000 (approximating cell ranger results), Number of 10x lanes = 1, Number of multiplexed samples = 1; Reference: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md #The number of doublet cells with 3.22% cutoff is 162 (4.6%) doublets called


identical(colnames(BM_WS_5_filt),colnames(sce_lung_dtSet)) #check that the column names are still the same prior to exporting the singlet/doublet information as a meta.data column in the Seurat object

BM_WS_5_filt$scDblFinder.class <- sce_lung_dtSet$scDblFinder.class
BM_WS_5_filt$scDblFinder.score <- sce_lung_dtSet$scDblFinder.score
#BM_WS_5_filt$scDblFinder.ratio <- sce_lung_dtSet$scDblFinder.ratio
BM_WS_5_filt$scDblFinder.weighted <- sce_lung_dtSet$scDblFinder.weighted
rm(sce_lung_dtSet)

BM_WS_5_filt$scDblFinder.class <- factor(BM_WS_5_filt$scDblFinder.class, levels=c("singlet","doublet"))

#pdf(paste0(fig_path, "/BM_WS_5_scdblfinder_Doublets.pdf"))
FeatureScatter(BM_WS_5_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by="scDblFinder.class", cols=c(alpha("blue",0.01),alpha("red",0.5))) + scale_y_continuous(trans="log10") + scale_x_continuous(trans="log10")
##dev.off()


#Source code: https://romanhaa.github.io/projects/scrnaseq_workflow/
#pdf(paste0(fig_path, "/BM_WS_5_doublets_scDblFinder_QC metrics.pdf"))
ggplot(as.data.frame(BM_WS_5_filt@meta.data), aes(x = scDblFinder.class, y = nFeature_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
ggplot(as.data.frame(BM_WS_5_filt@meta.data), aes(x = scDblFinder.class, y = nCount_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
##dev.off()

#exclude doublets
BM_WS_5_filt <- subset(BM_WS_5_filt, subset=scDblFinder.class=="singlet")
BM_WS_5_filt 

```


###  BM_WS_6
```{r}
#Source code: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md

DefaultAssay(BM_WS_6_filt) <- "RNA"

sce_lung_dtSet <- as.SingleCellExperiment(BM_WS_6_filt)
set.seed(100)
sce_lung_dtSet <- scDblFinder(sce_lung_dtSet, dbr=0.0322) #dbr, the doublet rate was calculated as 3.22%  from Seurat: https://satijalab.org/costpercell/ based on the parameters: number of cells desired = 4000 (approximating cell ranger results), Number of 10x lanes = 1, Number of multiplexed samples = 1; Reference: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md #The number of doublet cells with 3.22% cutoff is 106 (5.6%) doublets called


identical(colnames(BM_WS_6_filt),colnames(sce_lung_dtSet)) #check that the column names are still the same prior to exporting the singlet/doublet information as a meta.data column in the Seurat object

BM_WS_6_filt$scDblFinder.class <- sce_lung_dtSet$scDblFinder.class
BM_WS_6_filt$scDblFinder.score <- sce_lung_dtSet$scDblFinder.score
#BM_WS_6_filt$scDblFinder.ratio <- sce_lung_dtSet$scDblFinder.ratio
BM_WS_6_filt$scDblFinder.weighted <- sce_lung_dtSet$scDblFinder.weighted
rm(sce_lung_dtSet)

BM_WS_6_filt$scDblFinder.class <- factor(BM_WS_6_filt$scDblFinder.class, levels=c("singlet","doublet"))

#pdf(paste0(fig_path, "/BM_WS_6_scdblfinder_Doublets.pdf"))
FeatureScatter(BM_WS_6_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by="scDblFinder.class", cols=c(alpha("blue",0.01),alpha("red",0.5))) + scale_y_continuous(trans="log10") + scale_x_continuous(trans="log10")
##dev.off()


#Source code: https://romanhaa.github.io/projects/scrnaseq_workflow/
#pdf(paste0(fig_path, "/BM_WS_6_doublets_scDblFinder_QC metrics.pdf"))
ggplot(as.data.frame(BM_WS_6_filt@meta.data), aes(x = scDblFinder.class, y = nFeature_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
ggplot(as.data.frame(BM_WS_6_filt@meta.data), aes(x = scDblFinder.class, y = nCount_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
##dev.off()

#exclude doublets
BM_WS_6_filt <- subset(BM_WS_6_filt, subset=scDblFinder.class=="singlet")
BM_WS_6_filt 

```


###  BM_WS_8
```{r}
#Source code: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md

DefaultAssay(BM_WS_8_filt) <- "RNA"

sce_lung_dtSet <- as.SingleCellExperiment(BM_WS_8_filt)
set.seed(100)
sce_lung_dtSet <- scDblFinder(sce_lung_dtSet, dbr=0.0322) #dbr, the doublet rate was calculated as 3.22%  from Seurat: https://satijalab.org/costpercell/ based on the parameters: number of cells desired = 4000 (approximating cell ranger results), Number of 10x lanes = 1, Number of multiplexed samples = 1; Reference: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md #The number of doublet cells with 3.22% cutoff is 201 (5.1%) doublets called


identical(colnames(BM_WS_8_filt),colnames(sce_lung_dtSet)) #check that the column names are still the same prior to exporting the singlet/doublet information as a meta.data column in the Seurat object

BM_WS_8_filt$scDblFinder.class <- sce_lung_dtSet$scDblFinder.class
BM_WS_8_filt$scDblFinder.score <- sce_lung_dtSet$scDblFinder.score
#BM_WS_8_filt$scDblFinder.ratio <- sce_lung_dtSet$scDblFinder.ratio
BM_WS_8_filt$scDblFinder.weighted <- sce_lung_dtSet$scDblFinder.weighted
rm(sce_lung_dtSet)

BM_WS_8_filt$scDblFinder.class <- factor(BM_WS_8_filt$scDblFinder.class, levels=c("singlet","doublet"))

#pdf(paste0(fig_path, "/BM_WS_8_scdblfinder_Doublets.pdf"))
FeatureScatter(BM_WS_8_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by="scDblFinder.class", cols=c(alpha("blue",0.01),alpha("red",0.5))) + scale_y_continuous(trans="log10") + scale_x_continuous(trans="log10")
##dev.off()


#Source code: https://romanhaa.github.io/projects/scrnaseq_workflow/
#pdf(paste0(fig_path, "/BM_WS_8_doublets_scDblFinder_QC metrics.pdf"))
ggplot(as.data.frame(BM_WS_8_filt@meta.data), aes(x = scDblFinder.class, y = nFeature_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
ggplot(as.data.frame(BM_WS_8_filt@meta.data), aes(x = scDblFinder.class, y = nCount_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
##dev.off()

#exclude doublets
BM_WS_8_filt <- subset(BM_WS_8_filt, subset=scDblFinder.class=="singlet")
BM_WS_8_filt 

```


###  BM_WS_9
```{r}
#Source code: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md

DefaultAssay(BM_WS_9_filt) <- "RNA"

sce_lung_dtSet <- as.SingleCellExperiment(BM_WS_9_filt)
set.seed(100)
sce_lung_dtSet <- scDblFinder(sce_lung_dtSet, dbr=0.0322) #dbr, the doublet rate was calculated as 3.22%  from Seurat: https://satijalab.org/costpercell/ based on the parameters: number of cells desired = 4000 (approximating cell ranger results), Number of 10x lanes = 1, Number of multiplexed samples = 1; Reference: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md #The number of doublet cells with 3.22% cutoff is 142 (4.7%) doublets called


identical(colnames(BM_WS_9_filt),colnames(sce_lung_dtSet)) #check that the column names are still the same prior to exporting the singlet/doublet information as a meta.data column in the Seurat object

BM_WS_9_filt$scDblFinder.class <- sce_lung_dtSet$scDblFinder.class
BM_WS_9_filt$scDblFinder.score <- sce_lung_dtSet$scDblFinder.score
#BM_WS_9_filt$scDblFinder.ratio <- sce_lung_dtSet$scDblFinder.ratio
BM_WS_9_filt$scDblFinder.weighted <- sce_lung_dtSet$scDblFinder.weighted
rm(sce_lung_dtSet)

BM_WS_9_filt$scDblFinder.class <- factor(BM_WS_9_filt$scDblFinder.class, levels=c("singlet","doublet"))

#pdf(paste0(fig_path, "/BM_WS_9_scdblfinder_Doublets.pdf"))
FeatureScatter(BM_WS_9_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by="scDblFinder.class", cols=c(alpha("blue",0.01),alpha("red",0.5))) + scale_y_continuous(trans="log10") + scale_x_continuous(trans="log10")
##dev.off()


#Source code: https://romanhaa.github.io/projects/scrnaseq_workflow/
#pdf(paste0(fig_path, "/BM_WS_9_doublets_scDblFinder_QC metrics.pdf"))
ggplot(as.data.frame(BM_WS_9_filt@meta.data), aes(x = scDblFinder.class, y = nFeature_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
ggplot(as.data.frame(BM_WS_9_filt@meta.data), aes(x = scDblFinder.class, y = nCount_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
##dev.off()

#exclude doublets
BM_WS_9_filt <- subset(BM_WS_9_filt, subset=scDblFinder.class=="singlet")
BM_WS_9_filt 

```



###  BM_WS_10
```{r}
#Source code: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md

DefaultAssay(BM_WS_10_filt) <- "RNA"

sce_lung_dtSet <- as.SingleCellExperiment(BM_WS_10_filt)
set.seed(100)
sce_lung_dtSet <- scDblFinder(sce_lung_dtSet, dbr=0.0322) #dbr, the doublet rate was calculated as 3.22%  from Seurat: https://satijalab.org/costpercell/ based on the parameters: number of cells desired = 4000 (approximating cell ranger results), Number of 10x lanes = 1, Number of multiplexed samples = 1; Reference: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md #The number of doublet cells with 3.22% cutoff is 150 (4.9%) doublets called


identical(colnames(BM_WS_10_filt),colnames(sce_lung_dtSet)) #check that the column names are still the same prior to exporting the singlet/doublet information as a meta.data column in the Seurat object

BM_WS_10_filt$scDblFinder.class <- sce_lung_dtSet$scDblFinder.class
BM_WS_10_filt$scDblFinder.score <- sce_lung_dtSet$scDblFinder.score
#BM_WS_10_filt$scDblFinder.ratio <- sce_lung_dtSet$scDblFinder.ratio
BM_WS_10_filt$scDblFinder.weighted <- sce_lung_dtSet$scDblFinder.weighted
rm(sce_lung_dtSet)

BM_WS_10_filt$scDblFinder.class <- factor(BM_WS_10_filt$scDblFinder.class, levels=c("singlet","doublet"))

#pdf(paste0(fig_path, "/BM_WS_10_scdblfinder_Doublets.pdf"))
FeatureScatter(BM_WS_10_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by="scDblFinder.class", cols=c(alpha("blue",0.01),alpha("red",0.5))) + scale_y_continuous(trans="log10") + scale_x_continuous(trans="log10")
##dev.off()


#Source code: https://romanhaa.github.io/projects/scrnaseq_workflow/
#pdf(paste0(fig_path, "/BM_WS_10_doublets_scDblFinder_QC metrics.pdf"))
ggplot(as.data.frame(BM_WS_10_filt@meta.data), aes(x = scDblFinder.class, y = nFeature_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
ggplot(as.data.frame(BM_WS_10_filt@meta.data), aes(x = scDblFinder.class, y = nCount_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
##dev.off()

#exclude doublets
BM_WS_10_filt <- subset(BM_WS_10_filt, subset=scDblFinder.class=="singlet")
BM_WS_10_filt 

```



###  BM_WS_15
```{r}
#Source code: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md

DefaultAssay(BM_WS_15_filt) <- "RNA"

sce_lung_dtSet <- as.SingleCellExperiment(BM_WS_15_filt)
set.seed(100)
sce_lung_dtSet <- scDblFinder(sce_lung_dtSet, dbr=0.0322) #dbr, the doublet rate was calculated as 3.22%  from Seurat: https://satijalab.org/costpercell/ based on the parameters: number of cells desired = 4000 (approximating cell ranger results), Number of 10x lanes = 1, Number of multiplexed samples = 1; Reference: https://github.com/Terkild/CITE-seq_optimization/blob/master/Demux_Preprocess_Downsample.md #The number of doublet cells with 3.22% cutoff is 194 (5%) doublets called


identical(colnames(BM_WS_15_filt),colnames(sce_lung_dtSet)) #check that the column names are still the same prior to exporting the singlet/doublet information as a meta.data column in the Seurat object

BM_WS_15_filt$scDblFinder.class <- sce_lung_dtSet$scDblFinder.class
BM_WS_15_filt$scDblFinder.score <- sce_lung_dtSet$scDblFinder.score
#BM_WS_15_filt$scDblFinder.ratio <- sce_lung_dtSet$scDblFinder.ratio
BM_WS_15_filt$scDblFinder.weighted <- sce_lung_dtSet$scDblFinder.weighted
rm(sce_lung_dtSet)

BM_WS_15_filt$scDblFinder.class <- factor(BM_WS_15_filt$scDblFinder.class, levels=c("singlet","doublet"))

#pdf(paste0(fig_path, "/BM_WS_15_scdblfinder_Doublets.pdf"))
FeatureScatter(BM_WS_15_filt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by="scDblFinder.class", cols=c(alpha("blue",0.01),alpha("red",0.5))) + scale_y_continuous(trans="log10") + scale_x_continuous(trans="log10")
##dev.off()


#Source code: https://romanhaa.github.io/projects/scrnaseq_workflow/
#pdf(paste0(fig_path, "/BM_WS_15_doublets_scDblFinder_QC metrics.pdf"))
ggplot(as.data.frame(BM_WS_15_filt@meta.data), aes(x = scDblFinder.class, y = nFeature_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
ggplot(as.data.frame(BM_WS_15_filt@meta.data), aes(x = scDblFinder.class, y = nCount_RNA)) +
     geom_violin() +
     theme_bw() +
     coord_flip()
##dev.off()

#exclude doublets
BM_WS_15_filt <- subset(BM_WS_15_filt, subset=scDblFinder.class=="singlet")
BM_WS_15_filt 

```




# Integrating all wild type and il27raKO samples together
## Merging wild type and il27raKO datasets <br>

```{r}
#combining WT and il27raKO datasets
all_merged <- merge(x= BM_WS_5_filt, y = c(BM_WS_6_filt, BM_WS_8_filt, BM_WS_9_filt, BM_WS_10_filt, BM_WS_15_filt), add.cell.ids = c("wt_BM_WS_5", "wt_BM_WS_6", "wt_BM_WS_8", "il27raKO_BM_WS_9", "il27raKO_BM_WS_10", "il27raKO_BM_WS_15"), project = "wt_il27raKO")
all_merged

head(all_merged@meta.data)
tail(all_merged@meta.data)

```


## Normalization: RNA <br>
Notes: https://satijalab.org/seurat/articles/integration_introduction.html <br>
- Based on the recommendations from the Seurat developers (Reference: https://github.com/satijalab/seurat/issues/3890#issuecomment-756885252), integrating RNA and ADT assays separately. <br>
First, normalizing RNA individually using SCTransform that is able to better distinguish biological variation from technical noise (Reference: https://satijalab.org/seurat/articles/sctransform_vignette.html).

```{r, warning=FALSE}
#Normalization of RNA of each sample set/condition
#Split by samples
all_merged_list <- Seurat::SplitObject(object = all_merged, split.by = "sample_id")

#Normalize using SCTransform, regressing out any variation due to mitochondrial reads
all_merged_list <- lapply(X = all_merged_list, FUN =  SCTransform, assay = "RNA", vars.to.regress = c("percent.mt"), verbose = FALSE)


```


### Segregation of cells by cell cycle phase<br>
```{r}
#S and G2/M gene list and processing from Harvard Chan Bioinformatics Github (https://github.com/hbc/tinyatlas/tree/master/cell_cycle)
#S and G2/M gene list from Harvard Chan Bioinformatics Github (https://github.com/hbc/tinyatlas/tree/master/cell_cycle), ENSEMBL IDs converted by batch conversion from JAX
g2m_genes_mouse <- list("Ube2c", "Lbr", "Ctcf", "Cdc20", "Cbx5", "Kif11", "Anp32e", "Birc5", "Cdk1", "Tmpo", "Hmmr", "Jpt1", "Pimreg", "Aurkb", "Top2a", "Gtse1", "Rangap1", "Cdca3", "Ndc80", "Kif20b", "Cenpf", "Nek2", "Nuf2", "Nusap1", "Bub1", "Tpx2", "Aurka", "Ect2", "Cks1b", "Kif2c", "Cdca8", "Cenpa", "Mki67", "Ccnb2", "Kif23", "Smc4", "G2e3", "Tubb4b", "Anln", "Tacc3", "Dlgap5", "Ckap2", "Ncapd2", "Ttk", "Ckap5", "Cdc25c", "Hjurp", "Cenpe", "Ckap2l", "Cdca2", "Hmgb2", "Cks2", "Psrc1", "Gas2l3")

s_genes_mouse <- list("Cdc45", "Uhrf1", "Mcm2", "Slbp", "Mcm5", "Pola1", "Gmnn", "Cdc6", "Rrm2", "Atad2", "Dscc1", "Mcm4", "Chaf1b", "Rfc2", "Msh2", "Fen1", "Hells", "Prim1", "Tyms", "Mcm6", "Wdr76", "Rad51", "Pcna", "Ccne2", "Casp8ap2", "Usp1", "Nasp", "Rpa2", "Ung", "Rad51ap1", "Blm", "Pold3", "Rrm1", "Cenpu", "Gins2", "Tipin", "Brip1", "Dtl", "Exo1", "Ubr7", "Clspn", "E2f8", "Cdca7")


# Perform cell cycle scoring
all_merged_list <- lapply(X = all_merged_list, FUN =  CellCycleScoring, s.features = s_genes_mouse, g2m.features = g2m_genes_mouse, set.ident = TRUE)
   

#pdf(paste0(fig_path, "/all_merged_Cellcycle_phase.pdf"))
all_merged_list$BM_WS_5[[]] %>% data.frame() %>% ggplot(aes(x=S.Score, y=G2M.Score, color=Phase)) + geom_point() + ggtitle("WT: Cell Cycle Phase: BM_WS_5")
all_merged_list$BM_WS_6[[]] %>% data.frame() %>% ggplot(aes(x=S.Score, y=G2M.Score, color=Phase)) + geom_point() + ggtitle("WT: Cell Cycle Phase: BM_WS_6")
all_merged_list$BM_WS_8[[]] %>% data.frame() %>% ggplot(aes(x=S.Score, y=G2M.Score, color=Phase)) + geom_point() + ggtitle("WT: Cell Cycle Phase: BM_WS_8")

all_merged_list$BM_WS_9[[]] %>% data.frame() %>% ggplot(aes(x=S.Score, y=G2M.Score, color=Phase)) + geom_point() + ggtitle("C5aR DKO: Cell Cycle Phase: BM_WS_9")
all_merged_list$BM_WS_10[[]] %>% data.frame() %>% ggplot(aes(x=S.Score, y=G2M.Score, color=Phase)) + geom_point() + ggtitle("C5aR DKO: Cell Cycle Phase: BM_WS_10")
all_merged_list$BM_WS_15[[]] %>% data.frame() %>% ggplot(aes(x=S.Score, y=G2M.Score, color=Phase)) + geom_point() + ggtitle("C5aR DKO: Cell Cycle Phase: BM_WS_15")

##dev.off()

#Based on the above plot, as there is not much of a separation between the groups, cell cycle phase may only have a minor effect on the expression patterns (Ref: https://www.bioinformatics.babraham.ac.uk/training/10XRNASeq/seurat_workflow.html)
```

## Pre-Integration <br>
```{r, warning=FALSE}
#check if integration is necessary i.e. do the cells separate by sample_id, to determine if alignment will be helpful. To check this perform normalization of merged data
#create temporary object
tmp <- all_merged
DefaultAssay(tmp) <- "RNA"
tmp <- Seurat::SCTransform(tmp, vars.to.regress = "percent.mt", verbose = FALSE)

#Principal Component Analysis (PCA)
tmp <- Seurat::RunPCA(object = tmp, verbose = FALSE)

#Determine dimensionality of the dataset
#Elbow Plot
ElbowPlot(tmp, ndims = 50, reduction = "pca")
#ndims set to 50 based on warning message "The object only has information for 50 reductions" when elbow plot was run with ndims = 100

#Based on ElbowPlots for RNA , selecting 30 PCs which appear to include most of the variance in the data.
tmp <- FindNeighbors(tmp, dims = 1:30)
tmp <- FindClusters(tmp, resolution = 1) #selecting resolution of 1 arbitarily 

#Non-linear dimensional reduction
tmp <- Seurat::RunUMAP(tmp, dims = 1:30)


#pdf(paste0(fig_path, "/PCA_RNA_UMAP_bySample_allmerged.pdf"))

#Visualizing UMAP to determine if the clusters segregate by sample condition
DimPlot(tmp, reduction = "umap", group.by = "sample_id")

DimPlot(tmp, reduction = "pca", group.by = "sample_id")
##dev.off()


#Checking PCA for ADT
DefaultAssay(tmp) <- "ADT"
VariableFeatures(tmp) <- rownames(tmp[["ADT"]])
tmp <- Seurat::NormalizeData(tmp, normalization.method = "CLR", margin = 2, verbose = FALSE) %>% Seurat::ScaleData(do.scale = FALSE, vars.to.regress = "nCount_ADT") %>% Seurat::RunPCA(verbose = FALSE, reduction.name = "apca" )

#Determine dimensionality of the dataset
#Elbow Plot
ElbowPlot(tmp, ndims = 50, reduction = "apca")
#ndims set to 50 based on warning message "The object only has information for 50 reductions" when elbow plot was run with ndims = 100


rm(tmp)


#The clusters do not seem to separate specifically by sample_id; so Integration may not be needed within WT and C5aR DKO sample groups. However, performing Integration may help account for any minor sample-to-sample variation, within each group. Proceeding with Integration


```

# save workspace
```{r}
#save.image(paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_v1.1.RData"))
#load(paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_v1.1.RData"))
```


## Integration of wt and il27raKOo samples: RNA <br>
```{r, warning=FALSE}

# Integration of wt and il27raKOo samples: RNA
features_all_merged <- Seurat::SelectIntegrationFeatures(object.list = all_merged_list, nfeatures = 3000) #using 3000 as these are the number of features that are identified as most variable by default during normalization by SCTransform
all_merged_list <- Seurat::PrepSCTIntegration(object.list = all_merged_list, anchor.features = features_all_merged)
all_merged_list <- lapply(X = all_merged_list, FUN = RunPCA, features = features_all_merged) #running PCA, as a pre-requisite for RPCA reduction during integration

set.seed(42)
all_merged_anchors <- Seurat::FindIntegrationAnchors(object.list = all_merged_list, normalization.method = "SCT", anchor.features = features_all_merged, dims = 1:30, reduction = "rpca", k.anchor = 5, verbose = FALSE) #selecting 30 pcs based on pre-integration check, which showed that the variance is low beyond 30 pcs; selecting rpca as reduction method, and setting the k.anchor to the default 5 neighbors (References: https://github.com/satijalab/seurat/issues/3890#issuecomment-756885252; https://satijalab.org/seurat/articles/integration_rpca.html#performing-integration-on-datasets-normalized-with-sctransform-1)
set.seed(42)
all_integrated <- Seurat::IntegrateData(anchorset = all_merged_anchors, normalization.method = "SCT", dims = 1:30)

```


###Linear dimensional reduction of RNA assay all samples <br>
```{r}
#Linear dimensional reduction of RNA assay data
#Principal Component Analysis (PCA)

###Linear dimensional reduction of RNA assay data using Principal Component Analysis (PCA)
set.seed(42)
all_integrated <- Seurat::RunPCA(object = all_integrated, verbose = FALSE)

#Determine dimensionality of the dataset
#Elbow Plot
ElbowPlot(all_integrated, ndims = 50, reduction = "pca")
#ndims set to 50 based on warning message "The object only has information for 50 reductions" when elbow plot was run with ndims = 100

##ggsave(paste0(fig_path, "/ElbowPlot_RNA_PCA_all_integrated.pdf"), units = "in", width = 6, height = 6, dpi = 600)


#Visualizing cells and features that define the PCA
print(all_integrated[["pca"]], dims = 1:50, nfeatures = 5)

#DimPlot
DimPlot(all_integrated, reduction = "pca")
##ggsave(paste0(fig_path, "/DimPlot_RNA_PCA_all_integrated.pdf"), dpi = 600, units = "in", width = 5.5, height = 6)

#Visualizing top genes associated with reduction components
VizDimLoadings(all_integrated, dims = 1:5, reduction = "pca")
##ggsave(paste0(fig_path, "/Visual_top_genes_5_PCA_components_RNA_all_integrated.pdf"), units = "in", width = 12, height = 15, dpi = 600)

#Heatmaps
#pdf(paste0(fig_path, "/DimHeatmap_1-40PCA_50genes_RNA_all_integrated.pdf"), width = 15, height = 15, pointsize = 16)
DimHeatmap(all_integrated, dims = 1:5, cells = 500, reduction = "pca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 6:10, cells = 500, reduction = "pca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 11:15, cells = 500, reduction = "pca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 16:20, cells = 500, reduction = "pca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 21:25, cells = 500, reduction = "pca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 26:30, cells = 500, reduction = "pca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 31:35, cells = 500, reduction = "pca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 36:40, cells = 500, reduction = "pca", nfeatures = 50, balanced = TRUE)
##dev.off()

```

## Normalization and integration all samples: ADT <br>
```{r, warning=FALSE}
#Normalization of ADT of each sample set/condition (using CLR normalization method ; References: https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html#wnn-analysis-of-cite-seq-rna-adt-1 ;  https://cran.r-project.org/web/packages/SignacX/vignettes/signac-Seurat_CITE-seq.html). 
# setting variable features to include all ADT: rownames(wt_merged_list[[1]][["ADT"]])

#Normalization
all_merged_list <- lapply(X = all_merged_list, FUN = function(x){
  DefaultAssay(x) <- "ADT"
  VariableFeatures(x) <- rownames(x[["ADT"]])
  x <- NormalizeData(x, assay = "ADT", normalization.method = "CLR", margin = 2, verbose = FALSE)
}) #margin = 2 normalizes for differences in ADT depth across cells as suggested by the Seurat developers (References: https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html; https://github.com/satijalab/seurat/issues/3605)

# select features that are repeatedly variable across datasets for integration run PCA on each
# dataset using these features
features_all_merged_ADT <- rownames(all_merged_list[[1]][["ADT"]]) #this selects all 198 ADTs
all_merged_list <- lapply(X = all_merged_list, FUN = function(x) {
    x <- ScaleData(x, features = features_all_merged_ADT, do.scale = FALSE, vars.to.regress = "nCount_ADT", verbose = FALSE)
    x <- RunPCA(x, features = features_all_merged_ADT, verbose = FALSE)
}) #running ScaleData and PCA, as a pre-requisite for RPCA reduction during integration (Reference: https://satijalab.org/seurat/articles/integration_rpca.html#modifying-the-strength-of-integration-1; https://github.com/satijalab/seurat/issues/4815) #regressing out any technical variation due to sequencing depth

#Integration
set.seed(42)
all_merged_anchors_ADT <- Seurat::FindIntegrationAnchors(object.list = all_merged_list, normalization.method = "LogNormalize", anchor.features = features_all_merged_ADT, dims = 1:20, reduction = "rpca", k.anchor = 5, scale = FALSE, verbose = FALSE) #selecting 20 pcs based on pre-integration check, which showed that the variance is low beyond 20 pcs for ADT PCA, however can take more for integration; selecting rpca as reduction method based on the information from the Seurat development team (References: https://github.com/satijalab/seurat/issues/3890#issuecomment-756885252; https://satijalab.org/seurat/articles/integration_rpca.html#performing-integration-on-datasets-normalized-with-sctransform-1)

set.seed(42)
all_integrated_ADT <- Seurat::IntegrateData(anchorset = all_merged_anchors_ADT, normalization.method = "LogNormalize", dims = 1:20, new.assay.name = "integratedadt")

#Scale data (Reference: https://satijalab.org/seurat/articles/integration_rpca.html#perform-integration-1)
all_integrated_ADT <- Seurat::ScaleData(all_integrated_ADT, features = rownames(all_integrated_ADT), do.scale = FALSE, vars.to.regress = "nCount_ADT",  verbose = FALSE) #all ADT will be centered but not scaled


```

#Adding the integrated ADT data to the integrated RNA data: all samples
```{r}

#Adding the integrated ADT data to the integrated RNA data
#WT
all_integrated[["IADT"]] <- all_integrated_ADT[["integratedadt"]]
all_integrated


```

### Linear dimensional reduction of ADT assay data: all samples <br>
```{r}
#Linear dimensional reduction of RNA assay data
DefaultAssay(all_integrated) <- "IADT"

#Principal Component Analysis (PCA)

###Linear dimensional reduction of RNA assay data using Principal Component Analysis (PCA)
set.seed(42)
all_integrated <- Seurat::RunPCA(object = all_integrated, verbose = FALSE, reduction.name = "apca")

#Determine dimensionality of the dataset
#Elbow Plot
ElbowPlot(all_integrated, ndims = 50, reduction = "apca")
#ndims set to 50 based on warning message "The object only has information for 50 reductions" when elbow plot was run with ndims = 100

##ggsave(paste0(fig_path, "/ElbowPlot_adt-PCA_all_integrated_ADT.pdf"), units = "in", width = 6, height = 6, dpi = 600)


#Visualizing cells and features that define the PCA
print(all_integrated[["apca"]], dims = 1:50, nfeatures = 5)

#DimPlot
DimPlot(all_integrated, reduction = "apca")
##ggsave(paste0(fig_path, "/DimPlot_adt-PCA_all_integrated_ADT.pdf"), dpi = 600, units = "in", width = 5.5, height = 6)

#Visualizing top genes associated with reduction components
VizDimLoadings(all_integrated, dims = 1:5, reduction = "apca")
##ggsave(paste0(fig_path, "/Visual_top_genes_5_PCA_components_adt-all_integrated_ADT.pdf"), units = "in", width = 12, height = 15, dpi = 600)

#Heatmaps
#pdf(paste0(fig_path, "/DimHeatmap_1-40PCA_50genes_adt-all_integrated_ADT.pdf"), width = 15, height = 15, pointsize = 16)
DimHeatmap(all_integrated, dims = 1:5, cells = 500, reduction = "apca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 6:10, cells = 500, reduction = "apca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 11:15, cells = 500, reduction = "apca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 16:20, cells = 500, reduction = "apca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 21:25, cells = 500, reduction = "apca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 26:30, cells = 500, reduction = "apca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 31:35, cells = 500, reduction = "apca", nfeatures = 50, balanced = TRUE)
DimHeatmap(all_integrated, dims = 36:40, cells = 500, reduction = "apca", nfeatures = 50, balanced = TRUE)
##dev.off()



```

### Identify multimodal neighbours: all samples
Notes (Reference:  https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html): <br>
1.For each cell, calculate its closest neighbors in the dataset based on a weighted combination of RNA and protein similarities. <br>
2.The cell-specific modality weights and multimodal neighbors are calculated in a single function specify the dimensionality of each modality (similar to specifying the number of PCs to include in scRNA-seq clustering), but variation in these settings shows that small changes have minimal effect on the overall results.<br>
```{r}

#Source: https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html
# Identify multimodal neighbors. These will be stored in the neighbors slot, 
# and can be accessed using bm[['weighted.nn']]
# The WNN graph can be accessed at bm[["wknn"]], 
# and the SNN graph used for clustering at bm[["wsnn"]]
# Cell-specific modality weights can be accessed at bm$RNA.weight

all_integrated <- Seurat::FindMultiModalNeighbors(all_integrated, reduction.list = list("pca", "apca"), dims.list = list(1:30, 1:20), modality.weight.name = "SCT.weight")#using SCT.weight instead of RNA.weight as the RNA normalization and scaling were performed using SCT


```


### Clustering cells using SLM algorithm and UMAP (non-linear dimensional reduction): all samples<br>
Notes (Reference: https://satijalab.org/seurat/articles/pbmc3k_tutorial.html):
1. Clustering cells based on a weighted combination of RNA and protein data; checking with different resolutions from 0.5 to 4 for granularity <br>
2. The resolution limit of 4 was chosen as the number of cells is approximately 10000 per group, based on the note from Seurat developers that between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells, and with this dataset having 3 times that number, checking until resolution 4. <br>
```{r}
#Clustering cells based on a weighted combination of RNA and protein data; checking with different resolutions from 0.5 to 4 for granularity; resolution limit of 4 was chosen as the number of cells is approximately 10000 per group, based on the note from Seurat developers that between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells, and with this dataset having 3 times that number, checking until resolution 4.

#Clustering
all_integrated <- Seurat::FindClusters(all_integrated, graph.name = "wsnn", algorithm = 3, resolution = seq(from = 0.5, to = 4, by = 0.3), verbose = FALSE) #algorithm 3 = SLM algorithm

#Non-linear dimensional reduction using UMAP
all_integrated <- Seurat::RunUMAP(all_integrated, nn.name = "weighted.nn", reduction.name = "wnn.umap", reduction.key = "wnnUMAP_")


```


## Visualization of cluster through UMAP plot based on combined RNA and ADT data: all samples {.tabset .tabset-fade .tabset-pills} <br>
```{r}

#pdf(paste0(fig_path, "/UMAP_resolutions_0.5-4_by0.3_all_integrated.pdf"))

DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.0.5") + ggtitle(label = "UMAP (res = 0.5)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.0.8") + ggtitle(label = "UMAP (res = 0.8)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.1.1") + ggtitle(label = "UMAP (res = 1.1)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.1.4") + ggtitle(label = "UMAP (res = 1.4)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.1.7") + ggtitle(label = "UMAP (res = 1.7)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2") + ggtitle(label = "UMAP (res = 2)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2.3") + ggtitle(label = "UMAP (res = 2.3)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2.6") + ggtitle(label = "UMAP (res = 2.6)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2.9") + ggtitle(label = "UMAP (res = 2.9)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.3.2") + ggtitle(label = "UMAP (res = 3.2)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.3.5") + ggtitle(label = "UMAP (res = 3.5)")
DimPlot(object = all_integrated, reduction = "wnn.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.3.8") + ggtitle(label = "UMAP (res = 3.8)")

##dev.off()

```

##Visualization of clusters based only on RNA or ADT data <br>
- can also compute UMAP visualization based on only the RNA and protein data and compare.<br>
- RNA analysis may be more informative than the ADT analysis in identifying progenitor states (the ADT panel contains markers for differentiated cells), while the converse is true of T cell states (where the ADT analysis outperforms RNA) (Source: https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html).
###UMAP reduction of RNA and ADT data
```{r}

#RNA UMAP
all_integrated <- Seurat::RunUMAP(all_integrated, reduction = "pca", dims = 1:30, assay = "RNA", reduction.name = "rna.umap", reduction.key = "rnaUMAP_")

#ADT UMAP
all_integrated <- Seurat::RunUMAP(all_integrated,reduction = "apca", dims = 1:20, assay = "ADT", reduction.name = "adt.umap", reduction.key = "adtUMAP_")

#Visualization of RNA UMAP
#pdf(paste0(fig_path, "/RNA_UMAP_resolutions_0.5-4_by0.3_all_integrated.pdf"))

DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.0.5", repel = TRUE) + ggtitle(label = "UMAP (res = 0.5)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.0.8", repel = TRUE) + ggtitle(label = "UMAP (res = 0.8)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.1.1", repel = TRUE) + ggtitle(label = "UMAP (res = 1.1)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.1.4", repel = TRUE) + ggtitle(label = "UMAP (res = 1.4)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.1.7", repel = TRUE) + ggtitle(label = "UMAP (res = 1.7)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2", repel = TRUE) + ggtitle(label = "UMAP (res = 2)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2.3", repel = TRUE) + ggtitle(label = "UMAP (res = 2.3)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2.6", repel = TRUE) + ggtitle(label = "UMAP (res = 2.6)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2.9", repel = TRUE) + ggtitle(label = "UMAP (res = 2.9)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.3.2", repel = TRUE) + ggtitle(label = "UMAP (res = 3.2)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.3.5", repel = TRUE) + ggtitle(label = "UMAP (res = 3.5)")
DimPlot(object = all_integrated, reduction = "rna.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.3.8", repel = TRUE) + ggtitle(label = "UMAP (res = 3.8)")



##dev.off()


#Visualization of ADT UMAP

#pdf(paste0(fig_path, "/adt-UMAP_resolutions_0.5-4_by0.3_all_integrated.pdf"))

DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.0.5", repel = TRUE) + ggtitle(label = "UMAP (res = 0.5)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.0.8", repel = TRUE) + ggtitle(label = "UMAP (res = 0.8)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.1.1", repel = TRUE) + ggtitle(label = "UMAP (res = 1.1)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.1.4", repel = TRUE) + ggtitle(label = "UMAP (res = 1.4)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.1.7", repel = TRUE) + ggtitle(label = "UMAP (res = 1.7)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2", repel = TRUE) + ggtitle(label = "UMAP (res = 2)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2.3", repel = TRUE) + ggtitle(label = "UMAP (res = 2.3)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2.6", repel = TRUE) + ggtitle(label = "UMAP (res = 2.6)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.2.9", repel = TRUE) + ggtitle(label = "UMAP (res = 2.9)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.3.2", repel = TRUE) + ggtitle(label = "UMAP (res = 3.2)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.3.5", repel = TRUE) + ggtitle(label = "UMAP (res = 3.5)")
DimPlot(object = all_integrated, reduction = "adt.umap", pt.size = 0.1, label = TRUE, group.by = "wsnn_res.3.8", repel = TRUE) + ggtitle(label = "UMAP (res = 3.8)")


##dev.off()

```


##Identifying cluster resolution for downstream analysis<br>
###checking cluster stability at different resolutions using clustree <br>
Notes (Reference: https://cran.r-project.org/web/packages/clustree/clustree.pdf ) :<br>
1.clustree nodes are colored by sc3 stability index. <br>
2.The index varies from 0 to 1, where 1 suggests that a cluster is more stable across resolutions.<br>  
3.The size of each node is related to the number of samples in each cluster. <br>
4. Edges are coloured according to the number of cells they represent and the transparency shows the incoming node proportion, the number of cells in the edge divided by the number of cells in the node it points to.<br>
```{r}
#Number of cells at different resolutions
clustree(all_integrated, prefix = "wsnn_res.", node_colour = "sc3_stability")
##ggsave(paste0(fig_path, "/clustree_res_0.5-4_by0.3_all_integrated.pdf"), units = "in", width = 12,height = 8)


#Reference: https://cran.r-project.org/web/packages/clustree/clustree.pdf
#clustree nodes are colored by sc3 stability index. The index varies from 0 to 1, where 1 suggests that a cluster is more stable across resolutions.
#the size of each node is related to the number of samples in each cluster 

```

###Silhouette score <br>
Notes (Reference and Source code:https://github.com/satijalab/Integration2019/blob/master/analysis_code/integration/integration_metrics.R#L36 ; https://romanhaa.github.io/projects/scrnaseq_workflow/):<br>
Silhouette Coefficient or silhouette score is a metric used to calculate the goodness of a clustering technique. Its value ranges from -1 to 1 (Reference: https://towardsdatascience.com/silhouette-coefficient-validating-clustering-techniques-e976bb81d10c?gi=3a20424c8283).<br>
1: Means clusters are well apart from each other and clearly distinguished.<br>
0: Means clusters are indifferent, or we can say that the distance between clusters is not significant.<br>
-1: Means clusters are assigned in the wrong way.<br>
The Silhouette coefficient ranges from 1 to 1 and values close to zero indicate random clustering with regard to the specified indicator. (Reference: https://www.nature.com/articles/s41467-019-08831-9#MOESM1 )<br>
```{r}
#Silhouette score (Reference and Source code:https://github.com/satijalab/Integration2019/blob/master/analysis_code/integration/integration_metrics.R#L36 ; https://romanhaa.github.io/projects/scrnaseq_workflow/)
#Notes:
#Silhouette Coefficient or silhouette score is a metric used to calculate the goodness of a clustering technique. Its value ranges from -1 to 1 (Reference: https://towardsdatascience.com/silhouette-coefficient-validating-clustering-techniques-e976bb81d10c?gi=3a20424c8283).
##1: Means clusters are well apart from each other and clearly distinguished.
##0: Means clusters are indifferent, or we can say that the distance between clusters is not significant.
##-1: Means clusters are assigned in the wrong way.
# The Silhouette coefficient ranges from 1 to 1 and values close to zero indicate random clustering with regard to the specified indicator. (Reference: https://www.nature.com/articles/s41467-019-08831-9#MOESM1 )

library(cluster)

DefaultAssay(object = all_integrated) <- "RNA"

mean_silhouette_score <- mean(all_integrated@meta.data$silhouette_score_0.5)

for(i in seq(from = 0.5, to = 3.8, by = 0.3)){
  
  res <- paste0("wsnn_res.", i )
  Idents(all_integrated) <- res
  dist_mat <- dist(Embeddings(all_integrated[['pca']])[, 1:30])
  clusters <- all_integrated@meta.data[[res]]
  silhouette_res <- silhouette(as.numeric(clusters), dist = dist_mat)
  all_integrated@meta.data[[paste0("silhouette_score_", i)]] <- silhouette_res[,3]
  
  mean_silhouette_score <- mean(all_integrated@meta.data[[paste0("silhouette_score_", i)]])
  
  #pdf(paste0(fig_path, "/silhouette_score_res", i, "_all_integrated.pdf"), height = 20, width = 10)
  plot(silhouette_res, cex = 0.001, border=NA)
  ##dev.off()
  
}
rm(clusters, silhouette_res, mean_silhouette_score, res, dist_mat)

```

###Finding differentially expressed features and cluster markers <br>
Notes:<br>
According to Seurat Faq (4): https://github.com/satijalab/seurat/discussions/4032#discussioncomment-363994, it is better to perform differential expression analysis on RNA assay even when using SCTransform workflow
```{r, warning=FALSE}
#finding markers for every cluster compared to all remaining cells
#According to Seurat Faq (4): https://github.com/satijalab/seurat/discussions/4032#discussioncomment-363994, it is better to perform differential expression analysis on RNA assay even when using SCTransform workflow

#RNA based differential expression
# Select the RNA counts slot to be the default assay
DefaultAssay(object = all_integrated) <- "RNA"

# Normalize RNA data
all_integrated <- Seurat::NormalizeData(object = all_integrated, verbose = FALSE)

#Scale data
all_integrated <- Seurat::ScaleData(object = all_integrated, features = rownames(all_integrated), vars.to.regress="percent.mt", verbose = FALSE)

#parallelization of FindAllMarkers
#plan("multiprocess", workers = 4)

#Find Markers for clusters in different resolutions using the wilcoxon rank test
#for (i in seq(from = 0.5, to = 4, by = 0.3)){
#  Idents(all_integrated) <- paste0("wsnn_res.", toString(i))
#   assign(paste("all_integrated_RNAmarkers",i, sep= "_res_"), FindAllMarkers(object = all_integrated, min.pct = 0.1, logfc.threshold = 0.2, only.pos = TRUE, test.use = "wilcox", verbose = FALSE))# %>% #writexl::write_xlsx(paste0(tab_path, "/all_integrated_RNAmarkers_res_", i, "_by_cluster.xlsx"))
#}


#Top 20 markers for every cluster
#all_integrated_RNAmarkers_res_0.5_top20 <- as.data.frame(all_integrated_RNAmarkers_res_0.5 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_0.5_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_0.5_top20.xlsx"))

#all_integrated_RNAmarkers_res_0.8_top20 <- as.data.frame(all_integrated_RNAmarkers_res_0.8 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_0.8_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_0.8_top20.xlsx"))

#all_integrated_RNAmarkers_res_1.1_top20 <- as.data.frame(all_integrated_RNAmarkers_res_1.1 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_1.1_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_1.1_top20.xlsx"))

#all_integrated_RNAmarkers_res_1.4_top20 <- as.data.frame(all_integrated_RNAmarkers_res_1.4 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_1.4_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_1.4_top20.xlsx"))

#all_integrated_RNAmarkers_res_1.7_top20 <- as.data.frame(all_integrated_RNAmarkers_res_1.7 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_1.7_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_1.7_top20.xlsx"))

#all_integrated_RNAmarkers_res_2_top20 <- as.data.frame(all_integrated_RNAmarkers_res_2 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_2_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_2_top20.xlsx"))


#all_integrated_RNAmarkers_res_2.3_top20 <- as.data.frame(all_integrated_RNAmarkers_res_2.3 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_2.3_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_2.3_top20.xlsx"))


#all_integrated_RNAmarkers_res_2.6_top20 <- as.data.frame(all_integrated_RNAmarkers_res_2.6 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_2.6_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_2.6_top20.xlsx"))


#all_integrated_RNAmarkers_res_2.9_top20 <- as.data.frame(all_integrated_RNAmarkers_res_2.9 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_2.9_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_2.9_top20.xlsx"))


#all_integrated_RNAmarkers_res_3.2_top20 <- as.data.frame(all_integrated_RNAmarkers_res_3.2 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_3.2_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_3.2_top20.xlsx"))

#all_integrated_RNAmarkers_res_3.5_top20 <- as.data.frame(all_integrated_RNAmarkers_res_3.5 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_3.5_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_3.5_top20.xlsx"))


#all_integrated_RNAmarkers_res_3.8_top20 <- as.data.frame(all_integrated_RNAmarkers_res_3.8 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_RNAmarkers_res_3.8_top20, paste0(tab_path, "/all_integrated_RNAmarkers_res_3.8_top20.xlsx"))


#ADT based differential expression
# Select the ADT counts slot to be the default assay
DefaultAssay(object = all_integrated) <- "ADT"

# Normalize ADT data (Re-normalizing because normalized data does not seem to be carried forward while integrating)
all_integrated <- Seurat::NormalizeData(object = all_integrated, normalization.method = "CLR", margin = 2, verbose = FALSE)

#Scale data
all_integrated <- Seurat::ScaleData(object = all_integrated, features = rownames(all_integrated), do.scale = FALSE, vars.to.regress = "nCount_ADT", verbose = FALSE)



#parallelization of FindAllMarkers
#plan("multiprocess", workers = 4)

#Find Markers for clusters in different resolutions using the  wilcoxon rank test
#for (i in seq(from = 0.5, to = 4, by = 0.3)){
#  Idents(all_integrated) <- paste0("wsnn_res.", toString(i))
#   assign(paste("all_integrated_ADTmarkers",i, sep= "_res_"), FindAllMarkers(object = all_integrated, min.pct = 0.1, logfc.threshold = 0.2, only.pos = TRUE, test.use = "wilcox", verbose = FALSE))# %>% #writexl::write_xlsx(paste0(tab_path, "/all_integrated_ADTmarkers_res_", i, "_by_cluster.xlsx"))
#}


#Top 20 markers for every cluster
#all_integrated_ADTmarkers_res_0.5_top20 <- as.data.frame(all_integrated_ADTmarkers_res_0.5 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_0.5_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_0.5_top20.xlsx"))

#all_integrated_ADTmarkers_res_0.8_top20 <- as.data.frame(all_integrated_ADTmarkers_res_0.8 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_0.8_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_0.8_top20.xlsx"))

#all_integrated_ADTmarkers_res_1.1_top20 <- as.data.frame(all_integrated_ADTmarkers_res_1.1 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_1.1_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_1.1_top20.xlsx"))

#all_integrated_ADTmarkers_res_1.4_top20 <- as.data.frame(all_integrated_ADTmarkers_res_1.4 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_1.4_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_1.4_top20.xlsx"))

#all_integrated_ADTmarkers_res_1.7_top20 <- as.data.frame(all_integrated_ADTmarkers_res_1.7 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_1.7_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_1.7_top20.xlsx"))

#all_integrated_ADTmarkers_res_2_top20 <- as.data.frame(all_integrated_ADTmarkers_res_2 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_2_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_2_top20.xlsx"))


#all_integrated_ADTmarkers_res_2.3_top20 <- as.data.frame(all_integrated_ADTmarkers_res_2.3 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_2.3_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_2.3_top20.xlsx"))


#all_integrated_ADTmarkers_res_2.6_top20 <- as.data.frame(all_integrated_ADTmarkers_res_2.6 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_2.6_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_2.6_top20.xlsx"))


#all_integrated_ADTmarkers_res_2.9_top20 <- as.data.frame(all_integrated_ADTmarkers_res_2.9 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_2.9_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_2.9_top20.xlsx"))


#all_integrated_ADTmarkers_res_3.2_top20 <- as.data.frame(all_integrated_ADTmarkers_res_3.2 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_3.2_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_3.2_top20.xlsx"))

#all_integrated_ADTmarkers_res_3.5_top20 <- as.data.frame(all_integrated_ADTmarkers_res_3.5 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_3.5_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_3.5_top20.xlsx"))

#all_integrated_ADTmarkers_res_3.8_top20 <- as.data.frame(all_integrated_ADTmarkers_res_3.8 %>% group_by(cluster) %>% filter(p_val_adj < 0.05)  %>% top_n(n = 20, wt = avg_log2FC))
#writexl::write_xlsx(all_integrated_ADTmarkers_res_3.8_top20, paste0(tab_path, "/all_integrated_ADTmarkers_res_3.8_top20.xlsx"))

```

# IL27RA RNA expression at res0.8
```{r}
#pdf(paste0(fig_path, "/Il27ra_RNAexpression_res0.8_all_integrated.pdf"), height = 10, width = 10)
VlnPlot(all_integrated, c("rna_Il27ra","adt-CD4", "adt-CD8a", "adt-CD8b", "rna_Cd19", "adt-NK-1.1", "adt-CD27", "adt-CD11b"), split.by = "group", pt.size = 0.1, group.by = "wsnn_res.0.8", stack = TRUE)
##dev.off()

```


#Checking for ambient RNA contamination
```{r, eval=FALSE, echo=FALSE}
xy <- all_integrated
DefaultAssay(xy) <- "RNA"
tmp <- as.SingleCellExperiment(DietSeurat(xy))
library(celda)

tmp <- celda::decontX(x = tmp, z = tmp$wsnn_res.0.8)

plotDecontXContamination(tmp)

celda::plotDecontXMarkerPercentage(tmp,
markers = list( "cd4_markers" ="Cd4",  "cd8_markers" = c("Cd8a", "Cd8b1"), "nk_markers"= "Ncr1", "b_markers" = "Cd19", "alveolar_macrophage_markers" = c("Marco", "Siglecf"), "pDC_markers" = c("Siglech", "Bst2") ), assayName = "counts")

celda::plotDecontXMarkerPercentage(tmp,
markers = list( "cd4_markers" ="Cd4",  "cd8_markers" = c("Cd8a", "Cd8b1"), "nk_markers"= "Ncr1", "b_markers" = "Cd19", "alveolar_macrophage_markers" = c("Marco", "Siglecf"), "pDC_markers" = c("Siglech", "Bst2") ), assayName = "decontXcounts")

rm(xy, tmp)
```



### Checking conserved markers across clusters for cluster identity <br>
Notes:<br>
(Reference: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md; <br> https://satijalab.org/seurat/articles/integration_introduction.html)
1. Conserved markers across groups/conditions can be helpful in identification of cell types. <br>
2. Seurat's FindConservedMarkers function internally separates out cells by sample group/condition, and then performs differential gene expression testing for a single specified cluster against all other clusters (or a second cluster, if specified). Gene-level p-values are computed for each condition and then combined across groups. <br>
3. Selecting resolution 0.8 to check cluster cell types as it appears to be stable. <br>
```{r}
#Finding conserved markers across clusters (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)
# Assigning Seurat object to temporary variable
tmp <- all_integrated
Idents(tmp) <- "wsnn_res.0.8" #assigning cluster resolution 0.8 as the identity

# RNA based markers
get_conserved_markers_tmp <- function(cluster){
    Seurat::FindConservedMarkers(object = tmp, ident.1 = cluster, grouping.var = "group", verbose = FALSE, assay = "RNA") %>%
        rownames_to_column(var = "gene") %>%
        cbind(cluster_id = cluster, .)
}

conserved_markers_tmp_RNA <- map_dfr(c(0:20), get_conserved_markers_tmp)

#max_pval:	largest p value of p value calculated by each group/condition
#minimump_p_val: combined p value


conserved_markers_tmp_RNA$cluster_id <- factor(conserved_markers_tmp_RNA$cluster_id)
conserved_markers_tmp_RNA$wt_p_val <- round(conserved_markers_tmp_RNA$wt_p_val, digits = 5)
conserved_markers_tmp_RNA$wt_avg_log2FC <- round(conserved_markers_tmp_RNA$wt_avg_log2FC, digits = 5)
conserved_markers_tmp_RNA$wt_p_val_adj <- round(conserved_markers_tmp_RNA$wt_p_val_adj, digits = 5)
conserved_markers_tmp_RNA$il27raKO_p_val <- round(conserved_markers_tmp_RNA$il27raKO_p_val, digits = 5)
conserved_markers_tmp_RNA$il27raKO_avg_log2FC <- round(conserved_markers_tmp_RNA$il27raKO_avg_log2FC, digits = 5)
conserved_markers_tmp_RNA$il27raKO_p_val_adj <-  round(conserved_markers_tmp_RNA$il27raKO_p_val_adj, digits = 5)
conserved_markers_tmp_RNA$max_pval <- round(conserved_markers_tmp_RNA$max_pval, digits = 5)
conserved_markers_tmp_RNA$minimump_p_val <- round(conserved_markers_tmp_RNA$minimump_p_val, digits = 5)

library(DT)

DT::datatable(conserved_markers_tmp_RNA, filter = "top", options = list(pageLength = 20))

#writexl::write_xlsx(conserved_markers_tmp_RNA, paste0(tab_path, "Conserved_RNAmarkers_res0.8_bycluster.xlsx"))

#ADT
get_conserved_markers_tmp <- function(cluster){
    Seurat::FindConservedMarkers(object = tmp, ident.1 = cluster, grouping.var = "group", verbose = FALSE, assay = "ADT") %>%
        rownames_to_column(var = "gene") %>%
        cbind(cluster_id = cluster, .)
}

conserved_markers_tmp_ADT <- map_dfr(c(0:20), get_conserved_markers_tmp)

#max_pval:	largest p value of p value calculated by each group/condition
#minimump_p_val: combined p value


conserved_markers_tmp_ADT$cluster_id <- factor(conserved_markers_tmp_ADT$cluster_id)
conserved_markers_tmp_ADT$wt_p_val <- round(conserved_markers_tmp_ADT$wt_p_val, digits = 5)
conserved_markers_tmp_ADT$wt_avg_log2FC <- round(conserved_markers_tmp_ADT$wt_avg_log2FC, digits = 5)
conserved_markers_tmp_ADT$wt_p_val_adj <- round(conserved_markers_tmp_ADT$wt_p_val_adj, digits = 5)
conserved_markers_tmp_ADT$il27raKO_p_val <- round(conserved_markers_tmp_ADT$il27raKO_p_val, digits = 5)
conserved_markers_tmp_ADT$il27raKO_avg_log2FC <- round(conserved_markers_tmp_ADT$il27raKO_avg_log2FC, digits = 5)
conserved_markers_tmp_ADT$il27raKO_p_val_adj <-  round(conserved_markers_tmp_ADT$il27raKO_p_val_adj, digits = 5)
conserved_markers_tmp_ADT$max_pval <- round(conserved_markers_tmp_ADT$max_pval, digits = 5)
conserved_markers_tmp_ADT$minimump_p_val <- round(conserved_markers_tmp_ADT$minimump_p_val, digits = 5)

library(DT)

DT::datatable(conserved_markers_tmp_ADT, filter = "top", options = list(pageLength = 20))

#writexl::write_xlsx(conserved_markers_tmp_ADT, paste0(tab_path, "Conserved_ADTmarkers_res0.8_bycluster.xlsx"))


rm(tmp, conserved_markers_tmp_RNA, conserved_markers_tmp_ADT)

```


```{r, echo=FALSE, eval=FALSE, warning=FALSE}
library(SingleR)
library(celldex)
ref.data <- celldex::ImmGenData()

tmp <- all_integrated
DefaultAssay(tmp) <- "RNA"
Idents(tmp) <- "wsnn_res.0.8"

#on dataset as a whole
predictions <- SingleR(test=as.SingleCellExperiment(tmp), ref=ref.data, labels=ref.data$label.main)
tmp$SingleR.pruned.calls.main<- predictions$pruned.labels
tmp$SingleR.calls.main <- predictions$labels
DimPlot(tmp, group.by = "SingleR.calls.main", label = TRUE)

predictions <- SingleR(test=as.SingleCellExperiment(tmp), ref=ref.data, labels=ref.data$label.fine)
tmp$SingleR.pruned.calls.fine <- predictions$pruned.labels
tmp$SingleR.calls.fine <- predictions$labels
DimPlot(tmp, group.by = "SingleR.calls.fine", label = TRUE, repel = TRUE, label.size = 2.5) & NoLegend()

#specifying cluster information through meta.data
predictions <- SingleR(test=as.SingleCellExperiment(tmp), ref=ref.data, labels=ref.data$label.main, clusters = tmp@meta.data$wsnn_res.0.8)
tmp[["SingleR.cluster.labels"]] <- predictions$labels[match(tmp[[]][["wsnn_res.0.8"]], rownames(predictions))]
DimPlot(tmp, group.by = "SingleR.cluster.labels", label = TRUE)

#rm(tmp, predictions)

```

#Plot all ADT
```{r, echo=FALSE, eval=FALSE}
tmp <- all_integrated
Idents(tmp) <- "wsnn_res.0.8"

adt_all <- row.names(tmp[["ADT"]])



#pdf(paste0(fig_path, "/clusters_All_adt-res0.8.pdf"), width = 15, height = 20 )
Seurat::FeaturePlot(tmp, features = adt_all[1:20], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
Seurat::FeaturePlot(tmp, features = adt_all[21:40], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
Seurat::FeaturePlot(tmp, features = adt_all[41:60], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
Seurat::FeaturePlot(tmp, features = adt_all[61:80], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
Seurat::FeaturePlot(tmp, features = adt_all[81:100], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
Seurat::FeaturePlot(tmp, features = adt_all[101:120], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
Seurat::FeaturePlot(tmp, features = adt_all[121:140], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
Seurat::FeaturePlot(tmp, features = adt_all[141:160], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
Seurat::FeaturePlot(tmp, features = adt_all[161:180], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
Seurat::FeaturePlot(tmp, features = adt_all[181:198], reduction = 'wnn.umap', max.cutoff = 2, cols = c("lightgrey","darkgreen"), label = TRUE, label.size = 2, ncol = 3, repel = TRUE)
##dev.off()


#pdf(paste0(fig_path, "/all_integrated_clusters_bySample_res0.8.pdf"), width = 12, height = 8 )
DimPlot(tmp, group.by = "wsnn_res.0.8", split.by = "sample_id", ncol = 4, label = TRUE, label.size = 2)
##dev.off()

rm(tmp)

```


```{r, echo=FALSE, eval=FALSE, message=FALSE}

# Checking expression of top-level cell type markers <br>
#Notes:<br>
#Cell type annotation was based on expression of markers from literature: Markers_for_cellIdentity_scRNAseq_TotalseqA_ADT.xlsx<br>

markers_list <- list(
  "NK\nRNA\n" = c("rna_Klra4", "rna_Klra8", "rna_Klrb1a", "rna_Klrb1b", "rna_Klrb1c", "rna_Klrb1f",  "rna_Gzma", "rna_Gzmb", "rna_Ncr1", "rna_Nkg7", "rna_Itga2"),
  
  "NK\nADT\n" = c("adt-CD335", "adt-KLRG1", "adt-NK-1.1"),
  
  "NK\nsubset\n" = c("adt-CD27", "adt-CD11b", "rna_Cd27", "rna_Itgam"), 
  
  "T\nRNA\n" = c("rna_Cd3d", "rna_Cd3g", "rna_Cd3e", "rna_Trbc1", "rna_Trbc2"),
  
  "T\nADT\n" = c("adt-CD3"),
  
  "CD4\nRNA\n" = c("rna_Cd4"),
  
  "CD4\nADT\n" = c("adt-CD4"),
  
  "CD8\nRNA\n" = c("rna_Cd8a", "rna_Cd8b1"),
  
  "CD8\nADT\n" = c("adt-CD8a", "adt-CD8b"),
  
  "gamma-deltaT\nRNA\n" = c("rna_Il22", "rna_Il17f", "rna_Il17a", "rna_Blk", "rna_Tcrg-C1", "rna_Trgv2", "rna_Trdv4", "rna_Tcrg-V4"),
  
  "gamma-deltaT\nADT\n" = c("adt-TCR-gamma-delta", "adt-TCR-V-gamma-2"),
  
  "Treg\nRNA\n" = c("rna_Foxp3", "rna_Ctla4", "rna_Il2ra"),
  
  "Treg\nADT\n" = c("adt-CD25", "adt-CD39", "adt-CD127", "adt-CD152", "adt-CD62L", "adt-CD134", "adt-CD357", "adt-CD278", "adt-CD304", "adt-CD279"),
  
  "Alveolar_macrophages\nRNA\n" = c("rna_Ear1", "rna_Ear2", "rna_Itgax", "rna_Siglec1", "rna_Siglecf", "rna_Cd200r1", "rna_Lgals3", "rna_Marco", "rna_Chil3"),
  
  "Alveolar_macrophages\nADT\n" = c("adt-CD11c", "adt-CD169-SIGLEC1", "adt-CD170", "adt-CD200R", "adt-MAC-2"),
  
  "Monocytes\nRNA\n" = c("rna_F13a1", "rna_Ly6c2", "rna_Cd14", "rna_Fcgr4", "rna_Fcgr1", "rna_Mafb"),
  
   "Monocytes\nADT\n" = c("adt-Ly-6C", "adt-CD14", "adt-CD16-32", "adt-CD115"),
  
  "DCs\nRNA\n" = c("rna_Cd40", "rna_Cd24a","rna_Dpp4", "rna_H2-Ab1", "rna_H2-Eb1", "rna_Itgae", "rna_Esam", "rna_Flt3", "rna_Sirpa", "rna_Xcr1", "rna_Clec9a", "rna_Ccr7", "rna_Cst3", "rna_Irf4", "rna_Irf8", "rna_S100a4", "rna_Zbtb46", "rna_Cd74", "rna_Cd86", "rna_Ly86"),
  
  "DCs\nADT\n" = c("adt-CD40", "adt-CD24","adt-CD26", "adt-I-A-I-E", "adt-CD103", "adt-ESAM","adt-CD135", "adt-CD172a","adt-XCR1", "adt-CD370", "adt-CD197"),
  
  "pDCs\nRNA\n" = c("rna_Siglech", "rna_Ccr9", "rna_Bst2", "rna_Tcf4", "rna_Clec4b1", "rna_Irf7", "rna_Pld4"),
  
  "pDCs\nADT\n" = c("adt-SIGLECH", "adt-CD199", "adt-CD317"),
  
  "B\nRNA\n" = c("rna_Cd19", "rna_Cd79a", "rna_Cd79b", "rna_Ighm", "rna_Ighd", "rna_Igkc", "rna_Igha", "rna_Ms4a1", "rna_Mzb1"),
  
   "B\nADT\n" = c("adt-CD19", "adt-CD79b", "adt-IgM", "adt-IgD"),
  
  "Proliferating" = c("rna_Mki67", "rna_Pclaf", "rna_Nusap1", "rna_Birc5", "rna_Top2a")
  
)


#assigning Seurat object to a temporary variable
x <- all_integrated

#pdf(paste0(fig_path, "Dotplot_celltypemarkers_all_integrated_res0.5-2_by0.3.pdf"), width = 35, height = 15)

#checking expression of markers at res 0.5
Idents(x) <- "wsnn_res.0.5"
Idents(x) <- factor(Idents(x), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16))

grid::grid.newpage()
grid::grid.text("resolution = 0.5")
DotPlot(x, features = markers_list) + scale_color_gradient2(midpoint = 0, low = "navy", mid = "lightgrey", high = "red")+ theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))


#checking expression of markers at res 0.8
Idents(x) <- "wsnn_res.0.8"
Idents(x) <- factor(Idents(x), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20))

grid::grid.newpage()
grid::grid.text("resolution = 0.8")
DotPlot(x, features = markers_list) + scale_color_gradient2(midpoint = 0, low = "navy", mid = "lightgrey", high = "red")+ theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

#checking expression of markers at res 1.1
Idents(x) <- "wsnn_res.1.1"
Idents(x) <- factor(Idents(x), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24))

grid::grid.newpage()
grid::grid.text("resolution = 1.1")
DotPlot(x, features = markers_list) + scale_color_gradient2(midpoint = 0, low = "navy", mid = "lightgrey", high = "red")+ theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

#checking expression of markers at res 1.4
Idents(x) <- "wsnn_res.1.4"
Idents(x) <- factor(Idents(x), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29))

grid::grid.newpage()
grid::grid.text("resolution = 1.4")
DotPlot(x, features = markers_list) + scale_color_gradient2(midpoint = 0, low = "navy", mid = "lightgrey", high = "red")+ theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

#checking expression of markers at res 1.7
Idents(x) <- "wsnn_res.1.7"
Idents(x) <- factor(Idents(x), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31))

grid::grid.newpage()
grid::grid.text("resolution = 1.7")
DotPlot(x, features = markers_list) + scale_color_gradient2(midpoint = 0, low = "navy", mid = "lightgrey", high = "red")+ theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

#checking expression of markers at res 2
Idents(x) <- "wsnn_res.2"
Idents(x) <- factor(Idents(x), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31))

grid::grid.newpage()
grid::grid.text("resolution = 2")
DotPlot(x, features = markers_list) + scale_color_gradient2(midpoint = 0, low = "navy", mid = "lightgrey", high = "red")+ theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))
##dev.off()
rm(x)


# separate dotplot for res0.8 for each marker category
#assigning Seurat object to a temporary variable
tmp <- all_integrated

#checking expression of markers at res 0.8
Idents(tmp) <- "wsnn_res.0.8"
tmp <- FindSubCluster(tmp, cluster = 18, graph.name = "wsnn", algorithm = 3, resolution = 0.2)
#cluster 18 shows expression of both pDC and B cell markers. Performing subclustering to check if this cluster identity can be resolved.

Idents(tmp) <- "sub.cluster"

Idents(tmp) <- factor(Idents(tmp), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, "18_0", "18_1", 19, 20))

#pdf(paste0(fig_path, "Dotplot_celltypemarkers_all_integrated_res0.8.pdf"), width = 10, height = 8)

for(i in 1:length(markers_list)){
title1 <- names(markers_list[i])
  grid::grid.newpage()
  grid::grid.text(paste0(title1, "  resolution = 0.8") )
print(DotPlot(tmp, features = markers_list[i]) + scale_color_gradient2(midpoint = 0, low = "navy", mid = "lightgrey", high = "red")+ theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10)))
}

##dev.off()
rm(tmp)


```



## Resolving cluster identities <br>
*Notes: <br> 
- sub.cluster 18_0 expresses a few B cell markers at low levels, and also a few monocyte and DC markers. <br>
- checking expression of these markers to resolve cell type identity for this sub.cluster. <br>
```{r}
#DC subset markers:
tmp <- all_integrated
Idents(tmp) <- "wsnn_res.0.8"
#tmp <- FindSubCluster(tmp, cluster = 18, graph.name = "wsnn", algorithm = 3, resolution = 0.2)
tmp <- FindSubCluster(tmp, cluster = 18, graph.name = "wsnn", algorithm = 3, resolution = 0.4) #cluster 18 shows expression of both pDC and B cell markers. Performing subclustering to check if this cluster identity can be resolved.#with res 0.4, a separate cluster of B cells emerges, so proceeding with this resolution for sub-clustering
Idents(tmp) <- "sub.cluster"
Idents(tmp) <- factor(Idents(tmp), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, "18_0", "18_1", "18_2", 19, 20))


#B or Mo-DC:
#VlnPlot(tmp, c("rna_Cd79a", "rna_Cd79b", "rna_Cd19", "rna_Ms4a1", "rna_Siglecg", "rna_Bank1", "rna_Ebf1","rna_Ly6d", "rna_Fcer1g"), assay="RNA", stack = TRUE, same.y.lims = TRUE, pt.size = 0)#Markers from: https://research.cchmc.org/pbge/lunggens/tools/lung_at_glance.html?tab=cell&species=Mouse&cell=T039

#pdf(paste0(fig_path, "VlnPlot_B-MoDC_all_integrated_res0.8_Upd.pdf"), width = 20, height = 8)

#B or Mo-DC:
VlnPlot(tmp, c("adt-CD103", "rna_Itgae", "adt-CD8a","adt-XCR1","adt-CD14", "rna_C3ar1", "rna_Ptger2", "adt-MERTK","adt-CD135", "rna_Flt3", "adt-SIGLECH", "adt-CD317", "adt-Ly-6C", "adt-CD370", "rna_Clec9a", "adt-CD24", "adt-CD172a", "rna_Sirpa","adt-CD11b", "adt-CD8a", "adt-I-A-I-E", "adt-CD26", "adt-CD170", "rna_Irf4", "rna_Irf8", "rna_Fcgr1", "rna_C5ar1", "rna_Zbtb46", "rna_Cd209a", "rna_Ccl17", "adt-ESAM", "adt-CX3CR1"), pt.size = 0.0, stack = TRUE) & NoLegend()

#B or Mo-DC:
VlnPlot(tmp, c("rna_Cd79a","adt-CD79b",  "rna_Cd79b","adt-CD19",  "rna_Cd19", "rna_Ms4a1", "rna_Siglecg", "rna_Bank1", "rna_Ebf1","rna_Ly6d", "rna_Fcer1g", "adt-CD14", "adt-CD86", "adt-F4-80", "adt-CX3CR1", "rna_Adgre1", "rna_Cx3cr1", "adt-MERTK", "rna_Mertk", "rna_Cst3", "adt-CD103", "adt-CD40", "adt-CD26", "adt-CD11c", "adt-Ly-6C","rna_Ly6c2", "adt-CD11b", "rna_Mafb", "rna_F13a1", "rna_Nos2", "rna_Il12a", "rna_Fcgr1", "rna_Tnf"), stack = TRUE, same.y.lims = TRUE, pt.size = 0) & NoLegend()

VlnPlot(tmp, c("adt-CD79b", "adt-CX3CR1", "adt-SIGLECH", "adt-CD317"), stack = TRUE)
##dev.off()
# Based on discussion with Johannes after visualization of these marker genes, sub.cluster 18_1 is more likely to be monocyte-derived dendritic cells; new sub-cluster 18_2 may be B cells


# Checking expression of subset NK cell markers (update: 1 Feb 2022)

#pdf(paste0(fig_path, "VlnPlot_NKsubsetmarkers_all_integrated_res0.8_Upd.pdf"), width = 10, height = 6)
VlnPlot(tmp, c("adt-CD27", "adt-CD11b", "adt-NK-1.1", "adt-KLRG1", "adt-CD335"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE)
VlnPlot(tmp, c("rna_Cd27", "rna_Itgam", "rna_Klrb1c", "rna_Klrg1", "rna_Ncr1"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE)

VlnPlot(subset(tmp, sub.cluster %in% c(0, 2, 3, 4, 11, 14, 16, 20)), c("adt-CD27", "adt-CD11b", "rna_Cd27", "rna_Itgam"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE)

VlnPlot(subset(tmp, sub.cluster %in% c(0, 2, 3, 4, 11, 14, 16, 20)), c("rna_Mki67", "rna_Pclaf", "rna_Nusap1", "rna_Birc5", "rna_Top2a", "adt-NK-1.1", "rna_Klrb1c", "adt-KLRG1", "rna_Klrg1"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE)

##dev.off()

#pdf(paste0(fig_path, "VlnPlot_gdTcellsubsetmarkers_all_integrated_res0.8_Upd.pdf"), width = 20, height = 8)
#distinguish gamma-delta T cells based on CD27, Ifng, Il17 expression (Reference from Johannes: https://www.nature.com/articles/ni.1717)
VlnPlot(tmp, c("rna_Ifng", "rna_Il17a", "rna_Il4","adt-CD27", "rna_Cd27", "rna_Ltbr", "adt-CD122", "rna_Il2rb", "adt-CD25", "rna_Il2ra", "adt-CD44", "rna_Cd44", "adt-CD62L", "rna_Sell", "rna_Rorc", "adt-CD24", "adt-CD2", "adt-CD5", "rna_Runx1", "rna_Tbx21", "rna_Crem", "rna_Nr4a2", "rna_Rgs2", "rna_Rgs1"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE) & NoLegend() #CD27+ and CD27- gdT cells
##dev.off()

#VlnPlot(subset(tmp, sub.cluster %in% c(1, 8, 13, 9, 10) ), c("adt-CD11a", "adt-CD44", "adt-CD62L", "adt-CD197", "adt-CD127","adt-KLRG1", "rna_Itgal", "rna_Cd44", "rna_Sell", "rna_Ccr7", "rna_Il17ra", "rna_Klrg1", "rna_Ifng", "adt-CD4", "adt-CD8a", "adt-CD8b", "rna_Cd4", "rna_Cd8a", "rna_Cd8b1"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE) #TCM, TEM, Naive T cells  #https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5508124/

#pdf(paste0(fig_path, "VlnPlot_Tcellsubsetmarkers_all_integrated_res0.8_Upd.pdf"), width = 10, height = 6)
VlnPlot(subset(tmp, sub.cluster %in% c(1, 8, 13, 9, 10) ), c("adt-CD44", "adt-CD62L", "rna_Cd44", "rna_Sell", "rna_Mki67", "rna_Pclaf", "rna_Nusap1", "rna_Birc5", "rna_Top2a"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE) & NoLegend()#TCM, TEM, Naive T cells 
#https://jitc.biomedcentral.com/articles/10.1186/s40425-017-0235-4 #CD44lowCD62L Naive, CD44HighCD62L+ Central memory, CD44highCD62Lneg Effector-effector memory + proliferating cell markers

VlnPlot(subset(tmp, sub.cluster %in% c(1, 8, 13, 9, 10) ), c("adt-CD4", "rna_Cd4","adt-CD8a", "adt-CD8b", "rna_Cd4", "rna_Cd8a", "rna_Cd8b1"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE) & NoLegend()

VlnPlot(tmp, c("rna_Cxcr3", "rna_Cxcr6", "rna_Ccr5", "rna_Ccr6", "rna_Cx3cr1", "adt-CX3CR1", "adt-CD103","rna_Itgae"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE) & NoLegend()#TRM
##dev.off()
#rm(tmp)

#Checking if clusters 11, 20 may be NKT cells (#Markers reference: https://pubmed.ncbi.nlm.nih.gov/32905761/ and https://pubmed.ncbi.nlm.nih.gov/33173202/ ; and T cell markers from Markers excel file)
#pdf(paste0(fig_path, "VlnPlot_NKTcellmarkers_all_integrated_res0.8_Upd.pdf"), width = 10, height = 6)
VlnPlot(tmp, c("rna_Cd3e", "rna_Cd3d", "rna_Trbc1", "rna_Trbc2", "rna_Klra1", "rna_Klra6", "rna_Cd8a", "rna_Cd8b1", "rna_Nkg7", "rna_Klrb1c", "rna_Zbtb16", "rna_Tbx21", "rna_Ifng", "rna_Il4", "rna_Il17a", "rna_Rorc", "adt-CD1d", "adt-NK-1.1", "adt-CD335", "adt-CD44", "adt-CD186", "rna_Cxcr6"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE) & NoLegend()

VlnPlot(tmp, c("rna_Lef1", "rna_Itm2a", "rna_Ccr9", "rna_Id3", "rna_Ldhb", "rna_Zbtb16", "rna_Plac8", "rna_Il17rb", "rna_Icos", "rna_Il1r1", "rna_Rorc", "rna_Ccr6", "rna_Tmem176a", "rna_Emb", "rna_Id2", "rna_Nkg7", "rna_Klrb1c", "rna_Klrd1", "rna_Il2rb", "rna_Ifit3", "rna_Isg15", "rna_Stat1", "rna_Irf7"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE) & NoLegend()

VlnPlot(tmp, c("rna_Pdcd1", "adt-CD279", "rna_Cd27", "adt-CD27", "rna_Cd28", "rna_Slamf6", "rna_Cd81"), pt.size = 0.0, stack = TRUE, same.y.lims = TRUE) & NoLegend()

##dev.off()

# Clusters 5 and 15 exhibit Monocyte markers. Checking if there are any differences between the clusters, or if they can be combined
x <- FindMarkers(tmp, ident.1 = 5, ident.2 = 15, assay = "RNA", min.diff.pct = 0.3)
x$gene <- row.names(x)
y <- FindMarkers(tmp, ident.1 = 5, ident.2 = 15, assay = "ADT", min.diff.pct = 0.3)
y$protein <- row.names(y)

monocyte_clus5_15 <- list("clust_5_15_RNA" = x,
                          "clust_5_15_ADT" = y)

#writexl::write_xlsx(monocyte_clus5_15, paste0(tab_path, "Monocyte_Cluster5_15_check_res0.8.xlsx")) #As there are not much expression differences between clusters 5 and 15
rm(x, y)

```



## Inflammatory molecule levels; differential exp between IL27RA KO and WT
```{r}
#Inflammatory molecule levels; differential exp between IL27RA KO and WT
tmp <- all_integrated
Idents(tmp) <- "wsnn_res.0.8"
tmp <- FindSubCluster(tmp, cluster = 18, graph.name = "wsnn", algorithm = 3, resolution = 0.4)
DimPlot(tmp, group.by = "sub.cluster", split.by = "group", label = TRUE, repel = TRUE) & NoLegend()

Idents(tmp) <- "sub.cluster"
Idents(tmp) <- factor(Idents(tmp), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, "18_0","18_1", "18_2", 19, 20))

#pdf(paste0(fig_path, "VlnPlot_InflammRecept_all_integrated_res0.8.pdf"), width = 10, height = 6)
VlnPlot(tmp, c("adt-CD279","rna_Pdcd1", "rna_Havcr2", "rna_Lag3", "rna_Tigit", "rna_Prdm1", "rna_Maf", "rna_Il27", "adt-CD274", "rna_Cd274", "adt-CD152", "rna_Ctla4", "adt-CD25", "adt-CD69", "adt-CD366", "adt-CD86"), split.by = "group", stack = TRUE) # Source: References from literature shared by Johannes Platten: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6130914/ and https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6994206/).
##dev.off()

```





## Assigning cell identity
```{r}
all_integrated_rnm <- all_integrated
Idents(all_integrated_rnm) <- "wsnn_res.0.8"
all_integrated_rnm <- FindSubCluster(all_integrated_rnm, cluster = 18, graph.name = "wsnn", algorithm = 3, resolution = 0.4)
DimPlot(all_integrated_rnm, group.by = "sub.cluster", split.by = "group", label = TRUE, repel = TRUE) & NoLegend()
Idents(all_integrated_rnm) <- "sub.cluster"

all_integrated_rnm$cluster <- all_integrated_rnm$sub.cluster #creating a column with the sub.cluster information

Idents(all_integrated_rnm) <- "cluster"

# Assigning cell type identities based on marker expression
all_integrated_rnm <- RenameIdents(object = all_integrated_rnm,
                      "0" = "CD27+CD11blo_NK",
                      "1" = "CD8_Nave",
                      "2" = "CD27+CD11blo_NK",
                      "3" = "CD27loCD11b+_NK",
                      "4" = "CD27loCD11b+_NK",
                      "5" = "Monocyte",
                      "6" = "CD27+_gdT",
                      "7" = "Alveolar_macrophage",
                      "8" = "CD8_TEM",
                      "9" = "CD4_TEM",
                      "10" = "CD4_TEM",
                      "11" = "mito_NK",
                      "12" = "cDC",
                      "13" = "CD8_TCM",
                      "14" = "CD27+CD11b+_NK",
                      "15" = "Monocyte",
                      "16" = "CD27+CD11b+_NK",
                      "17" = "Treg",
                      "18_0" = "pDC",
                      "18_1" = "moDC",
                      "18_2" = "B_cell",
                      "19" = "CD27-_gdT",
                      "20" = "NK_other"
                                   )
#pdf(paste0(fig_path, "/DimPlot_all_integrated_rnm_res0.8.pdf"), width = 8, height = 6)
DimPlot(all_integrated_rnm, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 3) + NoLegend() | DimPlot(all_integrated_rnm, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 3, group.by = "group") + NoLegend()
DimPlot(all_integrated_rnm, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 2.5, split.by = "group") + NoLegend()
##dev.off()

#pdf(paste0(fig_path, "/DimPlot_bySampleID_all_integrated_rnm_res0.8.pdf"), width = 15, height = 10)
all_integrated_rnm$sample_id <- factor(x = all_integrated_rnm$sample_id, levels = c("BM_WS_9", "BM_WS_10", "BM_WS_15", "BM_WS_5", "BM_WS_6", "BM_WS_8"))
DimPlot(all_integrated_rnm, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 2, split.by = "sample_id", ncol = 3) + NoLegend()
##dev.off()

all_integrated_rnm$celltype <- Idents(all_integrated_rnm) #creating a column for celltype annotation
Idents(all_integrated_rnm) <- "celltype"


```


# Conserved and differentially expressed genes and proteins across conditions <br>
## Conserved genes and proteins <br>
```{r}
## Conserved genes and proteins
library('multtest')
library('metap')

#wt and il27raKO combined integrated

#genes
DefaultAssay(object = all_integrated_rnm) <- "RNA"

get_conserved_markers <- function(celltype){
  Seurat::FindConservedMarkers(object = all_integrated_rnm, ident.1 = celltype, grouping.var = "group", verbose = FALSE, assay = "RNA") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

conserved_markers <- map_dfr(tmp_celltype_list, get_conserved_markers)

#writexl::write_xlsx(conserved_markers, paste0(tab_path, "/conserved_RNAmarkers_all_integrated_by_cluster_res0.8.xlsx"))


#ADT
DefaultAssay(object = all_integrated_rnm) <- "ADT"

get_conserved_markers <- function(celltype){
  Seurat::FindConservedMarkers(object = all_integrated_rnm, ident.1 = celltype, grouping.var = "group", verbose = FALSE, assay = "ADT") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

conserved_markers <- map_dfr(tmp_celltype_list, get_conserved_markers)

#writexl::write_xlsx(conserved_markers, paste0(tab_path, "/conserved_ADTmarkers_all_integrated_by_cluster_res0.8.xlsx"))

```


## Differentially expressed genes <br>
Notes:<b>
(Reference: https://satijalab.org/seurat/articles/integration_introduction.html) <br>
1. Identify what genes change in different groups for cells of the same type using Wilcoxon Rank Sum Test and, MAST (GLM-based Hurdle model
). <br>
```{r}
## Differentially expressed genes between groups using Wilcoxon Rank Sum Test
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html

#First, create a column in the meta.data slot to hold both the cell type and group information and switch the current ident to that column. 
all_integrated_rnm$cluster0.8.group <- paste(Idents(all_integrated_rnm), all_integrated_rnm$group, sep = "_")
all_integrated_rnm$cluster0.8 <- Idents(all_integrated_rnm)
Idents(all_integrated_rnm) <- "cluster0.8.group"

#genes
DefaultAssay(object = all_integrated_rnm) <- "RNA"

#Finding differentially expressed genes across clusters (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)

get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_genes_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_genes_all_integrated_rnm, paste0(tab_path, "/differential_genes_RNA_all_integrated_by_cluster_res0.8.xlsx"))


#ADT
DefaultAssay(object = all_integrated_rnm) <- "ADT"
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "ADT", logfc.threshold = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_ADT_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_ADT_all_integrated_rnm, paste0(tab_path, "/differential_genes_ADT_all_integrated_by_cluster_res0.8_NoLFCThresh.xlsx"))


##===Differential testing using MAST framework=================##
#genes
DefaultAssay(object = all_integrated_rnm) <- "RNA"
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Finding differentially expressed genes across clusters (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)

get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", test.use = "MAST") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_genes_all_integrated_rnm_MAST <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_genes_all_integrated_rnm, paste0(tab_path, "/differential_genes_RNA_all_integrated_byCluster_res0.8_MAST.xlsx"))


#ADT
DefaultAssay(object = all_integrated_rnm) <- "ADT"
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "ADT", logfc.threshold = 0, test.use = "MAST") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_ADT_all_integrated_rnm_MAST <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_ADT_all_integrated_rnm, paste0(tab_path, "/differential_genes_ADT_all_integrated_byCluster_res0.8_NoLFCThresh_MAST.xlsx"))

```


##Correlation between RNA and ADT markers
```{r, Correlation between RNA and ADT markers}

#pdf(paste0(fig_path, "/res0.8_clusters_correlation_ADT_RNA_Markers_all_integrated.pdf"), width = 8, height = 6 )
Idents(all_integrated_rnm) <- "celltype"
FeatureScatter(all_integrated_rnm, "adt-Ly-6C", "rna_Ly6c2")
FeatureScatter(all_integrated_rnm, "adt-CD11b", "rna_Itgam")
FeatureScatter(all_integrated_rnm, "adt-CD11c", "rna_Itgax")
FeatureScatter(all_integrated_rnm, "adt-CD8a", "rna_Cd8a")
FeatureScatter(all_integrated_rnm, "adt-CD4", "rna_Cd4")
FeatureScatter(all_integrated_rnm, "adt-NK-1.1", "rna_Klrb1b")
FeatureScatter(all_integrated_rnm, "adt-NK-1.1", "rna_Klrb1c")
FeatureScatter(all_integrated_rnm, "adt-CD170", "rna_Siglecf")
FeatureScatter(all_integrated_rnm, "adt-CD169-SIGLEC1", "rna_Siglec1")
##dev.off()

#visualize the modality weights that were learned for each cell
VlnPlot(all_integrated_rnm, features = "integrated.weight", sort = TRUE, pt.size = 0.1) + NoLegend()
VlnPlot(all_integrated_rnm, features = "IADT.weight", sort = TRUE, pt.size = 0.1) + NoLegend()

```

#Rename levels for violin and dot plots
```{r, echo=FALSE, message=FALSE}
all_integrated_rnm_VlnPlot <- all_integrated_rnm

levels_VlnPlot <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell") 

#factor(Idents(all_integrated_rnm_VlnPlot), levels = all_integrated_rnm_levels_VlnPlot)
Idents(all_integrated_rnm_VlnPlot) <- factor(Idents(all_integrated_rnm_VlnPlot), levels = levels_VlnPlot)

all_integrated_rnm_DotPlot <- all_integrated_rnm
levels_DotPlot <- rev(levels_VlnPlot) 

#factor(Idents(all_integrated_rnm_DotPlot), levels = all_integrated_rnm_levels_DotPlot)
Idents(all_integrated_rnm_DotPlot) <- factor(Idents(all_integrated_rnm_DotPlot), levels = levels_DotPlot) 
```

##Number of cells per cluster per group and per sample_id

```{r}
#Number of cells per cluster per group and per sample_id
Idents(all_integrated_rnm) <- "celltype"
n_cells_per_cluster <- FetchData(all_integrated_rnm, vars = c("ident", "orig.ident", "group", "sample_id")) %>% dplyr::count(ident, orig.ident, group, sample_id) %>% tidyr::spread(ident, n)
#writexl::write_xlsx(n_cells_per_cluster, paste0(tab_path, "/cellNumbers_all_integrated_by_cluster_res0.8.xlsx"))

#plotting the number of cells

library(dittoSeq)
#pdf(paste0(fig_path, "/res0.8_Cells_per_cluster.pdf"), width = 8, height = 6 )
dittoBarPlot(all_integrated_rnm, "group", group.by = "celltype", color.panel = c("#F8766D", "#00BFC4")) + theme(axis.text.x =  element_text(angle = 90))

all_integrated_rnm$sample_id <- factor(x = all_integrated_rnm$sample_id, levels = c("BM_WS_15","BM_WS_10", "BM_WS_9", "BM_WS_8", "BM_WS_6", "BM_WS_5"))

dittoBarPlot(all_integrated_rnm, "sample_id", group.by = "celltype",  retain.factor.levels = TRUE,color.panel = c("#00A9FF", "#00BFC4", "#00BE67", "#7CAE00", "#CD9600", "#F8766D")) + theme(axis.text.x =  element_text(angle = 90))
##dev.off()


# retrieve the absolute number of cells by sample_id
cells_by_sampleID <- dittoBarPlot(all_integrated_rnm, "sample_id", group.by = "celltype", retain.factor.levels = TRUE, scale = "count")
  
#writexl::write_xlsx(cells_by_sampleID$data, paste0(tab_path, "/Absolute_cellNumbers_all_integrated_by_cluster_res0.8.xlsx"))



```



#Muscat differential state analysis
*Notes:<br>
- Performing differential state analysis using muscat R package to assess to cell-type-specific responses across samples (Reference: https://www.bioconductor.org/packages/devel/bioc/vignettes/muscat/inst/doc/analysis.html) <br>
- This has been developed for scRNA seq specifically, so trying with RNA assay for this analysis. <br>

```{r}
# Differential state analysis based on muscat
#- Performing differential state analysis using muscat R package to assess to cell-type-specific responses across samples (Reference: https://www.bioconductor.org/packages/devel/bioc/vignettes/muscat/inst/doc/analysis.html) <br>
#- This has been developed for scRNA seq specifically, so trying with RNA assay for this analysis. <br>



#preparing data for muscat
library(limma)
library(muscat)
library(purrr)
library(dplyr)

#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(all_integrated_rnm, assay = "RNA")

tmp_sce$id <- paste0(tmp_sce$group,"_", tmp_sce$sample_id) #combining condition and sample_id

tmp_sce <- muscat::prepSCE(tmp_sce, 
    kid = "celltype", # cluster numbers or celltype
    gid = "group",  # group IDs (il27raKO/wt)
    sid = "id",   # sample IDs (il27raKO/wt+sample_id)
    drop = TRUE)  # drop all other colData columns
tmp_sce$sample_id <- factor(x = tmp_sce$sample_id, levels = c("il27raKO_BM_WS_9", "il27raKO_BM_WS_10", "il27raKO_BM_WS_15", "wt_BM_WS_5",  "wt_BM_WS_6", "wt_BM_WS_8"))

nk <- length(kids <- levels(tmp_sce$celltype)) 
ns <- length(sids <- levels(tmp_sce$sample_id))
names(kids) <- kids; names(sids) <- sids


#Differential State (DS) analysis
pseudobulk_sce <- muscat::aggregateData(tmp_sce,
    assay = "counts", fun = "sum",
    by = c("cluster_id", "sample_id")) #cluster_id refers to cell type


# one sheet per subpopulation
assayNames(pseudobulk_sce)

#Pseudobulk-level MDS plot
pb_sce_mds <- muscat::pbMDS(pseudobulk_sce) 
#warning: Removing 4 rows with all zero counts


#pdf(paste0(fig_path, "/muscatDS_MDS_RNA_res0.8.pdf"), width = 10, height = 8)
pb_sce_mds
##dev.off()

#test for DS using pbDS. By default, a group_id model is fit, and the last coefficient of the linear model is tested to be equal to zero.

# construct design & contrast matrix
ei <- metadata(tmp_sce)$experiment_info
mm <- model.matrix(~ 0 + ei$group_id)
dimnames(mm) <- list(ei$sample_id, levels(ei$group_id))
contrast <- makeContrasts("il27raKO-wt", levels = mm)

# run DS analysis
res_sce <- muscat::pbDS(pseudobulk_sce,  method = "edgeR", design = mm, contrast = contrast, verbose = FALSE)
# access results table for 1st comparison
tbl_sce <- res_sce$table[[1]]
# one data.frame per cluster
names(tbl_sce)

# view results for 1st cluster
k1_sce <- tbl_sce[[1]]
head(format(k1_sce[, -ncol(k1_sce)], digits = 2))

#sort by adjusted p value
tbl_fil_all_RNA <- lapply(tbl_sce, function(u) {
  dplyr::arrange(u, p_adj.loc)
})


#filter results to retain hits with FDR < 5% and abs(logFC) > 0.58, and count the number and frequency of differential findings by cluster and view the top hits (lowest adj. p-value) in each cluster
# filter FDR < 5%, abs(logFC) > 1 & sort by adj. p-value
tbl_fil_sig <- lapply(tbl_sce, function(u) {
  u <- dplyr::filter(u, p_adj.loc < 0.05, abs(logFC) > 0.58)
  dplyr::arrange(u, p_adj.loc)
})

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil_sig, nrow, numeric(1))
p_de <- format(n_de / nrow(tmp_sce) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)

# view top 2 hits in each cluster
top2_sig <- bind_rows(lapply(tbl_fil_sig, top_n, 2, p_adj.loc))
format(top2_sig[, -ncol(top2_sig)], digits = 2)

#results table with expression frequency of each gene by sample and overall fold change between napc2 vs control (untreated)
res_tbl <- muscat::resDS(tmp_sce, res_sce, frq = TRUE)

#Between-cluster concordance: UpSet-plot that visualizes the number of DE genes that are shared across or unique to certain clusters
library(UpSetR)
de_gs_by_k <- map(tbl_fil_sig, "gene")
upset(fromList(de_gs_by_k))

#writexl::write_xlsx(tbl_fil_all_RNA,paste0(tab_path, "muscatDS_byCellType_RNA_all_integrated_res0.8.xlsx"))

library(scater)
#UMAP
.plot_dr <- function(tmp_sce, dr, col)
  plotReducedDim(tmp_sce, dimred = dr, colour_by = col) +
    guides(fill = guide_legend(override.aes = list(alpha = 1, size = 3))) +
    theme_minimal() + theme(aspect.ratio = 1)




# downsample to max. 100 cells per cluster
cs_by_k <- split(colnames(tmp_sce), tmp_sce$cluster_id)
cs100 <- unlist(sapply(cs_by_k, function(u) 
  sample(u, min(length(u), 100))))

# plot t-SNE & UMAP colored by cluster & group ID
for (dr in c("WNN.UMAP"))
  for (col in c("cluster_id", "group_id"))
    .plot_dr(tmp_sce[, cs100], dr, col)


# Violin plots of top 5 genes across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_RNA_VlnPlot_byCelltype_res0.8.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_RNA)){
print(plotExpression(tmp_sce[, tmp_sce$cluster_id == i],
                features = tbl_fil_all_RNA[[i]]$gene[seq_len(5)],
                x = "sample_id", colour_by = "group_id", ncol = 3) + scale_color_manual(values = c("#F8766D", "#00BFC4"), aesthetics = "color") +
     guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) +
     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ggtitle(paste0("Violin plot of top 5 muscat differential state genes; ", i) )
)
}
##dev.off()

# Heatmap of top 20 genes (or less; based on adjusted p value) across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_RNA_Heatmap_byCelltype_res0.8.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_RNA)){
  if( ! (i %in% c("NK_other", "CD27-_gdT", "moDC", "B_cell") ) ){ #excluding "NK_other", "CD27-_gdT", "moDC", "B_cell"as there are no genes with FDR < 0.05 and including this causes an error in generating the heatmap
tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, k = i, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp_htmap@column_title <- paste0("Heatmap of top 20 (or less) genes from muscat DS analysis in ", i)
print(tmp_htmap)
  }
}
##dev.off()

#------------------------------------------------------------------------

# Assay ADT
#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(all_integrated_rnm, assay = "ADT")

tmp_sce$id <- paste0(tmp_sce$group,"_", tmp_sce$sample_id) #combining condition and sample_id

tmp_sce <- muscat::prepSCE(tmp_sce, 
    kid = "celltype", # cluster numbers or celltype
    gid = "group",  # group IDs (il27raKO/wt)
    sid = "id",   # sample IDs (il27raKO/wt+sample_id)
    drop = TRUE)  # drop all other colData columns
tmp_sce$sample_id <- factor(x = tmp_sce$sample_id, levels = c("il27raKO_BM_WS_9", "il27raKO_BM_WS_10", "il27raKO_BM_WS_15", "wt_BM_WS_5",  "wt_BM_WS_6", "wt_BM_WS_8"))

nk <- length(kids <- levels(tmp_sce$celltype)) 
ns <- length(sids <- levels(tmp_sce$sample_id))
names(kids) <- kids; names(sids) <- sids


#Differential State (DS) analysis
pseudobulk_sce <- muscat::aggregateData(tmp_sce,
    assay = "counts", fun = "sum",
    by = c("cluster_id", "sample_id")) #cluster_id refers to cell type


# one sheet per subpopulation
assayNames(pseudobulk_sce)

#Pseudobulk-level MDS plot
pb_sce_mds <- muscat::pbMDS(pseudobulk_sce) 


#pdf(paste0(fig_path, "/muscatDS_MDS_ADT_res0.8.pdf"), width = 10, height = 8)
pb_sce_mds
##dev.off()

#test for DS using pbDS. By default, a group_id model is fit, and the last coefficient of the linear model is tested to be equal to zero.

# construct design & contrast matrix
ei <- metadata(tmp_sce)$experiment_info
mm <- model.matrix(~ 0 + ei$group_id)
dimnames(mm) <- list(ei$sample_id, levels(ei$group_id))
contrast <- makeContrasts("il27raKO-wt", levels = mm)

# run DS analysis
res_sce <- muscat::pbDS(pseudobulk_sce,  method = "edgeR", design = mm, contrast = contrast, verbose = FALSE)
# access results table for 1st comparison
tbl_sce <- res_sce$table[[1]]
# one data.frame per cluster
names(tbl_sce)

# view results for 1st cluster
k1_sce <- tbl_sce[[1]]
head(format(k1_sce[, -ncol(k1_sce)], digits = 2))

#sort by adjusted p value
tbl_fil_all_ADT <- lapply(tbl_sce, function(u) {
  dplyr::arrange(u, p_adj.loc)
})


#filter results to retain hits with FDR < 5% and abs(logFC) > 0.58, and count the number and frequency of differential findings by cluster and view the top hits (lowest adj. p-value) in each cluster
# filter FDR < 5%, abs(logFC) > 1 & sort by adj. p-value
tbl_fil_sig <- lapply(tbl_sce, function(u) {
  u <- dplyr::filter(u, p_adj.loc < 0.05, abs(logFC) > 0.58)
  dplyr::arrange(u, p_adj.loc)
})

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil_sig, nrow, numeric(1))
p_de <- format(n_de / nrow(tmp_sce) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)

# view top 2 hits in each cluster
top2_sig <- bind_rows(lapply(tbl_fil_sig, top_n, 2, p_adj.loc))
format(top2_sig[, -ncol(top2_sig)], digits = 2)

#results table with expression frequency of each gene by sample and overall fold change between napc2 vs control (untreated)
res_tbl <- muscat::resDS(tmp_sce, res_sce, frq = TRUE)

#Between-cluster concordance: UpSet-plot that visualizes the number of DE genes that are shared across or unique to certain clusters
de_gs_by_k <- map(tbl_fil_sig, "gene")
upset(fromList(de_gs_by_k))

#writexl::write_xlsx(tbl_fil_all_ADT,paste0(tab_path, "muscatDS_byCellType_ADT_all_integrated_res0.8.xlsx"))


#UMAP
.plot_dr <- function(tmp_sce, dr, col)
  plotReducedDim(tmp_sce, dimred = dr, colour_by = col) +
    guides(fill = guide_legend(override.aes = list(alpha = 1, size = 3))) +
    theme_minimal() + theme(aspect.ratio = 1)




# downsample to max. 100 cells per cluster
cs_by_k <- split(colnames(tmp_sce), tmp_sce$cluster_id)
cs100 <- unlist(sapply(cs_by_k, function(u) 
  sample(u, min(length(u), 100))))

# plot t-SNE & UMAP colored by cluster & group ID
for (dr in c("WNN.UMAP"))
  for (col in c("cluster_id", "group_id"))
    .plot_dr(tmp_sce[, cs100], dr, col)


# Violin plots of top 5 genes across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_ADT_VlnPlot_byCelltype_res0.8.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_ADT)){
print(plotExpression(tmp_sce[, tmp_sce$cluster_id == i],
                features = tbl_fil_all_ADT[[i]]$gene[seq_len(5)],
                x = "sample_id", colour_by = "group_id", ncol = 3) + scale_color_manual(values = c("#F8766D", "#00BFC4"), aesthetics = "color") +
     guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) +
     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ggtitle(paste0("Violin plot of top 5 muscat differential state proteins; ", i) )
)
}
##dev.off()

# Heatmap of top 20 (or less) proteins (based on adjusted p < 0.05) across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_ADT_Heatmap_byCelltype_res0.8.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_ADT)){
  if(! (i %in% c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "B_cell") )){ #excluding "CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "B_cell"  as there are no proteins with FDR < 0.05 and including this causes an error in generating the heatmap
tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, k = i, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp_htmap@column_title <- paste0("Heatmap of top 20 (or less) proteins from muscat DS analysis in ", i)
print(tmp_htmap)
  }
}
##dev.off()
#c("#F8766D", "#00BFC4")




```


# Gene ontology and pathway enrichment analysis
## Seurat DE genes
```{r, message=FALSE}
# Select background for each analysis based on genes which are expressed in the particular cluster of cells
#Source code: https://ucdavis-bioinformatics-training.github.io/2021-August-Single-Cell-RNA-Seq-Analysis/data_analysis/scRNA_Workshop-PART6_fixed

#TopGO analysis
library(pcaExplorer)
library(topGO)
library(org.Mm.eg.db)
library(AnnotationDbi)

topgoDE_cluster <- list()
tmp <- all_integrated_rnm
DefaultAssay(tmp) <- "RNA"


for(i in c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK",  "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC")){ #excluding "NK_other", "CD27-_gdT", "B_cell" as there are no statistically significant DE genes based on adjusted p value
  clust <- tmp[, tmp$celltype %in% i] #subset to a particular cell type
  expr_clust <- clust[Matrix::rowSums(clust[["RNA"]]@counts > 0) > 0, ] #selecting background as genes/proteins which are atleast expressed in a particular cluster/celltype
  background.genes <- rownames(expr_clust)
  clust_deg <- subset(differential_genes_all_integrated_rnm, subset = cell_type %in% i & p_val_adj < 0.05)$gene
  topgoDE_cluster[[i]] <-
  pcaExplorer::topGOtable(DEgenes =  clust_deg,
                          BGgenes = background.genes,
                          ontology = "BP",
                          mapping = "org.Mm.eg.db",
                          geneID = "symbol",
                          topTablerows = 500)

}


#writexl::write_xlsx(topgoDE_cluster, paste0(tab_path, "topGO-DE_Seurat_DEgenes_res0.8.xlsx"))

```

# topgo: muscat DS gene results
```{r}
# topgo: muscat DS gene results
topgoDE_cluster_muscatDS <- list()

#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(all_integrated_rnm, assay = "RNA")

differential_genes_muscatDS <- bind_rows(tbl_fil_all_RNA)


for(i in c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK",  "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "Monocyte", "Alveolar_macrophage", "cDC", "pDC")){ #excluding "CD27-_gdT", "NK_other", "moDC" and "B_cell"  as there are no statistically significant DE genes based on adjusted p value
  clust <- tmp_sce[, tmp_sce$celltype %in% i] #subset to a particular cell type
  expr_clust <- clust[rowSums(counts(clust) > 0) > 0, ] #selecting background as genes/proteins which are atleast expressed in a particular cluster/celltype
  background.genes <- rownames(expr_clust)
  clust_deg <- subset(differential_genes_muscatDS, subset = cluster_id %in% i & p_adj.loc < 0.05)$gene
  topgoDE_cluster_muscatDS[[i]] <-
  pcaExplorer::topGOtable(DEgenes =  clust_deg,
                          BGgenes = background.genes,
                          ontology = "BP",
                          mapping = "org.Mm.eg.db",
                          geneID = "symbol",
                          topTablerows = 500)

}


#writexl::write_xlsx(topgoDE_cluster_muscatDS, paste0(tab_path, "topGO-DE_muscat_DSgenes_res0.8.xlsx"))



```


# EnrichR gene set enrichment
```{r}
library(enrichR)

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

#RNA
# GO BP
enrichr_gobp_all_integrated_rnm <- list()
#pdf(paste0(fig_path, "/Seurat_DEenrichR_GO-BP_RNA_all_integrated_res0.8.pdf"), width = 20, height = 8)

for(i in tmp_celltype_list){
  plot_enrichr <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "GO_Biological_Process_2021", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA") #p.val.cutoff	: Cutoff to select DE genes.
  print(plot_enrichr)
  enrichr_gobp_all_integrated_rnm[[i]] <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "GO_Biological_Process_2021", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 100, verbose = FALSE, assay = "RNA", return.gene.list = TRUE)
  }

##dev.off()

library(data.table)
#writexl::write_xlsx(lapply(enrichr_gobp_all_integrated_rnm, rbindlist, idcol = TRUE), paste0(tab_path, "enrichr_gobp_all_integrated_rnm_res0.8.xlsx"))


# KEGG
enrichr_KEGG_all_integrated_rnm <- list()
#pdf(paste0(fig_path, "/Seurat_DEenrichR_KEGG_RNA_all_integrated_res0.8.pdf"), width = 20, height = 8)

for(i in tmp_celltype_list){
  plot_enrichr <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "KEGG_2019_Mouse", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA")
  print(plot_enrichr)
  enrichr_KEGG_all_integrated_rnm[[i]] <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "KEGG_2019_Mouse", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA", return.gene.list = TRUE)
}

##dev.off()
#writexl::write_xlsx(lapply(enrichr_KEGG_all_integrated_rnm, rbindlist, idcol = TRUE), paste0(tab_path, "enrichr_KEGG_all_integrated_rnm_res0.8.xlsx"))


# Reactome
enrichr_Reactome_all_integrated_rnm <- list()
#pdf(paste0(fig_path, "/Seurat_DEenrichR_Reactome_RNA_all_integrated_res0.8.pdf"), width = 20, height = 8)

for(i in tmp_celltype_list){
  plot_enrichr <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "Reactome_2016", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA")
  print(plot_enrichr)
  enrichr_Reactome_all_integrated_rnm[[i]] <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "Reactome_2016", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA", return.gene.list = TRUE)
}

##dev.off()
#writexl::write_xlsx(lapply(enrichr_Reactome_all_integrated_rnm, rbindlist, idcol = TRUE), paste0(tab_path, "enrichr_Reactome_all_integrated_rnm_res0.8.xlsx"))



# MSigDB_Hallmark_2020
enrichr_MSigDB_all_integrated_rnm <- list()

#pdf(paste0(fig_path, "/Seurat_DEenrichR_MSigDB_Hallmark_RNA_all_integrated_res0.8.pdf"), width = 20, height = 8)

for(i in tmp_celltype_list){
  plot_enrichr <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "MSigDB_Hallmark_2020", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA")
  print(plot_enrichr)
  enrichr_MSigDB_all_integrated_rnm[[i]] <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "MSigDB_Hallmark_2020", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA", return.gene.list = TRUE)
}

##dev.off()
#writexl::write_xlsx(lapply(enrichr_MSigDB_all_integrated_rnm, rbindlist, idcol = TRUE), paste0(tab_path, "enrichr_MSigDB_all_integrated_rnm_res0.8.xlsx"))

```

# Differential abundance (DA) analysis
*Notes <br>
- Differential abundance (DA) analysis compares the proportions of cell types across experimental conditions and aims to highlight populations that are present at different ratios. (Reference: http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/muscWorkshop__vignette/ ; https://bioconductor.org/books/release/OSCA/multi-sample-comparisons.html)
```{r}
Idents(all_integrated_rnm) <- "celltype"
# Differential abundance based on RNA
#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(all_integrated_rnm, assay = "RNA")

tmp_sce$id <- paste0(tmp_sce$group,"_", tmp_sce$sample_id) #combining condition and sample_id

tmp_sce <- muscat::prepSCE(tmp_sce, 
    kid = "celltype", # cluster numbers or celltype
    gid = "group",  # group IDs (il27raKO/wt)
    sid = "id",   # sample IDs (il27raKO/wt+sample_id)
    drop = TRUE)  # drop all other colData columns
tmp_sce$sample_id <- factor(x = tmp_sce$sample_id, levels = c("wt_BM_WS_5", "wt_BM_WS_6", "wt_BM_WS_8", "il27raKO_BM_WS_9", "il27raKO_BM_WS_10", "il27raKO_BM_WS_15")) #order of sample_ids with wt first, il27rako next, as this would help estimate the differential abundance correctly. ordering the il27rako first for differential abundance gives criss-crossed estimates.

nk <- length(kids <- levels(tmp_sce$cluster_id)) 
ns <- length(sids <- levels(tmp_sce$sample_id))
names(kids) <- kids; names(sids) <- sids

ei <- metadata(tmp_sce)$experiment_info

# calculate cluster-sample cell counts
n_cells <- table(tmp_sce$cluster_id, tmp_sce$sample_id)

# calculate cluster proportions across samples
freqs <- prop.table(n_cells, margin = 1) #margin = 1 utilizes as denominator the row sums of each celltype i.e. total number of cells for a celltype in both il27raKO and wt


# prep. data.frame for plotting
df <- data.frame(
    frequency = as.numeric(freqs), 
    cluster_id = rep(kids, ns),
    sample_id = rep(sids, each = nk))

m <- match(df$sample_id, ei$sample_id)
df$group_id <- ei$group_id[m]

#pdf(paste0(fig_path, "/Differential_abundance_all_integrated_rnm_res0.8.pdf"), width = 12, height = 8)

# barplot of relative cluster-abundances (Barplot of subpopulation frequencies; stratified by sample_id and group)
ggplot(df, aes(x = sample_id, y = frequency, fill = cluster_id)) +
    geom_bar(stat = "identity", position = "fill") + 
    facet_wrap(~ group_id, scales = "free_x") +
    theme_classic()  + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ggtitle("Barplot of subpopulation frequencies (relative abundance); stratified by sample_id and group")


# boxplot of relative cluster-abundances (Boxplot of subpopulation frequencies; stratified by sample_id)
ggplot(df, aes(x = group_id, y = frequency, color = group_id)) +
    geom_boxplot(outlier.colour = NA) +  geom_jitter() +
    facet_wrap(~ cluster_id, scales = "free_y", ncol = 4) +
    theme_classic() + ggtitle("Boxplot of subpopulation frequencies (relative abundance); stratified by sample_id and group")


# plot absolute number of cells by sample_id
tmp_df <- data.frame(abs_numb_cells = table(tmp_sce$cluster_id, tmp_sce$sample_id), 
     cluster_id = rep(kids, ns),
     sample_id = rep(sids, each = nk))

mt <- match(tmp_df$sample_id, ei$sample_id)
tmp_df$group_id <- ei$group_id[mt]

# boxplot of absolute cluster-abundances (Boxplot of subpopulation frequencies; stratified by sample_id)
ggplot(tmp_df , aes(x = group_id, y = abs_numb_cells.Freq, color = group_id)) +
     geom_boxplot(outlier.colour = NA) +  geom_jitter() +
     facet_wrap(~ cluster_id, scales = "free_y", ncol = 4) + ylab("absolute cell number") +
     theme_classic() + ggtitle("Boxplot of absolute cell type abundance; stratified by sample_id and group")


##dev.off()


library(edgeR)
# DA analysis
y <- DGEList(counts = n_cells)
y <- estimateDisp(y, design = mm, trend = "none")
fit <- glmQLFit(y, design = mm)
fit <- glmLRT(fit, contrast = contrast)
topTags(fit, n = Inf)$table %>% round(2)  #logFC = frequency of cells

fit_ab_RNA <- glmQLFit(y, design = mm, robust=TRUE, abundance.trend=FALSE)
res_DA_RNA <- glmQLFTest(fit_ab_RNA, contrast = contrast)
topTags(res_DA_RNA)
res_DA_RNA$table$celltype <- row.names(res_DA_RNA$table)
#writexl::write_xlsx(res_DA_RNA$table, paste0(tab_path, "Differential_abundance_all_integrated_rnm_res0.8.xlsx"))

```


##Expression of complement pathway genes
```{r, Expression of complement pathway genes}
#Source list of complement genes: Ricklin D, et al.  Nat Immunol. 2010 Sep;11(9):785-97; https://pubmed.ncbi.nlm.nih.gov/20720586/

#Human to mouse gene symbol conversion : http://www.informatics.jax.org/batch/summary

#1. Pattern recognition complement genes:
#C1qa = C1qa; C1qb = C1qb; MBL = Mbl2; Ficolin-2 = Fcnb; Properdin = Cfp; CRP = Crp; CFHR4 = Cfhr1
genesComplement_PatternRecog <- c("C1qa", "C1qb", "Cfp", "Fcnb") #, Mbl2, Crp and Cfhr1 are not present in the dataset, so excluded

#2. Complement proteases:
#C1r = C1ra; C1r = C1rb; C1s = C1s1; C1s = C1s2; MASP1 = Masp1; MASP2 = Masp2; C2 = C2; CFB = Cfb; CFD = Cfd; CFI = Cfi
genesComplement_Proteases <- c("C1ra", "C1s1", "Masp2", "C2", "Cfb") #C1rb, C1s2, Masp1, Cfd and Cfi are not present in the dataset, so excluded


#3. Complement components:
#C3 = C3; C4 = C4a; C4 = C4b; C5 = Hc; C6 = C6; C7 = C7; C8A = C8a; C8B = C8b; C8G =C8g; C9 = C9
genesComplement_Components <- c("C3", "C4a", "C4b", "Hc", "C8g") #C6, C7, C8a, C8b and C9 are not present in the dataset, so excluded

#4. Complement receptors:
#CR1 = Cr1l; CR2 = Cr2; CR3 = Itgam + Itgb2; CR4 = Itgax + Itgb2; C3aR = C3ar1; C5aR = C5ar1; C5L2 = C5ar2, CRIg = Vsig4; cC1qR = Calr; C1qbp = C1qbp; C1qRp = Cd93
genesComplement_Receptors <- c("Cr1l", "Cr2", "Itgam", "Itgax", "Itgb2", "C3ar1", "C5ar1", "C5ar2", "Calr", "C1qbp", "Cd93") #"Vsig4" not present in the dataset, so excluded

#5. Complement regulators:
#SERPING1 = Serping1; sMAP = Masp2; MAP1 = Masp1; C4BP = C4bp; CFH = Cfh; CFHL1 = Cfhr1; MCP = Cd46; DAF = Cd55; CD59 = Cd59a; CD59 = Cd59b; Vitronectin = Vtn; Clusterin = Clu; Carboxypeptidase-N = Cpn1; Carboxypeptidase-N = Cpn2 

genesComplement_Regulators <- c("Serping1", "Masp2", "Cfh", "Cd46", "Cd55", "Cd59a", "Clu", "Cd59b", "Cpn1") #, Masp1, C4bp, Cfhr1, Vtn, Cpn2 and Cpb2 are not present in the dataset, so excluded

#Figures
#Violin Plot
#pdf(paste0(fig_path, "/res0.8_ViolinPlot_ComplementPathwayGenes_all_integrated.pdf"), width = 9, height = 6 )
title1 <- "Pattern recognition genes"
grid::grid.newpage()
grid::grid.text(title1)
for (gene in genesComplement_PatternRecog){
  VlnPlot_PatternRecog <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_PatternRecog) 
}
title2 <- "Proteases"
grid::grid.newpage()
grid::grid.text(title2)
for (gene in genesComplement_Proteases){
  VlnPlot_Proteases <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_Proteases)
}
title3 <- "Complement components"
grid::grid.newpage()
grid::grid.text(title3)
for (gene in genesComplement_Components){
  VlnPlot_Components <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_Components)
}
title4 <- "Receptors"
grid::grid.newpage()
grid::grid.text(title4)
for (gene in genesComplement_Receptors){
  VlnPlot_Receptors <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_Receptors)
 }
title5 <- "Regulators"
grid::grid.newpage()
grid::grid.text(title5)
for (gene in genesComplement_Regulators){
  VlnPlot_Regulators <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_Regulators)
 }
##dev.off()

#Dotplots
#pdf(paste0(fig_path, "/res0.8_DotPlot_ComplementPathwayGenes_all_integrated.pdf"), width = 8, height = 10 )
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_PatternRecog, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) + labs(title = "Pattern recognition genes\n")
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_Proteases, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) + labs(title = "Proteases\n")
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_Components, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) + labs(title = "Complement components\n")
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_Receptors, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) +  labs(title = "Receptors\n")
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_Regulators, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) + labs(title = "Regulators\n")
##dev.off()

#DotPlot(all_integrated_rnm_DotPlot, features =c("C1qa", "C1qb", "Cfp", "C1ra", "C1s1", "Masp2", "C2", "Cfb", "C3", "C4a", "C4b", "Hc", "C8g", "Cr1l", "Cr2", "Itgam", "Itgax", "Itgb2", "C3ar1", "C5ar1", "C5ar2", "Calr", "C1qbp", "Cd93", "Vsig4", "Serping1", "Cfh", "Cd46", "Cd55", "Cd59a"), assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))


```



##Visualizing expression of Depdc7, Leng8, Ccdc9, Tmco1, Gnb2, Pacc1 (Tmem206), Myadm, Gapdh, Tlr4

```{r, Visualizing expression of Depdc7, Leng8, Ccdc9, Tmco1, Gnb2, Pacc1 (Tmem206), Myadm, Gapdh, Tlr4}
#Visualizing expression of Visualizing expression of Depdc7, Leng8, Ccdc9, Tmco1, Gnb2, Pacc1 (Tmem206), Myadm, Gapdh, Tlr4 in different clusters in the seurat object with renamed identities

#Violin plots: expression levels represent log normalized values

#pdf(paste0(fig_path, "/res1.1_ViolinPlot_11genes_all_integrated.pdf"), width = 8, height = 6 )

VlnPlot(all_integrated_rnm_VlnPlot, features = c("Depdc7"), pt.size = 0.1, assay = "RNA", split.by = "group") &  theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Leng8"), pt.size = 0.1, assay = "RNA", split.by = "group") &  theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Ccdc9"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Tmco1"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Gnb2"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Pacc1"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Myadm"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Gapdh"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Tlr4"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Il27ra"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Il6st"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))

##dev.off()

#pdf(paste0(fig_path, "/res1.1_DotPlot_11genes_all_integrated.pdf"), width = 10, height = 12 )
DotPlot(all_integrated_rnm_DotPlot, features = c("Depdc7", "Leng8", "Ccdc9", "Tmco1", "Gnb2",  "Pacc1", "Myadm", "Gapdh", "Tlr4", "Il27ra", "Il6st"), assay = "RNA",  split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank())
##dev.off()

```


##Expression of Miwi2 and Saravanan\'s research genes
```{r, Expression of Miwi2 and Saravanan\'s research genes}
#1.	Piwil4 = MIWI2 
#pdf(paste0(fig_path, "/res0.8_DotPlot_MIWI2_all_integrated.pdf"), width = 10, height = 10)
DotPlot(all_integrated_rnm_DotPlot, features = c("Piwil4"), assay = "RNA",  split.by = "group", cols = c("red", "grey")) + theme(axis.title.x = element_blank(), axis.title.y = element_blank())
##dev.off()

#pdf(paste0(fig_path, "/res0.8_ViolinPlot_MIWI2_all_integrated.pdf"), width = 8, height = 6 )
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Piwil4"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(),axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
##dev.off()

#2. F2r = PAR1; F2rl1 = PAR2; F3 = Tissue factor; F10 = coagulation factor X; F7 = coagulation factor VII; Procr = EPCR 
#pdf(paste0(fig_path, "/res0.8_DotPlot_additnlGenes_Saravanan_Research_all_integrated.pdf"), width = 10, height = 10)
DotPlot(all_integrated_rnm_DotPlot, features = c("F2r", "F2rl1", "F3", "F10", "F7", "Procr"), assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.title.x = element_blank(), axis.title.y = element_blank())
##dev.off()



#pdf(paste0(fig_path, "/res0.8_ViolinPlot_additnlGenes_Saravanan_Research_all_integrated.pdf"), width = 8, height = 6 )

VlnPlot(all_integrated_rnm_VlnPlot, features = c("F2r"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("F2rl1"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("F3"), pt.size = 0.1, assay = "RNA", split.by = "group") & NoLegend() & NoLegend() & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("F10"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("F7"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Procr"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
##dev.off()
```

#Mfsd1 expression
```{r}
#1.	Mfsd1
#pdf(paste0(fig_path, "/res0.8_DotPlot_Mfsd1_all_integrated.pdf"), width = 10, height = 10)
DotPlot(all_integrated_rnm_DotPlot, features = c("Mfsd1"), assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.title.x = element_blank(), axis.title.y = element_blank())
##dev.off()

#pdf(paste0(fig_path, "/res0.8_ViolinPlot_Mfsd1_all_integrated.pdf"), width = 8, height = 6 )
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Mfsd1"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(),axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
##dev.off()
```



## Statistical difference in Ifng expression across all clusters betweeen groups
### Ifng: Between il27raKO and wt
```{r}

#Finding differential expression of Ifng across clusters between il27raKO vs wt (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Ifng differential expression in clusters between il27raKO and wt
#RNA
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", features = "Ifng", logfc.threshold = 0, min.pct = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_Ifng_RNA_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_Ifng_RNA_all_integrated_rnm, paste0(tab_path, "/differential_Ifng_all_integrated_res0.8.xlsx"))


```


# IL27RA RNA expression
```{r}
#pdf(paste0(fig_path, "/Il27ra_RNAexpression_byCelltype_res0.8_all_integrated.pdf"), height = 6, width = 10)
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Il27ra"), split.by = "group", pt.size = 0.1, assay = "RNA", ) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
##dev.off()

```


# 7Feb2022: update: generate heatmaps for GO BP "(cellular) response to Interferon-gamma" and "cytokine-mediated signaling pathway" based on inputs from Johannes
## Heatmaps from GO analysis: 
```{r}
#Seurat-enrichr GO BP data
#assign the GO BP results from Seurat-enrichr to a temporary variable
library(data.table)
tmp <- lapply(enrichr_gobp_all_integrated_rnm, rbindlist, idcol = TRUE)
tmp <- bind_rows(tmp, .id = "celltype")

#extract genes from all clusters for GO BP: "cellular response to interferon-gamma (GO:0071346)"
genes_ifng_response <- tmp$GO_Biological_Process_2021.Genes[grepl("(GO:0071346)", tmp$GO_Biological_Process_2021.Term)]
genes_ifng_response <- str_to_title(unique(unlist(strsplit(genes_ifng_response, split=';', fixed=TRUE) ) ) )
#length(genes_ifng_response) = 25

#pdf(paste0(fig_path, "/Heatmap_Seu_Enrichr_Resp_IFNgamma_pathway_res0.8_all_integrated.pdf"), height = 12, width = 25)
DoHeatmap(subset(all_integrated_rnm, downsample = 100), features = genes_ifng_response, assay = "RNA", angle = 90, size = 3, group.by = "cluster0.8.group")+ scale_fill_gradient2(midpoint = 0, low = "navy", mid = "white",  high = "red", na.value = "white") + theme(axis.text.y.left = element_text(colour = "black")) + guides(color = "none")
##dev.off()

#extract genes from all clusters for GO BP: "cytokine-mediated signaling pathway (GO:0019221)"
genes_cytokine <- tmp$GO_Biological_Process_2021.Genes[grepl("(GO:0019221)", tmp$GO_Biological_Process_2021.Term)]
genes_cytokine <- str_to_title(unique(unlist(strsplit(genes_cytokine, split=';', fixed=TRUE) ) ) )
#length(genes_cytokine) = 114

#pdf(paste0(fig_path, "/Heatmap_Seu_Enrichr_CytokineSignal_res0.8_all_integrated.pdf"), height = 15, width = 25)
DoHeatmap(subset(all_integrated_rnm, downsample = 100), features = genes_cytokine, assay = "RNA", angle = 90, size = 3, group.by = "cluster0.8.group")+ scale_fill_gradient2(midpoint = 0, low = "navy", mid = "white",  high = "red", na.value = "white") + theme(axis.text.y.left = element_text(colour = "black")) + guides(color = "none")
##dev.off()


#pseudobulk heatmap visualization
#Cellular response to IFNgamma
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifng_response, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 25, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm") #adjusting the heatmap cell size
tmp1@matrix_param$height <- unit(805, "mm") #adjusting the heatmap cell size
#pdf(paste0(fig_path, "/Heatmap_MuscatPB_Resp_IFNgamma_pathway_res0.8_all_integrated.pdf"), height = 40, width = 10)
tmp1
##dev.off()

#Cytokine-mediated signaling pathway
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_cytokine, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 114, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm")
tmp1@matrix_param$height <- unit(4055, "mm")
#pdf(paste0(fig_path, "/Heatmap_MuscatPB_CytokineSignal_res0.8_all_integrated.pdf"), height = 65, width = 10)
tmp1
##dev.off()


#TopGO data: Seurat DEGs:
tmp <- bind_rows(topgoDE_cluster, .id = "celltype")

#extract genes from all clusters for GO BP: "cellular response to interferon-gamma (GO:0071346)"
genes_ifng_response <- tmp$genes[grepl("(GO:0071346)", tmp$GO.ID)]
genes_ifng_response <- unique(unlist(strsplit(genes_ifng_response, split=',', fixed=TRUE) ) )

#pdf(paste0(fig_path, "/Heatmap_TopGO_Seurat_Resp_IFNgamma_pathway_res0.8_all_integrated.pdf"), height = 12, width = 25)
DoHeatmap(subset(all_integrated_rnm, downsample = 100), features = genes_ifng_response, assay = "RNA", angle = 90, size = 3, group.by = "cluster0.8.group")+scale_fill_gradient2(midpoint = 0, low = "navy", mid = "white",  high = "red", na.value = "white") + theme(axis.text.y.left = element_text(colour = "black")) + guides(color = "none")
##dev.off()


#extract genes from all clusters for GO BP: "cytokine-mediated signaling pathway (GO:0019221)"
genes_cytokine <- tmp$genes[grepl("(GO:0019221)", tmp$GO.ID)]
genes_cytokine <- unique(unlist(strsplit(genes_cytokine, split=',', fixed=TRUE) ) )

#pdf(paste0(fig_path, "/Heatmap_TopGO_Seurat_CytokineSignal_res0.8_all_integrated.pdf"), height = 15, width = 25)
DoHeatmap(subset(all_integrated_rnm, downsample = 100), features = genes_cytokine, assay = "RNA", angle = 90, size = 3, group.by = "cluster0.8.group")+ scale_fill_gradient2(midpoint = 0, low = "navy", mid = "white",  high = "red", na.value = "white") + theme(axis.text.y.left = element_text(colour = "black")) + guides(color = "none")
##dev.off()


#pseudobulk heatmap visualization
#Cellular response to IFNgamma
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifng_response, fdr = 1, lfc= 0, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm") #adjusting the heatmap cell size
tmp1@matrix_param$height <- unit(405, "mm") #adjusting the heatmap cell size
#pdf(paste0(fig_path, "/Heatmap_MuscatPB_TopGO_Seurat_Resp_IFNgamma_pathway_res0.8.pdf"), height = 20, width = 10)
tmp1
##dev.off()

#Cytokine-mediated signaling pathway
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_cytokine, fdr = 1, lfc= 0, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm")
tmp1@matrix_param$height <- unit(555, "mm")
#pdf(paste0(fig_path, "/Heatmap_MuscatPB_TopGO_Seurat_CytokineSignal_res0.8.pdf"), height = 25, width = 10)
tmp1
##dev.off()


#TopGO data: MuscatDS DEGs:
tmp <- bind_rows(topgoDE_cluster_muscatDS, .id = "celltype")

#extract genes from all clusters for GO BP: "cellular response to interferon-gamma (GO:0071346)"
genes_ifng_response <- tmp$genes[grepl("(GO:0071346)", tmp$GO.ID)]
genes_ifng_response <- unique(unlist(strsplit(genes_ifng_response, split=',', fixed=TRUE) ) )

#extract genes from all clusters for GO BP: "cytokine-mediated signaling pathway (GO:0019221)"
genes_cytokine <- tmp$genes[grepl("(GO:0019221)", tmp$GO.ID)]
genes_cytokine <- unique(unlist(strsplit(genes_cytokine, split=',', fixed=TRUE) ) )


#pseudobulk heatmap visualization
#Cellular response to IFNgamma
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifng_response, fdr = 1, lfc= 0, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm") #adjusting the heatmap cell size
tmp1@matrix_param$height <- unit(455, "mm") #adjusting the heatmap cell size
#pdf(paste0(fig_path, "/Heatmap_MuscatDS_PB_TopGO_Resp_IFNgamma_pathway_res0.8.pdf"), height = 20, width = 10)
tmp1
##dev.off()

#Cytokine-mediated signaling pathway
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_cytokine, fdr = 1, lfc= 0, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm")
tmp1@matrix_param$height <- unit(655, "mm")
#pdf(paste0(fig_path, "/Heatmap_MuscatDS_PB_TopGO_CytokineSignal_res0.8.pdf"), height = 30, width = 10)
tmp1
##dev.off()

```
## Heatmaps from JAX GO terms:
```{r}
#The genes mapped in the Gene Ontology database for cellular response to interferon-gamma (GO:0071346; http://www.informatics.jax.org/go/term/GO:0071346) and  cytokine-mediated signaling pathway (GO:0019221; http://www.informatics.jax.org/go/term/GO:0019221 ) were retrieved (7 Feb 2022) and a heatmap was plotted using the muscatDS pseudobulk heatmap function
library(readxl)
tmp <- readxl::read_excel("GO_JAX_tables_input/GO_term_summary_20220207_152905_GO0071346_CellularResp_Ifngamma_JAX.xlsx")

genes_ifng_response_Jax <- unique(tmp$Symbol)

genes_ifng <- subset(genes_ifng_response_Jax, genes_ifng_response_Jax %in% row.names(tmp_sce)) #subset to keep only those genes from the JAX database response to IFNgamma GO that are expressed in the current dataset 
#length(genes_tmp) = 113
htmap_plots <- list()


for(i in names(tbl_fil_all_RNA)){
  tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifng, k = i, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 113, col = colorRampPalette(c("blue", "white", "red"))(50))
  tmp_htmap@column_title <- paste0("Heatmap of genes from the\nJAX GO BP response to Ifn gamma pathway in ", i, "\n")
  tmp_htmap@matrix_param$width <- unit(35, "mm")
  tmp_htmap@matrix_param$height <- unit(405, "mm")
  tmp_htmap@row_names_param$gp[[1]] <- 9
  htmap_plots[[i]] <- tmp_htmap
}

#pdf(paste0(fig_path, "/Heatmap_PB_JAX_GO_RespIFNgamma_res0.8.pdf"), height = 25, width = 10)
for(i in names(tbl_fil_all_RNA)){
  print(htmap_plots[[i]])
}

##dev.off()

rm(tmp, genes_tmp,genes_ifng, htmap_plots, tmp_htmap)


#Cytokine-mediated signaling
tmp <- readxl::read_excel("GO_JAX_tables_input/GO_term_summary_20220207_153037_GO0019221_cytokineSignal_JAX.xlsx")

genes_cytokineSignal_Jax <- unique(tmp$Symbol)

genes_cytok <- subset(genes_cytokineSignal_Jax, genes_cytokineSignal_Jax %in% row.names(tmp_sce)) #subset to keep only those genes from the JAX database Cytokine mediated signaling GO that are expressed in the current dataset 
#length(genes_tmp) = 325

htmap_plots <- list()


for(i in names(tbl_fil_all_RNA)){
  tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_cytok, k = i, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 325, col = colorRampPalette(c("blue", "white", "red"))(50))
  tmp_htmap@column_title <- paste0("Heatmap of genes from the\nJAX GO BP cytokine-mediated signaling pathway in ", i, "\n")
  tmp_htmap@matrix_param$width <- unit(35, "mm")
  tmp_htmap@matrix_param$height <- unit(805, "mm")
  tmp_htmap@row_names_param$gp[[1]] <- 8
  htmap_plots[[i]] <- tmp_htmap
}

#pdf(paste0(fig_path, "/Heatmap_PB_JAX_GO_CytokineSignal_res0.8.pdf"), height = 40, width = 10)
for(i in names(tbl_fil_all_RNA)){
  print(htmap_plots[[i]])
}

##dev.off()


```

# 8Feb2022: update: generate heatmap for GO BP "cellular response to type I interferon (GO:0071357)" based on inputs from Johannes
## Heatmaps from JAX GO terms:
```{r}
#The genes mapped in the Gene Ontology database for cellular response to type I interferon (GO:0071357; http://www.informatics.jax.org/go/term/GO:0071357) was retrieved (8 Feb 2022) and a heatmap was plotted using the muscatDS pseudobulk heatmap function
library(readxl)
tmp <- readxl::read_excel("GO_JAX_tables_input/GO_term_summary_20220208_092916_respTypeI_IFN_GO0071357_JAX.xlsx")

genes_type1_ifn_response_Jax <- unique(tmp$Symbol)

genes_ifn1 <- subset(genes_type1_ifn_response_Jax, genes_type1_ifn_response_Jax %in% row.names(tmp_sce)) #subset to keep only those genes from the JAX database type 1 response to interferon GO that are expressed in the current dataset 
#length(genes_tmp) = 46
htmap_plots <- list()


for(i in names(tbl_fil_all_RNA)){
  tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifn1, k = i, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 46, col = colorRampPalette(c("blue", "white", "red"))(50))
  tmp_htmap@column_title <- paste0("Heatmap of genes from the\nJAX GO BP cellular response to type I interferon in ", i, "\n")
  if(nrow(tmp_htmap@matrix) > 10){
  tmp_htmap@matrix_param$width <- unit(35, "mm")
  tmp_htmap@matrix_param$height <- unit(255, "mm")
  tmp_htmap@row_names_param$gp[[1]] <- 8
  }
  htmap_plots[[i]] <- tmp_htmap
}

#pdf(paste0(fig_path, "/Heatmap_PB_JAX_GO_type1_IFNresp_res0.8.pdf"), height = 20, width = 10)
for(i in names(tbl_fil_all_RNA)){
  print(htmap_plots[[i]])
}

##dev.off()

#trying heatmap without arranging genes by p_adj.loc values to see if a pattern can be observed
htmap_plots_noOrder <- list()
for(i in names(tbl_fil_all_RNA)){
  tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifn1, k = i, fdr = 1, lfc= 0, top_n = 46, col = colorRampPalette(c("blue", "white", "red"))(50))
  tmp_htmap@column_title <- paste0("Heatmap of genes from the\nJAX GO BP cellular response to type I interferon in ", i, "\n")
  if(nrow(tmp_htmap@matrix) > 10){
  tmp_htmap@matrix_param$width <- unit(35, "mm")
  tmp_htmap@matrix_param$height <- unit(105, "mm")
  tmp_htmap@row_names_param$gp[[1]] <- 8
  }
  htmap_plots_noOrder[[i]] <- tmp_htmap
}

#pdf(paste0(fig_path, "/Heatmap_NoOrd_PB_JAX_GO_type1_IFNresp_res0.8.pdf"), height = 15, width = 10)
for(i in names(tbl_fil_all_RNA)){
  print(htmap_plots_noOrder[[i]])
}

##dev.off()

rm(tmp, genes_tmp, htmap_plots, tmp_htmap)


```

## Plotting averageexp heatmap: for GO BP "cellular response to type I interferon (GO:0071357)"
```{r}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm #assign seurat object to a temporary variable

#pdf(paste0(fig_path, "AvgExpr_Heatmap_type1IFN.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_type1_ifn <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_type1_ifn) <- c("il27raKO", "wt")
  cluster.averages_type1_ifn <- ScaleData(cluster.averages_type1_ifn)
 print(Seurat::DoHeatmap(cluster.averages_type1_ifn, features = genes_ifn1, size = 3, draw.lines = FALSE, angle = 90) + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 6.5, colour = "black", margin = margin(r=-15))) + ggtitle(paste0("Heatmap of average expression of type I interferon response genes in\n",i, "\n") ) )
}
##dev.off()


#Plotting with dittoHeatmap. Some genes that have zero values will not be plotted.
#Adding metadata to variable storing the average expression values to enable annotation of group (Source code: https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/AddMetaData) 
tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_type1IFN.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_type1_ifn <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_type1_ifn) <- c("il27raKO", "wt")
  cluster.averages_type1_ifn <- ScaleData(cluster.averages_type1_ifn)
  group <- Idents(cluster.averages_type1_ifn)
  names(group) <- colnames(cluster.averages_type1_ifn)
  cluster.averages_type1_ifn <- AddMetaData(object = cluster.averages_type1_ifn, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_type1_ifn, genes = genes_ifn1, annot.by = "group", main = paste0("Heatmap of average expression of type I interferon response genes in\n",i, "\n") )) 
}
##dev.off()
```

## Plotting averageexp heatmap: for GO BP "cellular response to interferon gamma (GO:0071346)"
```{r}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm #assign seurat object to a temporary variable

#pdf(paste0(fig_path, "AvgExpr_Heatmap_IFNgamma.pdf"), height = 15)
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages) <- c("il27raKO", "wt")
  cluster.averages <- ScaleData(cluster.averages)
 print(Seurat::DoHeatmap(cluster.averages, features = genes_ifng, size = 3, draw.lines = FALSE, angle = 90) + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 6.5, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10)) + ggtitle(paste0("Heatmap of average expression of\ncellular response to interferon-gamma genes in\n",i, "\n") ) )
}
##dev.off()


#Plotting with dittoHeatmap. Some genes that have zero values will not be plotted.
#Adding metadata to variable storing the average expression values to enable annotation of group (Source code: https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/AddMetaData) 
tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_IFNgamma.pdf"), height = 15)
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages) <- c("il27raKO", "wt")
  cluster.averages <- ScaleData(cluster.averages)
  group <- Idents(cluster.averages)
  names(group) <- colnames(cluster.averages)
  cluster.averages <- AddMetaData(object = cluster.averages, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages, genes = genes_ifng, annot.by = "group", main = paste0("Heatmap of average expression of\ncellular response to interferon gamma genes in\n",i, "\n") )) 
}
##dev.off()
```


## Plotting averageexp heatmap: for GO BP "cytokine-mediated signaling (GO:0019221)"
```{r}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm #assign seurat object to a temporary variable

#pdf(paste0(fig_path, "AvgExpr_Heatmap_CytokineSignal.pdf"), height = 25)
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages) <- c("il27raKO", "wt")
  cluster.averages <- ScaleData(cluster.averages)
 print(Seurat::DoHeatmap(cluster.averages, features = genes_cytok, size = 3, draw.lines = FALSE, angle = 90) + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 6, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10)) + ggtitle(paste0("Heatmap of average expression of cytokine-mediated signaling genes in\n",i, "\n") ) )
}
##dev.off()


#Plotting with dittoHeatmap. Some genes that have zero values will not be plotted.
#Adding metadata to variable storing the average expression values to enable annotation of group (Source code: https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/AddMetaData) 
tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_CytokineSignal.pdf"), height = 30)
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages) <- c("il27raKO", "wt")
  cluster.averages <- ScaleData(cluster.averages)
  group <- Idents(cluster.averages)
  names(group) <- colnames(cluster.averages)
  cluster.averages <- AddMetaData(object = cluster.averages, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages, genes = genes_cytok, annot.by = "group", fontsize = 8, main = paste0("Heatmap of average expression of\ncytokine-mediated signaling genes in\n",i, "\n") )) 
}
##dev.off()
```



## Plotting averageexp heatmap from the gene list for GO BP "cellular response to type I interferon (GO:0071357)" only for genes that are statistically significant (Adjusted p < 0.05)
```{r}
#Plotting with dittoHeatmap. Some genes that have zero values will not be plotted.
#Plotting averageexp heatmap from the gene list for GO BP "cellular response to type I interferon (GO:0071357)" only for genes that are statistically significant (Adjusted p < 0.05)
#Adding metadata to variable storing the average expression values to enable annotation of group (Source code: https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/AddMetaData) 
tmp <- all_integrated_rnm

#subset differentially expressed genes list to include only genes that are statistically significant
pgenes <- subset(differential_genes_all_integrated_rnm, p_val_adj < 0.05)

#subset list of type I interferon genes to only include genes that are statistically significant (Adjusted p < 0.05)
genes_ifn1_plot <- subset(genes_ifn1, genes_ifn1 %in% pgenes$gene)

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_type1IFN_signifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_type1_ifn <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_type1_ifn) <- c("il27raKO", "wt")
  cluster.averages_type1_ifn <- ScaleData(cluster.averages_type1_ifn)
  group <- Idents(cluster.averages_type1_ifn)
  names(group) <- colnames(cluster.averages_type1_ifn)
  cluster.averages_type1_ifn <- AddMetaData(object = cluster.averages_type1_ifn, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_type1_ifn, genes = genes_ifn1_plot, annot.by = "group", main = paste0("Heatmap of average expression of statistically significant (adj p < 0.05)\ntype I interferon response genes in\n",i, "\n") )) 
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_type1IFN_signifGenes.pdf"), width = 8)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_ifn1_plot, split.by = "group", stack = TRUE)
##dev.off()
```


## Plotting averageexp heatmap from the gene list for GO BP "cellular response to interferon gamma (GO:0071346)" only for genes that are statistically significant (Adjusted p < 0.05)
```{r}
#subset list of cellular response to interferon-gamma genes (GO:0071346) to only include genes that are statistically significant (Adjusted p < 0.05)
genes_ifng_plot <- subset(genes_ifng, genes_ifng %in% pgenes$gene)

tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_IFNgamma_signifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_type1_ifn <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_type1_ifn) <- c("il27raKO", "wt")
  cluster.averages_type1_ifn <- ScaleData(cluster.averages_type1_ifn)
  group <- Idents(cluster.averages_type1_ifn)
  names(group) <- colnames(cluster.averages_type1_ifn)
  cluster.averages_type1_ifn <- AddMetaData(object = cluster.averages_type1_ifn, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_type1_ifn, genes = genes_ifng_plot, annot.by = "group", main = paste0("  Heatmap of average expression of statistically significant (adj p < 0.05)\ncellular response to interferon-gamma genes in\n",i, "\n") )) 
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_IFNgamma_signifGenes.pdf"), width = 8)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_ifng_plot, split.by = "group", stack = TRUE)
##dev.off()
```

## Plotting averageexp heatmap from the gene list for GO BP cytokine-mediated signaling pathway (GO:0019221)" only for genes that are statistically significant (Adjusted p < 0.05)
```{r}
#subset list of cellular response to cytokine-mediated signaling pathway (GO:0019221) genes to only include genes that are statistically significant (Adjusted p < 0.05)
genes_cytokineSig_plot <- subset(genes_cytok, genes_cytok %in% pgenes$gene)

tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_CytokineSignal_signifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_type1_ifn <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_type1_ifn) <- c("il27raKO", "wt")
  cluster.averages_type1_ifn <- ScaleData(cluster.averages_type1_ifn)
  group <- Idents(cluster.averages_type1_ifn)
  names(group) <- colnames(cluster.averages_type1_ifn)
  cluster.averages_type1_ifn <- AddMetaData(object = cluster.averages_type1_ifn, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_type1_ifn, genes = genes_cytokineSig_plot, annot.by = "group", main = paste0("  Heatmap of average expression of statistically significant (adj p < 0.05)\ncytokine-mediated signaling pathway genes in\n",i, "\n") )) 
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_CytokineSignal_signifGenes.pdf"), width = 12)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_cytokineSig_plot, split.by = "group", stack = TRUE)
##dev.off()
```


## Dotplots of significant ifn gamma response, cytokine signaling, type I ifn response genes
```{r}
# Separate dot plot for each cluster
tmp <- all_integrated_rnm_DotPlot
tmp_cell_list <-unique(rev(levels(tmp)))#specifying reverse to ensure order of cell types

# type I interferon response
dotplot_ifn1 <- list()
for(i in tmp_cell_list){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  dotplot_ifn1[[i]] <- DotPlot(tmp_clus, genes_ifn1_plot, assay = "RNA", dot.scale = 6) + scale_color_gradientn(colors = c("navy", "white", "red")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), plot.title = element_text(size = 10), axis.title = element_blank()) + ggtitle(paste0("Dotplot showing expression of statistically significant (adj p < 0.05)\ntype I interferon response genes in\n",i, "\n"))
}


#pdf(paste0(fig_path, "Dotplot_type1IFN_signifGenes.pdf"), width = 12, height = 12)
for(i in seq(1,length(dotplot_ifn1), by = 6) ){
  print(wrap_plots(dotplot_ifn1[[i]], dotplot_ifn1[[i+1]], dotplot_ifn1[[i+2]], dotplot_ifn1[[i+3]], dotplot_ifn1[[i+4]], dotplot_ifn1[[i+5]], ncol = 2, nrow = 3))
    }
##dev.off()

# cellular response to interferon-gamma
dotplot_ifngamma <- list()
for(i in tmp_cell_list){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  dotplot_ifngamma[[i]] <- DotPlot(tmp_clus, genes_ifng_plot, assay = "RNA", dot.scale = 6) + scale_color_gradientn(colors = c("navy", "white", "red")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), plot.title = element_text(size = 10), axis.title = element_blank()) + ggtitle(paste0("Dotplot showing expression of statistically significant (adj p < 0.05)\ncellular response to interferon-gamma genes in\n",i, "\n"))
}


#pdf(paste0(fig_path, "Dotplot_resp_IFNgamma_signifGenes.pdf"), width = 12, height = 12)
for(i in seq(1,length(dotplot_ifngamma), by = 6) ){
  print(wrap_plots(dotplot_ifngamma[[i]], dotplot_ifngamma[[i+1]], dotplot_ifngamma[[i+2]], dotplot_ifngamma[[i+3]], dotplot_ifngamma[[i+4]], dotplot_ifngamma[[i+5]], ncol = 2, nrow = 3))
    }
##dev.off()


# cytokine mediated signaling pathway
dotplot_cytokineSig <- list()
for(i in tmp_cell_list){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  dotplot_cytokineSig[[i]] <- DotPlot(tmp_clus, genes_cytokineSig_plot, assay = "RNA", dot.scale = 6) + scale_color_gradientn(colors = c("navy", "white", "red")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), plot.title = element_text(size = 10), axis.title = element_blank()) + ggtitle(paste0("Dotplot showing expression of statistically significant (adj p < 0.05)\ncytokine-mediated signaling genes in\n",i, "\n"))
}


#pdf(paste0(fig_path, "Dotplot_resp_CytokineSignal_signifGenes.pdf"), width = 15, height = 12)
for(i in seq(1,length(dotplot_cytokineSig), by = 6) ){
  print(wrap_plots(dotplot_cytokineSig[[i]], dotplot_cytokineSig[[i+1]], dotplot_cytokineSig[[i+2]], dotplot_cytokineSig[[i+3]], dotplot_cytokineSig[[i+4]], dotplot_cytokineSig[[i+5]], ncol = 2, nrow = 3))
    }
##dev.off()

```


```{r}
#VlnPlot(all_integrated_rnm, c("Prdm1","Maf", "Il27", "Il10", "Il1b", "Il18"), assay = "RNA", pt.size = 0, split.by = "group", stack = TRUE) #(Reference: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7771052/)

tmp <- all_integrated_rnm
DefaultAssay(tmp) <- "RNA"
#pdf(paste0(fig_path, "ModuleScore_GOdb_topGO_signifGenes.pdf"), width = 12, height = 8)
tmp <- AddModuleScore(tmp, features = list(genes_ifn1_plot), name = "type1_IFN")
FeaturePlot(tmp, features = "type1_IFN1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))

tmp <- AddModuleScore(tmp, features = list(genes_ifng_plot), name = "IFN_gamma_resp")
FeaturePlot(tmp, features = "IFN_gamma_resp1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))

tmp <- AddModuleScore(tmp, features = list(genes_cytokineSig_plot), name = "cytokine_Signal")
FeaturePlot(tmp, features = "cytokine_Signal1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))

#FeaturePlot(tmp, features = c("type1_IFN1", "IFN_gamma_resp1"), label = TRUE, repel = TRUE, split.by = "group", order=TRUE, blend = TRUE)

genes_pyroptosis <- c("Gzmb", "Zbp1", "Casp4",  "Casp1")
tmp <- AddModuleScore(tmp, features = list(genes_pyroptosis), name = "pyroptosis")
FeaturePlot(tmp, features = "pyroptosis1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))


xt <- bind_rows(topgoDE_cluster, .id = "celltype")
genes_reg_immuneResp <-  xt$genes[grepl("(GO:0050776)", xt$GO.ID)]
genes_reg_immuneResp <- unique(unlist(strsplit(genes_reg_immuneResp, split=',', fixed=TRUE) ) )
tmp <- AddModuleScore(tmp, features = list(genes_reg_immuneResp), name = "immune_resp_reg")
FeaturePlot(tmp, features = "immune_resp_reg1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))


genes_apoptotic_proc <-  xt$genes[grepl("(GO:0006915)", xt$GO.ID)]
genes_apoptotic_proc <- unique(unlist(strsplit(genes_apoptotic_proc, split=',', fixed=TRUE) ) )
tmp <- AddModuleScore(tmp, features = list(genes_apoptotic_proc), name = "apoptotic_proc")
FeaturePlot(tmp, features = "apoptotic_proc1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))


genes_defense_resp <- xt$genes[grepl("(GO:0006952)", xt$GO.ID)]
genes_defense_resp <- unique(unlist(strsplit(genes_defense_resp, split=',', fixed=TRUE) ) )
tmp <- AddModuleScore(tmp, features = list(genes_defense_resp), name = "defense_resp")
FeaturePlot(tmp, features = "defense_resp1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))

##dev.off()

```

# Upd 10Feb2022: checking for ILC2 cells based on Johannes and Prof. Bosmann's inputs
```{r}
#References for ILC2 markers (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6025753/; https://research.cchmc.org/pbge/lunggens/tools/lung_at_glance.html?tab=cell&species=Mouse&cell=T062 ; https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6763593/ ; https://pubmed.ncbi.nlm.nih.gov/32917510/ )

#pdf(paste0(fig_path, "VlnPlots_ILCmarkers_Unsubclustered.pdf"), width = 12, height = 8)

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD90.2", "adt-CD335", "adt-NK-1.1", "adt-CD49a", "adt-CD49b", "adt-CD253", "adt-CD69", "adt-CD183", "adt-CD186","rna_Tbx21", "rna_Eomes"), stack = TRUE, pt.size = 0, same.y.lims = TRUE)


VlnPlot(subset(all_integrated_rnm_VlnPlot, sub.cluster %in% c(0, 2, 3, 4, 11, 14, 16, 20)), c("adt-CD90.2", "adt-CD335", "adt-NK-1.1", "adt-CD49a", "adt-CD49b", "adt-CD253", "adt-CD69", "adt-CD183", "adt-CD186", "adt-CD127", "adt-CD117", "adt-anti-rat-CD90-mouse-CD90.1","rna_Tbx21", "rna_Eomes", "rna_Areg", "rna_Prf1", "rna_Fcgr4", "rna_Klrb1c", "rna_Itga1", "rna_Thy1", "rna_Il7r", "rna_Kit", "rna_Il1r1","rna_Rorc", "rna_Il17a"), stack = TRUE, pt.size = 0, same.y.lims = TRUE, group.by = "sub.cluster") & NoLegend()

##dev.off()

tmp <- all_integrated_rnm
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(0), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(2), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(3), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(4), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(11), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(14), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(16), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(20), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
Idents(tmp) <- factor(Idents(tmp), levels = c("0_0", "0_1", 1, "2_0", "2_1", "2_2", "3_0","3_1", "4_0","4_1", 5, 6, 7, 8, 9, 10, "11_0", 12, 13, "14_0", "14_1", 15, "16_0","16_1", 17, "18_0", "18_1", "18_2", 19, "20_0","20_1"))

#pdf(paste0(fig_path, "VlnPlots_ILCmarkers_subclustered.pdf"), width = 12, height = 8)
VlnPlot(subset(tmp, sub.cluster %in% c("0_0", "0_1", "2_0", "2_1", "2_2", "3_0","3_1", "4_0","4_1", "11_0", "14_0", "14_1", "16_0","16_1", "20_0","20_1")), c("adt-CD90.2", "adt-CD335", "adt-NK-1.1", "adt-CD49a", "adt-CD49b", "adt-CD253", "adt-CD69", "adt-CD183", "adt-CD186", "adt-CD127", "adt-CD117", "adt-anti-rat-CD90-mouse-CD90.1", "adt-CD103", "adt-CD200R","rna_Tbx21", "rna_Eomes", "rna_Areg", "rna_Prf1", "rna_Fcgr4", "rna_Klrb1c", "rna_Itga1", "rna_Thy1", "rna_Il7r", "rna_Kit", "rna_Il1r1","rna_Rorc", "rna_Il17a", "rna_Ifng", "rna_Il12a"), stack = TRUE, pt.size = 0, same.y.lims = TRUE) & NoLegend()

VlnPlot(subset(tmp, sub.cluster %in% c("0_0", "0_1", "2_0", "2_1", "2_2", "3_0","3_1", "4_0","4_1", "11_0", "14_0", "14_1", "16_0","16_1", "20_0","20_1")), c("adt-CD200R", "adt-CD49b", "adt-CD253", "rna_Tbx21", "rna_Eomes", "rna_Tnfsf10", "rna_Zfp683"), pt.size = 0, ncol = 2) & NoLegend()

VlnPlot(subset(tmp, sub.cluster %in% c("0_0", "0_1", "2_0", "2_1", "2_2", "3_0","3_1", "4_0","4_1", "11_0", "14_0", "14_1", "16_0","16_1", "20_0","20_1")), c("rna_Il1rl1", "rna_Il17rb", "rna_Il18r1", "adt-KLRG1", "rna_Klrg1", "adt-CD69", "rna_Nmur1", "rna_Vipr1", "rna_Vipr2", "adt-CD194", "rna_Ccr4", "adt-CD198", "rna_Ccr8"), pt.size = 0, ncol = 2, stack = TRUE, same.y.lims = TRUE) & NoLegend()
##dev.off()
```

# Checking GNB2, IL10 and IL27-EBI3 expression based on Johannes and Prof. Bosmann's inputs
```{r}
#pdf(paste0(fig_path, "VlnPlots_Gnb2_Il10_Il27Ebi3_Il27p28.pdf"))

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Gnb2", "rna_Il10","rna_Ebi3", "rna_Il27"), pt.size = 0, stack = TRUE, same.y.lims = TRUE) & NoLegend()
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Gnb2", "rna_Il10","rna_Ebi3", "rna_Il27"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group")

##dev.off()
```

## Statistical difference in "Gnb2", "Il10", "Il27", "Ebi3" expression across all clusters betweeen groups
### "Gnb2", "Il10", "Il27", "Ebi3": Between il27raKO and wt
```{r}

#Finding differential expression of "Gnb2", "Il10", "Il27", "Ebi3" across clusters between il27raKO vs wt (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Gnb2, Il10, Il27, Ebi3 differential expression in clusters between il27raKO and wt
#RNA
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", features = c("Gnb2", "Il10", "Il27", "Ebi3"), logfc.threshold = 0, min.pct = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_4genes_RNA_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_4genes_RNA_all_integrated_rnm, paste0(tab_path, "/differential_Gnb2_Il10_Il27_Ebi3_all_integrated_res0.8.xlsx"))

rm(differential_4genes_RNA_all_integrated_rnm, tmp_celltype_list, get_differential_genes)

#Finding differential expression of "Gnb2", "Il10", "Il27", "Ebi3" between clusters 
#Gnb2, Il10, Il27, Ebi3 differential expression in clusters between il27raKO and wt
#RNA
differential_4genes_RNA_btwn_clusters <- FindAllMarkers(all_integrated_rnm_VlnPlot, features = c("Gnb2", "Il10", "Il27", "Ebi3"), logfc.threshold = 0, min.pct = 0, assay = "RNA")

#writexl::write_xlsx(differential_4genes_RNA_btwn_clusters, paste0(tab_path, "/differential_Gnb2_Il10_Il27_Ebi3_betweenCellTypes_res0.8.xlsx"))


```


```{r}
#pdf(paste0(fig_path, "VlnPlots_Legionella_pathoGenes.pdf"), width = 12)

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-TLR4",  "rna_Tlr1","rna_Tlr2", "rna_Tlr3", "rna_Tlr4", "rna_Tlr5", "rna_Tlr6", "rna_Tlr7", "rna_Tlr8", "rna_Tlr9"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Irf5", "rna_Irf3", "rna_Ddx58", "rna_Ifih1", "rna_Mavs", "rna_Depdc7", "rna_Gpr65", "rna_Fpr1","adt-CD127"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank()) #https://www.frontiersin.org/articles/10.3389/fcimb.2021.733564/full
VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD127", "rna_Nlrp3", "rna_Casp1", "rna_Casp3", "rna_Scaf11", "rna_Casp4","rna_Il1b", "rna_Il18", "rna_Nlrc4", "rna_Myd88", "rna_Nod1", "rna_Nod2", "rna_Ripk2"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Ifi208", "rna_Prdx2", "rna_Wdfy1", "rna_Stat2", "rna_Irf9"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Irgm2", "rna_Usp18", "rna_Gbp3", "rna_Rtp4", "rna_Trim25", "rna_Rsad2", "rna_Bst2", "rna_Ifi27", "rna_Isg15", "rna_Fst", "rna_Xaf1", "rna_Lgals3bp", "rna_Stat1", "rna_Ifit1", "rna_Wdfy1", "rna_Spp1", "rna_C1ra", "rna_C1rb", "rna_Ifitm1", "rna_Pik3r1", "rna_Cd59a", "rna_Vcam1"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank()) #https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5392577/
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Ifng", "rna_Csf2", "rna_Ifnb1", "rna_Tnf", "rna_Il6", "rna_Il1b", "rna_Il4", "rna_Tnfsf12", "rna_Il15", "rna_Il12b", "rna_Csf1", "rna_Il18", "rna_Il13", "rna_Cxcl2", "rna_Ccl13"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4734697/
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Isg15", "rna_Ifi27", "rna_Ifit1", "rna_Ifit3", "rna_Oasl2", "rna_Rsad2", "rna_Trim25"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank()) #https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5392577/

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Stat1", "rna_Irf3", "rna_Irf7", "rna_Nfatc2", "rna_Irf5", "rna_Nfkbia", "rna_Stat4", "rna_Rela", "rna_Nfkb1", "rna_Irf1", "rna_Cebpe", "rna_Irf8", "rna_Egr1", "rna_Cebpb"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank()) #https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4734697/

##dev.off()
```

# Upd 11Feb2022: checking GATA-3 expression as an ILC2 marker based on inputs from Johannes (Reference: https://pubmed.ncbi.nlm.nih.gov/25521670/)
```{r}
#checking GATA-3 expression as an ILC2 marker based on inputs from Johannes (Reference: https://pubmed.ncbi.nlm.nih.gov/25521670/)

#pdf(paste0(fig_path, "VlnPlots_ILC_GATA3_Unsubclustered.pdf"), width = 8)

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD335", "adt-NK-1.1", "adt-CD49a", "adt-CD49b", "adt-CD253", "adt-CD69", "adt-CD183", "adt-CD186","adt-CD127", "adt-CD117", "adt-CD103", "adt-IL-33Ralpha", "adt-CD94","rna_Tbx21", "rna_Eomes", "rna_Gata3", "rna_Ptgdr2", "rna_Klrb1", "rna_Il23r"), stack = TRUE, pt.size = 0, same.y.lims = TRUE) & NoLegend() & ylab("")

Idents(all_integrated_rnm_VlnPlot) <- "sub.cluster"
Idents(all_integrated_rnm_VlnPlot) <- factor(Idents(all_integrated_rnm_VlnPlot), levels = c(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, "18_0", "18_1", "18_2", 19, 20))

VlnPlot(subset(all_integrated_rnm_VlnPlot, sub.cluster %in% c(0, 2, 3, 4, 11, 14, 16, 20)), c("adt-CD335", "adt-NK-1.1", "adt-CD49a", "adt-CD49b", "adt-CD253", "adt-CD69", "adt-CD183", "adt-CD186","adt-CD127", "adt-CD117", "adt-CD103", "adt-IL-33Ralpha", "adt-CD94","rna_Tbx21", "rna_Eomes", "rna_Gata3", "rna_Ptgdr2", "rna_Klrb1", "rna_Il23r"), stack = TRUE, pt.size = 0, same.y.lims = TRUE) & NoLegend()
##dev.off()

tmp <- all_integrated_rnm
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(0), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(2), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(3), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(4), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(11), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(14), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(16), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
tmp <- FindSubCluster(tmp, cluster = c(20), graph.name = "wsnn", algorithm = 3, resolution = 0.2)
Idents(tmp) <- "sub.cluster"
Idents(tmp) <- factor(Idents(tmp), levels = c("0_0", "0_1", 1, "2_0", "2_1", "2_2", "3_0","3_1", "4_0","4_1", 5, 6, 7, 8, 9, 10, "11_0", 12, 13, "14_0", "14_1", 15, "16_0","16_1", 17, "18_0", "18_1", "18_2", 19, "20_0","20_1"))

#pdf(paste0(fig_path, "VlnPlots_ILC_GATA3_subclustered.pdf"), width = 12, height = 8)
VlnPlot(subset(tmp, sub.cluster %in% c("0_0", "0_1", "2_0", "2_1", "2_2", "3_0","3_1", "4_0","4_1", "11_0", "14_0", "14_1", "16_0","16_1", "20_0","20_1")), c("adt-CD335", "adt-NK-1.1", "adt-CD49a", "adt-CD49b", "adt-CD253", "adt-CD69", "adt-CD183", "adt-CD186","adt-CD127", "adt-CD117", "adt-CD103","adt-IL-33Ralpha", "adt-CD94","rna_Tbx21", "rna_Eomes", "rna_Gata3", "rna_Ptgdr2", "rna_Klrb1", "rna_Il23r"), stack = TRUE, pt.size = 0, same.y.lims = TRUE) & NoLegend() & ylab("")
##dev.off()
```



# upd 22 Feb 2022: plotting IL-24, gp130 (IL6ST) and Stat3 expression based on inputs from Johannes
```{r}
#pdf(paste0(fig_path, "VlnPlots_Il6st_Stat3.pdf"), width = 6)

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Il24", "rna_Il6st", "rna_Stat3"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())
##dev.off()


#Finding differential expression of "Il6st", "Stat3" across clusters between il27raKO vs wt (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Il6st, Stat3 differential expression in clusters between il27raKO and wt
#RNA
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", features = c("Il6st", "Stat3"), logfc.threshold = 0, min.pct = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27+CD11blo_NK","CD27+CD11b+_NK", "mito_NK", "NK_other", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_2genes_RNA_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_2genes_RNA_all_integrated_rnm, paste0(tab_path, "/differential_Il6st_Stat3_all_integrated_res0.8.xlsx"))

rm(differential_2genes_RNA_all_integrated_rnm, tmp_celltype_list, get_differential_genes)

Idents(all_integrated_rnm) <- "celltype"

```



#==================Upd 23Feb2022===================================#

# Upd: NK cell cluster labels updated based on inputs from Prof. Bosmann and Johannes
```{r}
#checking for ADT markers between the NK cell clusters
#comparing each NK cluster with all other NK clusters

tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

#checking differential expression between neutrophil marker expressing clusters

NK_group_DE <- list()

for(i in c("CD27+CD11blo_NK", "CD27loCD11b+_NK", "CD27+CD11b+_NK", "mito_NK", "NK_other")){
  for(j in c("CD27+CD11blo_NK", "CD27loCD11b+_NK", "CD27+CD11b+_NK", "mito_NK", "NK_other")){
    if(i != j){
  RNAmarkers <- FindMarkers(tmp, ident.1 = i, ident.2 = j, only.pos = FALSE, assay = "RNA")
  RNAmarkers$gene <- row.names(RNAmarkers)
  RNAmarkers$comparison <- paste0(i, "-", j)
  NK_group_DE[[paste0("RNA_markers", i, "_", j)]] <- RNAmarkers
  ADTmarkers <- FindMarkers(tmp, ident.1 = i, ident.2 = j, only.pos = FALSE, assay = "ADT")
  ADTmarkers$protein <- row.names(ADTmarkers)
  ADTmarkers$comparison <- paste0(i, "-", j)
  NK_group_DE[[paste0("ADT_markers", i, "_", j)]] <- ADTmarkers
    }
  }
}

NK_group_DE <- bind_rows(NK_group_DE)

#writexl::write_xlsx(NK_group_DE, paste0(tab_path, "/all_integrated_NK_group_DE_res0.8.xlsx"))


compare_list_NK <- list("CD27+CD11blo_NK" = c("CD27loCD11b+_NK", "CD27+CD11b+_NK", "mito_NK", "NK_other"), "CD27loCD11b+_NK" = c("CD27+CD11blo_NK", "CD27+CD11b+_NK", "mito_NK", "NK_other"), "CD27+CD11b+_NK" = c("CD27+CD11blo_NK", "CD27loCD11b+_NK", "mito_NK", "NK_other"), "mito_NK" = c("CD27+CD11blo_NK", "CD27loCD11b+_NK", "CD27+CD11b+_NK", "NK_other"), "NK_other" = c("CD27+CD11blo_NK", "CD27loCD11b+_NK", "CD27+CD11b+_NK", "mito_NK"))

NK_byGroup_DE <- list()

for(i in c("CD27+CD11blo_NK", "CD27loCD11b+_NK", "CD27+CD11b+_NK", "mito_NK", "NK_other")){
      RNAmarkers <- FindMarkers(tmp, ident.1 = i, ident.2 = as.vector(unlist(compare_list_NK[[i]])), only.pos = FALSE, assay = "RNA")
  RNAmarkers$gene <- row.names(RNAmarkers)
  NK_byGroup_DE[[paste0("RNA_markers", i)]] <- RNAmarkers
  ADTmarkers <- FindMarkers(tmp, ident.1 = i, ident.2 = as.vector(unlist(compare_list_NK[[i]])), only.pos = FALSE, assay = "ADT")
  ADTmarkers$protein <- row.names(ADTmarkers)
  NK_byGroup_DE[[paste0("ADT_markers", i)]] <- ADTmarkers
      }

#writexl::write_xlsx(NK_byGroup_DE, paste0(tab_path, "/all_integrated_NK_byGroup_DE_res0.8.xlsx"))


VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD27+CD11blo_NK", "CD27loCD11b+_NK", "CD27+CD11b+_NK", "mito_NK", "NK_other")), c("adt-CD117", "adt-CD39", "adt-CD71",  "adt-CD11a", "adt-CD62L", "adt-CD55", "adt-Ly49D", "adt-CD94", "adt-CD25", "adt-CD61", "adt-CD11b", "adt-CD27", "adt-CD86", "adt-I-A-I-E"), stack = TRUE, same.y.lims = TRUE) & NoLegend()

FeaturePlot(all_integrated_rnm, c("adt-Ly49D"), label = TRUE, label.size = 2, cols = c("white", "darkgreen"), order = TRUE) + theme(plot.background = element_rect(fill = "lightgrey"))
```

# save workspace
```{r}
#save.image(paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_v1.1.RData"))
#load(paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_v1.1.RData"))

#saveRDS(all_integrated,paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_v1.1_all_integrated.rds")) #unannotated cell types, will not change in v1.2

#saveRDS(all_integrated,paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_v1.1_all_integrated.rds"))
```


# Updating the NK clusters after confirmation from Johannes
```{r}
# Updating the NK clusters 

all_integrated_rnm <- RenameIdents(all_integrated_rnm,
                                   "CD27+CD11blo_NK" = "CD27loCD11b+_NK",
                                   "CD27+CD11blo_NK" = "CD27loCD11b+_NK",
                                   "CD27loCD11b+_NK" = "CD27loCD11bhi_NK",
                                   "CD27loCD11b+_NK" = "CD27loCD11bhi_NK",
                                   "mito_NK" = "CD27loCD11b+_mito_NK",
                                   "CD27+CD11b+_NK" = "CD27hiCD11b+_NK",
                                   "CD27+CD11b+_NK" = "CD27hiCD11b+_NK",
                                   "NK_other" = "CD27loCD11blo_NK")
all_integrated_rnm$celltype <- Idents(all_integrated_rnm)
all_integrated_rnm$cluster0.8.group <- paste(Idents(all_integrated_rnm), all_integrated_rnm$group, sep = "_")
all_integrated_rnm$cluster0.8 <- Idents(all_integrated_rnm)
Idents(all_integrated_rnm) <- "cluster0.8.group"

```


```{r}
#specifying data directories
#fig_path <- file.path(paste0(output_dir, "SW0100_P239_figures_v1.2/")) #changed to v1.2 for figures after re-annotation of NK clusters
#tab_path <- file.path(paste0(output_dir, "SW0100_P239_tables_v1.2/")) #changed to v1.2 for tables after re-annotation of NK clusters
#data_path <- file.path(paste0(output_dir, "SW0100_P239_Rdata_v1.2/")) #changed to v1.2 for data after re-annotation of NK clusters
```


# New UMAP dimplots
```{r}
Idents(all_integrated_rnm) <- "celltype"

#pdf(paste0(fig_path, "/DimPlot_all_integrated_rnm_res0.8.pdf"), width = 8, height = 6)
DimPlot(all_integrated_rnm, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 3) + NoLegend() | DimPlot(all_integrated_rnm, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 3, group.by = "group")
DimPlot(all_integrated_rnm, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 2.5, split.by = "group") + NoLegend()
##dev.off()

#pdf(paste0(fig_path, "/DimPlot_bySampleID_all_integrated_rnm_res0.8.pdf"), width = 15, height = 10)
all_integrated_rnm$sample_id <- factor(x = all_integrated_rnm$sample_id, levels = c("BM_WS_9", "BM_WS_10", "BM_WS_15", "BM_WS_5", "BM_WS_6", "BM_WS_8"))
DimPlot(all_integrated_rnm, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 2, split.by = "sample_id", ncol = 3) + NoLegend()
##dev.off()


```


#Rename levels for violin and dot plots
```{r, echo=FALSE, message=FALSE}
all_integrated_rnm_VlnPlot <- all_integrated_rnm

levels_VlnPlot <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell") 

#factor(Idents(all_integrated_rnm_VlnPlot), levels = all_integrated_rnm_levels_VlnPlot)
Idents(all_integrated_rnm_VlnPlot) <- factor(Idents(all_integrated_rnm_VlnPlot), levels = levels_VlnPlot)

all_integrated_rnm_DotPlot <- all_integrated_rnm
levels_DotPlot <- rev(levels_VlnPlot) 

#factor(Idents(all_integrated_rnm_DotPlot), levels = all_integrated_rnm_levels_DotPlot)
Idents(all_integrated_rnm_DotPlot) <- factor(Idents(all_integrated_rnm_DotPlot), levels = levels_DotPlot) 
```



# Checking GNB2, IL10 and IL27-EBI3 expression based on Johannes and Prof. Bosmann's inputs
```{r}
#pdf(paste0(fig_path, "DotPlots_Gnb2_Il10_Il27Ebi3_Il27p28.pdf"), height = 9)
DotPlot(all_integrated_rnm_DotPlot, c("rna_Gnb2", "rna_Il10","rna_Ebi3", "rna_Il27"), split.by = "group", assay = "RNA", cols = c("navy", "red")) + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))
##dev.off()
```


# Conserved and differentially expressed genes and proteins across conditions <br>
## Conserved genes and proteins <br>
```{r}
## Conserved genes and proteins
library('multtest')
library('metap')

#wt and il27raKO combined integrated

#genes
DefaultAssay(object = all_integrated_rnm) <- "RNA"

get_conserved_markers <- function(celltype){
  Seurat::FindConservedMarkers(object = all_integrated_rnm, ident.1 = celltype, grouping.var = "group", verbose = FALSE, assay = "RNA") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

conserved_markers <- map_dfr(tmp_celltype_list, get_conserved_markers)

#writexl::write_xlsx(conserved_markers, paste0(tab_path, "/conserved_RNAmarkers_all_integrated_by_cluster_res0.8.xlsx"))


#ADT
DefaultAssay(object = all_integrated_rnm) <- "ADT"

get_conserved_markers <- function(celltype){
  Seurat::FindConservedMarkers(object = all_integrated_rnm, ident.1 = celltype, grouping.var = "group", verbose = FALSE, assay = "ADT") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

conserved_markers <- map_dfr(tmp_celltype_list, get_conserved_markers)

#writexl::write_xlsx(conserved_markers, paste0(tab_path, "/conserved_ADTmarkers_all_integrated_by_cluster_res0.8.xlsx"))

```


## Differentially expressed genes <br>
Notes:<b>
(Reference: https://satijalab.org/seurat/articles/integration_introduction.html) <br>
1. Identify what genes change in different groups for cells of the same type using Wilcoxon Rank Sum Test and, MAST (GLM-based Hurdle model
). <br>
```{r}
## Differentially expressed genes between groups using Wilcoxon Rank Sum Test
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html

#First, create a column in the meta.data slot to hold both the cell type and group information and switch the current ident to that column. 
all_integrated_rnm$cluster0.8.group <- paste(Idents(all_integrated_rnm), all_integrated_rnm$group, sep = "_")
all_integrated_rnm$cluster0.8 <- Idents(all_integrated_rnm)
Idents(all_integrated_rnm) <- "cluster0.8.group"

#genes
DefaultAssay(object = all_integrated_rnm) <- "RNA"

#Finding differentially expressed genes across clusters (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)

get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_genes_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_genes_all_integrated_rnm, paste0(tab_path, "/differential_genes_RNA_all_integrated_by_cluster_res0.8.xlsx"))


#ADT
DefaultAssay(object = all_integrated_rnm) <- "ADT"
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "ADT", logfc.threshold = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_ADT_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_ADT_all_integrated_rnm, paste0(tab_path, "/differential_genes_ADT_all_integrated_by_cluster_res0.8_NoLFCThresh.xlsx"))


##===Differential testing using MAST framework=================##
#genes
DefaultAssay(object = all_integrated_rnm) <- "RNA"
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Finding differentially expressed genes across clusters (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)

get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", test.use = "MAST") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_genes_all_integrated_rnm_MAST <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_genes_all_integrated_rnm, paste0(tab_path, "/differential_genes_RNA_all_integrated_byCluster_res0.8_MAST.xlsx"))


#ADT
DefaultAssay(object = all_integrated_rnm) <- "ADT"
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "ADT", logfc.threshold = 0, test.use = "MAST") %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_ADT_all_integrated_rnm_MAST <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_ADT_all_integrated_rnm, paste0(tab_path, "/differential_genes_ADT_all_integrated_byCluster_res0.8_NoLFCThresh_MAST.xlsx"))

```


##Number of cells per cluster per group and per sample_id
```{r}
#Number of cells per cluster per group and per sample_id
Idents(all_integrated_rnm) <- "celltype"
n_cells_per_cluster <- FetchData(all_integrated_rnm, vars = c("ident", "orig.ident", "group", "sample_id")) %>% dplyr::count(ident, orig.ident, group, sample_id) %>% tidyr::spread(ident, n)
#writexl::write_xlsx(n_cells_per_cluster, paste0(tab_path, "/cellNumbers_all_integrated_by_cluster_res0.8.xlsx"))

#plotting the number of cells

library(dittoSeq)
#pdf(paste0(fig_path, "/res0.8_Cells_per_cluster.pdf"), width = 8, height = 6 )
dittoBarPlot(all_integrated_rnm, "group", group.by = "celltype", color.panel = c("#F8766D", "#00BFC4")) + theme(axis.text.x =  element_text(angle = 90))

all_integrated_rnm$sample_id <- factor(x = all_integrated_rnm$sample_id, levels = c("BM_WS_15","BM_WS_10", "BM_WS_9", "BM_WS_8", "BM_WS_6", "BM_WS_5"))

dittoBarPlot(all_integrated_rnm, "sample_id", group.by = "celltype",  retain.factor.levels = TRUE,color.panel = c("#00A9FF", "#00BFC4", "#00BE67", "#7CAE00", "#CD9600", "#F8766D")) + theme(axis.text.x =  element_text(angle = 90))
##dev.off()


# retrieve the absolute number of cells by sample_id
cells_by_sampleID <- dittoBarPlot(all_integrated_rnm, "sample_id", group.by = "celltype", retain.factor.levels = TRUE, scale = "count")
  
#writexl::write_xlsx(cells_by_sampleID$data, paste0(tab_path, "/Absolute_cellNumbers_all_integrated_by_cluster_res0.8.xlsx"))



```


#Muscat differential state analysis
*Notes:<br>
- Performing differential state analysis using muscat R package to assess to cell-type-specific responses across samples (Reference: https://www.bioconductor.org/packages/devel/bioc/vignettes/muscat/inst/doc/analysis.html) <br>
- This has been developed for scRNA seq specifically, so trying with RNA assay for this analysis. <br>

```{r}
# Differential state analysis based on muscat
#- Performing differential state analysis using muscat R package to assess to cell-type-specific responses across samples (Reference: https://www.bioconductor.org/packages/devel/bioc/vignettes/muscat/inst/doc/analysis.html) <br>
#- This has been developed for scRNA seq specifically, so trying with RNA assay for this analysis. <br>



#preparing data for muscat
library(limma)
library(muscat)
library(purrr)
library(dplyr)

#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(all_integrated_rnm, assay = "RNA")

tmp_sce$id <- paste0(tmp_sce$group,"_", tmp_sce$sample_id) #combining condition and sample_id

tmp_sce <- muscat::prepSCE(tmp_sce, 
    kid = "celltype", # cluster numbers or celltype
    gid = "group",  # group IDs (il27raKO/wt)
    sid = "id",   # sample IDs (il27raKO/wt+sample_id)
    drop = TRUE)  # drop all other colData columns
tmp_sce$sample_id <- factor(x = tmp_sce$sample_id, levels = c("il27raKO_BM_WS_9", "il27raKO_BM_WS_10", "il27raKO_BM_WS_15", "wt_BM_WS_5",  "wt_BM_WS_6", "wt_BM_WS_8"))

nk <- length(kids <- levels(tmp_sce$celltype)) 
ns <- length(sids <- levels(tmp_sce$sample_id))
names(kids) <- kids; names(sids) <- sids


#Differential State (DS) analysis
pseudobulk_sce <- muscat::aggregateData(tmp_sce,
    assay = "counts", fun = "sum",
    by = c("cluster_id", "sample_id")) #cluster_id refers to cell type


# one sheet per subpopulation
assayNames(pseudobulk_sce)

#Pseudobulk-level MDS plot
pb_sce_mds <- muscat::pbMDS(pseudobulk_sce) 
#warning: Removing 4 rows with all zero counts


#pdf(paste0(fig_path, "/muscatDS_MDS_RNA_res0.8.pdf"), width = 10, height = 8)
pb_sce_mds
##dev.off()

#test for DS using pbDS. By default, a group_id model is fit, and the last coefficient of the linear model is tested to be equal to zero.

# construct design & contrast matrix
ei <- metadata(tmp_sce)$experiment_info
mm <- model.matrix(~ 0 + ei$group_id)
dimnames(mm) <- list(ei$sample_id, levels(ei$group_id))
contrast <- makeContrasts("il27raKO-wt", levels = mm)

# run DS analysis
res_sce <- muscat::pbDS(pseudobulk_sce,  method = "edgeR", design = mm, contrast = contrast, verbose = FALSE)
# access results table for 1st comparison
tbl_sce <- res_sce$table[[1]]
# one data.frame per cluster
names(tbl_sce)

# view results for 1st cluster
k1_sce <- tbl_sce[[1]]
head(format(k1_sce[, -ncol(k1_sce)], digits = 2))

#sort by adjusted p value
tbl_fil_all_RNA <- lapply(tbl_sce, function(u) {
  dplyr::arrange(u, p_adj.loc)
})


#filter results to retain hits with FDR < 5% and abs(logFC) > 0.58, and count the number and frequency of differential findings by cluster and view the top hits (lowest adj. p-value) in each cluster
# filter FDR < 5%, abs(logFC) > 1 & sort by adj. p-value
tbl_fil_sig <- lapply(tbl_sce, function(u) {
  u <- dplyr::filter(u, p_adj.loc < 0.05, abs(logFC) > 0.58)
  dplyr::arrange(u, p_adj.loc)
})

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil_sig, nrow, numeric(1))
p_de <- format(n_de / nrow(tmp_sce) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)

# view top 2 hits in each cluster
top2_sig <- bind_rows(lapply(tbl_fil_sig, top_n, 2, p_adj.loc))
format(top2_sig[, -ncol(top2_sig)], digits = 2)

#results table with expression frequency of each gene by sample and overall fold change between napc2 vs control (untreated)
res_tbl <- muscat::resDS(tmp_sce, res_sce, frq = TRUE)

#Between-cluster concordance: UpSet-plot that visualizes the number of DE genes that are shared across or unique to certain clusters
library(UpSetR)
de_gs_by_k <- map(tbl_fil_sig, "gene")
upset(fromList(de_gs_by_k))

#writexl::write_xlsx(tbl_fil_all_RNA,paste0(tab_path, "muscatDS_byCellType_RNA_all_integrated_res0.8.xlsx"))

library(scater)
#UMAP
.plot_dr <- function(tmp_sce, dr, col)
  plotReducedDim(tmp_sce, dimred = dr, colour_by = col) +
    guides(fill = guide_legend(override.aes = list(alpha = 1, size = 3))) +
    theme_minimal() + theme(aspect.ratio = 1)




# downsample to max. 100 cells per cluster
cs_by_k <- split(colnames(tmp_sce), tmp_sce$cluster_id)
cs100 <- unlist(sapply(cs_by_k, function(u) 
  sample(u, min(length(u), 100))))

# plot t-SNE & UMAP colored by cluster & group ID
for (dr in c("WNN.UMAP"))
  for (col in c("cluster_id", "group_id"))
    .plot_dr(tmp_sce[, cs100], dr, col)


# Violin plots of top 5 genes across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_RNA_VlnPlot_byCelltype_res0.8.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_RNA)){
print(plotExpression(tmp_sce[, tmp_sce$cluster_id == i],
                features = tbl_fil_all_RNA[[i]]$gene[seq_len(5)],
                x = "sample_id", colour_by = "group_id", ncol = 3) + scale_color_manual(values = c("#F8766D", "#00BFC4"), aesthetics = "color") +
     guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) +
     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ggtitle(paste0("Violin plot of top 5 muscat differential state genes; ", i) )
)
}
##dev.off()

# Heatmap of top 20 genes (or less; based on adjusted p value) across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_RNA_Heatmap_byCelltype_res0.8.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_RNA)){
  if( ! (i %in% c("CD27loCD11blo_NK", "CD27-_gdT", "moDC", "B_cell") ) ){ #excluding "CD27loCD11blo_NK", "CD27-_gdT", "moDC", "B_cell"as there are no genes with FDR < 0.05 and including this causes an error in generating the heatmap
tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, k = i, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp_htmap@column_title <- paste0("Heatmap of top 20 (or less) genes from muscat DS analysis in ", i)
print(tmp_htmap)
  }
}
##dev.off()

#------------------------------------------------------------------------

# Assay ADT
#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(all_integrated_rnm, assay = "ADT")

tmp_sce$id <- paste0(tmp_sce$group,"_", tmp_sce$sample_id) #combining condition and sample_id

tmp_sce <- muscat::prepSCE(tmp_sce, 
    kid = "celltype", # cluster numbers or celltype
    gid = "group",  # group IDs (il27raKO/wt)
    sid = "id",   # sample IDs (il27raKO/wt+sample_id)
    drop = TRUE)  # drop all other colData columns
tmp_sce$sample_id <- factor(x = tmp_sce$sample_id, levels = c("il27raKO_BM_WS_9", "il27raKO_BM_WS_10", "il27raKO_BM_WS_15", "wt_BM_WS_5",  "wt_BM_WS_6", "wt_BM_WS_8"))

nk <- length(kids <- levels(tmp_sce$celltype)) 
ns <- length(sids <- levels(tmp_sce$sample_id))
names(kids) <- kids; names(sids) <- sids


#Differential State (DS) analysis
pseudobulk_sce <- muscat::aggregateData(tmp_sce,
    assay = "counts", fun = "sum",
    by = c("cluster_id", "sample_id")) #cluster_id refers to cell type


# one sheet per subpopulation
assayNames(pseudobulk_sce)

#Pseudobulk-level MDS plot
pb_sce_mds <- muscat::pbMDS(pseudobulk_sce) 


#pdf(paste0(fig_path, "/muscatDS_MDS_ADT_res0.8.pdf"), width = 10, height = 8)
pb_sce_mds
##dev.off()

#test for DS using pbDS. By default, a group_id model is fit, and the last coefficient of the linear model is tested to be equal to zero.

# construct design & contrast matrix
ei <- metadata(tmp_sce)$experiment_info
mm <- model.matrix(~ 0 + ei$group_id)
dimnames(mm) <- list(ei$sample_id, levels(ei$group_id))
contrast <- makeContrasts("il27raKO-wt", levels = mm)

# run DS analysis
res_sce_ADT <- muscat::pbDS(pseudobulk_sce,  method = "edgeR", design = mm, contrast = contrast, verbose = FALSE)
# access results table for 1st comparison
tbl_sce_ADT <- res_sce_ADT$table[[1]]
# one data.frame per cluster
names(tbl_sce_ADT)

# view results for 1st cluster
k1_sce_ADT <- tbl_sce_ADT[[1]]
head(format(k1_sce[, -ncol(k1_sce)], digits = 2))

#sort by adjusted p value
tbl_fil_all_ADT <- lapply(tbl_sce_ADT, function(u) {
  dplyr::arrange(u, p_adj.loc)
})


#filter results to retain hits with FDR < 5% and abs(logFC) > 0.58, and count the number and frequency of differential findings by cluster and view the top hits (lowest adj. p-value) in each cluster
# filter FDR < 5%, abs(logFC) > 1 & sort by adj. p-value
tbl_fil_sig_ADT <- lapply(tbl_sce_ADT, function(u) {
  u <- dplyr::filter(u, p_adj.loc < 0.05, abs(logFC) > 0.58)
  dplyr::arrange(u, p_adj.loc)
})

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil_sig_ADT, nrow, numeric(1))
p_de <- format(n_de / nrow(tmp_sce) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)

# view top 2 hits in each cluster
top2_sig_ADT <- bind_rows(lapply(tbl_fil_sig_ADT, top_n, 2, p_adj.loc))
format(top2_sig[, -ncol(top2_sig_ADT)], digits = 2)

#results table with expression frequency of each gene by sample and overall fold change between napc2 vs control (untreated)
res_tbl_ADT <- muscat::resDS(tmp_sce, res_sce_ADT, frq = TRUE)

#Between-cluster concordance: UpSet-plot that visualizes the number of DE genes that are shared across or unique to certain clusters
de_gs_by_k <- map(tbl_fil_sig_ADT, "gene")
upset(fromList(de_gs_by_k))

#writexl::write_xlsx(tbl_fil_all_ADT,paste0(tab_path, "muscatDS_byCellType_ADT_all_integrated_res0.8.xlsx"))


#UMAP
.plot_dr <- function(tmp_sce, dr, col)
  plotReducedDim(tmp_sce, dimred = dr, colour_by = col) +
    guides(fill = guide_legend(override.aes = list(alpha = 1, size = 3))) +
    theme_minimal() + theme(aspect.ratio = 1)




# downsample to max. 100 cells per cluster
cs_by_k <- split(colnames(tmp_sce), tmp_sce$cluster_id)
cs100 <- unlist(sapply(cs_by_k, function(u) 
  sample(u, min(length(u), 100))))

# plot t-SNE & UMAP colored by cluster & group ID
for (dr in c("WNN.UMAP"))
  for (col in c("cluster_id", "group_id"))
    .plot_dr(tmp_sce[, cs100], dr, col)


# Violin plots of top 5 genes across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_ADT_VlnPlot_byCelltype_res0.8.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_ADT)){
print(plotExpression(tmp_sce[, tmp_sce$cluster_id == i],
                features = tbl_fil_all_ADT[[i]]$gene[seq_len(5)],
                x = "sample_id", colour_by = "group_id", ncol = 3) + scale_color_manual(values = c("#F8766D", "#00BFC4"), aesthetics = "color") +
     guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) +
     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ggtitle(paste0("Violin plot of top 5 muscat differential state proteins; ", i) )
)
}
##dev.off()

# Heatmap of top 20 (or less) proteins (based on adjusted p < 0.05) across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_ADT_Heatmap_byCelltype_res0.8.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_ADT)){
  if(! (i %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK","CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "B_cell") )){ #"CD27loCD11b+_NK", "CD27loCD11bhi_NK","CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK",, "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "B_cell"  as there are no proteins with FDR < 0.05 and including this causes an error in generating the heatmap
tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce_ADT, k = i, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp_htmap@column_title <- paste0("Heatmap of top 20 (or less) proteins from muscat DS analysis in ", i)
print(tmp_htmap)
  }
}
##dev.off()
#c("#F8766D", "#00BFC4")




```



# Gene ontology and pathway enrichment analysis: topGO
## Seurat DE genes
```{r, message=FALSE}
# Select background for each analysis based on genes which are expressed in the particular cluster of cells
#Source code: https://ucdavis-bioinformatics-training.github.io/2021-August-Single-Cell-RNA-Seq-Analysis/data_analysis/scRNA_Workshop-PART6_fixed

#TopGO analysis
library(pcaExplorer)
library(topGO)
library(org.Mm.eg.db)
library(AnnotationDbi)

topgoDE_cluster <- list()
tmp <- all_integrated_rnm
DefaultAssay(tmp) <- "RNA"


for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC")){ #excluding "CD27loCD11blo_NK", "CD27-_gdT", "B_cell" as there are no statistically significant DE genes based on adjusted p value
  clust <- tmp[, tmp$celltype %in% i] #subset to a particular cell type
  expr_clust <- clust[Matrix::rowSums(clust[["RNA"]]@counts > 0) > 0, ] #selecting background as genes/proteins which are atleast expressed in a particular cluster/celltype
  background.genes <- rownames(expr_clust)
  clust_deg <- subset(differential_genes_all_integrated_rnm, subset = cell_type %in% i & p_val_adj < 0.05)$gene
  topgoDE_cluster[[i]] <-
  pcaExplorer::topGOtable(DEgenes =  clust_deg,
                          BGgenes = background.genes,
                          ontology = "BP",
                          mapping = "org.Mm.eg.db",
                          geneID = "symbol",
                          topTablerows = 500)

}


#writexl::write_xlsx(topgoDE_cluster, paste0(tab_path, "topGO-DE_Seurat_DEgenes_res0.8.xlsx"))

```

# topgo: muscat DS gene results
```{r}
# topgo: muscat DS gene results
topgoDE_cluster_muscatDS <- list()

#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(all_integrated_rnm, assay = "RNA")

differential_genes_muscatDS <- bind_rows(tbl_fil_all_RNA)


for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "Monocyte", "Alveolar_macrophage", "cDC", "pDC")){ #excluding "CD27-_gdT", "CD27loCD11blo_NK", "moDC" and "B_cell"  as there are no statistically significant DE genes based on adjusted p value
  clust <- tmp_sce[, tmp_sce$celltype %in% i] #subset to a particular cell type
  expr_clust <- clust[rowSums(counts(clust) > 0) > 0, ] #selecting background as genes/proteins which are atleast expressed in a particular cluster/celltype
  background.genes <- rownames(expr_clust)
  clust_deg <- subset(differential_genes_muscatDS, subset = cluster_id %in% i & p_adj.loc < 0.05)$gene
  topgoDE_cluster_muscatDS[[i]] <-
  pcaExplorer::topGOtable(DEgenes =  clust_deg,
                          BGgenes = background.genes,
                          ontology = "BP",
                          mapping = "org.Mm.eg.db",
                          geneID = "symbol",
                          topTablerows = 500)

}


#writexl::write_xlsx(topgoDE_cluster_muscatDS, paste0(tab_path, "topGO-DE_muscat_DSgenes_res0.8.xlsx"))



```


# EnrichR gene set enrichment
```{r}
library(enrichR)

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

Idents(all_integrated_rnm) <- "cluster0.8.group"

#RNA
# GO BP
enrichr_gobp_all_integrated_rnm <- list()
#pdf(paste0(fig_path, "/Seurat_DEenrichR_GO-BP_RNA_all_integrated_res0.8.pdf"), width = 20, height = 8)

for(i in tmp_celltype_list){
  plot_enrichr <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "GO_Biological_Process_2021", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA") #p.val.cutoff	: Cutoff to select DE genes.
  print(plot_enrichr)
  enrichr_gobp_all_integrated_rnm[[i]] <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "GO_Biological_Process_2021", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 100, verbose = FALSE, assay = "RNA", return.gene.list = TRUE)
  }

##dev.off()

library(data.table)
#writexl::write_xlsx(lapply(enrichr_gobp_all_integrated_rnm, rbindlist, idcol = TRUE), paste0(tab_path, "enrichr_gobp_all_integrated_rnm_res0.8.xlsx"))


# KEGG
enrichr_KEGG_all_integrated_rnm <- list()
#pdf(paste0(fig_path, "/Seurat_DEenrichR_KEGG_RNA_all_integrated_res0.8.pdf"), width = 20, height = 8)

for(i in tmp_celltype_list){
  plot_enrichr <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "KEGG_2019_Mouse", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA")
  print(plot_enrichr)
  enrichr_KEGG_all_integrated_rnm[[i]] <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "KEGG_2019_Mouse", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA", return.gene.list = TRUE)
}

##dev.off()
#writexl::write_xlsx(lapply(enrichr_KEGG_all_integrated_rnm, rbindlist, idcol = TRUE), paste0(tab_path, "enrichr_KEGG_all_integrated_rnm_res0.8.xlsx"))


# Reactome
enrichr_Reactome_all_integrated_rnm <- list()
#pdf(paste0(fig_path, "/Seurat_DEenrichR_Reactome_RNA_all_integrated_res0.8.pdf"), width = 20, height = 8)

for(i in tmp_celltype_list){
  plot_enrichr <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "Reactome_2016", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA")
  print(plot_enrichr)
  enrichr_Reactome_all_integrated_rnm[[i]] <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "Reactome_2016", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA", return.gene.list = TRUE)
}

##dev.off()
#writexl::write_xlsx(lapply(enrichr_Reactome_all_integrated_rnm, rbindlist, idcol = TRUE), paste0(tab_path, "enrichr_Reactome_all_integrated_rnm_res0.8.xlsx"))



# MSigDB_Hallmark_2020
enrichr_MSigDB_all_integrated_rnm <- list()

#pdf(paste0(fig_path, "/Seurat_DEenrichR_MSigDB_Hallmark_RNA_all_integrated_res0.8.pdf"), width = 20, height = 8)

for(i in tmp_celltype_list){
  plot_enrichr <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "MSigDB_Hallmark_2020", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA")
  print(plot_enrichr)
  enrichr_MSigDB_all_integrated_rnm[[i]] <- Seurat::DEenrichRPlot(object = all_integrated_rnm, ident.1 = paste0(i, "_il27raKO"), ident.2 = paste0(i, "_wt"), enrich.database = "MSigDB_Hallmark_2020", test.use = "wilcox", max.genes = Inf, balanced = TRUE, p.val.cutoff = 0.05, num.pathway = 20, verbose = FALSE, assay = "RNA", return.gene.list = TRUE)
}

##dev.off()
#writexl::write_xlsx(lapply(enrichr_MSigDB_all_integrated_rnm, rbindlist, idcol = TRUE), paste0(tab_path, "enrichr_MSigDB_all_integrated_rnm_res0.8.xlsx"))

```

# Differential abundance (DA) analysis
*Notes <br>
- Differential abundance (DA) analysis compares the proportions of cell types across experimental conditions and aims to highlight populations that are present at different ratios. (Reference: http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/muscWorkshop__vignette/ ; https://bioconductor.org/books/release/OSCA/multi-sample-comparisons.html)
```{r}
Idents(all_integrated_rnm) <- "celltype"
# Differential abundance based on RNA
#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(all_integrated_rnm, assay = "RNA")

tmp_sce$id <- paste0(tmp_sce$group,"_", tmp_sce$sample_id) #combining condition and sample_id

tmp_sce <- muscat::prepSCE(tmp_sce, 
    kid = "celltype", # cluster numbers or celltype
    gid = "group",  # group IDs (il27raKO/wt)
    sid = "id",   # sample IDs (il27raKO/wt+sample_id)
    drop = TRUE)  # drop all other colData columns
tmp_sce$sample_id <- factor(x = tmp_sce$sample_id, levels = c("wt_BM_WS_5", "wt_BM_WS_6", "wt_BM_WS_8", "il27raKO_BM_WS_9", "il27raKO_BM_WS_10", "il27raKO_BM_WS_15")) #order of sample_ids with wt first, il27rako next, as this would help estimate the differential abundance correctly. ordering the il27rako first for differential abundance gives criss-crossed estimates.

nk <- length(kids <- levels(tmp_sce$cluster_id)) 
ns <- length(sids <- levels(tmp_sce$sample_id))
names(kids) <- kids; names(sids) <- sids

ei <- metadata(tmp_sce)$experiment_info

# calculate cluster-sample cell counts
n_cells <- table(tmp_sce$cluster_id, tmp_sce$sample_id)

# calculate cluster proportions across samples
freqs <- prop.table(n_cells, margin = 1) #margin = 1 utilizes as denominator the row sums of each celltype i.e. total number of cells for a celltype in both il27raKO and wt


# prep. data.frame for plotting
df <- data.frame(
    frequency = as.numeric(freqs), 
    cluster_id = rep(kids, ns),
    sample_id = rep(sids, each = nk))

m <- match(df$sample_id, ei$sample_id)
df$group_id <- ei$group_id[m]

#pdf(paste0(fig_path, "/Differential_abundance_all_integrated_rnm_res0.8.pdf"), width = 12, height = 8)

# barplot of relative cluster-abundances (Barplot of subpopulation frequencies; stratified by sample_id and group)
ggplot(df, aes(x = sample_id, y = frequency, fill = cluster_id)) +
    geom_bar(stat = "identity", position = "fill") + 
    facet_wrap(~ group_id, scales = "free_x") +
    theme_classic()  + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ggtitle("Barplot of subpopulation frequencies (relative abundance); stratified by sample_id and group")


# boxplot of relative cluster-abundances (Boxplot of subpopulation frequencies; stratified by sample_id)
ggplot(df, aes(x = group_id, y = frequency, color = group_id)) +
    geom_boxplot(outlier.colour = NA) +  geom_jitter() +
    facet_wrap(~ cluster_id, scales = "free_y", ncol = 4) +
    theme_classic() + ggtitle("Boxplot of subpopulation frequencies (relative abundance); stratified by sample_id and group")


# plot absolute number of cells by sample_id
tmp_df <- data.frame(abs_numb_cells = table(tmp_sce$cluster_id, tmp_sce$sample_id), 
     cluster_id = rep(kids, ns),
     sample_id = rep(sids, each = nk))

mt <- match(tmp_df$sample_id, ei$sample_id)
tmp_df$group_id <- ei$group_id[mt]

# boxplot of absolute cluster-abundances (Boxplot of subpopulation frequencies; stratified by sample_id)
ggplot(tmp_df , aes(x = group_id, y = abs_numb_cells.Freq, color = group_id)) +
     geom_boxplot(outlier.colour = NA) +  geom_jitter() +
     facet_wrap(~ cluster_id, scales = "free_y", ncol = 4) + ylab("absolute cell number") +
     theme_classic() + ggtitle("Boxplot of absolute cell type abundance; stratified by sample_id and group")


##dev.off()


library(edgeR)
# DA analysis
y <- DGEList(counts = n_cells)
y <- estimateDisp(y, design = mm, trend = "none")
fit <- glmQLFit(y, design = mm)
fit <- glmLRT(fit, contrast = contrast)
topTags(fit, n = Inf)$table %>% round(2)  #logFC = frequency of cells

fit_ab_RNA <- glmQLFit(y, design = mm, robust=TRUE, abundance.trend=FALSE)
res_DA_RNA <- glmQLFTest(fit_ab_RNA, contrast = contrast)
topTags(res_DA_RNA)
res_DA_RNA$table$celltype <- row.names(res_DA_RNA$table)
#writexl::write_xlsx(res_DA_RNA$table, paste0(tab_path, "Differential_abundance_all_integrated_rnm_res0.8.xlsx"))

```


##Expression of complement pathway genes
```{r}
#Source list of complement genes: Ricklin D, et al.  Nat Immunol. 2010 Sep;11(9):785-97; https://pubmed.ncbi.nlm.nih.gov/20720586/

#Human to mouse gene symbol conversion : http://www.informatics.jax.org/batch/summary

#1. Pattern recognition complement genes:
#C1qa = C1qa; C1qb = C1qb; MBL = Mbl2; Ficolin-2 = Fcnb; Properdin = Cfp; CRP = Crp; CFHR4 = Cfhr1
genesComplement_PatternRecog <- c("C1qa", "C1qb", "Cfp", "Fcnb") #, Mbl2, Crp and Cfhr1 are not present in the dataset, so excluded

#2. Complement proteases:
#C1r = C1ra; C1r = C1rb; C1s = C1s1; C1s = C1s2; MASP1 = Masp1; MASP2 = Masp2; C2 = C2; CFB = Cfb; CFD = Cfd; CFI = Cfi
genesComplement_Proteases <- c("C1ra", "C1s1", "Masp2", "C2", "Cfb") #C1rb, C1s2, Masp1, Cfd and Cfi are not present in the dataset, so excluded


#3. Complement components:
#C3 = C3; C4 = C4a; C4 = C4b; C5 = Hc; C6 = C6; C7 = C7; C8A = C8a; C8B = C8b; C8G =C8g; C9 = C9
genesComplement_Components <- c("C3", "C4a", "C4b", "Hc", "C8g") #C6, C7, C8a, C8b and C9 are not present in the dataset, so excluded

#4. Complement receptors:
#CR1 = Cr1l; CR2 = Cr2; CR3 = Itgam + Itgb2; CR4 = Itgax + Itgb2; C3aR = C3ar1; C5aR = C5ar1; C5L2 = C5ar2, CRIg = Vsig4; cC1qR = Calr; C1qbp = C1qbp; C1qRp = Cd93
genesComplement_Receptors <- c("Cr1l", "Cr2", "Itgam", "Itgax", "Itgb2", "C3ar1", "C5ar1", "C5ar2", "Calr", "C1qbp", "Cd93") #"Vsig4" not present in the dataset, so excluded

#5. Complement regulators:
#SERPING1 = Serping1; sMAP = Masp2; MAP1 = Masp1; C4BP = C4bp; CFH = Cfh; CFHL1 = Cfhr1; MCP = Cd46; DAF = Cd55; CD59 = Cd59a; CD59 = Cd59b; Vitronectin = Vtn; Clusterin = Clu; Carboxypeptidase-N = Cpn1; Carboxypeptidase-N = Cpn2 

genesComplement_Regulators <- c("Serping1", "Masp2", "Cfh", "Cd46", "Cd55", "Cd59a", "Clu", "Cd59b", "Cpn1") #, Masp1, C4bp, Cfhr1, Vtn, Cpn2 and Cpb2 are not present in the dataset, so excluded

#Figures
#Violin Plot
#pdf(paste0(fig_path, "/res0.8_ViolinPlot_ComplementPathwayGenes_all_integrated.pdf"), width = 9, height = 6 )
title1 <- "Pattern recognition genes"
grid::grid.newpage()
grid::grid.text(title1)
for (gene in genesComplement_PatternRecog){
  VlnPlot_PatternRecog <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_PatternRecog) 
}
title2 <- "Proteases"
grid::grid.newpage()
grid::grid.text(title2)
for (gene in genesComplement_Proteases){
  VlnPlot_Proteases <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_Proteases)
}
title3 <- "Complement components"
grid::grid.newpage()
grid::grid.text(title3)
for (gene in genesComplement_Components){
  VlnPlot_Components <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_Components)
}
title4 <- "Receptors"
grid::grid.newpage()
grid::grid.text(title4)
for (gene in genesComplement_Receptors){
  VlnPlot_Receptors <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_Receptors)
 }
title5 <- "Regulators"
grid::grid.newpage()
grid::grid.text(title5)
for (gene in genesComplement_Regulators){
  VlnPlot_Regulators <- VlnPlot(all_integrated_rnm_VlnPlot, features = gene, pt.size = 0.1, assay = "RNA", split.by = "group")  & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
  print(VlnPlot_Regulators)
 }
##dev.off()

#Dotplots
#pdf(paste0(fig_path, "/res0.8_DotPlot_ComplementPathwayGenes_all_integrated.pdf"), width = 8, height = 10 )
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_PatternRecog, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) + labs(title = "Pattern recognition genes\n")
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_Proteases, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) + labs(title = "Proteases\n")
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_Components, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) + labs(title = "Complement components\n")
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_Receptors, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) +  labs(title = "Receptors\n")
DotPlot(all_integrated_rnm_DotPlot, features =genesComplement_Regulators, assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank()) + labs(title = "Regulators\n")
##dev.off()

#DotPlot(all_integrated_rnm_DotPlot, features =c("C1qa", "C1qb", "Cfp", "C1ra", "C1s1", "Masp2", "C2", "Cfb", "C3", "C4a", "C4b", "Hc", "C8g", "Cr1l", "Cr2", "Itgam", "Itgax", "Itgb2", "C3ar1", "C5ar1", "C5ar2", "Calr", "C1qbp", "Cd93", "Vsig4", "Serping1", "Cfh", "Cd46", "Cd55", "Cd59a"), assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5))

#pdf(paste0(fig_path, "/res0.8_GroupedViolinPlot_ComplementPathwayGenes_all_integrated.pdf"), width = 25, height = 10 )
VlnPlot(all_integrated_rnm_VlnPlot, c(genesComplement_PatternRecog, genesComplement_Proteases, genesComplement_Components, genesComplement_Receptors, genesComplement_Regulators), stack = TRUE, same.y.lims = TRUE, split.by = "group", assay = "RNA") & ylab("")
##dev.off()

```



##Visualizing expression of Depdc7, Leng8, Ccdc9, Tmco1, Gnb2, Pacc1 (Tmem206), Myadm, Gapdh, Tlr4

```{r}
#Visualizing expression of Visualizing expression of Depdc7, Leng8, Ccdc9, Tmco1, Gnb2, Pacc1 (Tmem206), Myadm, Gapdh, Tlr4 in different clusters in the seurat object with renamed identities

#Violin plots: expression levels represent log normalized values

#pdf(paste0(fig_path, "/res1.1_ViolinPlot_11genes_all_integrated.pdf"), width = 8, height = 6 )

VlnPlot(all_integrated_rnm_VlnPlot, features = c("Depdc7"), pt.size = 0.1, assay = "RNA", split.by = "group") &  theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Leng8"), pt.size = 0.1, assay = "RNA", split.by = "group") &  theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Ccdc9"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Tmco1"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Gnb2"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Pacc1"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Myadm"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Gapdh"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Tlr4"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Il27ra"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Il6st"), pt.size = 0.1, assay = "RNA", split.by = "group") &   theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))

##dev.off()

#pdf(paste0(fig_path, "/res1.1_DotPlot_11genes_all_integrated.pdf"), width = 10, height = 12 )
DotPlot(all_integrated_rnm_DotPlot, features = c("Depdc7", "Leng8", "Ccdc9", "Tmco1", "Gnb2",  "Pacc1", "Myadm", "Gapdh", "Tlr4", "Il27ra", "Il6st"), assay = "RNA",  split.by = "group", cols = c("red", "grey")) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.y = element_blank())
##dev.off()

```


##Expression of Miwi2 and Saravanan\'s research genes
```{r}
#1.	Piwil4 = MIWI2 
#pdf(paste0(fig_path, "/res0.8_DotPlot_MIWI2_all_integrated.pdf"), width = 10, height = 10)
DotPlot(all_integrated_rnm_DotPlot, features = c("Piwil4"), assay = "RNA",  split.by = "group", cols = c("red", "grey")) + theme(axis.title.x = element_blank(), axis.title.y = element_blank())
##dev.off()

#pdf(paste0(fig_path, "/res0.8_ViolinPlot_MIWI2_all_integrated.pdf"), width = 8, height = 6 )
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Piwil4"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(),axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
##dev.off()

#2. F2r = PAR1; F2rl1 = PAR2; F3 = Tissue factor; F10 = coagulation factor X; F7 = coagulation factor VII; Procr = EPCR 
#pdf(paste0(fig_path, "/res0.8_DotPlot_additnlGenes_Saravanan_Research_all_integrated.pdf"), width = 10, height = 10)
DotPlot(all_integrated_rnm_DotPlot, features = c("F2r", "F2rl1", "F3", "F10", "F7", "Procr"), assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.title.x = element_blank(), axis.title.y = element_blank())
##dev.off()



#pdf(paste0(fig_path, "/res0.8_ViolinPlot_additnlGenes_Saravanan_Research_all_integrated.pdf"), width = 8, height = 6 )

VlnPlot(all_integrated_rnm_VlnPlot, features = c("F2r"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("F2rl1"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("F3"), pt.size = 0.1, assay = "RNA", split.by = "group") & NoLegend() & NoLegend() & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("F10"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("F7"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Procr"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5),plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
##dev.off()
```

#Mfsd1 expression
```{r}
#1.	Mfsd1
#pdf(paste0(fig_path, "/res0.8_DotPlot_Mfsd1_all_integrated.pdf"), width = 10, height = 10)
DotPlot(all_integrated_rnm_DotPlot, features = c("Mfsd1"), assay = "RNA", split.by = "group", cols = c("red", "grey")) + theme(axis.title.x = element_blank(), axis.title.y = element_blank())
##dev.off()

#pdf(paste0(fig_path, "/res0.8_ViolinPlot_Mfsd1_all_integrated.pdf"), width = 8, height = 6 )
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Mfsd1"), pt.size = 0.1, assay = "RNA", split.by = "group") & theme(axis.title.x = element_blank(),axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
##dev.off()
```



## Statistical difference in Ifng expression across all clusters betweeen groups
### Ifng: Between il27raKO and wt
```{r}

#Finding differential expression of Ifng across clusters between il27raKO vs wt (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Ifng differential expression in clusters between il27raKO and wt
#RNA
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", features = "Ifng", logfc.threshold = 0, min.pct = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_Ifng_RNA_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_Ifng_RNA_all_integrated_rnm, paste0(tab_path, "/differential_Ifng_all_integrated_res0.8.xlsx"))


```


# IL27RA RNA expression
```{r}
Idents(all_integrated_rnm) <- "celltype"
#pdf(paste0(fig_path, "/Il27ra_RNAexpression_byCelltype_res0.8_all_integrated.pdf"), height = 6, width = 10)
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Il27ra"), split.by = "group", pt.size = 0.1, assay = "RNA", ) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, vjust = 0.5), plot.margin = unit(c(0.1,0.1,0.1,0.1), "in"))
#dev.off()

```


# generate heatmaps for GO BP "(cellular) response to Interferon-gamma" and "cytokine-mediated signaling pathway" based on inputs from Johannes
## Heatmaps from GO analysis: 
```{r}
#Seurat-enrichr GO BP data
#assign the GO BP results from Seurat-enrichr to a temporary variable
library(data.table)
tmp <- lapply(enrichr_gobp_all_integrated_rnm, rbindlist, idcol = TRUE)
tmp <- bind_rows(tmp, .id = "celltype")

#extract genes from all clusters for GO BP: "cellular response to interferon-gamma (GO:0071346)"
genes_ifng_response <- tmp$GO_Biological_Process_2021.Genes[grepl("(GO:0071346)", tmp$GO_Biological_Process_2021.Term)]
genes_ifng_response <- str_to_title(unique(unlist(strsplit(genes_ifng_response, split=';', fixed=TRUE) ) ) )
#length(genes_ifng_response) = 25

#pdf(paste0(fig_path, "/Heatmap_Seu_Enrichr_Resp_IFNgamma_pathway_res0.8_all_integrated.pdf"), height = 12, width = 25)
DoHeatmap(subset(all_integrated_rnm, downsample = 100), features = genes_ifng_response, assay = "RNA", angle = 90, size = 3, group.by = "cluster0.8.group")+ scale_fill_gradient2(midpoint = 0, low = "navy", mid = "white",  high = "red", na.value = "white") + theme(axis.text.y.left = element_text(colour = "black")) + guides(color = "none")
##dev.off()

#extract genes from all clusters for GO BP: "cytokine-mediated signaling pathway (GO:0019221)"
genes_cytokine <- tmp$GO_Biological_Process_2021.Genes[grepl("(GO:0019221)", tmp$GO_Biological_Process_2021.Term)]
genes_cytokine <- str_to_title(unique(unlist(strsplit(genes_cytokine, split=';', fixed=TRUE) ) ) )
#length(genes_cytokine) = 114

#pdf(paste0(fig_path, "/Heatmap_Seu_Enrichr_CytokineSignal_res0.8_all_integrated.pdf"), height = 15, width = 25)
DoHeatmap(subset(all_integrated_rnm, downsample = 100), features = genes_cytokine, assay = "RNA", angle = 90, size = 3, group.by = "cluster0.8.group")+ scale_fill_gradient2(midpoint = 0, low = "navy", mid = "white",  high = "red", na.value = "white") + theme(axis.text.y.left = element_text(colour = "black")) + guides(color = "none")
##dev.off()


#pseudobulk heatmap visualization
#Cellular response to IFNgamma #run muscatDS again to set res_sce to RNA data
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifng_response, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 25, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm") #adjusting the heatmap cell size
tmp1@matrix_param$height <- unit(805, "mm") #adjusting the heatmap cell size
#pdf(paste0(fig_path, "/Heatmap_MuscatPB_Resp_IFNgamma_pathway_res0.8_all_integrated.pdf"), height = 40, width = 10)
tmp1
##dev.off()

#Cytokine-mediated signaling pathway
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_cytokine, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 114, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm")
tmp1@matrix_param$height <- unit(4055, "mm")
#pdf(paste0(fig_path, "/Heatmap_MuscatPB_CytokineSignal_res0.8_all_integrated.pdf"), height = 65, width = 10)
tmp1
##dev.off()


#TopGO data: Seurat DEGs:
tmp <- bind_rows(topgoDE_cluster, .id = "celltype")

#extract genes from all clusters for GO BP: "cellular response to interferon-gamma (GO:0071346)"
genes_ifng_response <- tmp$genes[grepl("(GO:0071346)", tmp$GO.ID)]
genes_ifng_response <- unique(unlist(strsplit(genes_ifng_response, split=',', fixed=TRUE) ) )

#pdf(paste0(fig_path, "/Heatmap_TopGO_Seurat_Resp_IFNgamma_pathway_res0.8_all_integrated.pdf"), height = 12, width = 25)
DoHeatmap(subset(all_integrated_rnm, downsample = 100), features = genes_ifng_response, assay = "RNA", angle = 90, size = 3, group.by = "cluster0.8.group")+scale_fill_gradient2(midpoint = 0, low = "navy", mid = "white",  high = "red", na.value = "white") + theme(axis.text.y.left = element_text(colour = "black")) + guides(color = "none")
##dev.off()


#extract genes from all clusters for GO BP: "cytokine-mediated signaling pathway (GO:0019221)"
genes_cytokine <- tmp$genes[grepl("(GO:0019221)", tmp$GO.ID)]
genes_cytokine <- unique(unlist(strsplit(genes_cytokine, split=',', fixed=TRUE) ) )

#pdf(paste0(fig_path, "/Heatmap_TopGO_Seurat_CytokineSignal_res0.8_all_integrated.pdf"), height = 15, width = 25)
DoHeatmap(subset(all_integrated_rnm, downsample = 100), features = genes_cytokine, assay = "RNA", angle = 90, size = 3, group.by = "cluster0.8.group")+ scale_fill_gradient2(midpoint = 0, low = "navy", mid = "white",  high = "red", na.value = "white") + theme(axis.text.y.left = element_text(colour = "black")) + guides(color = "none")
##dev.off()


#pseudobulk heatmap visualization
#Cellular response to IFNgamma
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifng_response, fdr = 1, lfc= 0, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm") #adjusting the heatmap cell size
tmp1@matrix_param$height <- unit(405, "mm") #adjusting the heatmap cell size
#pdf(paste0(fig_path, "/Heatmap_MuscatPB_TopGO_Seurat_Resp_IFNgamma_pathway_res0.8.pdf"), height = 20, width = 10)
tmp1
##dev.off()

#Cytokine-mediated signaling pathway
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_cytokine, fdr = 1, lfc= 0, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm")
tmp1@matrix_param$height <- unit(555, "mm")
#pdf(paste0(fig_path, "/Heatmap_MuscatPB_TopGO_Seurat_CytokineSignal_res0.8.pdf"), height = 25, width = 10)
tmp1
##dev.off()


#TopGO data: MuscatDS DEGs:
tmp <- bind_rows(topgoDE_cluster_muscatDS, .id = "celltype")

#extract genes from all clusters for GO BP: "cellular response to interferon-gamma (GO:0071346)"
genes_ifng_response <- tmp$genes[grepl("(GO:0071346)", tmp$GO.ID)]
genes_ifng_response <- unique(unlist(strsplit(genes_ifng_response, split=',', fixed=TRUE) ) )

#extract genes from all clusters for GO BP: "cytokine-mediated signaling pathway (GO:0019221)"
genes_cytokine <- tmp$genes[grepl("(GO:0019221)", tmp$GO.ID)]
genes_cytokine <- unique(unlist(strsplit(genes_cytokine, split=',', fixed=TRUE) ) )


#pseudobulk heatmap visualization
#Cellular response to IFNgamma
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifng_response, fdr = 1, lfc= 0, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm") #adjusting the heatmap cell size
tmp1@matrix_param$height <- unit(455, "mm") #adjusting the heatmap cell size
#pdf(paste0(fig_path, "/Heatmap_MuscatDS_PB_TopGO_Resp_IFNgamma_pathway_res0.8.pdf"), height = 20, width = 10)
tmp1
##dev.off()

#Cytokine-mediated signaling pathway
tmp1 <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_cytokine, fdr = 1, lfc= 0, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp1@matrix_param$width <- unit(25, "mm")
tmp1@matrix_param$height <- unit(655, "mm")
#pdf(paste0(fig_path, "/Heatmap_MuscatDS_PB_TopGO_CytokineSignal_res0.8.pdf"), height = 30, width = 10)
tmp1
##dev.off()

```
## Heatmaps from JAX GO terms:
```{r}
#The genes mapped in the Gene Ontology database for cellular response to interferon-gamma (GO:0071346; http://www.informatics.jax.org/go/term/GO:0071346) and  cytokine-mediated signaling pathway (GO:0019221; http://www.informatics.jax.org/go/term/GO:0019221 ) were retrieved (7 Feb 2022) and a heatmap was plotted using the muscatDS pseudobulk heatmap function
library(readxl)
tmp <- readxl::read_excel("GO_JAX_tables_input/GO_term_summary_20220207_152905_GO0071346_CellularResp_Ifngamma_JAX.xlsx")

genes_ifng_response_Jax <- unique(tmp$Symbol)

genes_ifng <- subset(genes_ifng_response_Jax, genes_ifng_response_Jax %in% row.names(tmp_sce)) #subset to keep only those genes from the JAX database response to IFNgamma GO that are expressed in the current dataset 
#length(genes_tmp) = 113
htmap_plots <- list()


for(i in names(tbl_fil_all_RNA)){
  tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifng, k = i, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 113, col = colorRampPalette(c("blue", "white", "red"))(50))
  tmp_htmap@column_title <- paste0("Heatmap of genes from the\nJAX GO BP response to Ifn gamma pathway in ", i, "\n")
  tmp_htmap@matrix_param$width <- unit(35, "mm")
  tmp_htmap@matrix_param$height <- unit(405, "mm")
  tmp_htmap@row_names_param$gp[[1]] <- 9
  htmap_plots[[i]] <- tmp_htmap
}

#pdf(paste0(fig_path, "/Heatmap_PB_JAX_GO_RespIFNgamma_res0.8.pdf"), height = 25, width = 10)
for(i in names(tbl_fil_all_RNA)){
  print(htmap_plots[[i]])
}

##dev.off()

rm(tmp, genes_tmp,genes_ifng, htmap_plots, tmp_htmap)


#Cytokine-mediated signaling
tmp <- readxl::read_excel("GO_JAX_tables_input/GO_term_summary_20220207_153037_GO0019221_cytokineSignal_JAX.xlsx")

genes_cytokineSignal_Jax <- unique(tmp$Symbol)

genes_cytok <- subset(genes_cytokineSignal_Jax, genes_cytokineSignal_Jax %in% row.names(tmp_sce)) #subset to keep only those genes from the JAX database Cytokine mediated signaling GO that are expressed in the current dataset 
#length(genes_tmp) = 325

htmap_plots <- list()


for(i in names(tbl_fil_all_RNA)){
  tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_cytok, k = i, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 325, col = colorRampPalette(c("blue", "white", "red"))(50))
  tmp_htmap@column_title <- paste0("Heatmap of genes from the\nJAX GO BP cytokine-mediated signaling pathway in ", i, "\n")
  tmp_htmap@matrix_param$width <- unit(35, "mm")
  tmp_htmap@matrix_param$height <- unit(805, "mm")
  tmp_htmap@row_names_param$gp[[1]] <- 8
  htmap_plots[[i]] <- tmp_htmap
}

#pdf(paste0(fig_path, "/Heatmap_PB_JAX_GO_CytokineSignal_res0.8.pdf"), height = 40, width = 10)
for(i in names(tbl_fil_all_RNA)){
  print(htmap_plots[[i]])
}

##dev.off()


```

#generate heatmap for GO BP "cellular response to type I interferon (GO:0071357)" based on inputs from Johannes
## Heatmaps from JAX GO terms:
```{r}
#The genes mapped in the Gene Ontology database for cellular response to type I interferon (GO:0071357; http://www.informatics.jax.org/go/term/GO:0071357) was retrieved (8 Feb 2022) and a heatmap was plotted using the muscatDS pseudobulk heatmap function
library(readxl)
tmp <- readxl::read_excel("GO_JAX_tables_input/GO_term_summary_20220208_092916_respTypeI_IFN_GO0071357_JAX.xlsx")

genes_type1_ifn_response_Jax <- unique(tmp$Symbol)

genes_ifn1 <- subset(genes_type1_ifn_response_Jax, genes_type1_ifn_response_Jax %in% row.names(tmp_sce)) #subset to keep only those genes from the JAX database type 1 response to interferon GO that are expressed in the current dataset 
#length(genes_tmp) = 46
htmap_plots <- list()


for(i in names(tbl_fil_all_RNA)){
  tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifn1, k = i, fdr = 1, lfc= 0, sort_by = "p_adj.loc", decreasing = TRUE, top_n = 46, col = colorRampPalette(c("blue", "white", "red"))(50))
  tmp_htmap@column_title <- paste0("Heatmap of genes from the\nJAX GO BP cellular response to type I interferon in ", i, "\n")
  if(nrow(tmp_htmap@matrix) > 10){
  tmp_htmap@matrix_param$width <- unit(35, "mm")
  tmp_htmap@matrix_param$height <- unit(255, "mm")
  tmp_htmap@row_names_param$gp[[1]] <- 8
  }
  htmap_plots[[i]] <- tmp_htmap
}

#pdf(paste0(fig_path, "/Heatmap_PB_JAX_GO_type1_IFNresp_res0.8.pdf"), height = 20, width = 10)
for(i in names(tbl_fil_all_RNA)){
  print(htmap_plots[[i]])
}

##dev.off()

#trying heatmap without arranging genes by p_adj.loc values to see if a pattern can be observed
htmap_plots_noOrder <- list()
for(i in names(tbl_fil_all_RNA)){
  tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, g = genes_ifn1, k = i, fdr = 1, lfc= 0, top_n = 46, col = colorRampPalette(c("blue", "white", "red"))(50))
  tmp_htmap@column_title <- paste0("Heatmap of genes from the\nJAX GO BP cellular response to type I interferon in ", i, "\n")
  if(nrow(tmp_htmap@matrix) > 10){
  tmp_htmap@matrix_param$width <- unit(35, "mm")
  tmp_htmap@matrix_param$height <- unit(105, "mm")
  tmp_htmap@row_names_param$gp[[1]] <- 8
  }
  htmap_plots_noOrder[[i]] <- tmp_htmap
}

#pdf(paste0(fig_path, "/Heatmap_NoOrd_PB_JAX_GO_type1_IFNresp_res0.8.pdf"), height = 15, width = 10)
for(i in names(tbl_fil_all_RNA)){
  print(htmap_plots_noOrder[[i]])
}

##dev.off()

rm(tmp, genes_tmp, htmap_plots, tmp_htmap)


```

## Plotting averageexp heatmap: for GO BP "cellular response to type I interferon (GO:0071357)"
```{r}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm #assign seurat object to a temporary variable

#pdf(paste0(fig_path, "AvgExpr_Heatmap_type1IFN.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_type1_ifn <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_type1_ifn) <- c("il27raKO", "wt")
  cluster.averages_type1_ifn <- ScaleData(cluster.averages_type1_ifn)
 print(Seurat::DoHeatmap(cluster.averages_type1_ifn, features = genes_ifn1, size = 3, draw.lines = FALSE, angle = 90) + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 6.5, colour = "black", margin = margin(r=-15))) + ggtitle(paste0("Heatmap of average expression of type I interferon response genes in\n",i, "\n") ) )
}
##dev.off()


#Plotting with dittoHeatmap. Some genes that have zero values will not be plotted.
#Adding metadata to variable storing the average expression values to enable annotation of group (Source code: https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/AddMetaData) 
tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_type1IFN.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_type1_ifn <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_type1_ifn) <- c("il27raKO", "wt")
  cluster.averages_type1_ifn <- ScaleData(cluster.averages_type1_ifn)
  group <- Idents(cluster.averages_type1_ifn)
  names(group) <- colnames(cluster.averages_type1_ifn)
  cluster.averages_type1_ifn <- AddMetaData(object = cluster.averages_type1_ifn, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_type1_ifn, genes = genes_ifn1, annot.by = "group", main = paste0("Heatmap of average expression of type I interferon response genes in\n",i, "\n") )) 
}
##dev.off()
```

## Plotting averageexp heatmap: for GO BP "cellular response to interferon gamma (GO:0071346)"
```{r}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm #assign seurat object to a temporary variable

#pdf(paste0(fig_path, "AvgExpr_Heatmap_IFNgamma.pdf"), height = 15)
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages) <- c("il27raKO", "wt")
  cluster.averages <- ScaleData(cluster.averages)
 print(Seurat::DoHeatmap(cluster.averages, features = genes_ifng, size = 3, draw.lines = FALSE, angle = 90) + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 6.5, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10)) + ggtitle(paste0("Heatmap of average expression of\ncellular response to interferon-gamma genes in\n",i, "\n") ) )
}
##dev.off()


#Plotting with dittoHeatmap. Some genes that have zero values will not be plotted.
#Adding metadata to variable storing the average expression values to enable annotation of group (Source code: https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/AddMetaData) 
tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_IFNgamma.pdf"), height = 15)
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages) <- c("il27raKO", "wt")
  cluster.averages <- ScaleData(cluster.averages)
  group <- Idents(cluster.averages)
  names(group) <- colnames(cluster.averages)
  cluster.averages <- AddMetaData(object = cluster.averages, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages, genes = genes_ifng, annot.by = "group", main = paste0("Heatmap of average expression of\ncellular response to interferon gamma genes in\n",i, "\n") )) 
}
##dev.off()
```


## Plotting averageexp heatmap: for GO BP "cytokine-mediated signaling (GO:0019221)"
```{r}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm #assign seurat object to a temporary variable

#pdf(paste0(fig_path, "AvgExpr_Heatmap_CytokineSignal.pdf"), height = 25)
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages) <- c("il27raKO", "wt")
  cluster.averages <- ScaleData(cluster.averages)
 print(Seurat::DoHeatmap(cluster.averages, features = genes_cytok, size = 3, draw.lines = FALSE, angle = 90) + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 6, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10)) + ggtitle(paste0("Heatmap of average expression of cytokine-mediated signaling genes in\n",i, "\n") ) )
}
##dev.off()


#Plotting with dittoHeatmap. Some genes that have zero values will not be plotted.
#Adding metadata to variable storing the average expression values to enable annotation of group (Source code: https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/AddMetaData) 
tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_CytokineSignal.pdf"), height = 30)
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages) <- c("il27raKO", "wt")
  cluster.averages <- ScaleData(cluster.averages)
  group <- Idents(cluster.averages)
  names(group) <- colnames(cluster.averages)
  cluster.averages <- AddMetaData(object = cluster.averages, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages, genes = genes_cytok, annot.by = "group", fontsize = 8, main = paste0("Heatmap of average expression of\ncytokine-mediated signaling genes in\n",i, "\n") )) 
}
##dev.off()
```



## Plotting averageexp heatmap from the gene list for GO BP "cellular response to type I interferon (GO:0071357)" only for genes that are statistically significant (Adjusted p < 0.05)
```{r}
#Plotting with dittoHeatmap. Some genes that have zero values will not be plotted.
#Plotting averageexp heatmap from the gene list for GO BP "cellular response to type I interferon (GO:0071357)" only for genes that are statistically significant (Adjusted p < 0.05)
#Adding metadata to variable storing the average expression values to enable annotation of group (Source code: https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/AddMetaData) 
tmp <- all_integrated_rnm

#subset differentially expressed genes list to include only genes that are statistically significant
pgenes <- subset(differential_genes_all_integrated_rnm, p_val_adj < 0.05)

#subset list of type I interferon genes to only include genes that are statistically significant (Adjusted p < 0.05)
genes_ifn1_plot <- subset(genes_ifn1, genes_ifn1 %in% pgenes$gene)

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_type1IFN_signifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_tmp <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_tmp) <- c("il27raKO", "wt")
  cluster.averages_tmp <- ScaleData(cluster.averages_tmp)
  group <- Idents(cluster.averages_tmp)
  names(group) <- colnames(cluster.averages_tmp)
  cluster.averages_tmp <- AddMetaData(object = cluster.averages_tmp, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_tmp, genes = genes_ifn1_plot, annot.by = "group", main = paste0("Heatmap of average expression of statistically significant (adj p < 0.05)\ntype I interferon response genes in\n",i, "\n") )) 
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_type1IFN_signifGenes.pdf"), width = 8)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_ifn1_plot, split.by = "group", stack = TRUE, assay = "RNA")
##dev.off()
```


## Plotting averageexp heatmap from the gene list for GO BP "cellular response to interferon gamma (GO:0071346)" only for genes that are statistically significant (Adjusted p < 0.05)
```{r}
#subset list of cellular response to interferon-gamma genes (GO:0071346) to only include genes that are statistically significant (Adjusted p < 0.05)
genes_ifng_plot <- subset(genes_ifng, genes_ifng %in% pgenes$gene)

tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_IFNgamma_signifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_tmp <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_tmp) <- c("il27raKO", "wt")
  cluster.averages_tmp <- ScaleData(cluster.averages_tmp)
  group <- Idents(cluster.averages_tmp)
  names(group) <- colnames(cluster.averages_tmp)
  cluster.averages_tmp <- AddMetaData(object = cluster.averages_tmp, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_tmp, genes = genes_ifng_plot, annot.by = "group", main = paste0("  Heatmap of average expression of statistically significant (adj p < 0.05)\ncellular response to interferon-gamma genes in\n",i, "\n") )) 
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_IFNgamma_signifGenes.pdf"), width = 8)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_ifng_plot, split.by = "group", stack = TRUE, assay = "RNA")
##dev.off()
```

## Plotting averageexp heatmap from the gene list for GO BP cytokine-mediated signaling pathway (GO:0019221)" only for genes that are statistically significant (Adjusted p < 0.05)
```{r}
#subset list of cellular response to cytokine-mediated signaling pathway (GO:0019221) genes to only include genes that are statistically significant (Adjusted p < 0.05)
genes_cytokineSig_plot <- subset(genes_cytok, genes_cytok %in% pgenes$gene)

tmp <- all_integrated_rnm

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_CytokineSignal_signifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_type1_ifn <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_type1_ifn) <- c("il27raKO", "wt")
  cluster.averages_type1_ifn <- ScaleData(cluster.averages_type1_ifn)
  group <- Idents(cluster.averages_type1_ifn)
  names(group) <- colnames(cluster.averages_type1_ifn)
  cluster.averages_type1_ifn <- AddMetaData(object = cluster.averages_type1_ifn, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_type1_ifn, genes = genes_cytokineSig_plot, annot.by = "group", main = paste0("  Heatmap of average expression of statistically significant (adj p < 0.05)\ncytokine-mediated signaling pathway genes in\n",i, "\n") )) 
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_CytokineSignal_signifGenes.pdf"), width = 12)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_cytokineSig_plot, split.by = "group", stack = TRUE, assay = "RNA")
##dev.off()
```


## Dotplots of significant ifn gamma response, cytokine signaling, type I ifn response genes
```{r}
# Separate dot plot for each cluster
tmp <- all_integrated_rnm_DotPlot
tmp_cell_list <-unique(rev(levels(tmp)))#specifying reverse to ensure order of cell types

# type I interferon response
dotplot_ifn1 <- list()
for(i in tmp_cell_list){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  dotplot_ifn1[[i]] <- DotPlot(tmp_clus, genes_ifn1_plot, assay = "RNA", dot.scale = 6) + scale_color_gradientn(colors = c("navy", "white", "red")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), plot.title = element_text(size = 10), axis.title = element_blank()) + ggtitle(paste0("Dotplot showing expression of statistically significant (adj p < 0.05)\ntype I interferon response genes in\n",i, "\n"))
}


#pdf(paste0(fig_path, "Dotplot_type1IFN_signifGenes.pdf"), width = 12, height = 12)
for(i in seq(1,length(dotplot_ifn1), by = 6) ){
  print(wrap_plots(dotplot_ifn1[[i]], dotplot_ifn1[[i+1]], dotplot_ifn1[[i+2]], dotplot_ifn1[[i+3]], dotplot_ifn1[[i+4]], dotplot_ifn1[[i+5]], ncol = 2, nrow = 3))
    }
##dev.off()

# cellular response to interferon-gamma
dotplot_ifngamma <- list()
for(i in tmp_cell_list){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  dotplot_ifngamma[[i]] <- DotPlot(tmp_clus, genes_ifng_plot, assay = "RNA", dot.scale = 6) + scale_color_gradientn(colors = c("navy", "white", "red")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), plot.title = element_text(size = 10), axis.title = element_blank()) + ggtitle(paste0("Dotplot showing expression of statistically significant (adj p < 0.05)\ncellular response to interferon-gamma genes in\n",i, "\n"))
}


#pdf(paste0(fig_path, "Dotplot_resp_IFNgamma_signifGenes.pdf"), width = 12, height = 12)
for(i in seq(1,length(dotplot_ifngamma), by = 6) ){
  print(wrap_plots(dotplot_ifngamma[[i]], dotplot_ifngamma[[i+1]], dotplot_ifngamma[[i+2]], dotplot_ifngamma[[i+3]], dotplot_ifngamma[[i+4]], dotplot_ifngamma[[i+5]], ncol = 2, nrow = 3))
    }
##dev.off()


# cytokine mediated signaling pathway
dotplot_cytokineSig <- list()
for(i in tmp_cell_list){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  dotplot_cytokineSig[[i]] <- DotPlot(tmp_clus, genes_cytokineSig_plot, assay = "RNA", dot.scale = 6) + scale_color_gradientn(colors = c("navy", "white", "red")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), plot.title = element_text(size = 10), axis.title = element_blank()) + ggtitle(paste0("Dotplot showing expression of statistically significant (adj p < 0.05)\ncytokine-mediated signaling genes in\n",i, "\n"))
}


#pdf(paste0(fig_path, "Dotplot_resp_CytokineSignal_signifGenes.pdf"), width = 15, height = 12)
for(i in seq(1,length(dotplot_cytokineSig), by = 6) ){
  print(wrap_plots(dotplot_cytokineSig[[i]], dotplot_cytokineSig[[i+1]], dotplot_cytokineSig[[i+2]], dotplot_cytokineSig[[i+3]], dotplot_cytokineSig[[i+4]], dotplot_cytokineSig[[i+5]], ncol = 2, nrow = 3))
    }
##dev.off()

```


```{r}
#VlnPlot(all_integrated_rnm, c("Prdm1","Maf", "Il27", "Il10", "Il1b", "Il18"), assay = "RNA", pt.size = 0, split.by = "group", stack = TRUE) #(Reference: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7771052/)

tmp <- all_integrated_rnm
DefaultAssay(tmp) <- "RNA"
#pdf(paste0(fig_path, "ModuleScore_GOdb_topGO_signifGenes.pdf"), width = 12, height = 8)
tmp <- AddModuleScore(tmp, features = genes_ifn1_plot, name = "type1_IFN")
FeaturePlot(tmp, features = "type1_IFN1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))

tmp <- AddModuleScore(tmp, features = genes_ifng_plot, name = "IFN_gamma_resp")
FeaturePlot(tmp, features = "IFN_gamma_resp1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))

tmp <- AddModuleScore(tmp, features = genes_cytokineSig_plot, name = "cytokine_Signal")
FeaturePlot(tmp, features = "cytokine_Signal1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))

#FeaturePlot(tmp, features = c("type1_IFN1", "IFN_gamma_resp1"), label = TRUE, repel = TRUE, split.by = "group", order=TRUE, blend = TRUE)

genes_pyroptosis <- c("Gzmb", "Zbp1", "Casp4",  "Casp1")
tmp <- AddModuleScore(tmp, features = genes_pyroptosis, name = "pyroptosis")
FeaturePlot(tmp, features = "pyroptosis1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))


xt <- bind_rows(topgoDE_cluster, .id = "celltype")
genes_reg_immuneResp <-  xt$genes[grepl("(GO:0050776)", xt$GO.ID)]
genes_reg_immuneResp <- unique(unlist(strsplit(genes_reg_immuneResp, split=',', fixed=TRUE) ) )
tmp <- AddModuleScore(tmp, features = genes_reg_immuneResp, name = "immune_resp_reg")
FeaturePlot(tmp, features = "immune_resp_reg1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))


genes_apoptotic_proc <-  xt$genes[grepl("(GO:0006915)", xt$GO.ID)]
genes_apoptotic_proc <- unique(unlist(strsplit(genes_apoptotic_proc, split=',', fixed=TRUE) ) )
tmp <- AddModuleScore(tmp, features = genes_apoptotic_proc, name = "apoptotic_proc")
FeaturePlot(tmp, features = "apoptotic_proc1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))


genes_defense_resp <- xt$genes[grepl("(GO:0006952)", xt$GO.ID)]
genes_defense_resp <- unique(unlist(strsplit(genes_defense_resp, split=',', fixed=TRUE) ) )
tmp <- AddModuleScore(tmp, features = genes_defense_resp, name = "defense_resp")
FeaturePlot(tmp, features = "defense_resp1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))

##dev.off()

```



# Checking GNB2, IL10 and IL27-EBI3 expression based on Johannes and Prof. Bosmann's inputs
```{r}
#pdf(paste0(fig_path, "VlnPlots_Gnb2_Il10_Il27Ebi3_Il27p28.pdf"))

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Gnb2", "rna_Il10","rna_Ebi3", "rna_Il27"), pt.size = 0, stack = TRUE, same.y.lims = TRUE) & NoLegend()
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Gnb2", "rna_Il10","rna_Ebi3", "rna_Il27"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group")

##dev.off()
```

## Statistical difference in "Gnb2", "Il10", "Il27", "Ebi3" expression across all clusters betweeen groups
### "Gnb2", "Il10", "Il27", "Ebi3": Between il27raKO and wt
```{r}

#Finding differential expression of "Gnb2", "Il10", "Il27", "Ebi3" across clusters between il27raKO vs wt (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Gnb2, Il10, Il27, Ebi3 differential expression in clusters between il27raKO and wt
#RNA
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", features = c("Gnb2", "Il10", "Il27", "Ebi3"), logfc.threshold = 0, min.pct = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_4genes_RNA_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_4genes_RNA_all_integrated_rnm, paste0(tab_path, "/differential_Gnb2_Il10_Il27_Ebi3_all_integrated_res0.8.xlsx"))

rm(differential_4genes_RNA_all_integrated_rnm, tmp_celltype_list, get_differential_genes)

#Finding differential expression of "Gnb2", "Il10", "Il27", "Ebi3" between clusters 
#Gnb2, Il10, Il27, Ebi3 differential expression in clusters between il27raKO and wt
#RNA
differential_4genes_RNA_btwn_clusters <- FindAllMarkers(all_integrated_rnm_VlnPlot, features = c("Gnb2", "Il10", "Il27", "Ebi3"), logfc.threshold = 0, min.pct = 0, assay = "RNA")

#writexl::write_xlsx(differential_4genes_RNA_btwn_clusters, paste0(tab_path, "/differential_Gnb2_Il10_Il27_Ebi3_betweenCellTypes_res0.8.xlsx"))


```



```{r}
#pdf(paste0(fig_path, "VlnPlots_Legionella_pathoGenes.pdf"), width = 12)

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-TLR4",  "rna_Tlr1","rna_Tlr2", "rna_Tlr3", "rna_Tlr4", "rna_Tlr5", "rna_Tlr6", "rna_Tlr7", "rna_Tlr8", "rna_Tlr9"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Irf5", "rna_Irf3", "rna_Ddx58", "rna_Ifih1", "rna_Mavs", "rna_Depdc7", "rna_Gpr65", "rna_Fpr1","adt-CD127"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank()) #https://www.frontiersin.org/articles/10.3389/fcimb.2021.733564/full
VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD127", "rna_Nlrp3", "rna_Casp1", "rna_Casp3", "rna_Scaf11", "rna_Casp4","rna_Il1b", "rna_Il18", "rna_Nlrc4", "rna_Myd88", "rna_Nod1", "rna_Nod2", "rna_Ripk2"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Ifi208", "rna_Prdx2", "rna_Wdfy1", "rna_Stat2", "rna_Irf9"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Irgm2", "rna_Usp18", "rna_Gbp3", "rna_Rtp4", "rna_Trim25", "rna_Rsad2", "rna_Bst2", "rna_Ifi27", "rna_Isg15", "rna_Fst", "rna_Xaf1", "rna_Lgals3bp", "rna_Stat1", "rna_Ifit1", "rna_Wdfy1", "rna_Spp1", "rna_C1ra", "rna_C1rb", "rna_Ifitm1", "rna_Pik3r1", "rna_Cd59a", "rna_Vcam1"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank()) #https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5392577/
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Ifng", "rna_Csf2", "rna_Ifnb1", "rna_Tnf", "rna_Il6", "rna_Il1b", "rna_Il4", "rna_Tnfsf12", "rna_Il15", "rna_Il12b", "rna_Csf1", "rna_Il18", "rna_Il13", "rna_Cxcl2", "rna_Ccl13"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4734697/
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Isg15", "rna_Ifi27", "rna_Ifit1", "rna_Ifit3", "rna_Oasl2", "rna_Rsad2", "rna_Trim25"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank()) #https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5392577/

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Stat1", "rna_Irf3", "rna_Irf7", "rna_Nfatc2", "rna_Irf5", "rna_Nfkbia", "rna_Stat4", "rna_Rela", "rna_Nfkb1", "rna_Irf1", "rna_Cebpe", "rna_Irf8", "rna_Egr1", "rna_Cebpb"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank()) #https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4734697/

##dev.off()
```


# plotting IL-24, gp130 (IL6ST) and Stat3 expression based on inputs from Johannes
```{r}
#pdf(paste0(fig_path, "VlnPlots_Il6st_Stat3.pdf"), width = 6)

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Il24", "rna_Il6st", "rna_Stat3"), pt.size = 0, stack = TRUE, same.y.lims = TRUE, split.by = "group") & theme(axis.title.y.left = element_blank())
##dev.off()


#Finding differential expression of "Il6st", "Stat3" across clusters between il27raKO vs wt (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Il6st, Stat3 differential expression in clusters between il27raKO and wt
#RNA
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", features = c("Il6st", "Stat3"), logfc.threshold = 0, min.pct = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_2genes_RNA_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

#writexl::write_xlsx(differential_2genes_RNA_all_integrated_rnm, paste0(tab_path, "/differential_Il6st_Stat3_all_integrated_res0.8.xlsx"))

rm(differential_2genes_RNA_all_integrated_rnm, tmp_celltype_list, get_differential_genes)

Idents(all_integrated_rnm) <- "celltype"

```


# Find differentially expressed genes and proteins between NK cell clusters based on inputs from Prof. Bosmann and Johannes
```{r}
#checking for genes and ADT markers between the NK cell clusters
#comparing each NK cluster with all other NK clusters

tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

#checking differential expression between neutrophil marker expressing clusters

NK_group_DE <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")){
  for(j in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")){
    if(i != j){
  RNAmarkers <- FindMarkers(tmp, ident.1 = i, ident.2 = j, only.pos = FALSE, assay = "RNA")
  RNAmarkers$gene <- row.names(RNAmarkers)
  RNAmarkers$comparison <- paste0(i, "-", j)
  NK_group_DE[[paste0("RNA_markers", i, "_", j)]] <- RNAmarkers
  ADTmarkers <- FindMarkers(tmp, ident.1 = i, ident.2 = j, only.pos = FALSE, assay = "ADT")
  ADTmarkers$protein <- row.names(ADTmarkers)
  ADTmarkers$comparison <- paste0(i, "-", j)
  NK_group_DE[[paste0("ADT_markers", i, "_", j)]] <- ADTmarkers
    }
  }
}

NK_group_DE <- bind_rows(NK_group_DE)

#writexl::write_xlsx(NK_group_DE, paste0(tab_path, "/all_integrated_NK_group_DE_res0.8.xlsx"))


compare_list_NK <- list("CD27loCD11b+_NK" = c("CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), "CD27loCD11bhi_NK"  = c("CD27loCD11b+_NK",  "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"),  "CD27hiCD11b+_NK" = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), "CD27loCD11b+_mito_NK" = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11blo_NK"),  "CD27loCD11blo_NK" = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK"))

NK_byGroup_DE <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")){
      RNAmarkers <- FindMarkers(tmp, ident.1 = i, ident.2 = as.vector(unlist(compare_list_NK[[i]])), only.pos = FALSE, assay = "RNA", logfc.threshold = 0)
  RNAmarkers$gene <- row.names(RNAmarkers)
  NK_byGroup_DE[[paste0("RNA_markers", i)]] <- RNAmarkers
  ADTmarkers <- FindMarkers(tmp, ident.1 = i, ident.2 = as.vector(unlist(compare_list_NK[[i]])), only.pos = FALSE, assay = "ADT", logfc.threshold = 0)
  ADTmarkers$protein <- row.names(ADTmarkers)
  NK_byGroup_DE[[paste0("ADT_markers", i)]] <- ADTmarkers
      }

#writexl::write_xlsx(NK_byGroup_DE, paste0(tab_path, "/all_integrated_NK_byGroup_DE_res0.8.xlsx"))

#pdf(paste0(fig_path, "/NKCellSubsets_differenentiatingMarkers_res0.8.pdf"), width = 12, height = 8)
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")), c("adt-CD117", "adt-CD39", "adt-CD71",  "adt-CD11a", "adt-CD62L", "adt-CD55", "adt-Ly49D", "adt-CD94", "adt-CD25", "adt-CD61", "adt-CD11b", "adt-CD27", "adt-CD86", "adt-I-A-I-E"), stack = TRUE, same.y.lims = TRUE) & NoLegend()

FeaturePlot(all_integrated_rnm, c("adt-Ly49D"), label = TRUE, label.size = 2, cols = c("white", "darkgreen"), order = TRUE) + theme(plot.background = element_rect(fill = "lightgrey"))

##dev.off()

#pdf(paste0(fig_path, "/NKCellSubsets_FeaturePlot_Ly49D_res0.8.pdf"), width = 10, height = 8)
FeaturePlot(all_integrated_rnm, c("adt-Ly49D"), label = TRUE, label.size = 3, repel = TRUE ,cols = c("white", "darkgreen"), order = TRUE) + theme(plot.background = element_rect(fill = "lightgrey"))
##dev.off()



#pdf(paste0(fig_path, "/NKCellSubsets_diffMarkers_res0.8.pdf"), width = 12, height = 6)
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")), c("adt-CD117", "adt-CD39", "adt-CD71",  "adt-CD11a", "adt-CD62L", "adt-CD55", "adt-Ly49D", "adt-CD94", "adt-CD25", "adt-CD61", "adt-CD11b", "adt-CD27", "adt-CD86", "adt-I-A-I-E", "adt-NK-1.1"), stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("")
##dev.off()

#pdf(paste0(fig_path, "/NKCellSubsets_diffMarkers_singleADTView_res0.8.pdf"), width = 12, height = 20)
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")), c("adt-CD117", "adt-CD39", "adt-CD71",  "adt-CD11a", "adt-CD62L", "adt-CD55", "adt-Ly49D", "adt-CD94", "adt-CD25", "adt-CD61", "adt-CD11b", "adt-CD27", "adt-CD86", "adt-I-A-I-E", "adt-NK-1.1", "adt-CD73", "adt-CD335"), split.by = "group", pt.size = 0.1, ncol = 4) & theme(axis.text.x = element_text(angle = 90),  axis.title.x = element_blank())
##dev.off()


#pdf(paste0(fig_path, "/NKCellSubsets_diffMarkers_singleRNAView_res0.8.pdf"), width = 15, height = 15)
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")), c("rna_Emb", "rna_Prnp", "rna_Mki67", "rna_Fcer1g", "rna_Rack1", "rna_Ctla2b", "rna_Ifngr1", "rna_Cd53", "rna_Cd82", "rna_Pdcd4", "rna_Pik3r1", "rna_Cd247", "rna_Klrb1b"), split.by = "group", pt.size = 0.1, ncol = 5) & theme(axis.text.x = element_text(angle = 90),  axis.title.x = element_blank())
##dev.off()

#Il18r1

```




## Plot of co-stimulatory and co-inhibitory genes and proteins, as suggested by Johannes (Reference: Chihara N, et al. Nature. 2018;558(7710):454-459. Figure 1, Supplementary Table 1: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6130914/#SD2)
```{r}
## Plot of co-stimulatory and co-inhibitory genes and proteins, as suggested by Johannes (Reference: Chihara N, et al. Nature. 2018;558(7710):454-459.Figure 1, Supplementary Table 1: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6130914/#SD2)
coStim_Inhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD357", "adt-CD160", "adt-CD272", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD85k", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD94", "adt-CD365")

coStim_Inhib_genes <- c("rna_Pdcd1", "rna_Havcr2", "rna_Lag3", "rna_Tigit", "rna_Ctla4", "rna_Tnfrsf18", "rna_Cd160", "rna_Btla", "rna_Cd2", "rna_Cd27", "rna_Lilrb4a", "rna_Icos", "rna_Tnfrsf9", "rna_Tnfrsf4", "rna_Slamf6", "rna_Cd226", "rna_Tnfrsf14", "rna_Klrd1", "rna_Havcr1") #rna_Cd28 not present in this data

#pdf(paste0(fig_path, "/Co-inhibitoryGeneModule_allCellTypes_res0.8_Chihara_2018.pdf"), width = 15, height = 15)
VlnPlot(all_integrated_rnm_VlnPlot, features = coStim_Inhib_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("")
VlnPlot(all_integrated_rnm_VlnPlot, features = coStim_Inhib_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") #"Lilrb4b" not present in this dataset
##dev.off()

#pdf(paste0(fig_path, "/Co-inhibitoryGeneModule_TcellsSubsets_res0.8_Chihara_2018.pdf"), width = 15, height = 10)
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coStim_Inhib_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("")

VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coStim_Inhib_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") #"Lilrb4b" not present in this dataset
##dev.off()


# Heatmaps
tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"
library(dittoSeq)

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_ADT_Tcell_CoStimulatory_InhibitoryGenes_Chihara_2018.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_coStimInhb <- AverageExpression(tmp_clus, slot = "data", assays = "ADT", return.seurat = TRUE)
  levels(cluster.averages_coStimInhb) <- c("il27raKO", "wt")
  cluster.averages_coStimInhb <- ScaleData(cluster.averages_coStimInhb)
  group <- Idents(cluster.averages_coStimInhb)
  names(group) <- colnames(cluster.averages_coStimInhb)
  cluster.averages_coStimInhb <- AddMetaData(object = cluster.averages_coStimInhb, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_coStimInhb, genes = coStim_Inhib_ADT, annot.by = "group", main = paste0("  Heatmap of average expression of T cell co-stimulatory and co-inhibitory ADT\n(Chihara et al., 2018: PMC6130914) in\n",i, "\n") )) 
}
##dev.off()

coStim_Inhib_genes <- c("Pdcd1", "Havcr2", "Lag3", "Tigit", "Ctla4", "Tnfrsf18", "Cd160", "Btla", "Cd2", "Cd27", "Lilrb4a", "Icos", "Tnfrsf9", "Tnfrsf4", "Slamf6", "Cd226", "Tnfrsf14", "Klrd1", "Havcr1") #excluded the tag "rna_" to avoid plotting conflicts. As the assay is set to RNA, this tag is not needed

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_RNA_Tcell_CoStimulatory_InhibitoryGenes_Chihara_2018.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_coStimInhb <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_coStimInhb) <- c("il27raKO", "wt")
  cluster.averages_coStimInhb <- ScaleData(cluster.averages_coStimInhb)
  group <- Idents(cluster.averages_coStimInhb)
  names(group) <- colnames(cluster.averages_coStimInhb)
  cluster.averages_coStimInhb <- AddMetaData(object = cluster.averages_coStimInhb, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_coStimInhb, genes = coStim_Inhib_genes, annot.by = "group", main = paste0("  Heatmap of average expression of T cell co-stimulatory and co-inhibitory RNA\n(Chihara et al., 2018: PMC6130914) in\n",i, "\n") )) 
}
##dev.off()

#pdf(paste0(fig_path, "dittoSeq_Expr_Htmap_RNA_Tcell_CoStim_Inhibit_Chihara_2018_CellsDownSamp100.pdf"))

dittoHeatmap(subset(all_integrated_rnm_VlnPlot, downsample = 100 ), genes = coStim_Inhib_ADT, annot.by = c("celltype",  "group"), cluster_cols = FALSE, assay = "ADT" )
dittoHeatmap(subset(all_integrated_rnm_VlnPlot, downsample = 100 ), genes = coStim_Inhib_genes, annot.by = c("celltype",  "group"), cluster_cols = FALSE, assay = "RNA" )

##dev.off()



# Dotplot of cosimulatory and coinhibitory ADT and genes

# Separate dot plot for each cluster
tmp <- all_integrated_rnm_DotPlot
tmp_cell_list <-unique(rev(levels(tmp)))#specifying reverse to ensure order of cell types

#ADT
dotplot_coStimInhb <- list()
for(i in tmp_cell_list){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  dotplot_coStimInhb[[i]] <- DotPlot(tmp_clus, coStim_Inhib_ADT, assay = "ADT", dot.scale = 6) + scale_color_gradientn(colors = c("navy", "white", "red")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), plot.title = element_text(size = 10), axis.title = element_blank()) + ggtitle(paste0("Dotplot showing expression of T cell co-stimulatory and co-inhibitory ADT\n(Chihara et al., 2018: PMC6130914) in\n",i, "\n"))
}


#pdf(paste0(fig_path, "Dotplot_ADT_Tcell_CoStimulatory_InhibitoryGenes_Chihara_2018.pdf"), width = 12, height = 12)
for(i in seq(1,length(dotplot_coStimInhb), by = 6) ){
  print(wrap_plots(dotplot_coStimInhb[[i]], dotplot_coStimInhb[[i+1]], dotplot_coStimInhb[[i+2]], dotplot_coStimInhb[[i+3]], dotplot_coStimInhb[[i+4]], dotplot_coStimInhb[[i+5]], ncol = 2, nrow = 3))
    }
##dev.off()

#RNA
dotplot_coStimInhb <- list()
for(i in tmp_cell_list){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  dotplot_coStimInhb[[i]] <- DotPlot(tmp_clus, coStim_Inhib_genes, assay = "RNA", dot.scale = 6) + scale_color_gradientn(colors = c("navy", "white", "red")) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), plot.title = element_text(size = 10), axis.title = element_blank()) + ggtitle(paste0("Dotplot showing expression of T cell co-stimulatory and co-inhibitory ADT\n(Chihara et al., 2018: PMC6130914) in\n",i, "\n"))
}


#pdf(paste0(fig_path, "Dotplot_RNA_Tcell_CoStimulatory_InhibitoryGenes_Chihara_2018.pdf"), width = 12, height = 12)
for(i in seq(1,length(dotplot_coStimInhb), by = 6) ){
  print(wrap_plots(dotplot_coStimInhb[[i]], dotplot_coStimInhb[[i+1]], dotplot_coStimInhb[[i+2]], dotplot_coStimInhb[[i+3]], dotplot_coStimInhb[[i+4]], dotplot_coStimInhb[[i+5]], ncol = 2, nrow = 3))
    }
##dev.off()

# Module score
tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

coStim_Inhib_genes <- c("Pdcd1", "Havcr2", "Lag3", "Tigit", "Ctla4", "Tnfrsf18", "Cd160", "Btla", "Cd2", "Cd27", "Lilrb4a", "Icos", "Tnfrsf9", "Tnfrsf4", "Slamf6", "Cd226", "Tnfrsf14", "Klrd1", "Havcr1") #excluded the tag "rna_" to avoid plotting conflicts. As the assay is set to RNA, this tag is not needed

#pdf(paste0(fig_path, "ModuleScore_Tcell_CoStimulatory_InhibitoryGenes_Chihara_2018.pdf"), width = 12, height = 8)
#DefaultAssay(tmp) <- "ADT"
#tmp <- AddModuleScore(tmp, features = coStim_Inhib_ADT, name = "coStim_Inhib_ADT", assay = "ADT") #does nt analyze for ADT data


DefaultAssay(tmp) <- "RNA"
tmp <- AddModuleScore(tmp, features = list(coStim_Inhib_genes), name = "coStim_Inhib_RNA")

FeaturePlot(tmp, features = "coStim_Inhib_RNA1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))
##dev.off()




#Finding differential expression of costimulatory and coinhibitory genes and proteins across clusters between il27raKO vs wt (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)
Idents(all_integrated_rnm) <- "cluster0.8.group"

#ADT
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "ADT", features = coStim_Inhib_ADT, logfc.threshold = 0, min.pct = 0) %>%
    rownames_to_column(var = "ADT") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_coStim_Inhb_ADT_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)


#RNA
get_differential_genes <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", features = coStim_Inhib_genes, logfc.threshold = 0, min.pct = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_coStim_Inhb_RNA_all_integrated_rnm <- map_dfr(tmp_celltype_list, get_differential_genes)

differential_coStim_Inhb <- list( differential_coStim_Inhb_ADT = differential_coStim_Inhb_ADT_all_integrated_rnm,
                                   differential_coStim_Inhb_RNA = differential_coStim_Inhb_RNA_all_integrated_rnm
   
 )

#writexl::write_xlsx(differential_coStim_Inhb, paste0(tab_path, "/differential_coStim_Inhb_all_integrated_rnm_res0.8_Chihara_2018_Chen_Flies_2013.xlsx"))




#VlnPlot(all_integrated_rnm_VlnPlot, "adt-CD279", split.by = "group", pt.size = 0.1) + stat_compare_means(method = "wilcox", aes(label=..p.adj..))                        ) #https://github.com/kassambara/ggpubr/issues/65; http://www.sthda.com/english/articles/24-ggpubr-publication-ready-plots/76-add-p-values-and-significance-levels-to-ggplots/

```


# JAX GO BP genes for T cell activation (GO:0042110), T cell costimulation (GO:0031295), positive regulation of T cell apoptotic process (GO:0070234 )
## T cell activation (GO:0042110)
```{r}
#The genes mapped in the JAX Gene Ontology database for T cell activation (http://www.informatics.jax.org/vocab/gene_ontology/GO:0042110) were retrieved (4 Mar 2022) and a heatmap was plotted using dittoheatmap function
library(readxl)

# T cell activation
tmp <- readxl::read_excel("GO_JAX_tables_input/GO_term_summary_20220304_160635_Tcell_Activation.xlsx")

genes_Tcell_activation_Jax <- unique(tmp$Symbol)

genes_Tcell_activation <- subset(genes_Tcell_activation_Jax, genes_Tcell_activation_Jax %in% row.names(tmp_sce)) #subset to keep only those genes from the JAX database Tcell_activation GO that are expressed in the current dataset 


#subset differentially expressed genes list to include only genes that are statistically significant
pgenes <- subset(differential_genes_all_integrated_rnm, p_val_adj < 0.05)

#subset list of Tcell_activation genes to only include genes that are statistically significantly (Adjusted p < 0.05) differentially expressed  between IL27RAKO and WT in any of the cell types
genes_Tcell_activation_plot <- subset(genes_Tcell_activation, genes_Tcell_activation %in% pgenes$gene)


tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_Tcell_activation_signifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_tmp <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_tmp) <- c("il27raKO", "wt")
  cluster.averages_tmp <- ScaleData(cluster.averages_tmp)
  group <- Idents(cluster.averages_tmp)
  names(group) <- colnames(cluster.averages_tmp)
  cluster.averages_tmp <- AddMetaData(object = cluster.averages_tmp, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_tmp, genes = genes_Tcell_activation_plot, annot.by = "group", main = paste0("  Heatmap of average expression of statistically significant (adj p < 0.05)\nTcell activation GO BP (GO:0042110) genes in\n",i, "\n") )) 
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_genes_Tcell_activation_signifGenes.pdf"), width = 12)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_Tcell_activation_plot, split.by = "group", stack = TRUE, assay = "RNA") & theme(axis.title.y.left = element_blank(), axis.text.y.left = element_text(size = 10), strip.text.x = element_text(face = "plain", size = 10, hjust = 1))
##dev.off()

# Heatmap of ADT of genes_Tcell_activation_plot, where corresponding ADT is available
#available ADT: "adt-CD27", "adt-CD25", "adt-CD127", "adt-CD85k", "adt-Ly108"

tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"
adt_Tcell_activation_plot <- c("adt-CD27", "adt-CD25", "adt-CD127", "adt-CD85k", "adt-Ly108")

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_Tcell_activation_signifADT.pdf"))
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_tmp <- AverageExpression(tmp_clus, slot = "data", assays = "ADT", return.seurat = TRUE)
  levels(cluster.averages_tmp) <- c("il27raKO", "wt")
  cluster.averages_tmp <- ScaleData(cluster.averages_tmp)
  group <- Idents(cluster.averages_tmp)
  names(group) <- colnames(cluster.averages_tmp)
  cluster.averages_tmp <- AddMetaData(object = cluster.averages_tmp, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_tmp, genes = adt_Tcell_activation_plot, annot.by = "group", main = paste0("  Heatmap of average expression of statistically significant (adj p < 0.05)\nTcell activation GO BP (GO:0042110) ADT in\n",i, "\n") )) 
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_genes_Tcell_activation_signifADT.pdf"), width = 8)
VlnPlot(all_integrated_rnm_VlnPlot, features = adt_Tcell_activation_plot, split.by = "group", stack = TRUE, assay = "ADT") & theme(axis.title.y.left = element_blank(), axis.text.y.left = element_text(size = 10), strip.text.x = element_text(face = "plain", size = 10, hjust = 1))
##dev.off()

```


## T cell costimulation (GO:0031295)
```{r}
#The genes mapped in the JAX Gene Ontology database for T cell costimulation (http://www.informatics.jax.org/vocab/gene_ontology/GO:0031295) were retrieved (4 Mar 2022) and a heatmap was plotted using dittoheatmap function
library(readxl)

# T cell costimulation
tmp <- readxl::read_excel("GO_JAX_tables_input/GO_term_summary_20220304_160608_Tcell_costimulation.xlsx")

genes_Tcell_costimulation_Jax <- unique(tmp$Symbol)

genes_Tcell_costimulation <- subset(genes_Tcell_costimulation_Jax, genes_Tcell_costimulation_Jax %in% row.names(tmp_sce)) #subset to keep only those genes from the JAX database Tcell_costimulation GO that are expressed in the current dataset 


#subset differentially expressed genes list to include only genes that are statistically significant
pgenes <- subset(differential_genes_all_integrated_rnm, p_val_adj < 0.05)

#subset list of Tcell_costimulation genes to only include genes that are statistically significantly (Adjusted p < 0.05) differentially expressed  between IL27RAKO and WT in any of the cell types
genes_Tcell_costimulation_plot <- subset(genes_Tcell_costimulation, genes_Tcell_costimulation %in% pgenes$gene)



tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_Tcell_costimulation_signifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){ 
  if (i != "B_cell") { #excluding B_cell cluster as it does not map these genes
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_tmp <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_tmp) <- c("il27raKO", "wt")
  cluster.averages_tmp <- ScaleData(cluster.averages_tmp)
  group <- Idents(cluster.averages_tmp)
  names(group) <- colnames(cluster.averages_tmp)
  cluster.averages_tmp <- AddMetaData(object = cluster.averages_tmp, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_tmp, genes = genes_Tcell_costimulation_plot, annot.by = "group", main = paste0("  Heatmap of average expression of statistically significant (adj p < 0.05)\nTcell costimulation GO BP (GO:0031295) genes in\n",i, "\n") )) 
  }
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_genes_Tcell_costimulation_signifGenes.pdf"), width = 6)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_Tcell_costimulation_plot, split.by = "group", stack = TRUE, assay = "RNA") & theme(axis.title.y.left = element_blank(), axis.text.y.left = element_text(size = 10), strip.text.x = element_text(face = "plain", size = 10, hjust = 1))
##dev.off()


# Also, plotting the non-significant genes
tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

genes_Tcell_costimulation_nonSignif <- subset(genes_Tcell_costimulation, ! (genes_Tcell_costimulation %in% genes_Tcell_costimulation_plot)) #subsetting only the non-significant genes

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_Tcell_costimulation_NonsignifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){ 
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_tmp <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_tmp) <- c("il27raKO", "wt")
  cluster.averages_tmp <- ScaleData(cluster.averages_tmp)
  group <- Idents(cluster.averages_tmp)
  names(group) <- colnames(cluster.averages_tmp)
  cluster.averages_tmp <- AddMetaData(object = cluster.averages_tmp, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_tmp, genes = genes_Tcell_costimulation_nonSignif, annot.by = "group", main = paste0("  Heatmap of average expression of non-statistically significant (adj p > 0.05)\nTcell costimulation GO BP (GO:0031295) genes in\n",i, "\n") )) 
  }
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_genes_Tcell_costimulation_NonsignifGenes.pdf"), width = 12)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_Tcell_costimulation_nonSignif, split.by = "group", stack = TRUE, assay = "RNA") & theme(axis.title.y.left = element_blank(), axis.text.y.left = element_text(size = 10), strip.text.x = element_text(face = "plain", size = 10, hjust = 1))
##dev.off()


```




## positive regulation of T cell apoptotic process (GO:0070234)
```{r}
#The genes mapped in the JAX Gene Ontology database for positive regulation of T cell apoptotic process (http://www.informatics.jax.org/vocab/gene_ontology/GO:0070234) were retrieved (4 Mar 2022) and a heatmap was plotted using dittoheatmap function
library(readxl)

# positive regulation of T cell apoptotic process
tmp <- readxl::read_excel("GO_JAX_tables_input/GO_term_summary_20220304_160653_PosRegln_Tcell_Apoptosis.xlsx")

genes_posRegTcell_apoptosis_Jax <- unique(tmp$Symbol)

genes_posRegTcell_apoptosis <- subset(genes_posRegTcell_apoptosis_Jax, genes_posRegTcell_apoptosis_Jax %in% row.names(tmp_sce)) #subset to keep only those genes from the JAX database posRegTcell_apoptosis GO that are expressed in the current dataset 


#subset differentially expressed genes list to include only genes that are statistically significant
pgenes <- subset(differential_genes_all_integrated_rnm, p_val_adj < 0.05)

#subset list of posRegTcell_apoptosis genes to only include genes that are statistically significantly (Adjusted p < 0.05) differentially expressed  between IL27RAKO and WT in any of the cell types
genes_posRegTcell_apoptosis_plot <- subset(genes_posRegTcell_apoptosis, genes_posRegTcell_apoptosis %in% pgenes$gene)



tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_posRegTcell_apoptosis_signifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){
   if (i != "B_cell") { #excluding B_cell cluster as it does not map these genes
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_tmp <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_tmp) <- c("il27raKO", "wt")
  cluster.averages_tmp <- ScaleData(cluster.averages_tmp)
  group <- Idents(cluster.averages_tmp)
  names(group) <- colnames(cluster.averages_tmp)
  cluster.averages_tmp <- AddMetaData(object = cluster.averages_tmp, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_tmp, genes = genes_posRegTcell_apoptosis_plot, annot.by = "group", main = paste0("  Heatmap of average expression of statistically significant (adj p < 0.05)\npositive regulation of T cell apoptotic\nprocess GO BP (GO:0070234) genes in\n",i, "\n") ))
   }
}
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_genes_posRegTcell_apoptosis_signifGenes.pdf"), width = 6)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_posRegTcell_apoptosis_plot, split.by = "group", stack = TRUE, assay = "RNA") & theme(axis.title.y.left = element_blank(), axis.text.y.left = element_text(size = 10), strip.text.x = element_text(face = "plain", size = 10, hjust = 1))
##dev.off()


# Also, plotting the non-significant genes
tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

genes_posRegTcell_apoptosis_nonSignif <- subset(genes_posRegTcell_apoptosis, ! (genes_posRegTcell_apoptosis %in% genes_posRegTcell_apoptosis_plot)) #subsetting only the non-significant genes

#pdf(paste0(fig_path, "dittoSeq_AvgExpr_Heatmap_posRegTcell_apoptosis_NonsignifGenes.pdf"))
for(i in unique(tmp@meta.data$celltype)){ 
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages_tmp <- AverageExpression(tmp_clus, slot = "data", assays = "RNA", return.seurat = TRUE)
  levels(cluster.averages_tmp) <- c("il27raKO", "wt")
  cluster.averages_tmp <- ScaleData(cluster.averages_tmp)
  group <- Idents(cluster.averages_tmp)
  names(group) <- colnames(cluster.averages_tmp)
  cluster.averages_tmp <- AddMetaData(object = cluster.averages_tmp, metadata = group, col.name = "group")
  print(dittoHeatmap(cluster.averages_tmp, genes = genes_posRegTcell_apoptosis_nonSignif, annot.by = "group", main = paste0("  Heatmap of average expression of non-statistically significant (adj p > 0.05)\npositive regulation of T cell apoptotic\nprocess GO BP (GO:0031295) genes in\n",i, "\n") )) 
  }
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_genes_posRegTcell_apoptosis_NonsignifGenes.pdf"), width = 12)
VlnPlot(all_integrated_rnm_VlnPlot, features = genes_posRegTcell_apoptosis_nonSignif, split.by = "group", stack = TRUE, assay = "RNA") & theme(axis.title.y.left = element_blank(), axis.text.y.left = element_text(size = 10), strip.text.x = element_text(face = "plain", size = 10, hjust = 1))
##dev.off()



```



# T cell exhaustion markers
```{r}
# T cell exhaustion markers (Reference: https://www.rndsystems.com/product-highlights/adoptive-cell-transfer-monitor-t-cell-exhaustion)
#beta cc chemokines: https://www.sciencedirect.com/science/article/pii/B9780123744104003745

#pdf(paste0(fig_path, "VlnPlot_Tcellexhaustionmarkers_rndsystems_TCells.pdf"), width = 12)
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = c("rna_Il2","rna_Gzmb", "rna_Tnf", "rna_Ifng", "rna_Ccl2", "rna_Ccl8", "rna_Ccl7", "rna_Ccl3", "rna_Ccl4", "rna_Ccl5", "rna_Ccl1", "rna_Eomes", "rna_Gata3", "rna_Bcl6", "rna_Ikzf2", "rna_Il4", "rna_Il6", "rna_Il21"), stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & ggtitle("T cell exhaustion markers\n(Ref: https://www.rndsystems.com/product-highlights/adoptive-cell-transfer-monitor-t-cell-exhaustion)\n") & theme(plot.title = element_text(size = 12))
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_Tcellexhaustionmarkers_rndsystems_AllCells.pdf"), width = 12)
VlnPlot(all_integrated_rnm_VlnPlot, features = c("rna_Il2","rna_Gzmb", "rna_Tnf", "rna_Ifng", "rna_Ccl2", "rna_Ccl8", "rna_Ccl7", "rna_Ccl3", "rna_Ccl4", "rna_Ccl5", "rna_Ccl1", "rna_Eomes", "rna_Gata3", "rna_Bcl6", "rna_Ikzf2", "rna_Il4", "rna_Il6", "rna_Il21"), stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & ggtitle("T cell exhaustion markers\n(Ref: https://www.rndsystems.com/product-highlights/adoptive-cell-transfer-monitor-t-cell-exhaustion)\n") & theme(plot.title = element_text(size = 12))
##dev.off()


 # T cell exhaustion markers- Immune checkpoint coverage (Reference: https://nanostring.com/products/ncounter-assays-panels/immunology/immune-exhaustion/; https://nanostring.com/wp-content/uploads/PB_MK2848_ImmuneExhaustion_r4-1.pdf ; Immune Checkpoint Coverage)
#pdf(paste0(fig_path, "VlnPlot_ImmuneCheckpointCoverage_NanoString_AllCells.pdf"), width = 25)
VlnPlot(all_integrated_rnm_VlnPlot, features = c("rna_Cd28", "adt-CD28","rna_Icosl", "rna_Lag3", "rna_Tnfsf18", "rna_Icos", "adt-CD278", "rna_Cd40", "adt-CD40", "rna_Pdcd1", "adt-CD279", "rna_Havcr2", "adt-CD366", "rna_Vsir", "rna_Cd70", "rna_Cd80", "adt-CD80", "rna_Tnfsf4", "adt-CD252", "rna_Tnfsf9", "adt-CD137L", "rna_Cd27",  "adt-CD27", "rna_Cd276", "rna_Cd86", "adt-CD86", "rna_Cd274", "adt-CD274", "rna_Tnfrsf9", "adt-CD137", "rna_Cd40lg", "rna_Adora2a", "rna_Ctla4", "adt-CD152", "rna_Pdcd1lg2", "adt-CD273", "rna_Tnfrsf18", "adt-CD357", "rna_Tnfrsf4", "adt-CD134", "rna_Tigit", "adt-TIGIT"), stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & ggtitle("Immune checkpoint coverage genes\n(Ref: https://nanostring.com/products/ncounter-assays-panels/immunology/immune-exhaustion/)\n") & theme(plot.title = element_text(size = 12))
##dev.off()



```



```{r}
#https://www.nature.com/articles/s41590-021-00975-5
#https://www.frontiersin.org/articles/10.3389/fimmu.2018.02569/full
#https://www.frontiersin.org/articles/10.3389/fimmu.2021.691142/full#h3
#https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1009892
```

# checking expression of Gigyf1 based on Johannes and Prof. Bosmann's inputs
```{r}
#pdf(paste0(fig_path, "DotPlot_Gigyf1.pdf"), height = 9)
DotPlot(all_integrated_rnm_DotPlot, c("rna_Gigyf1"), split.by = "group", assay = "RNA", cols = c("navy", "red")) + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_Gigyf1.pdf"), height = 9)
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Gigyf1"), split.by = "group", assay = "RNA") + theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))
##dev.off()
```


# Upd 8Apr2022: Single cell pathway analysis (SCPA)
Notes:  (Reference: https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383; https://jackbibby1.github.io/SCPA/articles/comparing_two_populations.html#some-conclusions-about-scpa)
- "If doing a 2 sample comparison there will be a fold change column in the output from compare_pathways, which will tells the directionality."<br>
- "The fold change is calculated from population1 - population2, so a negative value will be higher in population2." <br> 
- "SCPA measures changes in the multivariate distribution of a pathway (reflecting multiple parameters including enrichment) as its primary outcome, so there will be pathways that show large qvals, but no fold change. These pathways are still importantly different. qval should be used as the primary statistic when judging biological relevance, and the pathway fold change/enirchment should only be a secondary informative value." <br>
- The authors/developers of SCPA "recommend using the qval as the measure of differential regulation of a pathway, rather than fold change. And just use the fold change to get an idea if there is enrichment of that pathway or not." <br>
```{r}
# Performing SCPA analysis (Source vignette: https://jackbibby1.github.io/SCPA/articles/disease_comparison.html; https://jackbibby1.github.io/SCPA/articles/systematic_tissue_comparison.html; https://igordot.github.io/msigdbr/articles/msigdbr-intro.html; https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383; https://jackbibby1.github.io/SCPA/articles/comparing_two_populations.html#some-conclusions-about-scpa)
#devtools::install_github("jackbibby1/SCPA")

# temprary vector of cell types
tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

# vector comprising pathways
library(msigdbr)
library(SCPA)

#pws <- c("bp", "cc", "mf", "kegg", "reactome")
#pathways <- msigdbr("Mus musculus") %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()

pws <- c("bp")
pathways <- msigdbr("Mus musculus") %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()




scpa_out <- list()

for (i in tmp_celltype_list) {
  
  wt <- seurat_extract(subset(all_integrated_rnm_VlnPlot, group %in% "wt"), 
                            meta1 = "celltype", value_meta1 = i)
  
  il27raKO <- seurat_extract(subset(all_integrated_rnm_VlnPlot, group %in% "il27raKO"), 
                          meta1 = "celltype", value_meta1 = i)
  
  print(paste("comparing", i))
  scpa_out[[i]] <- compare_pathways(samples = list(wt, il27raKO), pathways = pathways) %>% 
    select(Pathway, qval, FC) %>%
    set_colnames(c("Pathway", paste(i, "qval", sep = "_"), paste(i, "FC", sep = "_")))
  #fold change in pathways calculated from population 1 - population 2 (https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383)
}
#writexl::write_xlsx(scpa_out, "scpa_out.xlsx")
#saveRDS(scpa_out, "scpa_out.rds")
#scpa_out <- readRDS("scpa_out.rds")

scpa_out_fig <- scpa_out %>% 
  purrr::reduce(full_join, by = "Pathway") %>%  
  set_colnames(gsub(colnames(.), pattern = " ", replacement = "_")) %>%
  select(c("Pathway", grep("_qval", colnames(.)))) %>%
  filter_all(any_vars(. > 2)) %>%
  column_to_rownames("Pathway") #specify purrr::reduce to prevent error (Source code: https://stackoverflow.com/questions/66776517/how-do-i-solve-the-problem-with-reduce-function-in-r)


paths_to_visualize <- c("HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_INFLAMMATORY_RESPONSE",
                 "HALLMARK_COMPLEMENT", "HALLMARK_IL6_JAK_STAT3_SIGNALING",
                 "HALLMARK_IL2_STAT5_SIGNALING", "HALLMARK_INTERFERON_GAMMA_RESPONSE",
                  "HALLMARK_INTERFERON_ALPHA_RESPONSE", "HALLMARK_OXIDATIVE_PHOSPHORYLATION")
library(ComplexHeatmap)
library(circlize)
position <- which(rownames(scpa_out_fig) %in% paths_to_visualize)
row_an <- rowAnnotation(Genes = anno_mark(at = which(rownames(scpa_out_fig) %in% paths_to_visualize),
                                          labels = rownames(scpa_out_fig)[position],
                                          labels_gp = gpar(fontsize = 7),
                                          link_width = unit(2.5, "mm"),
                                          padding = unit(1, "mm"),
                                          link_gp = gpar(lwd = 0.5)))
col_hm <- colorRamp2(colors = c("blue", "white", "red"), breaks = c(0, 3, 6))

ComplexHeatmap::Heatmap(scpa_out_fig,
        col = col_hm,
        name = "Qval",
        show_row_names = FALSE,
        right_annotation = row_an,
        column_names_gp = gpar(fontsize = 8),
        border = TRUE,
        column_km = 3,
        row_km = 3,
        column_labels = tmp_celltype_list)

#15March2023:
scpa_out_fig_filt <- scpa_out_fig %>% filter(grepl("HALLMARK_", row.names(scpa_out_fig)))
scpa_out_fig_filt <-  scpa_out_fig_filt[, c(1:12)] #Filter to only keep columns that pertain to NK and T cells

htmap_scpa_out_fig_filt <- ComplexHeatmap::Heatmap(scpa_out_fig_filt ,
                        col = col_hm,
                        name = "Qval",
                        show_row_names = TRUE,
                        column_names_gp = gpar(fontsize = 8),row_names_gp = gpar(fontsize = 8),
                        border = TRUE,
                        column_km = 3, cluster_columns = FALSE,
                        row_km = 3,
                        column_labels = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), height = unit(3, "mm")*nrow(scpa_out_fig_filt), width = unit(5, "mm")*ncol(scpa_out_fig_filt)  )

calc_ht_size = function(ht, unit = "inch") {
    #pdf(NULL)
    ht = draw(ht)
    w = ComplexHeatmap:::width(ht)
    w = convertX(w, unit, valueOnly = TRUE)
    h = ComplexHeatmap:::height(ht)
    h = convertY(h, unit, valueOnly = TRUE)
    #dev.off()

    c(w, h)
}


size = calc_ht_size(htmap_scpa_out_fig_filt)

#pdf(paste0(fig_path, "tmp_SCPA_pathway_wt_vs_KO_Qval_NK_Tcell.pdf"), width = size[1] + 2, height = size[2])

draw(htmap_scpa_out_fig_filt,
    heatmap_legend_side = 'right',
    annotation_legend_side = 'right',
    row_sub_title_side = 'left',
    row_title_side = "right")

##dev.off()


#===========================================================================================================

scpa_out_fig <- scpa_out %>% 
  purrr::reduce(full_join, by = "Pathway") %>%  
  set_colnames(gsub(colnames(.), pattern = " ", replacement = "_")) %>%
  select(c("Pathway", grep("_FC", colnames(.)))) %>%
  column_to_rownames("Pathway") #specify purrr::reduce to prevent error (Source code: https://stackoverflow.com/questions/66776517/how-do-i-solve-the-problem-with-reduce-function-in-r)


paths_to_visualize <- c("HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_INFLAMMATORY_RESPONSE",
                 "HALLMARK_COMPLEMENT", "HALLMARK_IL6_JAK_STAT3_SIGNALING",
                 "HALLMARK_IL2_STAT5_SIGNALING", "HALLMARK_INTERFERON_GAMMA_RESPONSE",
                  "HALLMARK_INTERFERON_ALPHA_RESPONSE", "HALLMARK_OXIDATIVE_PHOSPHORYLATION")
library(ComplexHeatmap)
library(circlize)
position <- which(rownames(scpa_out_fig) %in% paths_to_visualize)
row_an <- rowAnnotation(Genes = anno_mark(at = which(rownames(scpa_out_fig) %in% paths_to_visualize),
                                          labels = rownames(scpa_out_fig)[position],
                                          labels_gp = gpar(fontsize = 7),
                                          link_width = unit(2.5, "mm"),
                                          padding = unit(1, "mm"),
                                          link_gp = gpar(lwd = 0.5)))
col_hm <- colorRamp2(colors = c("blue", "white", "red"), breaks = c(-10, 0, 10))

ComplexHeatmap::Heatmap(scpa_out_fig,
        col = col_hm,
        name = "Fold-change",
        show_row_names = FALSE,
        right_annotation = row_an,
        column_names_gp = gpar(fontsize = 8),
        border = TRUE,
        column_km = 3,
        row_km = 3,
        column_labels = tmp_celltype_list)


paths_to_visualize <- bind_rows(scpa_out) %>% filter(grepl("HALLMARK_", Pathway)) %>% select("Pathway") %>% unique() %>% unlist()


position <- which(rownames(scpa_out_fig) %in% paths_to_visualize)
row_an <- rowAnnotation(Genes = anno_mark(at = which(rownames(scpa_out_fig) %in% paths_to_visualize),
                                          labels = rownames(scpa_out_fig)[position],
                                          labels_gp = gpar(fontsize = 7),
                                          link_width = unit(2.5, "mm"),
                                          padding = unit(1, "mm"),
                                          link_gp = gpar(lwd = 0.5)))

col_hm <- colorRamp2(colors = c("blue", "white", "red"), breaks = c(-10, 0, 10))



scpa_htmap_fc <- ComplexHeatmap::Heatmap(scpa_out_fig,
                        col = col_hm,
                        name = "Fold-change",
                        show_row_names = FALSE,
                        right_annotation = row_an,
                        row_names_gp = gpar(fontsize = 3),
                        column_names_gp = gpar(fontsize = 8),
                        border = TRUE,
                        column_km = 0,
                        row_km = 0,
                        column_labels = tmp_celltype_list,
                        height = unit(0.5, "mm")*nrow(scpa_out_fig), width = unit(5, "mm")*ncol(scpa_out_fig)
                        )



#calculate heatmap figure size
#Source code: https://jokergoo.github.io/2020/05/11/set-cell-width/height-in-the-heatmap/

calc_ht_size = function(ht, unit = "inch") {
    #pdf(NULL)
    ht = draw(ht)
    w = ComplexHeatmap:::width(ht)
    w = convertX(w, unit, valueOnly = TRUE)
    h = ComplexHeatmap:::height(ht)
    h = convertY(h, unit, valueOnly = TRUE)
    #dev.off()

    c(w, h)
}


size = calc_ht_size(scpa_htmap_fc)

#pdf(paste0(fig_path, "tmp_SCPA_pathway_wt_vs_KO_FC.pdf"), width = size[1], height = size[2])

draw(scpa_htmap_fc,
    heatmap_legend_side = 'right',
    annotation_legend_side = 'right',
    row_sub_title_side = 'left',
    row_title_side = "right")

##dev.off()

#=====================================================
#Upd 20Mar2023: Hallmark pathways based on FC

# temprary vector of cell types
tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

# vector comprising pathways
library(msigdbr)
library(SCPA)

#pws <- c("bp", "cc", "mf", "kegg", "reactome")
#pathways <- msigdbr("Mus musculus") %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()

pws <- c("bp")
pathways_Hallmark <- msigdbr("Mus musculus", category = c("H")) %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()



scpa_out_il27rako_wt <- list()

for (i in tmp_celltype_list) {
  
  
  il27raKO <- seurat_extract(subset(all_integrated_rnm_VlnPlot, group %in% "il27raKO"), 
                          meta1 = "celltype", value_meta1 = i)
  
  wt <- seurat_extract(subset(all_integrated_rnm_VlnPlot, group %in% "wt"), 
                            meta1 = "celltype", value_meta1 = i)
  
  print(paste("comparing", i))
  scpa_out_il27rako_wt[[i]] <- compare_pathways(samples = list(il27raKO, wt), pathways = pathways_Hallmark) %>% 
    dplyr::select(Pathway, qval, FC) %>%
    set_colnames(c("Pathway", paste(i, "qval", sep = "_"), paste(i, "FC", sep = "_")))
  #fold change in pathways calculated from population 1 - population 2 (https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383)

}
  
scpa_out_il27rako_wt_fig_filt <- scpa_out_il27rako_wt %>% 
  purrr::reduce(full_join, by = "Pathway") %>%  
  set_colnames(gsub(colnames(.), pattern = " ", replacement = "_")) %>%
  dplyr::select(c("Pathway", grep("_FC", colnames(.)))) %>%
  column_to_rownames("Pathway") #specify purrr::reduce to prevent error (Source code: https://stackoverflow.com/questions/66776517/how-do-i-solve-the-problem-with-reduce-function-in-r)


scpa_out_il27rako_wt_fig_filt_NK_T <-  scpa_out_il27rako_wt_fig_filt[, c(1:12)] #Filter to only keep columns that pertain to NK and T cells


library(ComplexHeatmap)
library(circlize)

col_hm <- colorRamp2(colors = c("blue", "white", "red"), breaks = c(-10, 0, 10))


htmap_scpa_out_il27rako_wt_fig_filt_NK_T <- ComplexHeatmap::Heatmap(scpa_out_il27rako_wt_fig_filt_NK_T ,
                        col = col_hm,
                        name = "Fold-change",
                        show_row_names = TRUE,
                        column_names_gp = gpar(fontsize = 8),row_names_gp = gpar(fontsize = 8),
                        column_title = 'SCPA pathway heatmap of Hallmark pathways based on fold change\nin IL27RAKO vs WT\n',
                        border = TRUE,
                        column_km = 3, cluster_columns = FALSE,
                        row_km = 3,
                        column_labels = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), height = unit(3, "mm")*nrow(scpa_out_il27rako_wt_fig_filt_NK_T), width = unit(5, "mm")*ncol(scpa_out_il27rako_wt_fig_filt_NK_T)  )

calc_ht_size = function(ht, unit = "inch") {
    #pdf(NULL)
    ht = draw(ht)
    w = ComplexHeatmap:::width(ht)
    w = convertX(w, unit, valueOnly = TRUE)
    h = ComplexHeatmap:::height(ht)
    h = convertY(h, unit, valueOnly = TRUE)
    #dev.off()

    c(w, h)
}


size = calc_ht_size(htmap_scpa_out_il27rako_wt_fig_filt_NK_T)

#pdf(paste0(fig_path, "tmp_SCPA_pathway_KO_wt_vs_FoldChange_NK_Tcell.pdf"), width = size[1] + 2, height = size[2])

draw(htmap_scpa_out_il27rako_wt_fig_filt_NK_T,
    heatmap_legend_side = 'right',
    annotation_legend_side = 'right',
    row_sub_title_side = 'left',
    row_title_side = "right")

##dev.off()

scpa_out_il27rako_wt_fig_filt_qval <- scpa_out_il27rako_wt %>% 
  purrr::reduce(full_join, by = "Pathway") %>%  
  set_colnames(gsub(colnames(.), pattern = " ", replacement = "_")) %>%
  select(c("Pathway", grep("_qval", colnames(.)))) %>%
  column_to_rownames("Pathway") #specify purrr::reduce to prevent error (Source code: https://stackoverflow.com/questions/66776517/how-do-i-solve-the-problem-with-reduce-function-in-r)


scpa_out_il27rako_wt_fig_filt_NK_T_qval <-  scpa_out_il27rako_wt_fig_filt_qval[, c(1:12)] #Filter to only keep columns that pertain to NK and T cells


htmap_scpa_out_il27rako_wt_fig_filt_NK_T_Qval <- ComplexHeatmap::Heatmap(scpa_out_il27rako_wt_fig_filt_NK_T_qval ,
                        #col = col_hm,
                        name = "Qval",
                        show_row_names = TRUE,
                        column_names_gp = gpar(fontsize = 8),row_names_gp = gpar(fontsize = 8),
                        column_title = 'SCPA pathway heatmap of Hallmark pathways based on Q value\nin IL27RAKO vs WT\n',
                        border = TRUE,
                        column_km = 3, cluster_columns = FALSE,
                        row_km = 3,
                        column_labels = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), height = unit(3, "mm")*nrow(scpa_out_il27rako_wt_fig_filt_NK_T_qval), width = unit(5, "mm")*ncol(scpa_out_il27rako_wt_fig_filt_NK_T_qval)  )

size = calc_ht_size(htmap_scpa_out_il27rako_wt_fig_filt_NK_T_Qval)

#pdf(paste0(fig_path, "tmp_SCPA_pathway_KO_vs_wt_Qvalue_NK_Tcell.pdf"), width = size[1] + 2, height = size[2])

draw(htmap_scpa_out_il27rako_wt_fig_filt_NK_T_Qval,
    heatmap_legend_side = 'right',
    annotation_legend_side = 'right',
    row_sub_title_side = 'left',
    row_title_side = "right")

##dev.off()

```


#separting based on celltype and group
```{r}
# Performing SCPA analysis (Source vignette: https://jackbibby1.github.io/SCPA/articles/disease_comparison.html; https://jackbibby1.github.io/SCPA/articles/systematic_tissue_comparison.html; https://igordot.github.io/msigdbr/articles/msigdbr-intro.html; https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383; https://jackbibby1.github.io/SCPA/articles/comparing_two_populations.html#some-conclusions-about-scpa)
#devtools::install_github("jackbibby1/SCPA")

# temporary vector of cell types
tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

# vector comprising pathways
library(msigdbr)
library(SCPA)

#pws <- c("bp", "cc", "mf", "kegg", "reactome")
#pathways <- msigdbr("Mus musculus") %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()

pws <- c("bp")
pathways <- msigdbr("Mus musculus") %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()


scpa_wt <- seurat_extract(all_integrated_rnm_VlnPlot, 
                            meta1 = "group", value_meta1 = "wt") #https://jackbibby1.github.io/SCPA/articles/systematic_tissue_comparison.html
  
scpa_il27raKO <- seurat_extract(all_integrated_rnm_VlnPlot, 
                          meta1 = "group", value_meta1 = "il27raKO")
  
scpa_out_wt_il27rako <- compare_pathways(samples = list(scpa_wt, scpa_il27raKO), pathways = pathways) %>% 
    select(Pathway, qval, FC) %>%
    set_colnames(c("Pathway", paste(i, "qval", sep = "_"), paste(i, "FC", sep = "_")))
  #fold change in pathways calculated from population 1 - population 2 (https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383)

  
#writexl::write_xlsx(scpa_out_wt_il27rako, "scpa_out_wt_il27rako.xlsx")
#saveRDS(scpa_out_wt_il27rako, "scpa_out_wt_il27rako.rds")
#scpa_out_wt_il27rako <- readRDS("scpa_out_wt_il27rako.rds")


scpa_out_fig <- scpa_out_wt_il27rako %>% 
  purrr::reduce(full_join, by = "Pathway") %>%  
  set_colnames(gsub(colnames(.), pattern = " ", replacement = "_")) %>%
  select(c("Pathway", grep("_FC", colnames(.), grep("_qval", colnames(.))))) %>%
  filter_all(any_vars(qval > 2)) %>%
  column_to_rownames("Pathway") #specify purrr::reduce to prevent error (Source code: https://stackoverflow.com/questions/66776517/how-do-i-solve-the-problem-with-reduce-function-in-r)



paths_to_visualize <- bind_rows(scpa_out_wt_il27rako) %>% filter(grepl("HALLMARK_", Pathway)) %>% select("Pathway") %>% unique() %>% unlist()


position <- which(rownames(scpa_out_fig) %in% paths_to_visualize)
row_an <- rowAnnotation(Genes = anno_mark(at = which(rownames(scpa_out_fig) %in% paths_to_visualize),
                                          labels = rownames(scpa_out_fig)[position],
                                          labels_gp = gpar(fontsize = 7),
                                          link_width = unit(2.5, "mm"),
                                          padding = unit(1, "mm"),
                                          link_gp = gpar(lwd = 0.5)))

col_hm <- colorRamp2(colors = c("blue", "white", "red"), breaks = c(-10, 0, 10))



scpa_htmap_fc <- ComplexHeatmap::Heatmap(scpa_out_fig,
                        col = col_hm,
                        name = "Fold-change",
                        show_row_names = FALSE,
                        right_annotation = row_an,
                        row_names_gp = gpar(fontsize = 3),
                        column_names_gp = gpar(fontsize = 8),
                        border = TRUE,
                        column_km = 0,
                        row_km = 0,
                        column_labels = tmp_celltype_list,
                        height = unit(0.5, "mm")*nrow(scpa_out_fig), width = unit(5, "mm")*ncol(scpa_out_fig)
                        )



#calculate heatmap figure size
#Source code: https://jokergoo.github.io/2020/05/11/set-cell-width/height-in-the-heatmap/

calc_ht_size = function(ht, unit = "inch") {
    #pdf(NULL)
    ht = draw(ht)
    w = ComplexHeatmap:::width(ht)
    w = convertX(w, unit, valueOnly = TRUE)
    h = ComplexHeatmap:::height(ht)
    h = convertY(h, unit, valueOnly = TRUE)
    #dev.off()

    c(w, h)
}


size = calc_ht_size(scpa_htmap_fc)

#pdf(paste0(fig_path, "tmp2_SCPA_pathway_wt_vs_KO_FC.pdf"), width = size[1], height = size[2])

draw(scpa_htmap_fc,
    heatmap_legend_side = 'right',
    annotation_legend_side = 'right',
    row_sub_title_side = 'left',
    row_title_side = "right")

##dev.off()



```


#separting based on celltype and group
```{r}
# Performing SCPA analysis (Source vignette: https://jackbibby1.github.io/SCPA/articles/disease_comparison.html; https://jackbibby1.github.io/SCPA/articles/systematic_tissue_comparison.html; https://igordot.github.io/msigdbr/articles/msigdbr-intro.html; https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383; https://jackbibby1.github.io/SCPA/articles/comparing_two_populations.html#some-conclusions-about-scpa)
#devtools::install_github("jackbibby1/SCPA")

# temporary vector of cell types
#tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

nk_cells <- subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"))
Idents(nk_cells) <- "celltype"
nk_cells$celltype <- droplevels(x = nk_cells$celltype) #(drop unused cell type labels ; Reference: https://github.com/satijalab/seurat/issues/1385)
nk_cells$cluster0.8 <- droplevels(x = nk_cells$cluster0.8) 
Idents(nk_cells) <- "celltype" 
Idents(nk_cells) <- "group"

# vector comprising pathways
library(msigdbr)
library(SCPA)

#pws <- c("bp", "cc", "mf", "kegg", "reactome")
#pathways <- msigdbr("Mus musculus") %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()

pws <- c("bp")
pathways <- msigdbr("Mus musculus") %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()

nk_cell_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")

scpa_out_wt_il27rako_nk <- list()

for (i in nk_cell_list) {
  
  wt_nk <- seurat_extract(subset(nk_cells, group %in% "wt"), 
                            meta1 = "celltype", value_meta1 = i)
  
  il27raKO_nk <- seurat_extract(subset(nk_cells, group %in% "il27raKO"), 
                          meta1 = "celltype", value_meta1 = i)
  
  print(paste("comparing", i))
  scpa_out_wt_il27rako_nk[[i]] <- compare_pathways(samples = list(wt_nk, il27raKO_nk), pathways = pathways) %>% 
    select(Pathway, qval, FC) %>%
    set_colnames(c("Pathway", paste(i, "qval", sep = "_"), paste(i, "FC", sep = "_")))
  #fold change in pathways calculated from population 1 - population 2 (https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383)
}

  
#writexl::write_xlsx(scpa_out_wt_il27rako_nk, "scpa_out_wt_il27rako_nk.xlsx")
#saveRDS(scpa_out_wt_il27rako_nk, "scpa_out_wt_il27rako_nk.rds")
#scpa_out_wt_il27rako_nk <- readRDS("scpa_out_wt_il27rako_nk.rds")


scpa_out_fig_nk <- scpa_out_wt_il27rako_nk %>% 
  purrr::reduce(full_join, by = "Pathway") %>%  
  set_colnames(gsub(colnames(.), pattern = " ", replacement = "_")) %>%
  select(c("Pathway", grep("_FC", colnames(.), grep("_qval", colnames(.))))) %>%
  filter_all(any_vars(qval > 2)) %>%
  column_to_rownames("Pathway") #specify purrr::reduce to prevent error (Source code: https://stackoverflow.com/questions/66776517/how-do-i-solve-the-problem-with-reduce-function-in-r)



paths_to_visualize <- bind_rows(scpa_out_wt_il27rako) %>% filter(grepl("HALLMARK_", Pathway)) %>% select("Pathway") %>% unique() %>% unlist()


position <- which(rownames(scpa_out_fig) %in% paths_to_visualize)
row_an <- rowAnnotation(Genes = anno_mark(at = which(rownames(scpa_out_fig) %in% paths_to_visualize),
                                          labels = rownames(scpa_out_fig)[position],
                                          labels_gp = gpar(fontsize = 7),
                                          link_width = unit(2.5, "mm"),
                                          padding = unit(1, "mm"),
                                          link_gp = gpar(lwd = 0.5)))

col_hm <- colorRamp2(colors = c("blue", "white", "red"), breaks = c(-10, 0, 10))



scpa_htmap_fc <- ComplexHeatmap::Heatmap(scpa_out_fig,
                        col = col_hm,
                        name = "Fold-change",
                        show_row_names = FALSE,
                        right_annotation = row_an,
                        row_names_gp = gpar(fontsize = 3),
                        column_names_gp = gpar(fontsize = 8),
                        border = TRUE,
                        column_km = 0,
                        row_km = 0,
                        column_labels = tmp_celltype_list,
                        height = unit(0.5, "mm")*nrow(scpa_out_fig), width = unit(5, "mm")*ncol(scpa_out_fig)
                        )



#calculate heatmap figure size
#Source code: https://jokergoo.github.io/2020/05/11/set-cell-width/height-in-the-heatmap/

calc_ht_size = function(ht, unit = "inch") {
    #pdf(NULL)
    ht = draw(ht)
    w = ComplexHeatmap:::width(ht)
    w = convertX(w, unit, valueOnly = TRUE)
    h = ComplexHeatmap:::height(ht)
    h = convertY(h, unit, valueOnly = TRUE)
    #dev.off()

    c(w, h)
}


size = calc_ht_size(scpa_htmap_fc)

#pdf(paste0(fig_path, "tmp2_SCPA_pathway_wt_vs_KO_FC.pdf"), width = size[1], height = size[2])

draw(scpa_htmap_fc,
    heatmap_legend_side = 'right',
    annotation_legend_side = 'right',
    row_sub_title_side = 'left',
    row_title_side = "right")

##dev.off()



```



# 22Apr2022:UMAP dimplot for Johannes-conference
```{r}
tmp <- all_integrated_rnm
tmp$cluster_label <- tmp$sub.cluster
Idents(tmp) <- "cluster_label"
tmp <-  RenameIdents(object = tmp,
                      "0" = "0", 
                      "2" = "0", 
                     "1" = "1", 
                      "3" = "2", 
                      "4" = "2", 
                      "5" = "3", 
                      "6" = "4", 
                      "7" = "5", 
                      "8" = "6", 
                      "9" = "7", 
                      "10" = "7", 
                      "11" = "8", 
                      "12" = "9", 
                      "13" = "10", 
                      "14" = "11", 
                      "15" = "3", 
                      "16" = "11", 
                      "17" = "12", 
                      "18_0" = "13", 
                      "18_1" = "14", 
                      "18_2" = "15", 
                      "19" = "16", 
                      "20" = "17" 
                                   )
tmp$cluster_label <- Idents(tmp)


#pdf(paste0(fig_path, "/Johannes_ForConference_DimPlot_all_integrated_rnm_res0.8.pdf"), width = 8, height = 6)
DimPlot(tmp, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 5, split.by = "group") + NoLegend() + theme(axis.text = element_text(size = rel(1.25)), axis.title = element_text(size = rel(1.25)), )
##dev.off()




```


##================= Upd 9-11 May 2022: plots and analysis based on Prof. Bosmann's inputs===============================#
#1. Violin plots of Ifi208, Wdfy1, CD365, Ly49H, Ly49D, Gzmb, Satb1, adt-CD49d, adt-CD127, adt-CD314, adt-CD134, adt-CD124 

```{r}
#1. Violin plots of Ifi208, Wdfy1, CD365, Ly49H, Ly49D

#pdf(paste0(fig_path, "VlnPlot_ADT_Ifi208_Wdfy1_CD365_Ly49H_Ly49D.pdf"), height = 6, width = 6)
#VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Ifi208"), split.by = "group", assay = "RNA", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

#VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Wdfy1"), split.by = "group", assay = "RNA", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

#VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD365"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

#VlnPlot(all_integrated_rnm_VlnPlot, c("adt-Ly49H"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

#VlnPlot(all_integrated_rnm_VlnPlot, c("adt-Ly49D"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))
##dev.off()


#1. Violin plots of Ifi208, Wdfy1, CD365, Ly49H, Ly49D, Gzmb, Satb1, adt-CD49d, adt-CD127, adt-CD314, adt-CD134, adt-CD124
#pdf(paste0(fig_path, "VlnPlot_SelectedGenes_ADT_11May2022.pdf"), height = 6, width = 6)
VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Ifi208"), split.by = "group", assay = "RNA", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Wdfy1"), split.by = "group", assay = "RNA", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD365"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-Ly49H"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-Ly49D"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Gzmb"), split.by = "group", assay = "RNA", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("rna_Satb1"), split.by = "group", assay = "RNA", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD49d"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD127"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD314"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD134"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

VlnPlot(all_integrated_rnm_VlnPlot, c("adt-CD124"), split.by = "group", assay = "ADT", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))
##dev.off()

```


#2. Average expression levels of differentially expressed genes and ADT/proteins (DGE and DPE based on Seurat)
- Average non-log expression values of normalized counts
```{r}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm_VlnPlot #assign seurat object to a temporary variable

cluster_Average_exp_RNA <- list()

#RNA
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  genes_list <- differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type == i]
  cluster.averages <- AverageExpression(tmp_clus, features = genes_list, slot = "data", assays = "RNA")
  cluster.averages <- as.data.frame(cluster.averages$RNA)
  cluster.averages$Celltype <- i
  cluster.averages$gene <- row.names(cluster.averages)
  cluster_Average_exp_RNA[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_RNA, paste0(tab_path, "cluster_average_Exp_differentialRNA_il27raKO_vs_wt.xlsx"))


#ADT
cluster_Average_exp_ADT <- list()

for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  adt_list <- differential_ADT_all_integrated_rnm$gene[differential_ADT_all_integrated_rnm$cell_type == i]
  cluster.averages <- AverageExpression(tmp_clus, features = adt_list, slot = "data", assays = "ADT")
  cluster.averages <- as.data.frame(cluster.averages$ADT)
  cluster.averages$Celltype <- i
  cluster.averages$protein <- row.names(cluster.averages)
  cluster_Average_exp_ADT[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_ADT, paste0(tab_path, "cluster_average_Exp_differentialADT_il27raKO_vs_wt.xlsx"))


```


#3. Vennplot of DE genes (based on Seurat) from NK and T cell clusters
## Genes and ADT/proteins common or uniquely up or downregulated between NK and T cells
```{r}
suppressPackageStartupMessages({
library(ggpubr)
library(VennDetail)
library(VennDiagram)
})

#RNA
# Combine all NK clusters into one and all T cell clusters into one

NK_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

Tcell_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM", "Treg", "CD27+_gdT", "CD27-_gdT"), ]

vennPlot_DGE_NK_vs_T_upreg <- venndetail(list(
  NK_up = NK_genes$gene[NK_genes$avg_log2FC > 0 & NK_genes$p_val_adj < 0.05], #subset to upregulated genes (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
  T_up = Tcell_genes$gene[Tcell_genes$avg_log2FC > 0 & Tcell_genes$p_val_adj < 0.05]
))


vennPlot_DGE_NK_vs_T_downreg <- venndetail(list(
  NK_down = NK_genes$gene[NK_genes$avg_log2FC < 0 & NK_genes$p_val_adj < 0.05], #subset to downregulated genes (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  T_down = Tcell_genes$gene[Tcell_genes$avg_log2FC < 0 & Tcell_genes$p_val_adj < 0.05]

  ))

#ADT
# Combine all NK clusters into one and all T cell clusters into one

NK_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

Tcell_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM", "Treg", "CD27+_gdT", "CD27-_gdT"), ]

vennPlot_DPE_NK_vs_T_upreg <- venndetail(list(
  NK_up = NK_ADT$gene[NK_ADT$avg_log2FC > 0 & NK_ADT$p_val_adj < 0.05], #subset to upregulated proteins/ADT (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
  T_up = Tcell_ADT$gene[Tcell_ADT$avg_log2FC > 0 & Tcell_ADT$p_val_adj < 0.05]
))


vennPlot_DPE_NK_vs_T_downreg <- venndetail(list(
  NK_down = NK_ADT$gene[NK_ADT$avg_log2FC < 0 & NK_ADT$p_val_adj < 0.05], #subset to downregulated proteins/ADT (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  T_down = Tcell_ADT$gene[Tcell_ADT$avg_log2FC < 0 & Tcell_ADT$p_val_adj < 0.05]

  ))



#pdf(paste0(fig_path, "UpsetPlot_NK_vs_T.pdf"), height = 8, width = 6)
plot(vennPlot_DGE_NK_vs_T_upreg, type = "upset",  mainbar.y.label = paste0("Upregulated genes in IL27RAKO vs WT in NK cells and T cells\nIntersection Size\n"))
plot(vennPlot_DGE_NK_vs_T_downreg, type = "upset", mainbar.y.label = paste0("Downregulated genes in IL27RAKO vs WT in NK cells and T cells\nIntersection Size\n"))

plot(vennPlot_DPE_NK_vs_T_upreg, type = "upset",  mainbar.y.label = paste0("Upregulated ADT in IL27RAKO vs WT in NK cells and T cells\nIntersection Size\n"))
plot(vennPlot_DPE_NK_vs_T_downreg, type = "upset", mainbar.y.label = paste0("Downregulated ADT in IL27RAKO vs WT in NK cells and T cells\nIntersection Size\n"))
##dev.off()


vennPlot_list <- list(vennPlot_DGE_NK_vs_T_upreg = result(vennPlot_DGE_NK_vs_T_upreg, wide = TRUE), vennPlot_DGE_NK_vs_T_downreg = result(vennPlot_DGE_NK_vs_T_downreg, wide = TRUE),
                      vennPlot_DPE_NK_vs_T_upreg = result(vennPlot_DPE_NK_vs_T_upreg, wide = TRUE), vennPlot_DPE_NK_vs_T_downreg = result(vennPlot_DPE_NK_vs_T_downreg, wide = TRUE))

#writexl::write_xlsx(vennPlot_list, paste0(tab_path, "VennPlotdata_NK_vs_T.xlsx") )



# Venn diagram: RNA
vennDiagram_NK_vs_T_RNA <- draw.pairwise.venn(area1 = 22, area2 = 20, cross.area = 11, category = c("NK cells\n", "T cells\n"), lty = 1, fill = "white", alpha = 0.5, label.col = "white", cat.fontface = "plain", cat.cex = 1, lwd = 2, cat.fontfamily = "Arial", margin = 0.15, cat.dist = c(0.10, 0.10))

figure_vennDiagram_NK_vs_T_RNA <- ggarrange(vennDiagram_NK_vs_T_RNA)

figure_vennDiagram_NK_vs_T_RNA <- figure_vennDiagram_NK_vs_T_RNA + 
  annotate("text", x = 0.313, y = 0.50, label = paste0(sprintf('\u21e9'), "12"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n1 NK cells IL27RAKO vs WT upreg
  annotate("text", x = 0.30, y = 0.54, label = paste0(sprintf('\u21e7'), "6"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n1 NK cells IL27RAKO vs WT downreg
  
  annotate("text", x = 0.69, y = 0.50, label = paste0(sprintf('\u21e9'), "48"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n2 T cells IL27RAKO vs WT upreg
  annotate("text", x = 0.69, y = 0.54, label = paste0(sprintf('\u21e7'), "86"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n2 T cells IL27RAKO vs WT downreg
   
  annotate("text", x = 0.52, y = 0.50, label = paste0(sprintf('\u21e9'), "13"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n12 NK - T cells IL27RAKO vs WT upreg
  annotate("text", x = 0.51, y = 0.54, label = paste0(sprintf('\u21e7'), "8"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n12 NK - T cells IL27RAKO vs WT downreg
  
   annotate("text", x = 0.83, y = 0.22, label = paste0(sprintf('\u21e9'), "Down"), size = 4, colour = colorRampPalette(c("navy"))(50)) + 
  annotate("text", x = 0.805, y = 0.26, label = paste0(sprintf('\u21e7'), "Up"), size = 4, colour = colorRampPalette(c("red"))(50)) 

annotate_figure(figure_vennDiagram_NK_vs_T_RNA, top = text_grob("Differentially expressed genes:\nIL27RAKO vs WT\n", color = "black", face = "bold", size = 10) ) #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html

#ggsave(paste0(fig_path, "VennDiagram_NK_vs_T_DGE_RNA.pdf"), units ="in", width = 4, height = 4, dpi = 600, device = cairo_pdf)


# Venn diagram: ADT
vennDiagram_NK_vs_T_ADT <- draw.pairwise.venn(area1 = 22, area2 = 20, cross.area = 11, category = c("NK cells\n", "T cells\n"), lty = 1, fill = "white", alpha = 0.5, label.col = "white", cat.fontface = "plain", cat.cex = 1, lwd = 2, cat.fontfamily = "Arial", margin = 0.15, cat.dist = c(0.10, 0.10))

figure_vennDiagram_NK_vs_T_ADT <- ggarrange(vennDiagram_NK_vs_T_ADT) 

figure_vennDiagram_NK_vs_T_ADT <- figure_vennDiagram_NK_vs_T_ADT + 
  annotate("text", x = 0.31, y = 0.50, label = paste0(sprintf('\u21e9'), "35"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n1 NK cells IL27RAKO vs WT upreg
  annotate("text", x = 0.31, y = 0.54, label = paste0(sprintf('\u21e7'), "14"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n1 NK cells IL27RAKO vs WT downreg
  
  annotate("text", x = 0.70, y = 0.50, label = paste0(sprintf('\u21e9'), "9"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n2 T cells IL27RAKO vs WT upreg
  annotate("text", x = 0.71, y = 0.54, label = paste0(sprintf('\u21e7'), "17"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n2 T cells IL27RAKO vs WT downreg
   
  annotate("text", x = 0.50, y = 0.50, label = paste0(sprintf('\u21e9'), "46"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n12 NK - T cells IL27RAKO vs WT upreg
  annotate("text", x = 0.50, y = 0.54, label = paste0(sprintf('\u21e7'), "26"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n12 NK - T cells IL27RAKO vs WT downreg
  
   annotate("text", x = 0.83, y = 0.22, label = paste0(sprintf('\u21e9'), "Down"), size = 4, colour = colorRampPalette(c("navy"))(50)) + 
  annotate("text", x = 0.805, y = 0.26, label = paste0(sprintf('\u21e7'), "Up"), size = 4, colour = colorRampPalette(c("red"))(50)) 

annotate_figure(figure_vennDiagram_NK_vs_T_ADT, top = text_grob("Differentially expressed ADT (proteins):\nIL27RAKO vs WT\n", color = "black", face = "bold", size = 10) )

#ggsave(paste0(fig_path, "VennDiagram_NK_vs_T_DGE_ADT.pdf"), units ="in", width = 4, height = 4, dpi = 600, device = cairo_pdf)


#Upd: 14March2023
#Unique and commonly DE in NK and T cells
#Uniquely DE in NK cells
NK_vs_T_DE_up <- vennPlot_DGE_NK_vs_T_upreg@wide$Detail[vennPlot_DGE_NK_vs_T_upreg@wide$NK_up == 1 & !(vennPlot_DGE_NK_vs_T_upreg@wide$T_up == 1 )]
NK_vs_T_DE_down <- vennPlot_DGE_NK_vs_T_downreg@wide$Detail[vennPlot_DGE_NK_vs_T_downreg@wide$NK_down == 1 & !(vennPlot_DGE_NK_vs_T_downreg@wide$T_down == 1 )]


NK_vs_T_DPE_up <- vennPlot_DPE_NK_vs_T_upreg@wide$Detail[vennPlot_DPE_NK_vs_T_upreg@wide$NK_up == 1 & !(vennPlot_DPE_NK_vs_T_upreg@wide$T_up == 1 )]
NK_vs_T_DPE_down <- vennPlot_DPE_NK_vs_T_downreg@wide$Detail[vennPlot_DPE_NK_vs_T_downreg@wide$NK_down == 1 & !(vennPlot_DPE_NK_vs_T_downreg@wide$T_down == 1 )]


#Uniquely DE in T cells
T_vs_NK_DE_up <- vennPlot_DGE_NK_vs_T_upreg@wide$Detail[!(vennPlot_DGE_NK_vs_T_upreg@wide$NK_up == 1) & vennPlot_DGE_NK_vs_T_upreg@wide$T_up == 1]
T_vs_NK_DE_down <- vennPlot_DGE_NK_vs_T_downreg@wide$Detail[!(vennPlot_DGE_NK_vs_T_downreg@wide$NK_down == 1) & vennPlot_DGE_NK_vs_T_downreg@wide$T_down == 1]


T_vs_NK_DPE_up <- vennPlot_DPE_NK_vs_T_upreg@wide$Detail[!(vennPlot_DPE_NK_vs_T_upreg@wide$NK_up == 1) & vennPlot_DPE_NK_vs_T_upreg@wide$T_up == 1]
T_vs_NK_DPE_down <- vennPlot_DPE_NK_vs_T_downreg@wide$Detail[!(vennPlot_DPE_NK_vs_T_downreg@wide$NK_down == 1) & vennPlot_DPE_NK_vs_T_downreg@wide$T_down == 1]



unique_NK_vs_T_DEG <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") & differential_genes_all_integrated_rnm$gene %in%c(NK_vs_T_DE_up, NK_vs_T_DE_down),  ] 

unique_NK_vs_T_DEP <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") & differential_ADT_all_integrated_rnm$gene %in%c(NK_vs_T_DE_up, NK_vs_T_DE_down),  ] 



unique_T_vs_NK_DEG <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & differential_genes_all_integrated_rnm$gene %in% c(T_vs_NK_DE_up, T_vs_NK_DE_down),  ]

unique_T_vs_NK_DEP <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & differential_ADT_all_integrated_rnm$gene %in% c(T_vs_NK_DE_up, T_vs_NK_DE_down),  ]


#clusterProfiler enrichment
library(clusterProfiler)
goBP_unique_NK_vs_T_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(unique_NK_vs_T_DEG$gene, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )

goBP_unique_T_vs_NK_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(unique_T_vs_NK_DEG$gene, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )

#-------
goBP_unique_NK_vs_T_up_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(NK_vs_T_DE_up, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )
goBP_unique_NK_vs_T_down_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(NK_vs_T_DE_down, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )


goBP_unique_T_vs_NK_up_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(T_vs_NK_DE_up, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )


goBP_unique_T_vs_NK_down_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(T_vs_NK_DE_down, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )


# writexl::write_xlsx(list("goBP_NK_vs_T_up_DEG" = goBP_unique_NK_vs_T_up_DEG@result,
#                          "goBP_NK_vs_T_down_DEG" = goBP_unique_NK_vs_T_down_DEG@result,
#                          "goBP_T_vs_NK_up_DEG" = goBP_unique_T_vs_NK_up_DEG@result,
#                          "goBP_T_vs_NK_down_DEG" = goBP_unique_T_vs_NK_down_DEG@result),
#                     "clusterProfiler_GOBP_NKcellSpecific_TcellSpecific_DE_March2023.xlsx")




#pdf(paste0(fig_path, "ClusterProfiler_GO_BP_NK_vs_T.pdf"), height = 9, width = 7)
barplot(goBP_unique_NK_vs_T_up_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\nupregulated in NK cells vs T cells\n")

barplot(goBP_unique_NK_vs_T_down_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\ndownregulated in NK cells vs T cells\n")

barplot(goBP_unique_T_vs_NK_up_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\nupregulated in T cells vs NK cells \n")

barplot(goBP_unique_T_vs_NK_down_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\ndownregulated in T cells vs NK cells\n")
##dev.off()





#topGO enrichement
library(pcaExplorer)
library(topGO)
library(org.Mm.eg.db)
library(AnnotationDbi)

x <- pcaExplorer::topGOtable(DEgenes =  NK_vs_T_DE_up, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)

y <- pcaExplorer::topGOtable(DEgenes =  NK_vs_T_DE_down, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)

p <- pcaExplorer::topGOtable(DEgenes =  T_vs_NK_DE_up, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)

q <- pcaExplorer::topGOtable(DEgenes =  T_vs_NK_DE_down, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)


# writexl::write_xlsx(list("goBP_NK_vs_T_up_DEG" = x,
#                          "goBP_NK_vs_T_down_DEG" = y,
#                          "goBP_T_vs_NK_up_DEG" = p,
#                          "goBP_T_vs_NK_down_DEG" = q),
#                     "topGO_GOBP_NKcellSpecific_TcellSpecific_DE_March2023.xlsx")



#pdf(paste0(fig_path, "topGO_GO_BP_NK_vs_T.pdf"), height = 9, width = 15)
ggplot(x[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\nupregulated in NK cells vs T cells\n")

ggplot(y[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\ndownregulated in NK cells vs T cells\n")

ggplot(p[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\nupregulated in T cells vs NK cells \n")

ggplot(q[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\ndownregulated in T cells vs NK cells\n")
##dev.off()


```


## Genes and ADT/proteins (based on Seurat) upregulated in NK and downregulated in T cells and vice versa
```{r}

#RNA
#genes upregulated in NK cells and downregulated in T cells
vennPlot_DGE_NK_up_vs_T_down <- venndetail(list(
  NK_up = NK_genes$gene[NK_genes$avg_log2FC > 0 & NK_genes$p_val_adj < 0.05], #subset to upregulated genes (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
  T_down = Tcell_genes$gene[Tcell_genes$avg_log2FC < 0 & Tcell_genes$p_val_adj < 0.05]
))

#genes downregulated in NK cells and upregulated in T cells
vennPlot_DGE_NK_down_vs_T_up <- venndetail(list(
  NK_down = NK_genes$gene[NK_genes$avg_log2FC < 0 & NK_genes$p_val_adj < 0.05], #subset to downregulated genes (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  T_up = Tcell_genes$gene[Tcell_genes$avg_log2FC > 0 & Tcell_genes$p_val_adj < 0.05]

  ))

#ADT

#ADT/proteins upregulated in NK cells and downregulated in T cells
vennPlot_DPE_NK_up_vs_T_down <- venndetail(list(
  NK_up = NK_ADT$gene[NK_ADT$avg_log2FC > 0 & NK_ADT$p_val_adj < 0.05], #subset to upregulated proteins/ADT (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
  T_down = Tcell_ADT$gene[Tcell_ADT$avg_log2FC < 0 & Tcell_ADT$p_val_adj < 0.05]
))


#ADT/proteins downregulated in NK cells and upregulated in T cells
vennPlot_DPE_NK_vs_T_downreg <- venndetail(list(
  NK_down = NK_ADT$gene[NK_ADT$avg_log2FC < 0 & NK_ADT$p_val_adj < 0.05], #subset to downregulated proteins/ADT (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  T_up = Tcell_ADT$gene[Tcell_ADT$avg_log2FC > 0 & Tcell_ADT$p_val_adj < 0.05]

  ))



#pdf(paste0(fig_path, "UpsetPlot_Inversely_correlated_NK_vs_T.pdf"), height = 8, width = 6)
plot(vennPlot_DGE_NK_up_vs_T_down, type = "upset",  mainbar.y.label = paste0("Upregulated genes in IL27RAKO vs WT in NK cells\nand downregulated T cells\nIntersection Size\n"))
plot(vennPlot_DGE_NK_down_vs_T_up, type = "upset", mainbar.y.label = paste0("Downregulated genes in IL27RAKO vs WT in NK cells\n and upregulated in T cells\nIntersection Size\n"))

plot(vennPlot_DPE_NK_up_vs_T_down, type = "upset",  mainbar.y.label = paste0("Upregulated ADT in IL27RAKO vs WT in NK cells\nand downregulated in T cells\nIntersection Size\n"))
plot(vennPlot_DPE_NK_vs_T_downreg, type = "upset", mainbar.y.label = paste0("Downregulated ADT in IL27RAKO vs WT in NK cells\nand upregulated in T cells\nIntersection Size\n"))
##dev.off()


vennPlot_list_inverse_correlated <- list(vennPlot_DGE_NK_up_vs_T_down = result(vennPlot_DGE_NK_up_vs_T_down, wide = TRUE), vennPlot_DGE_NK_down_vs_T_up = result(vennPlot_DGE_NK_down_vs_T_up, wide = TRUE),
                      vennPlot_DPE_NK_up_vs_T_down = result(vennPlot_DPE_NK_up_vs_T_down, wide = TRUE), vennPlot_DPE_NK_vs_T_downreg = result(vennPlot_DPE_NK_vs_T_downreg, wide = TRUE))

#writexl::write_xlsx(vennPlot_list_inverse_correlated, paste0(tab_path, "VennPlotdata_NK_vs_T_InverselyCorrelated.xlsx") )



```

#4. Correlation analysis: NK vs T cells (Seurat)
```{r}
#RNA
#subset the NK_genes and Tcell_genes objects to only include significant genes (adjusted p_val < 0.05) 

NK_genes_subset <- NK_genes[NK_genes$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))
Tcell_genes_subset <- Tcell_genes[Tcell_genes$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))

df_NK_T_corr <- Reduce(function(...) merge(..., by = "gene", all = TRUE), list(
                data.frame(gene = NK_genes_subset$gene,
                          NKcell_log2FC = round(as.numeric(NK_genes_subset$avg_log2FC), 2),
                           stringsAsFactors = FALSE),
                   data.frame(gene = Tcell_genes_subset$gene,
                              Tcell_log2FC = round(as.numeric(Tcell_genes_subset$avg_log2FC), 2),
                              stringsAsFactors = FALSE)
                                     ))



row.names(df_NK_T_corr) <- unique(df_NK_T_corr$gene)
df_NK_T_corr$gene <- NULL
df_NK_T_corr$onecol <- rep("", nrow(df_NK_T_corr))


#pdf(paste0(fig_path, "Correlation_plot_NK_Tcells_SeuratDGE.pdf"))
#Plotting correlation based on avg_log2FC
sp_logFC_NK_T <- ggscatter(df_NK_T_corr, x = "NKcell_log2FC", y = "Tcell_log2FC",
                add = "reg.line",  # Add regression line
                add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
                conf.int = TRUE, # Add confidence interval
                label = row.names(df_NK_T_corr),
                repel = TRUE, 
                font.label = c(14, "plain"), size = 0.8, cor.coef.size = 14
)
# Add correlation coefficient
sp_logFC_NK_T + stat_cor(method = "pearson") + xlab("NK cells (avg_log2FC)") + ylab("T cells (avg_log2FC)") + theme(axis.text = element_text(size = rel(1.05)), axis.title = element_text(size = rel(1.15)))

#dev.off()


#ADT
#subset the NK_genes and Tcell_genes objects to only include significant genes (adjusted p_val < 0.05) 

NK_ADT_subset <- NK_ADT[NK_ADT$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))
Tcell_ADT_subset <- Tcell_ADT[Tcell_ADT$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))

df_NK_T_corr_ADT <- Reduce(function(...) merge(..., by = "gene", all = TRUE), list(
                data.frame(gene = NK_ADT_subset$gene,
                          NKcell_log2FC = round(as.numeric(NK_ADT_subset$avg_log2FC), 2),
                           stringsAsFactors = FALSE),
                   data.frame(gene = Tcell_ADT_subset$gene,
                              Tcell_log2FC = round(as.numeric(Tcell_ADT_subset$avg_log2FC), 2),
                              stringsAsFactors = FALSE)
                                     ))



row.names(df_NK_T_corr_ADT) <- unique(df_NK_T_corr_ADT$gene)
df_NK_T_corr_ADT$gene <- NULL
df_NK_T_corr_ADT$onecol <- rep("", nrow(df_NK_T_corr_ADT))


#pdf(paste0(fig_path, "Correlation_plot_NK_Tcells_SeuratDPE.pdf"), width = 9.5, height = 9)
#Plotting correlation based on avg_log2FC
sp_logFC_NK_T_ADT <- ggscatter(df_NK_T_corr_ADT, x = "NKcell_log2FC", y = "Tcell_log2FC",
                add = "reg.line",  # Add regression line
                add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
                conf.int = TRUE, # Add confidence interval
                label = row.names(df_NK_T_corr_ADT),
                font.label = c(9, "plain"), size = 0.8, cor.coef.size = 14,
                repel = TRUE
                )
# Add correlation coefficient
sp_logFC_NK_T_ADT + stat_cor(method = "pearson") + xlab("NK cells (avg_log2FC)") + ylab("T cells (avg_log2FC)") + theme(axis.text = element_text(size = rel(1.05)), axis.title = element_text(size = rel(1.15)))

#dev.off()




```

#5. Average expression levels of differentially expressed genes and ADT/proteins (DGE and DPE based on muscatDS)
- Average non-log expression values of normalized counts
```{r}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm_VlnPlot #assign seurat object to a temporary variable

cluster_Average_exp_RNA <- list()

dge_muscatDS <- bind_rows(tbl_fil_all_RNA) #unlist and combine the nested muscatDS results

#RNA
for(i in unique(dge_muscatDS$cluster_id)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  genes_list <- dge_muscatDS$gene[dge_muscatDS$cluster_id == i]
  cluster.averages <- AverageExpression(tmp_clus, features = genes_list, slot = "data", assays = "RNA")
  cluster.averages <- as.data.frame(cluster.averages$RNA)
  cluster.averages$Celltype <- i
  cluster.averages$gene <- row.names(cluster.averages)
  cluster_Average_exp_RNA[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_RNA, paste0(tab_path, "cluster_average_Exp_muscatDGE_il27raKO_vs_wt.xlsx"))


#ADT
cluster_Average_exp_ADT <- list()

dpe_muscatDS <- bind_rows(tbl_fil_all_ADT) #unlist and combine the nested muscatDS results

for(i in unique(dpe_muscatDS$cluster_id)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  adt_list <- dpe_muscatDS$gene[dpe_muscatDS$cluster_id == i]
  cluster.averages <- AverageExpression(tmp_clus, features = adt_list, slot = "data", assays = "ADT")
  cluster.averages <- as.data.frame(cluster.averages$ADT)
  cluster.averages$Celltype <- i
  cluster.averages$protein <- row.names(cluster.averages)
  cluster_Average_exp_ADT[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_ADT, paste0(tab_path, "cluster_average_Exp_muscatDPE_il27raKO_vs_wt.xlsx"))


```

#6. Vennplot of DE genes (based on muscatDS) from NK can T cell clusters
## Genes and ADT/proteins common or uniquely up or downregulated between NK and T cells
```{r}
suppressPackageStartupMessages({
library(ggpubr)
library(VennDetail)
library(VennDiagram)
})

#RNA
# Combine all NK clusters into one and all T cell clusters into one

NK_genes <- dge_muscatDS[dge_muscatDS$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

Tcell_genes <- dge_muscatDS[dge_muscatDS$cluster_id %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM", "Treg", "CD27+_gdT", "CD27-_gdT"), ]

vennPlot_DGE_NK_vs_T_upreg <- venndetail(list(
  NK_up = NK_genes$gene[NK_genes$logFC > 0 & NK_genes$p_adj.loc < 0.05], #subset to upregulated genes (logFC > 0) that are statistically significant (p_adj.loc < 0.05)
  T_up = Tcell_genes$gene[Tcell_genes$logFC > 0 & Tcell_genes$p_adj.loc < 0.05]
))


vennPlot_DGE_NK_vs_T_downreg <- venndetail(list(
  NK_down = NK_genes$gene[NK_genes$logFC < 0 & NK_genes$p_adj.loc < 0.05], #subset to downregulated genes (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  T_down = Tcell_genes$gene[Tcell_genes$logFC < 0 & Tcell_genes$p_adj.loc < 0.05]

  ))

#ADT : No statistically significant ADTs were present in the muscat DS analysis of NK cells
# Combine all NK clusters into one and all T cell clusters into one

#NK_ADT <- dpe_muscatDS[dpe_muscatDS$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

#Tcell_ADT <- dpe_muscatDS[dpe_muscatDS$cluster_id %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM", "Treg", "CD27+_gdT", "CD27-_gdT"), ]

#vennPlot_DPE_NK_vs_T_upreg <- venndetail(list(
#  NK_up = NK_ADT$gene[NK_ADT$logFC > 0 & NK_ADT$p_adj.loc < 0.05], #subset to upregulated proteins/ADT (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
#  T_up = Tcell_ADT$gene[Tcell_ADT$logFC > 0 & Tcell_ADT$p_adj.loc < 0.05]
#))


#vennPlot_DPE_NK_vs_T_downreg <- venndetail(list(
#  NK_down = NK_ADT$gene[NK_ADT$logFC < 0 & NK_ADT$p_adj.loc < 0.05], #subset to downregulated proteins/ADT (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
#  T_down = Tcell_ADT$gene[Tcell_ADT$logFC < 0 & Tcell_ADT$p_adj.loc < 0.05]
# ))



#pdf(paste0(fig_path, "UpsetPlot_NK_vs_T_muscatDS_DGE.pdf"), height = 8, width = 6)
plot(vennPlot_DGE_NK_vs_T_upreg, type = "upset",  mainbar.y.label = paste0("muscatDS\nUpregulated genes in IL27RAKO vs WT in NK cells and T cells\nIntersection Size\n"))
plot(vennPlot_DGE_NK_vs_T_downreg, type = "upset", mainbar.y.label = paste0("muscatDS\nDownregulated genes in IL27RAKO vs WT in NK cells and T cells\nIntersection Size\n"))
##dev.off()


vennPlot_list <- list(vennPlot_DGE_NK_vs_T_upreg = result(vennPlot_DGE_NK_vs_T_upreg, wide = TRUE), vennPlot_DGE_NK_vs_T_downreg = result(vennPlot_DGE_NK_vs_T_downreg, wide = TRUE))

#writexl::write_xlsx(vennPlot_list, paste0(tab_path, "VennPlotdata_NK_vs_T_muscatDS_DGE.xlsx") )



# Venn diagram: RNA
vennDiagram_NK_vs_T_RNA <- draw.pairwise.venn(area1 = 22, area2 = 20, cross.area = 11, category = c("NK cells\n", "T cells\n"), lty = 1, fill = "white", alpha = 0.5, label.col = "white", cat.fontface = "plain", cat.cex = 1, lwd = 2, cat.fontfamily = "Arial", margin = 0.15, cat.dist = c(0.10, 0.10))

figure_vennDiagram_NK_vs_T_RNA <- ggarrange(vennDiagram_NK_vs_T_RNA)

figure_vennDiagram_NK_vs_T_RNA <- figure_vennDiagram_NK_vs_T_RNA + 
  annotate("text", x = 0.30, y = 0.50, label = paste0(sprintf('\u21e9'), "35"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n1 NK cells IL27RAKO vs WT upreg
  annotate("text", x = 0.30, y = 0.54, label = paste0(sprintf('\u21e7'), "38"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n1 NK cells IL27RAKO vs WT downreg
  
  annotate("text", x = 0.70, y = 0.50, label = paste0(sprintf('\u21e9'), "62"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n2 T cells IL27RAKO vs WT upreg
  annotate("text", x = 0.70, y = 0.54, label = paste0(sprintf('\u21e7'), "40"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n2 T cells IL27RAKO vs WT downreg
   
  annotate("text", x = 0.51, y = 0.50, label = paste0(sprintf('\u21e9'), "20"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n12 NK - T cells IL27RAKO vs WT upreg
  annotate("text", x = 0.51, y = 0.54, label = paste0(sprintf('\u21e7'), "14"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n12 NK - T cells IL27RAKO vs WT downreg
  
   annotate("text", x = 0.83, y = 0.22, label = paste0(sprintf('\u21e9'), "Down"), size = 4, colour = colorRampPalette(c("navy"))(50)) + 
  annotate("text", x = 0.805, y = 0.26, label = paste0(sprintf('\u21e7'), "Up"), size = 4, colour = colorRampPalette(c("red"))(50)) 

annotate_figure(figure_vennDiagram_NK_vs_T_RNA, top = text_grob("Differentially expressed genes (muscatDS):\nIL27RAKO vs WT\n", color = "black", face = "bold", size = 10) ) #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html

#ggsave(paste0(fig_path, "VennDiagram_NK_vs_T_DGE_muscatDSRNA.pdf"), units ="in", width = 4, height = 4, dpi = 600, device = cairo_pdf)

```

## Genes and ADT/proteins (based on muscat) upregulated in NK and downregulated in T cells and vice versa
```{r}

#RNA
#genes upregulated in NK cells and downregulated in T cells
vennPlot_DGE_NK_up_vs_T_down <- venndetail(list(
  NK_up = NK_genes$gene[NK_genes$logFC > 0 & NK_genes$p_adj.loc < 0.05], #subset to upregulated genes (logFC > 0) that are statistically significant (p_adj.loc)
  T_down = Tcell_genes$gene[Tcell_genes$logFC < 0 & Tcell_genes$p_adj.loc < 0.05]
))

#genes downregulated in NK cells and upregulated in T cells
vennPlot_DGE_NK_down_vs_T_up <- venndetail(list(
  NK_down = NK_genes$gene[NK_genes$logFC < 0 & NK_genes$p_adj.loc < 0.05], #subset to downregulated genes (logFC < 0) that are statistically significant (p_adj.loc < 0.05)
  T_up = Tcell_genes$gene[Tcell_genes$logFC > 0 & Tcell_genes$p_adj.loc < 0.05]

  ))

#ADT: There are no statistically significant ADTs based on muscatDS analysis in the NK cell clusters.

#ADT/proteins upregulated in NK cells and downregulated in T cells
#vennPlot_DPE_NK_up_vs_T_down <- venndetail(list(
#  NK_up = NK_ADT$gene[NK_ADT$avg_log2FC > 0 & NK_ADT$p_val_adj < 0.05], #subset to upregulated proteins/ADT (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
#  T_down = Tcell_ADT$gene[Tcell_ADT$avg_log2FC < 0 & Tcell_ADT$p_val_adj < 0.05]
#))


#ADT/proteins downregulated in NK cells and upregulated in T cells
#vennPlot_DPE_NK_vs_T_downreg <- venndetail(list(
#  NK_down = NK_ADT$gene[NK_ADT$avg_log2FC < 0 & NK_ADT$p_val_adj < 0.05], #subset to downregulated proteins/ADT (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
 # T_up = Tcell_ADT$gene[Tcell_ADT$avg_log2FC > 0 & Tcell_ADT$p_val_adj < 0.05]
#  ))



#pdf(paste0(fig_path, "UpsetPlot_Inversely_correlated_NK_vs_T_muscatDSDGE.pdf"), height = 8, width = 6)
plot(vennPlot_DGE_NK_up_vs_T_down, type = "upset",  mainbar.y.label = paste0("muscatDS:\nUpregulated genes in IL27RAKO vs WT in NK cells\nand downregulated T cells\nIntersection Size\n"))
plot(vennPlot_DGE_NK_down_vs_T_up, type = "upset", mainbar.y.label = paste0("muscatDS:\nDownregulated genes in IL27RAKO vs WT in NK cells\n and upregulated in T cells\nIntersection Size\n"))

##dev.off()


vennPlot_list_inverse_correlated <- list(vennPlot_DGE_NK_up_vs_T_down = result(vennPlot_DGE_NK_up_vs_T_down, wide = TRUE), vennPlot_DGE_NK_down_vs_T_up = result(vennPlot_DGE_NK_down_vs_T_up, wide = TRUE))

#writexl::write_xlsx(vennPlot_list_inverse_correlated, paste0(tab_path, "VennPlotdata_NK_vs_T_InverselyCorrelated_muscatDSDGE.xlsx") )



```

#7. Correlation analysis: NK vs T cells (based on muscatDS)
```{r}

#subset the NK_genes and Tcell_genes objects to only include significant genes (adjusted p_adj.loc < 0.05) 

NK_genes_subset <- NK_genes[NK_genes$p_adj.loc < 0.05, ] %>% group_by(gene) %>%  summarize(logFC = mean(logFC, na.rm = TRUE))
Tcell_genes_subset <- Tcell_genes[Tcell_genes$p_adj.loc < 0.05, ] %>% group_by(gene) %>%  summarize(logFC = mean(logFC, na.rm = TRUE))


df_NK_T_corr <- Reduce(function(...) merge(..., by = "gene", all = TRUE), list(
                data.frame(gene = NK_genes_subset$gene,
                          NKcell_logFC = round(as.numeric(NK_genes_subset$logFC), 2),
                           stringsAsFactors = FALSE),
                   data.frame(gene = Tcell_genes_subset$gene,
                              Tcell_logFC = round(as.numeric(Tcell_genes_subset$logFC), 2),
                              stringsAsFactors = FALSE)
                                     ))



row.names(df_NK_T_corr) <- unique(df_NK_T_corr$gene)
df_NK_T_corr$gene <- NULL
df_NK_T_corr$onecol <- rep("", nrow(df_NK_T_corr))


#pdf(paste0(fig_path, "Correlation_plot_NK_Tcells_muscatDSDGE.pdf"))
#Plotting correlation based on avg_log2FC
sp_logFC_NK_T <- ggscatter(df_NK_T_corr, x = "NKcell_logFC", y = "Tcell_logFC",
                add = "reg.line",  # Add regression line
                add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
                conf.int = TRUE, # Add confidence interval
                label = row.names(df_NK_T_corr),
                repel = TRUE, 
                font.label = c(14, "plain"), size = 0.8, cor.coef.size = 14
)
# Add correlation coefficient
sp_logFC_NK_T + stat_cor(method = "pearson") + xlab("NK cells (logFC)") + ylab("T cells (logFC)") + theme(axis.text = element_text(size = rel(1.05)), axis.title = element_text(size = rel(1.15)))

#dev.off()


#ADT: No statistically significant ADTs in NK cells accoriding to muscat DS analysis


```


#8. Vennplot of DE genes (based on Seurat) from NK and Alveolar macrophage clusters
## Genes and ADT/proteins common or uniquely up or downregulated between NK and Alveolar macrophages
```{r}
suppressPackageStartupMessages({
library(ggpubr)
library(VennDetail)
library(VennDiagram)
})

#RNA
# Combine all NK clusters into one

NK_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

AM_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("Alveolar_macrophage"), ]

vennPlot_DGE_NK_vs_AM_upreg <- venndetail(list(
  NK_up = NK_genes$gene[NK_genes$avg_log2FC > 0 & NK_genes$p_val_adj < 0.05], #subset to upregulated genes (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
  AM_up = AM_genes$gene[AM_genes$avg_log2FC > 0 & AM_genes$p_val_adj < 0.05]
))


vennPlot_DGE_NK_vs_AM_downreg <- venndetail(list(
  NK_down = NK_genes$gene[NK_genes$avg_log2FC < 0 & NK_genes$p_val_adj < 0.05], #subset to downregulated genes (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  AM_down = AM_genes$gene[AM_genes$avg_log2FC < 0 & AM_genes$p_val_adj < 0.05]

  ))

#ADT
# Combine all NK clusters into one

NK_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

AM_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("Alveolar_macrophage"), ]

vennPlot_DPE_NK_vs_AM_upreg <- venndetail(list(
  NK_up = NK_ADT$gene[NK_ADT$avg_log2FC > 0 & NK_ADT$p_val_adj < 0.05], #subset to upregulated proteins/ADT (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
  AM_up = AM_ADT$gene[AM_ADT$avg_log2FC > 0 & AM_ADT$p_val_adj < 0.05]
))


vennPlot_DPE_NK_vs_AM_downreg <- venndetail(list(
  NK_down = NK_ADT$gene[NK_ADT$avg_log2FC < 0 & NK_ADT$p_val_adj < 0.05], #subset to downregulated proteins/ADT (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  AM_down = AM_ADT$gene[AM_ADT$avg_log2FC < 0 & AM_ADT$p_val_adj < 0.05]

  ))



#pdf(paste0(fig_path, "UpsetPlot_NK_vs_AM.pdf"), height = 8, width = 6)
plot(vennPlot_DGE_NK_vs_AM_upreg, type = "upset",  mainbar.y.label = paste0("Upregulated genes in IL27RAKO vs WT in\nNK cells and Alveolar macrophages\nIntersection Size\n"))
plot(vennPlot_DGE_NK_vs_AM_downreg, type = "upset", mainbar.y.label = paste0("Downregulated genes in IL27RAKO vs WT\nin NK cells and Alveolar macrophages\nIntersection Size\n"))

plot(vennPlot_DPE_NK_vs_AM_upreg, type = "upset",  mainbar.y.label = paste0("Upregulated ADT in IL27RAKO vs WT in\nNK cells and Alveolar macrophages\nIntersection Size\n"))
plot(vennPlot_DPE_NK_vs_AM_downreg, type = "upset", mainbar.y.label = paste0("Downregulated ADT in IL27RAKO vs WT\nin NK cells and Alveolar macrophages\nIntersection Size\n"))
##dev.off()


vennPlot_list <- list(vennPlot_DGE_NK_vs_AM_upreg = result(vennPlot_DGE_NK_vs_AM_upreg, wide = TRUE), vennPlot_DGE_NK_vs_AM_downreg = result(vennPlot_DGE_NK_vs_AM_downreg, wide = TRUE),
                      vennPlot_DPE_NK_vs_AM_upreg = result(vennPlot_DPE_NK_vs_AM_upreg, wide = TRUE), vennPlot_DPE_NK_vs_AM_downreg = result(vennPlot_DPE_NK_vs_AM_downreg, wide = TRUE))

#writexl::write_xlsx(vennPlot_list, paste0(tab_path, "VennPlotdata_NK_vs_AM_Seurat.xlsx") )



# Venn diagram: RNA
vennDiagram_NK_vs_AM_RNA <- draw.pairwise.venn(area1 = 22, area2 = 20, cross.area = 11, category = c("NK cells\n", "Alveolar\nmacrophages\n"), lty = 1, fill = "white", alpha = 0.5, label.col = "white", cat.fontface = "plain", cat.cex = 1, lwd = 2, cat.fontfamily = "Arial", margin = 0.15, cat.dist = c(0.10, 0.10))

figure_vennDiagram_NK_vs_AM_RNA <- ggarrange(vennDiagram_NK_vs_AM_RNA)

figure_vennDiagram_NK_vs_AM_RNA <- figure_vennDiagram_NK_vs_AM_RNA + 
  annotate("text", x = 0.30, y = 0.50, label = paste0(sprintf('\u21e9'), "22"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n1 NK cells IL27RAKO vs WT upreg
  annotate("text", x = 0.30, y = 0.54, label = paste0(sprintf('\u21e7'), "11"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n1 NK cells IL27RAKO vs WT downreg
  
  annotate("text", x = 0.69, y = 0.50, label = paste0(sprintf('\u21e9'), "8"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n2 AM IL27RAKO vs WT upreg
  annotate("text", x = 0.70, y = 0.54, label = paste0(sprintf('\u21e7'), "11"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n2 AM IL27RAKO vs WT downreg
   
  annotate("text", x = 0.52, y = 0.50, label = paste0(sprintf('\u21e9'), "3"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n12 NK - AM IL27RAKO vs WT upreg
  annotate("text", x = 0.52, y = 0.54, label = paste0(sprintf('\u21e7'), "3"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n12 NK - AM IL27RAKO vs WT downreg
  
   annotate("text", x = 0.83, y = 0.22, label = paste0(sprintf('\u21e9'), "Down"), size = 4, colour = colorRampPalette(c("navy"))(50)) + 
  annotate("text", x = 0.805, y = 0.26, label = paste0(sprintf('\u21e7'), "Up"), size = 4, colour = colorRampPalette(c("red"))(50)) 

annotate_figure(figure_vennDiagram_NK_vs_AM_RNA, top = text_grob("Differentially expressed genes (Seurat):\nIL27RAKO vs WT\n", color = "black", face = "bold", size = 10) ) #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html

#ggsave(paste0(fig_path, "VennDiagram_NK_vs_AM_DGE_RNA.pdf"), units ="in", width = 4, height = 4, dpi = 600, device = cairo_pdf)


# Venn diagram: ADT
vennDiagram_NK_vs_AM_ADT <- draw.pairwise.venn(area1 = 22, area2 = 20, cross.area = 11, category = c("NK cells\n", "Alveolar\nmacrophages\n"), lty = 1, fill = "white", alpha = 0.5, label.col = "white", cat.fontface = "plain", cat.cex = 1, lwd = 2, cat.fontfamily = "Arial", margin = 0.15, cat.dist = c(0.10, 0.10))

figure_vennDiagram_NK_vs_AM_ADT <- ggarrange(vennDiagram_NK_vs_AM_ADT) 

figure_vennDiagram_NK_vs_AM_ADT <- figure_vennDiagram_NK_vs_AM_ADT + 
  annotate("text", x = 0.31, y = 0.50, label = paste0(sprintf('\u21e9'), "70"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n1 NK cells IL27RAKO vs WT upreg
  annotate("text", x = 0.31, y = 0.54, label = paste0(sprintf('\u21e7'), "28"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n1 NK cells IL27RAKO vs WT downreg
  
  annotate("text", x = 0.70, y = 0.50, label = paste0(sprintf('\u21e9'), "1"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n2 AM IL27RAKO vs WT upreg
  annotate("text", x = 0.71, y = 0.54, label = paste0(sprintf('\u21e7'), "13"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n2 AM IL27RAKO vs WT downreg
   
  annotate("text", x = 0.50, y = 0.50, label = paste0(sprintf('\u21e9'), "11"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n12 NK - AM IL27RAKO vs WT upreg
  annotate("text", x = 0.50, y = 0.54, label = paste0(sprintf('\u21e7'), "12"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n12 NK - AM IL27RAKO vs WT downreg
  
   annotate("text", x = 0.83, y = 0.22, label = paste0(sprintf('\u21e9'), "Down"), size = 4, colour = colorRampPalette(c("navy"))(50)) + 
  annotate("text", x = 0.805, y = 0.26, label = paste0(sprintf('\u21e7'), "Up"), size = 4, colour = colorRampPalette(c("red"))(50)) 

annotate_figure(figure_vennDiagram_NK_vs_AM_ADT, top = text_grob("Differentially expressed ADT (proteins; Seurat):\nIL27RAKO vs WT\n", color = "black", face = "bold", size = 10) )

#ggsave(paste0(fig_path, "VennDiagram_NK_vs_AM_DGE_ADT.pdf"), units ="in", width = 4, height = 4, dpi = 600, device = cairo_pdf)



# write differential expression values for the common and unique genes and ADTs
vennPlot_list <- list(vennPlot_DGE_NK_vs_AM_upreg = result(vennPlot_DGE_NK_vs_AM_upreg, wide = TRUE), vennPlot_DGE_NK_vs_AM_downreg = result(vennPlot_DGE_NK_vs_AM_downreg, wide = TRUE),
                      vennPlot_DPE_NK_vs_AM_upreg = result(vennPlot_DPE_NK_vs_AM_upreg, wide = TRUE), vennPlot_DPE_NK_vs_AM_downreg = result(vennPlot_DPE_NK_vs_AM_downreg, wide = TRUE))

NK_unique_up <- vennPlot_list$vennPlot_DGE_NK_vs_AM_upreg$Detail[vennPlot_list$vennPlot_DGE_NK_vs_AM_upreg$AM_up == 0]
NK_unique_down <- vennPlot_list$vennPlot_DGE_NK_vs_AM_downreg$Detail[vennPlot_list$vennPlot_DGE_NK_vs_AM_downreg$AM_down == 0]


DE_NK_unique_vs_AM_RNA <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") & differential_genes_all_integrated_rnm$gene %in% c(NK_unique_up, NK_unique_down), ]


NK_unique_up <- vennPlot_list$vennPlot_DPE_NK_vs_AM_upreg$Detail[vennPlot_list$vennPlot_DPE_NK_vs_AM_upreg$AM_up == 0]
NK_unique_down <- vennPlot_list$vennPlot_DPE_NK_vs_AM_downreg$Detail[vennPlot_list$vennPlot_DPE_NK_vs_AM_downreg$AM_down == 0]

DE_NK_unique_vs_AM_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") & differential_ADT_all_integrated_rnm$gene %in% c(NK_unique_up, NK_unique_down), ]


DE_NK_unique_vs_AM_list <- list(DE_NK_unique_vs_AM_RNA = DE_NK_unique_vs_AM_RNA,
                                DE_NK_unique_vs_AM_ADT = DE_NK_unique_vs_AM_ADT)

#writexl::write_xlsx(DE_NK_unique_vs_AM_list, paste0(tab_path, "DE_NK_unique_vs_AM_Seurat.xlsx") )


#include cluster averages for these genes

tmp <- all_integrated_rnm_VlnPlot #assign seurat object to a temporary variable

cluster_Average_exp_RNA <- list()

#RNA
for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  genes_list <- DE_NK_unique_vs_AM_RNA$gene[DE_NK_unique_vs_AM_RNA$cell_type == i]
  cluster.averages <- AverageExpression(tmp_clus, features = genes_list, slot = "data", assays = "RNA")
  cluster.averages <- as.data.frame(cluster.averages$RNA)
  cluster.averages$Celltype <- i
  cluster.averages$gene <- row.names(cluster.averages)
  cluster_Average_exp_RNA[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(bind_rows(cluster_Average_exp_RNA), paste0(tab_path, "cluster_average_Exp_differentialRNA_il27raKO_vs_wt_NK_unique_vsAM.xlsx"))


#ADT
cluster_Average_exp_ADT <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  adt_list <- DE_NK_unique_vs_AM_ADT$gene[DE_NK_unique_vs_AM_ADT$cell_type == i]
  cluster.averages <- AverageExpression(tmp_clus, features = adt_list, slot = "data", assays = "ADT")
  cluster.averages <- as.data.frame(cluster.averages$ADT)
  cluster.averages$Celltype <- i
  cluster.averages$protein <- row.names(cluster.averages)
  cluster_Average_exp_ADT[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(bind_rows(cluster_Average_exp_ADT), paste0(tab_path, "cluster_average_Exp_differentialADT_il27raKO_vs_wt_NK_unique_vsAM.xlsx"))


#VlnPlot of the unique genes
#pdf(paste0(fig_path, "VlnPlot_unique_NK_vs_AM_Seurat.pdf"), height = 6, width = 8)
tmp_gene <- unique(DE_NK_unique_vs_AM_RNA$gene)
tmp_protein <- unique(DE_NK_unique_vs_AM_ADT$gene)

for(i in 1:length(tmp_gene)){

  print(VlnPlot(all_integrated_rnm_VlnPlot, assay = "RNA", features = tmp_gene[i], split.by = "group", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = rel(1.05), hjust = 1, vjust = 0.5), axis.text.y = element_text(size = rel(1.05)), legend.title = element_text(size=rel(1)), legend.text = element_text(size = rel(1))) )
}

for(i in 1:length(tmp_protein)){
  print(VlnPlot(all_integrated_rnm_VlnPlot, assay = "ADT",  features = tmp_protein[i], split.by = "group", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = rel(1.05), hjust = 1, vjust = 0.5), axis.text.y = element_text(size = rel(1.05)),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = rel(1))) )
}
##dev.off()




```


#9. Correlation analysis: NK vs Alveolar macrophages (Seurat)
```{r}
#RNA
#subset the NK_genes and AM_genes objects to only include significant genes (adjusted p_val < 0.05) 

NK_genes_subset <- NK_genes[NK_genes$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))
AM_genes_subset <- AM_genes[AM_genes$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))

df_NK_AM_corr <- Reduce(function(...) merge(..., by = "gene", all = TRUE), list(
                data.frame(gene = NK_genes_subset$gene,
                          NKcell_log2FC = round(as.numeric(NK_genes_subset$avg_log2FC), 2),
                           stringsAsFactors = FALSE),
                   data.frame(gene = AM_genes_subset$gene,
                              AM_log2FC = round(as.numeric(AM_genes_subset$avg_log2FC), 2),
                              stringsAsFactors = FALSE)
                                     ))



row.names(df_NK_AM_corr) <- unique(df_NK_AM_corr$gene)
df_NK_AM_corr$gene <- NULL
df_NK_AM_corr$onecol <- rep("", nrow(df_NK_AM_corr))


#pdf(paste0(fig_path, "Correlation_plot_NK_AlveolarMacrophages_SeuratDGE.pdf"))
#Plotting correlation based on avg_log2FC
sp_logFC_NK_AM <- ggscatter(df_NK_AM_corr, x = "NKcell_log2FC", y = "AM_log2FC",
                add = "reg.line",  # Add regression line
                add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
                conf.int = TRUE, # Add confidence interval
                label = row.names(df_NK_AM_corr),
                repel = TRUE, 
                font.label = c(14, "plain"), size = 0.8, cor.coef.size = 14
)
# Add correlation coefficient
sp_logFC_NK_AM + stat_cor(method = "pearson") + xlab("NK cells (avg_log2FC)") + ylab("Alveolar macrophages (avg_log2FC)") + theme(axis.text = element_text(size = rel(1.05)), axis.title = element_text(size = rel(1.15)))

#dev.off()


#ADT
#subset the NK_genes and Tcell_genes objects to only include significant genes (adjusted p_val < 0.05) 

NK_ADT_subset <- NK_ADT[NK_ADT$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))
AM_ADT_subset <- AM_ADT[AM_ADT$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))

df_NK_AM_corr_ADT <- Reduce(function(...) merge(..., by = "gene", all = TRUE), list(
                data.frame(gene = NK_ADT_subset$gene,
                          NKcell_log2FC = round(as.numeric(NK_ADT_subset$avg_log2FC), 2),
                           stringsAsFactors = FALSE),
                   data.frame(gene = AM_ADT_subset$gene,
                              AM_log2FC = round(as.numeric(AM_ADT_subset$avg_log2FC), 2),
                              stringsAsFactors = FALSE)
                                     ))



row.names(df_NK_AM_corr_ADT) <- unique(df_NK_AM_corr_ADT$gene)
df_NK_AM_corr_ADT$gene <- NULL
df_NK_AM_corr_ADT$onecol <- rep("", nrow(df_NK_AM_corr_ADT))


#pdf(paste0(fig_path, "Correlation_plot_NK_AlveolarMacrophages_SeuratDPE.pdf"), width = 9.5, height = 9)
#Plotting correlation based on avg_log2FC
sp_logFC_NK_AM_ADT <- ggscatter(df_NK_AM_corr_ADT, x = "NKcell_log2FC", y = "AM_log2FC",
                add = "reg.line",  # Add regression line
                add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
                conf.int = TRUE, # Add confidence interval
                label = row.names(df_NK_AM_corr_ADT),
                font.label = c(9, "plain"), size = 0.8, cor.coef.size = 14,
                repel = TRUE
                )
# Add correlation coefficient
sp_logFC_NK_AM_ADT + stat_cor(method = "pearson") + xlab("NK cells (avg_log2FC)") + ylab("Alveolar macrophages (avg_log2FC)") + theme(axis.text = element_text(size = rel(1.05)), axis.title = element_text(size = rel(1.15)))

#dev.off()




```


#10. Vennplot of DE genes (based on muscatDS) from NK and Alveolar macrophage clusters
## Genes and ADT/proteins common or uniquely up or downregulated between NK and Alveolar macrophages
```{r}
suppressPackageStartupMessages({
library(ggpubr)
library(VennDetail)
library(VennDiagram)
})

#RNA
# Combine all NK clusters into one and all T cell clusters into one
dge_muscatDS <- bind_rows(tbl_fil_all_RNA) #unlist and combine the nested muscatDS results
NK_genes <- dge_muscatDS[dge_muscatDS$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

AM_genes <- dge_muscatDS[dge_muscatDS$cluster_id %in% c("Alveolar_macrophage"), ]

vennPlot_DGE_NK_vs_AM_upreg <- venndetail(list(
  NK_up = NK_genes$gene[NK_genes$logFC > 0 & NK_genes$p_adj.loc < 0.05], #subset to upregulated genes (logFC > 0) that are statistically significant (p_adj.loc < 0.05)
  AM_up = AM_genes$gene[AM_genes$logFC > 0 & AM_genes$p_adj.loc < 0.05]
))


vennPlot_DGE_NK_vs_AM_downreg <- venndetail(list(
  NK_down = NK_genes$gene[NK_genes$logFC < 0 & NK_genes$p_adj.loc < 0.05], #subset to downregulated genes (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  AM_down = AM_genes$gene[AM_genes$logFC < 0 & AM_genes$p_adj.loc < 0.05]

  ))

#ADT : No statistically significant ADTs were present in the muscat DS analysis of NK cells



#pdf(paste0(fig_path, "UpsetPlot_NK_vs_AM_muscatDS_DGE.pdf"), height = 8, width = 6)
plot(vennPlot_DGE_NK_vs_AM_upreg, type = "upset",  mainbar.y.label = paste0("muscatDS\nUpregulated genes in IL27RAKO vs WT in\nNK cells and Alveolar macrophages\nIntersection Size\n"))
plot(vennPlot_DGE_NK_vs_AM_downreg, type = "upset", mainbar.y.label = paste0("muscatDS\nDownregulated genes in IL27RAKO vs WT in\nNK cells and Alveolar macrophages\nIntersection Size\n"))
##dev.off()


vennPlot_list <- list(vennPlot_DGE_NK_vs_AM_upreg = result(vennPlot_DGE_NK_vs_AM_upreg, wide = TRUE), vennPlot_DGE_NK_vs_AM_downreg = result(vennPlot_DGE_NK_vs_AM_downreg, wide = TRUE))

#writexl::write_xlsx(vennPlot_list, paste0(tab_path, "VennPlotdata_NK_vs_AM_muscatDS_DGE.xlsx") )



# Venn diagram: RNA
vennDiagram_NK_vs_AM_RNA <- draw.pairwise.venn(area1 = 22, area2 = 20, cross.area = 11, category = c("NK cells\n", "Alveolar\nmacrophages\n"), lty = 1, fill = "white", alpha = 0.5, label.col = "white", cat.fontface = "plain", cat.cex = 1, lwd = 2, cat.fontfamily = "Arial", margin = 0.15, cat.dist = c(0.10, 0.10))

figure_vennDiagram_NK_vs_AM_RNA <- ggarrange(vennDiagram_NK_vs_AM_RNA)

figure_vennDiagram_NK_vs_AM_RNA <- figure_vennDiagram_NK_vs_AM_RNA + 
  annotate("text", x = 0.30, y = 0.50, label = paste0(sprintf('\u21e9'), "51"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n1 NK cells IL27RAKO vs WT upreg
  annotate("text", x = 0.30, y = 0.54, label = paste0(sprintf('\u21e7'), "47"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n1 NK cells IL27RAKO vs WT downreg
  
  annotate("text", x = 0.70, y = 0.50, label = paste0(sprintf('\u21e9'), "1"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n2 AM IL27RAKO vs WT upreg
  annotate("text", x = 0.70, y = 0.54, label = paste0(sprintf('\u21e7'), "3"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n2 AM IL27RAKO vs WT downreg
   
  annotate("text", x = 0.51, y = 0.50, label = paste0(sprintf('\u21e9'), "4"), size = 4, colour = colorRampPalette(c("navy"))(50)) + #n12 NK - AM IL27RAKO vs WT upreg
  annotate("text", x = 0.51, y = 0.54, label = paste0(sprintf('\u21e7'), "5"), size = 4, colour = colorRampPalette(c("red"))(50)) + #n12 NK - AM IL27RAKO vs WT downreg
  
   annotate("text", x = 0.83, y = 0.22, label = paste0(sprintf('\u21e9'), "Down"), size = 4, colour = colorRampPalette(c("navy"))(50)) + 
  annotate("text", x = 0.805, y = 0.26, label = paste0(sprintf('\u21e7'), "Up"), size = 4, colour = colorRampPalette(c("red"))(50)) 

annotate_figure(figure_vennDiagram_NK_vs_AM_RNA, top = text_grob("Differentially expressed genes (muscatDS):\nIL27RAKO vs WT\n", color = "black", face = "bold", size = 10) ) #https://rpkgs.datanovia.com/ggpubr/reference/annotate_figure.html

#ggsave(paste0(fig_path, "VennDiagram_NK_vs_AM_DGE_muscatDSRNA.pdf"), units ="in", width = 4, height = 4, dpi = 600, device = cairo_pdf)


# write differential expression values for the common and unique genes and ADTs
vennPlot_list <- list(vennPlot_DGE_NK_vs_AM_upreg = result(vennPlot_DGE_NK_vs_AM_upreg, wide = TRUE), vennPlot_DGE_NK_vs_AM_downreg = result(vennPlot_DGE_NK_vs_AM_downreg, wide = TRUE))

NK_unique_up <- vennPlot_list$vennPlot_DGE_NK_vs_AM_upreg$Detail[vennPlot_list$vennPlot_DGE_NK_vs_AM_upreg$AM_up == 0]
NK_unique_down <- vennPlot_list$vennPlot_DGE_NK_vs_AM_downreg$Detail[vennPlot_list$vennPlot_DGE_NK_vs_AM_downreg$AM_down == 0]


DE_NK_unique_vs_AM_muscatDS <- dge_muscatDS[dge_muscatDS$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") & dge_muscatDS$gene %in% c(NK_unique_up, NK_unique_down), ]

#writexl::write_xlsx(DE_NK_unique_vs_AM_muscatDS, paste0(tab_path, "DE_NK_unique_vs_AM_muscatDS.xlsx") )


#include cluster averages for these genes

#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm_VlnPlot #assign seurat object to a temporary variable

cluster_Average_exp_RNA <- list()

dge_muscatDS <- bind_rows(tbl_fil_all_RNA) #unlist and combine the nested muscatDS results

#RNA
for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  genes_list <- DE_NK_unique_vs_AM_muscatDS$gene[DE_NK_unique_vs_AM_muscatDS$cluster_id == i]
  cluster.averages <- AverageExpression(tmp_clus, features = genes_list, slot = "data", assays = "RNA")
  cluster.averages <- as.data.frame(cluster.averages$RNA)
  cluster.averages$Celltype <- i
  cluster.averages$gene <- row.names(cluster.averages)
  cluster_Average_exp_RNA[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(bind_rows(cluster_Average_exp_RNA), paste0(tab_path, "cluster_AvgExp_DE_RNA_il27raKO_vs_wt_NK_unique_vsAM_muscatDS.xlsx"))


#VlnPlot of the unique genes
#pdf(paste0(fig_path, "VlnPlot_unique_NK_vs_AM_muscatDSDGE.pdf"), height = 6, width = 8)
tmp_gene <- unique(DE_NK_unique_vs_AM_muscatDS$gene)
print("muscatDS: unique_NK-Alveolar Macrophages: IL27RAko vs wt")
for(i in 1:length(tmp_gene)){

  print(VlnPlot(all_integrated_rnm_VlnPlot, assay = "RNA", features = tmp_gene[i], split.by = "group", pt.size = 0.1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = rel(1.05), hjust = 1, vjust = 0.5), axis.text.y = element_text(size = rel(1.05)), legend.title = element_text(size=rel(1)), legend.text = element_text(size = rel(1))) )
}
##dev.off()


```


#7. Correlation analysis: NK vs Alveolar macrophages (based on muscatDS)
```{r}

#subset the NK_genes and AM_genes objects to only include significant genes (adjusted p_adj.loc < 0.05) 

NK_genes_subset <- NK_genes[NK_genes$p_adj.loc < 0.05, ] %>% group_by(gene) %>%  summarize(logFC = mean(logFC, na.rm = TRUE))
AM_genes_subset <- AM_genes[AM_genes$p_adj.loc < 0.05, ] %>% group_by(gene) %>%  summarize(logFC = mean(logFC, na.rm = TRUE))


df_NK_AM_corr <- Reduce(function(...) merge(..., by = "gene", all = TRUE), list(
                data.frame(gene = NK_genes_subset$gene,
                          NKcell_logFC = round(as.numeric(NK_genes_subset$logFC), 2),
                           stringsAsFactors = FALSE),
                   data.frame(gene = AM_genes_subset$gene,
                              AM_logFC = round(as.numeric(AM_genes_subset$logFC), 2),
                              stringsAsFactors = FALSE)
                                     ))



row.names(df_NK_AM_corr) <- unique(df_NK_AM_corr$gene)
df_NK_AM_corr$gene <- NULL
df_NK_AM_corr$onecol <- rep("", nrow(df_NK_AM_corr))


#pdf(paste0(fig_path, "Correlation_plot_NK_AM_muscatDSDGE.pdf"))
#Plotting correlation based on avg_log2FC
sp_logFC_NK_AM <- ggscatter(df_NK_AM_corr, x = "NKcell_logFC", y = "AM_logFC",
                add = "reg.line",  # Add regression line
                add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
                conf.int = TRUE, # Add confidence interval
                label = row.names(df_NK_AM_corr),
                repel = TRUE, 
                font.label = c(14, "plain"), size = 0.8, cor.coef.size = 14
)
# Add correlation coefficient
sp_logFC_NK_AM + stat_cor(method = "pearson") + xlab("NK cells (logFC)") + ylab("Alveolar macrophages (logFC)") + theme(axis.text = element_text(size = rel(1.05)), axis.title = element_text(size = rel(1.15)))

#dev.off()


#ADT: No statistically significant ADTs in NK cells accoriding to muscat DS analysis


```





#========================================================================

#Average expression levels of differentially expressed genes and ADT/proteins (DGE and DPE based on Seurat)
- Average non-log expression values of normalized counts
```{r, eval=FALSE, echo=FALSE}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm_VlnPlot #assign seurat object to a temporary variable

Idents(tmp) <- tmp$celltype

# Rename all NK cells to one cell type and all T cells to one cell type
tmp <- RenameIdents(tmp,
                    "CD27loCD11bhi_NK" = "NK_cell", 
                    "CD27loCD11b+_NK" = "NK_cell", 
                    "CD27hiCD11b+_NK" = "NK_cell",
                    "CD27loCD11b+_mito_NK" = "NK_cell", 
                    "CD27loCD11blo_NK" = "NK_cell",
                    "CD8_TEM" = "T_cell",
                    "CD8_Nave"  = "T_cell", 
                    "CD27+_gdT" = "T_cell", 
                    "CD4_TEM" = "T_cell", 
                    "Treg" = "T_cell", 
                    "CD8_TCM" = "T_cell", 
                     "CD27-_gdT" = "T_cell")

tmp$broad_cell_type <- Idents(tmp)
tmp$broad_cell_type_group <- paste0(Idents(tmp), "_", tmp$group)


#genes
DefaultAssay(object = tmp) <- "RNA"
Idents(tmp) <- tmp$broad_cell_type_group
#Finding differentially expressed genes across clusters (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)

get_differential_genes <- function(cluster){
  Seurat::FindMarkers(object = tmp, ident.1 = paste0(cluster, "_il27raKO"), ident.2 = paste0(cluster, "_wt"), verbose = FALSE, assay = "RNA", only.pos = FALSE) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = cluster, .)
}

differential_RNA_tmp <- map_dfr(c("NK_cell", "T_cell", "cDC", "Monocyte", "Alveolar_macrophage", "pDC", "moDC", "B_cell"), get_differential_genes)

#writexl::write_xlsx(differential_RNA_tmp, paste0(tab_path, "/Tmp_differential_genes_RNA_byCelltype_broad.xlsx")) 


x <- Seurat::FindMarkers(object = tmp, ident.1 = "NK_cell_il27raKO", ident.2 = "T_cell_il27raKO", verbose = FALSE, assay = "RNA", only.pos = FALSE)
x$gene <- row.names(x)
x_0.05 <- subset(x, x$p_val_adj < 0.05)

y <- Seurat::FindMarkers(object = tmp, ident.1 = "NK_cell_wt", ident.2 = "T_cell_wt", verbose = FALSE, assay = "RNA", only.pos = FALSE)
y$gene <- row.names(y)
y_0.05 <- subset(y, y$p_val_adj < 0.05)

z <- Seurat::FindMarkers(object = tmp, ident.1 = "NK_cell_il27raKO", ident.2 = c("T_cell_il27raKO", "cDC_il27raKO", "Monocyte_il27raKO", "Alveolar_macrophage_il27raKO", "pDC_il27raKO", "moDC_il27raKO", "B_cell_il27raKO"), verbose = FALSE, assay = "RNA", only.pos = FALSE)
z$gene <- row.names(z)
z_0.05 <- subset(z, z$p_val_adj < 0.05)


v <- Seurat::FindMarkers(object = tmp, ident.1 = "NK_cell_wt", ident.2 = c("T_cell_wt", "cDC_wt", "Monocyte_wt", "Alveolar_macrophage_wt", "pDC_wt", "moDC_wt", "B_cell_wt"), verbose = FALSE, assay = "RNA", only.pos = FALSE)
v$gene <- row.names(v)
v_0.05 <- subset(v, v$p_val_adj < 0.05)

library(VennDetail)

zz <- venndetail(list(NK_vs_other_il27rako = z_0.05$gene, NK_vs_other_wt = v_0.05$gene))

#filter out genes not differentially expressed in NK_vs_other (v)
tbl <- result(zz, wide = TRUE)
tbl_gene <- tbl$Detail[tbl$NK_vs_other_wt == 0]

#pdf(paste0(fig_path, "tbl_gene.pdf"), width = 15, height = 25)
VlnPlot(tmp, tbl_gene[1:23], split.by = "group", group.by = "broad_cell_type", assay = "RNA", pt.size = 0.1, ncol = 5) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.15, 'in'), legend.key.width= unit(0.15, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10), legend.position = "top")
VlnPlot(tmp, tbl_gene[24:47], split.by = "group", group.by = "broad_cell_type", assay = "RNA", pt.size = 0.1, ncol = 5) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.15, 'in'), legend.key.width= unit(0.15, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10), legend.position = "top")
VlnPlot(tmp, tbl_gene[48:71], split.by = "group", group.by = "broad_cell_type", assay = "RNA", pt.size = 0.1, ncol = 5) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.15, 'in'), legend.key.width= unit(0.15, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10), legend.position = "top")
VlnPlot(tmp, tbl_gene[72:95], split.by = "group", group.by = "broad_cell_type", assay = "RNA", pt.size = 0.1, ncol = 5) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.15, 'in'), legend.key.width= unit(0.15, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10), legend.position = "top")
VlnPlot(tmp, tbl_gene[96:115], split.by = "group", group.by = "broad_cell_type", assay = "RNA", pt.size = 0.1, ncol = 5) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.15, 'in'), legend.key.width= unit(0.15, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10), legend.position = "top")
#dev.off()
tmp_list <- list("NK_vs_T_il27raKO" = x, "NK_vs_T_wt" = y, "NK_vs_OtherCellTy_il27rako"= z, "NK_vs_OtherCellTy_wt" = v)

#writexl::write_xlsx(tmp_list, paste0(tab_path, "RNA_NK_DE_comparisons.xlsx"))

cluster_Average_exp_RNA <- list()

#RNA
Idents(tmp) <- tmp$broad_cell_type
for(i in unique(tmp@meta.data$broad_cell_type)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "broad_cell_type_group"
  genes_list <- differential_RNA_tmp$gene[differential_RNA_tmp$cell_type == i]
  cluster.averages <- AverageExpression(tmp_clus, features = genes_list, slot = "data", assays = "RNA")
  cluster.averages <- as.data.frame(cluster.averages$RNA)
  cluster.averages$Celltype <- i
  cluster.averages$gene <- row.names(cluster.averages)
  cluster_Average_exp_RNA[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_RNA, paste0(tab_path, "RNA_cluster_average_Exp_BroadCellType_il27raKO_vs_wt.xlsx"))


#ADT

DefaultAssay(object = tmp) <- "ADT"
Idents(tmp) <- tmp$broad_cell_type_group
#Finding differentially expressed genes across clusters (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)

get_differential_genes <- function(cluster){
  Seurat::FindMarkers(object = tmp, ident.1 = paste0(cluster, "_il27raKO"), ident.2 = paste0(cluster, "_wt"), verbose = FALSE, assay = "ADT", only.pos = FALSE, logfc.threshold = 0) %>%
    rownames_to_column(var = "protein") %>%
    cbind(cell_type = cluster, .)
}

get_differential_ADT <- map_dfr(c("NK_cell", "T_cell", "cDC", "Monocyte", "Alveolar_macrophage", "pDC", "moDC", "B_cell"), get_differential_genes)

#writexl::write_xlsx(get_differential_ADT, paste0(tab_path, "/Tmp_differential_genes_ADT_byCelltype_broad.xlsx")) 


x <- Seurat::FindMarkers(object = tmp, ident.1 = "NK_cell_il27raKO", ident.2 = "T_cell_il27raKO", verbose = FALSE, assay = "ADT", only.pos = FALSE)
x$protein <- row.names(x)
x_0.05 <- subset(x, x$p_val_adj < 0.05)

y <- Seurat::FindMarkers(object = tmp, ident.1 = "NK_cell_wt", ident.2 = "T_cell_wt", verbose = FALSE, assay = "ADT", only.pos = FALSE)
y$protein <- row.names(y)
y_0.05 <- subset(y, y$p_val_adj < 0.05)

z <- Seurat::FindMarkers(object = tmp, ident.1 = "NK_cell_il27raKO", ident.2 = c("T_cell_il27raKO", "cDC_il27raKO", "Monocyte_il27raKO", "Alveolar_macrophage_il27raKO", "pDC_il27raKO", "moDC_il27raKO", "B_cell_il27raKO"), verbose = FALSE, assay = "ADT", only.pos = FALSE)
z$protein <- row.names(z)
z_0.05 <- subset(z, z$p_val_adj < 0.05)


v <- Seurat::FindMarkers(object = tmp, ident.1 = "NK_cell_wt", ident.2 = c("T_cell_wt", "cDC_wt", "Monocyte_wt", "Alveolar_macrophage_wt", "pDC_wt", "moDC_wt", "B_cell_wt"), verbose = FALSE, assay = "ADT", only.pos = FALSE)
v$protein <- row.names(v)
v_0.05 <- subset(v, v$p_val_adj < 0.05)

library(VennDetail)

zz <- venndetail(list(NK_vs_other_il27rako = z_0.05$protein, NK_vs_other_wt = v_0.05$protein))

#filter out genes not differentially expressed in NK_vs_other (v)
tbl <- result(zz, wide = TRUE)
tbl_ADT <- tbl$Detail[tbl$NK_vs_other_wt == 0]

#pdf(paste0(fig_path, "tbl_ADT.pdf"), width = 15, height = 15)
VlnPlot(tmp, tbl_ADT[1:8], split.by = "group", group.by = "broad_cell_type", assay = "ADT", pt.size = 0.1, ncol = 4) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.15, 'in'), legend.key.width= unit(0.15, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10), legend.position = "top")
#dev.off()

tmp_list <- list("NK_vs_T_il27raKO" = x, "NK_vs_T_wt" = y, "NK_vs_OtherCellTy_il27rako"= z, "NK_vs_OtherCellTy_wt" = v)

#writexl::write_xlsx(tmp_list, paste0(tab_path, "ADT_NK_DE_comparisons.xlsx"))





cluster_Average_exp_ADT <- list()

Idents(tmp) <- tmp$broad_cell_type
for(i in unique(tmp@meta.data$broad_cell_type)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "broad_cell_type_group"
  genes_list <- differential_ADT_tmp$gene[differential_ADT_tmp$cell_type == i]
  cluster.averages <- AverageExpression(tmp_clus, features = genes_list, slot = "data", assays = "ADT")
  cluster.averages <- as.data.frame(cluster.averages$ADT)
  cluster.averages$Celltype <- i
  cluster.averages$protein <- row.names(cluster.averages)
  cluster_Average_exp_ADT[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_ADT, paste0(tab_path, "ADT_cluster_average_Exp_BroadCellType_il27raKO_vs_wt.xlsx"))


```


```{r, eval=FALSE, echo=FALSE}
#preparing data for muscat
library(limma)
library(muscat)
library(purrr)
library(dplyr)
Idents(tmp) <- tmp$broad_cell_type
#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(tmp, assay = "RNA")

tmp_sce$id <- paste0(tmp_sce$group,"_", tmp_sce$sample_id) #combining condition and sample_id

tmp_sce <- muscat::prepSCE(tmp_sce, 
    kid = "broad_cell_type", # cluster numbers or celltype
    gid = "group",  # group IDs (il27raKO/wt)
    sid = "id",   # sample IDs (il27raKO/wt+sample_id)
    drop = TRUE)  # drop all other colData columns
tmp_sce$sample_id <- factor(x = tmp_sce$sample_id, levels = c("il27raKO_BM_WS_9", "il27raKO_BM_WS_10", "il27raKO_BM_WS_15", "wt_BM_WS_5",  "wt_BM_WS_6", "wt_BM_WS_8"))

nk <- length(kids <- levels(tmp_sce$cluster_id)) 
ns <- length(sids <- levels(tmp_sce$sample_id))
names(kids) <- kids; names(sids) <- sids


#Differential State (DS) analysis
pseudobulk_sce <- muscat::aggregateData(tmp_sce,
    assay = "counts", fun = "sum",
    by = c("cluster_id", "sample_id")) #cluster_id refers to cell type


# one sheet per subpopulation
assayNames(pseudobulk_sce)

#Pseudobulk-level MDS plot
pb_sce_mds <- muscat::pbMDS(pseudobulk_sce) 
#warning: Removing 4 rows with all zero counts


#pdf(paste0(fig_path, "/muscatDS_MDS_RNA_res0.8.pdf"), width = 10, height = 8)
pb_sce_mds
##dev.off()

#test for DS using pbDS. By default, a group_id model is fit, and the last coefficient of the linear model is tested to be equal to zero.

# construct design & contrast matrix
ei <- metadata(tmp_sce)$experiment_info
mm <- model.matrix(~ 0 + ei$group_id)
dimnames(mm) <- list(ei$sample_id, levels(ei$group_id))
contrast <- makeContrasts("il27raKO-wt", levels = mm)

# run DS analysis
res_sce <- muscat::pbDS(pseudobulk_sce,  method = "edgeR", design = mm, contrast = contrast, verbose = FALSE)
# access results table for 1st comparison
tbl_sce <- res_sce$table[[1]]
# one data.frame per cluster
names(tbl_sce)

# view results for 1st cluster
k1_sce <- tbl_sce[[1]]
head(format(k1_sce[, -ncol(k1_sce)], digits = 2))

#sort by adjusted p value
tbl_fil_all_RNA <- lapply(tbl_sce, function(u) {
  dplyr::arrange(u, p_adj.loc)
})


#filter results to retain hits with FDR < 5% and abs(logFC) > 0.58, and count the number and frequency of differential findings by cluster and view the top hits (lowest adj. p-value) in each cluster
# filter FDR < 5%, abs(logFC) > 1 & sort by adj. p-value
tbl_fil_sig <- lapply(tbl_sce, function(u) {
  u <- dplyr::filter(u, p_adj.loc < 0.05, abs(logFC) > 0.58)
  dplyr::arrange(u, p_adj.loc)
})

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil_sig, nrow, numeric(1))
p_de <- format(n_de / nrow(tmp_sce) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)

# view top 2 hits in each cluster
top2_sig <- bind_rows(lapply(tbl_fil_sig, top_n, 2, p_adj.loc))
format(top2_sig[, -ncol(top2_sig)], digits = 2)

#results table with expression frequency of each gene by sample and overall fold change between napc2 vs control (untreated)
res_tbl <- muscat::resDS(tmp_sce, res_sce, frq = TRUE)

#Between-cluster concordance: UpSet-plot that visualizes the number of DE genes that are shared across or unique to certain clusters
library(UpSetR)
de_gs_by_k <- map(tbl_fil_sig, "gene")
upset(fromList(de_gs_by_k))

#writexl::write_xlsx(tbl_fil_all_RNA,paste0(tab_path, "muscatDS_RNA_tmp.xlsx"))

library(scater)
#UMAP
.plot_dr <- function(tmp_sce, dr, col)
  plotReducedDim(tmp_sce, dimred = dr, colour_by = col) +
    guides(fill = guide_legend(override.aes = list(alpha = 1, size = 3))) +
    theme_minimal() + theme(aspect.ratio = 1)




# downsample to max. 100 cells per cluster
cs_by_k <- split(colnames(tmp_sce), tmp_sce$cluster_id)
cs100 <- unlist(sapply(cs_by_k, function(u) 
  sample(u, min(length(u), 100))))

# plot t-SNE & UMAP colored by cluster & group ID
for (dr in c("WNN.UMAP"))
  for (col in c("cluster_id", "group_id"))
    .plot_dr(tmp_sce[, cs100], dr, col)


# Violin plots of top 5 genes across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_RNA_VlnPlot_tmp.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_RNA)){
print(plotExpression(tmp_sce[, tmp_sce$cluster_id == i],
                features = tbl_fil_all_RNA[[i]]$gene[seq_len(5)],
                x = "sample_id", colour_by = "group_id", ncol = 3) + scale_color_manual(values = c("#F8766D", "#00BFC4"), aesthetics = "color") +
     guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) +
     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ggtitle(paste0("Violin plot of top 5 muscat differential state genes; ", i) )
)
}
##dev.off()

# Heatmap of top 20 genes (or less; based on adjusted p value) across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_RNA_Heatmap_Tmp.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_RNA)){
  if( ! (i %in% c("moDC", "B_cell") ) ){ #excluding "moDC", "B_cell"as there are no genes with FDR < 0.05 and including this causes an error in generating the heatmap
tmp_htmap <- muscat::pbHeatmap(x = tmp_sce, y = res_sce, k = i, col = colorRampPalette(c("blue", "white", "red"))(50))
tmp_htmap@column_title <- paste0("Heatmap of top 20 (or less) genes from muscat DS analysis in ", i)
print(tmp_htmap)
  }
}
##dev.off()

#------------------------------------------------------------------------

# Assay ADT
#converting the seurat object to a single cell experiment object as required by the muscat package, selecting RNA assay
tmp_sce <- as.SingleCellExperiment(tmp, assay = "ADT")

tmp_sce$id <- paste0(tmp_sce$group,"_", tmp_sce$sample_id) #combining condition and sample_id

tmp_sce <- muscat::prepSCE(tmp_sce, 
    kid = "broad_cell_type", # cluster numbers or celltype
    gid = "group",  # group IDs (il27raKO/wt)
    sid = "id",   # sample IDs (il27raKO/wt+sample_id)
    drop = TRUE)  # drop all other colData columns
tmp_sce$sample_id <- factor(x = tmp_sce$sample_id, levels = c("il27raKO_BM_WS_9", "il27raKO_BM_WS_10", "il27raKO_BM_WS_15", "wt_BM_WS_5",  "wt_BM_WS_6", "wt_BM_WS_8"))

nk <- length(kids <- levels(tmp_sce$cluster_id)) 
ns <- length(sids <- levels(tmp_sce$sample_id))
names(kids) <- kids; names(sids) <- sids


#Differential State (DS) analysis
pseudobulk_sce <- muscat::aggregateData(tmp_sce,
    assay = "counts", fun = "sum",
    by = c("cluster_id", "sample_id")) #cluster_id refers to cell type


# one sheet per subpopulation
assayNames(pseudobulk_sce)

#Pseudobulk-level MDS plot
pb_sce_mds <- muscat::pbMDS(pseudobulk_sce) 


#pdf(paste0(fig_path, "/muscatDS_MDS_ADT_res0.8.pdf"), width = 10, height = 8)
pb_sce_mds
##dev.off()

#test for DS using pbDS. By default, a group_id model is fit, and the last coefficient of the linear model is tested to be equal to zero.

# construct design & contrast matrix
ei <- metadata(tmp_sce)$experiment_info
mm <- model.matrix(~ 0 + ei$group_id)
dimnames(mm) <- list(ei$sample_id, levels(ei$group_id))
contrast <- makeContrasts("il27raKO-wt", levels = mm)

# run DS analysis
res_sce_ADT <- muscat::pbDS(pseudobulk_sce,  method = "edgeR", design = mm, contrast = contrast, verbose = FALSE)
# access results table for 1st comparison
tbl_sce_ADT <- res_sce_ADT$table[[1]]
# one data.frame per cluster
names(tbl_sce_ADT)

# view results for 1st cluster
k1_sce_ADT <- tbl_sce_ADT[[1]]
head(format(k1_sce[, -ncol(k1_sce)], digits = 2))

#sort by adjusted p value
tbl_fil_all_ADT <- lapply(tbl_sce_ADT, function(u) {
  dplyr::arrange(u, p_adj.loc)
})


#filter results to retain hits with FDR < 5% and abs(logFC) > 0.58, and count the number and frequency of differential findings by cluster and view the top hits (lowest adj. p-value) in each cluster
# filter FDR < 5%, abs(logFC) > 1 & sort by adj. p-value
tbl_fil_sig_ADT <- lapply(tbl_sce_ADT, function(u) {
  u <- dplyr::filter(u, p_adj.loc < 0.05, abs(logFC) > 0.58)
  dplyr::arrange(u, p_adj.loc)
})

# nb. of DS genes & % of total by cluster
n_de <- vapply(tbl_fil_sig_ADT, nrow, numeric(1))
p_de <- format(n_de / nrow(tmp_sce) * 100, digits = 3)
data.frame("#DS" = n_de, "%DS" = p_de, check.names = FALSE)

# view top 2 hits in each cluster
top2_sig_ADT <- bind_rows(lapply(tbl_fil_sig_ADT, top_n, 2, p_adj.loc))
format(top2_sig[, -ncol(top2_sig_ADT)], digits = 2)

#results table with expression frequency of each gene by sample and overall fold change between napc2 vs control (untreated)
res_tbl_ADT <- muscat::resDS(tmp_sce, res_sce_ADT, frq = TRUE)

#Between-cluster concordance: UpSet-plot that visualizes the number of DE genes that are shared across or unique to certain clusters
de_gs_by_k <- map(tbl_fil_sig_ADT, "gene")
upset(fromList(de_gs_by_k))

#writexl::write_xlsx(tbl_fil_all_ADT,paste0(tab_path, "muscatDS_byCellType_ADT_Tmp.xlsx"))


#UMAP
.plot_dr <- function(tmp_sce, dr, col)
  plotReducedDim(tmp_sce, dimred = dr, colour_by = col) +
    guides(fill = guide_legend(override.aes = list(alpha = 1, size = 3))) +
    theme_minimal() + theme(aspect.ratio = 1)




# downsample to max. 100 cells per cluster
cs_by_k <- split(colnames(tmp_sce), tmp_sce$cluster_id)
cs100 <- unlist(sapply(cs_by_k, function(u) 
  sample(u, min(length(u), 100))))

# plot t-SNE & UMAP colored by cluster & group ID
for (dr in c("WNN.UMAP"))
  for (col in c("cluster_id", "group_id"))
    .plot_dr(tmp_sce[, cs100], dr, col)


# Violin plots of top 5 genes across samples for each cell type with differential state analysis results available

#pdf(paste0(fig_path, "/muscatDS_ADT_VlnPlot_Tmp.pdf"), width = 10, height = 8)
for(i in names(tbl_fil_all_ADT)){
print(plotExpression(tmp_sce[, tmp_sce$cluster_id == i],
                features = tbl_fil_all_ADT[[i]]$gene[seq_len(5)],
                x = "sample_id", colour_by = "group_id", ncol = 3) + scale_color_manual(values = c("#F8766D", "#00BFC4"), aesthetics = "color") +
     guides(fill = guide_legend(override.aes = list(size = 5, alpha = 1))) +
     theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ggtitle(paste0("Violin plot of top 5 muscat differential state proteins; ", i) )
)
}
##dev.off()


##dev.off()
#c("#F8766D", "#00BFC4")
```


# Average expression levels of differentially expressed genes and ADT/proteins (DGE and DPE based on Seurat)
- Average non-log expression values of normalized counts
```{r, eval=FALSE, echo=FALSE}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm_VlnPlot #assign seurat object to a temporary variable

cluster_Average_exp_RNA <- list()

cluster_Aggregate_exp_RNA <- list()

#RNA
for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "sample_id"
  genes_list <- differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type == i]
  cluster.averages <- AverageExpression(tmp_clus, features = genes_list, slot = "data", assays = "RNA")
  cluster.averages <- as.data.frame(cluster.averages$RNA)
  cluster.averages$Celltype <- i
  cluster.averages$gene <- row.names(cluster.averages)
  cluster_Average_exp_RNA[[i]] <- cluster.averages
  
  cluster.aggregates <- AverageExpression(tmp_clus, features = genes_list, slot = "data", assays = "RNA")
  cluster.aggregates <- as.data.frame(cluster.aggregates$RNA)
  cluster.aggregates$Celltype <- i
  cluster.aggregates$gene <- row.names(cluster.aggregates)
  cluster_Aggregate_exp_RNA[[i]] <- cluster.aggregates
 
}


#ADT
cluster_Average_exp_ADT <- list()
cluster_Aggregate_exp_ADT <- list()


for(i in unique(tmp@meta.data$celltype)){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "sample_id"
  adt_list <- differential_ADT_all_integrated_rnm$gene[differential_ADT_all_integrated_rnm$cell_type == i]
  cluster.averages <- AverageExpression(tmp_clus, features = adt_list, slot = "data", assays = "ADT")
  cluster.averages <- as.data.frame(cluster.averages$ADT)
  cluster.averages$Celltype <- i
  cluster.averages$protein <- row.names(cluster.averages)
  cluster_Average_exp_ADT[[i]] <- cluster.averages
  
  cluster.aggregates <- AverageExpression(tmp_clus, features = adt_list, slot = "data", assays = "ADT")
  cluster.aggregates <- as.data.frame(cluster.aggregates$ADT)
  cluster.aggregates$Celltype <- i
  cluster.aggregates$protein <- row.names(cluster.aggregates)
  cluster_Aggregate_exp_ADT[[i]] <- cluster.aggregates
 
}



```





#Upd 31May2022: FOR GERMAN ANIMAL PROTOCOL PROPOSAL: Based on suggestion from Johannes generating violin plots for Tim-1 (CD365), Ly49D and Ly49H
```{r}

tmp_cell_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM", "Treg", "CD27+_gdT", "CD27-_gdT")



#pdf(paste0(fig_path, "VlnPlot_German_AnimProt_Tim1_Ly49D_Ly49H_TotalSeq_IL27RAKO_Johannes.pdf"), height = 6, width = 8)
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in%  tmp_cell_list), assay = "ADT",  features = c("adt-CD365"), split.by = "group", pt.size = 0.1, ncol = 1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = rel(1.05), hjust = 1, vjust = 0.5), axis.text.y = element_text(size = rel(1.05)),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = rel(1)))
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in%  tmp_cell_list), assay = "ADT",  features = c("adt-Ly49D"), split.by = "group", pt.size = 0.1, ncol = 1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = rel(1.05), hjust = 1, vjust = 0.5), axis.text.y = element_text(size = rel(1.05)),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = rel(1)))
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in%  tmp_cell_list), assay = "ADT",  features = c("adt-Ly49H"), split.by = "group", pt.size = 0.1, ncol = 1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = rel(1.05), hjust = 1, vjust = 0.5), axis.text.y = element_text(size = rel(1.05)),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = rel(1)))

VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in%  tmp_cell_list), assay = "RNA",  features = c("rna_Havcr1"), split.by = "group", pt.size = 0.1, ncol = 1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = rel(1.05), hjust = 1, vjust = 0.5), axis.text.y = element_text(size = rel(1.05)),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = rel(1)))
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in%  tmp_cell_list), assay = "RNA",  features = c("rna_Klra4"), split.by = "group", pt.size = 0.1, ncol = 1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = rel(1.05), hjust = 1, vjust = 0.5), axis.text.y = element_text(size = rel(1.05)),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = rel(1)))
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in%  tmp_cell_list), assay = "RNA",  features = c("rna_Klra8"), split.by = "group", pt.size = 0.1, ncol = 1) & theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90, size = rel(1.05), hjust = 1, vjust = 0.5), axis.text.y = element_text(size = rel(1.05)),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = rel(1)))
##dev.off()


```


# save workspace
```{r}
#save.image(paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_v1.2.RData")) #update based on change in NK cell type annotation
#load(paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_v1.2.RData"))

#saveRDS(all_integrated,paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_v1.2_all_integrated.rds")) #update based on change in NK cell type annotation
#saveRDS(all_integrated_rnm_VlnPlot, paste0(data_path, "SW0100_P239_TotalSeq_IL27RAKO_all_integrated_rnm_VlnPlot_v1.2.rds"))
```




# Upd 8Aug: Manuscript figures: Based on discussion with Johannes, Prof. Bosmann and Kevin, updating and generating figures for manuscript
## 1. Combined dotplot of markers
```{r}
#####awaiting inputs on markers to highlight from Johannes
markers_list_RNA <- c("rna_Klra4", "rna_Klra8", "rna_Klrb1a", "rna_Klrb1b", "rna_Klrb1c", "rna_Klrb1f",  "rna_Gzma", "rna_Gzmb", "rna_Ncr1", "rna_Nkg7", "rna_Itga2", "rna_Cd27", "rna_Itgam", "rna_Cd3d", "rna_Cd3g", "rna_Cd3e", "rna_Trbc1", "rna_Trbc2", "rna_Cd4", "rna_Cd8a", "rna_Cd8b1", "rna_Il22", "rna_Il17f", "rna_Il17a", "rna_Blk", "rna_Tcrg-C1", "rna_Trgv2", "rna_Trdv4", "rna_Tcrg-V4", "rna_Foxp3", "rna_Ctla4", "rna_Il2ra", "rna_Ear1", "rna_Ear2", "rna_Itgax", "rna_Siglec1", "rna_Siglecf", "rna_Cd200r1", "rna_Lgals3", "rna_Marco", "rna_Chil3","rna_F13a1", "rna_Ly6c2", "rna_Cd14", "rna_Fcgr4", "rna_Fcgr1", "rna_Mafb",
  "rna_Cd40", "rna_Cd24a","rna_Dpp4", "rna_H2-Ab1", "rna_H2-Eb1", "rna_Itgae", "rna_Esam", "rna_Flt3", "rna_Sirpa", "rna_Xcr1", "rna_Clec9a", "rna_Ccr7", "rna_Cst3", "rna_Irf4", "rna_Irf8", "rna_S100a4", "rna_Zbtb46", "rna_Cd74", "rna_Cd86", "rna_Ly86", "rna_Siglech", "rna_Ccr9", "rna_Bst2", "rna_Tcf4", "rna_Clec4b1", "rna_Irf7", "rna_Pld4","rna_Cd19", "rna_Cd79a", "rna_Cd79b", "rna_Ighm", "rna_Ighd", "rna_Igkc", "rna_Igha", "rna_Ms4a1", "rna_Mzb1", "rna_Mki67", "rna_Pclaf", "rna_Nusap1", "rna_Birc5", "rna_Top2a")


markers_list_ADT <- c("adt-CD335", "adt-KLRG1", "adt-NK-1.1", "adt-CD27", "adt-CD11b", "adt-CD3","adt-CD4", "adt-CD8a", "adt-CD8b", "adt-TCR-gamma-delta", "adt-TCR-V-gamma-2", "adt-CD25", "adt-CD39", "adt-CD127", "adt-CD152", "adt-CD62L", "adt-CD134", "adt-CD357", "adt-CD278", "adt-CD304", "adt-CD279", "adt-CD11c", "adt-CD169-SIGLEC1", "adt-CD170", "adt-CD200R", "adt-MAC-2", "adt-Ly-6C", "adt-CD14", "adt-CD16-32", "adt-CD115", "adt-CD40", "adt-CD24","adt-CD26", "adt-I-A-I-E", "adt-CD103", "adt-ESAM","adt-CD135", "adt-CD172a","adt-XCR1", "adt-CD370", "adt-CD197", "adt-SIGLECH", "adt-CD199", "adt-CD317", "adt-CD19", "adt-CD79b", "adt-IgM", "adt-IgD")


#pdf(paste0(fig_path, "Manuscript_Dotplot_celltypemarkers_all_integrated_res0.8.pdf"), width = 20, height = 8)
DotPlot(all_integrated_rnm_DotPlot, features = markers_list_RNA, assay = "RNA") + scale_color_gradient2(low = "navy", mid = "white", high = "red", midpoint = 0) + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

DotPlot(all_integrated_rnm_DotPlot, features = markers_list_ADT, assay = "ADT") + scale_color_gradient2(low = "navy", mid = "white", high = "red", midpoint = 0)  + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10))

##dev.off()



```

# 2. Volcano plots for each cell type
```{r}
library(EnhancedVolcano)

#retrieve differential expression analysis with logfc.threshold = 0 for volcano plot
#genes
DefaultAssay(object = all_integrated_rnm) <- "RNA"
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Finding differentially expressed genes across clusters (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)

get_differential_genes_volcano <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", logfc.threshold = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_genes_volcano <- map_dfr(tmp_celltype_list, get_differential_genes_volcano)


#cairo_pdf(file = paste0(fig_path, "Volcano_Plot_with_DEGs_all_integrated_res0.8.pdf"), width = 8, height = 5, fallback_resolution = 600, onefile = TRUE)
for(i in unique(differential_genes_volcano$cell_type)){
  x <- differential_genes_volcano[differential_genes_volcano$cell_type %in% i, ]
  x_up <- x[x$p_val_adj < 0.05 & x$avg_log2FC >0, ]
  x_down <- x[x$p_val_adj < 0.05 & x$avg_log2FC < 0, ]
  selectlab <- na.omit(c(x_up$gene[1:10], x_down$gene[1:10]))
 plot_volc <- EnhancedVolcano(x, lab = x$gene, x = 'avg_log2FC', y = 'p_val_adj', selectLab = selectlab, legendPosition = "right", legendLabSize = 10, legendIconSize = 2, title = paste0("Volcano Plot: Genes -\nIL27RA KO vs wt :",  i, "\n"), subtitle = " ", gridlines.major = FALSE, gridlines.minor = FALSE, caption = " ", FCcutoff = 0.25, pCutoff = 0.05, drawConnectors = TRUE, arrowheads = FALSE, ylab = bquote(~-Log[10] ~ adjusted ~ italic(P))) 

 print(plot_volc + annotate("text", x = min(x$avg_log2FC) - 1, y = max(-log10(x$p_val_adj)) + 3 , label = paste0(sprintf('\u21e9'), nrow(x_down)), size = 6, fontface = "plain") + annotate("text", x =  max(x$avg_log2FC) + 1, y = max(-log10(x$p_val_adj)) + 3, label = paste0(sprintf('\u21e7'), nrow(x_up)), size = 6, fontface = "plain") ) 

}
#dev.off()

rm(x, x_up, x_down, plot_volc)


#==========================ADT
#ADT
#ADT differential expression was already calculated with logfc.threshold; using differential_ADT_all_integrated_rnm for volano plots

#cairo_pdf(file = paste0(fig_path, "Volcano_Plot_with_diffADTs_all_integrated_res0.8.pdf"), width = 8, height = 6.5, fallback_resolution = 600, onefile = TRUE)
for(i in unique(differential_ADT_all_integrated_rnm$cell_type)){
  x <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% i, ]
  x_up <- x[x$p_val_adj < 0.05 & x$avg_log2FC >0, ]
  x_down <- x[x$p_val_adj < 0.05 & x$avg_log2FC < 0, ]
  selectlab <- na.omit(c(x_up$gene[1:10], x_down$gene[1:10]))
 plot_volc <- EnhancedVolcano(x, lab = x$gene, x = 'avg_log2FC', y = 'p_val_adj', selectLab = selectlab, legendPosition = "right", legendLabSize = 10, legendIconSize = 2, title = paste0("Volcano Plot: ADTs -\nIL27RA KO vs wt :",  i, "\n"), subtitle = " ", gridlines.major = FALSE, gridlines.minor = FALSE, caption = " ", FCcutoff = 0.25, pCutoff = 0.05, drawConnectors = TRUE, arrowheads = FALSE, ylab = bquote(~-Log[10] ~ adjusted ~ italic(P))) 

 print(plot_volc + annotate("text", x = min(x$avg_log2FC) - 1, y = max(-log10(x$p_val_adj)) + 3 , label = paste0(sprintf('\u21e9'), nrow(x_down)), size = 6, fontface = "plain") + annotate("text", x =  max(x$avg_log2FC) + 1, y = max(-log10(x$p_val_adj)) + 3, label = paste0(sprintf('\u21e7'), nrow(x_up)), size = 6, fontface = "plain") ) 

}
#dev.off()

rm(x, x_up, x_down, plot_volc)


```



# Plotting co-stimulatory and co-inhibitory genes and proteins separately (Reference: Chihara N, et al. Nature. 2018;558(7710):454-459. Figure 1, Supplementary Table 1: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6130914/#SD2)
-Classification from Chen L, Flies DB. Nat Rev Immunol. 2013 Apr;13(4):227-42. PMID: 23470321https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3786574/<br>
-Upd 3Mar2023: t o repalce CD200 with CD2 and CD28; 13March2023: added CD27, Ly49D, Ly49H, Ly49A, Ly49G <br>
```{r}
## Plot of co-stimulatory and co-inhibitory genes and proteins, as suggested by Johannes (Reference: Chihara N, et al. Nature. 2018;558(7710):454-459.Figure 1, Supplementary Table 1: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6130914/#SD2)
## Classification from Chen L, Flies DB. Nat Rev Immunol. 2013 Apr;13(4):227-42. PMID: 23470321https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3786574/<br>

coStim_ADT <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD90.2", "adt-CD365") #"adt-CD365" is added from Chen and Flies

coInhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94")


coStim_genes <- c( "rna_Tnfrsf18", "rna_Cd2", "rna_Cd27", "rna_Icos", "rna_Tnfrsf9", "rna_Tnfrsf4", "rna_Slamf6", "rna_Cd226", "rna_Tnfrsf14", "rna_Thy1", "rna_Havcr1") #"Havcr1" == Tim1 is added from Chen and Flies; Cd28 RNA not present in this data

coInhib_genes <- c("rna_Pdcd1", "rna_Havcr2", "rna_Lag3", "rna_Tigit", "rna_Ctla4", "rna_Cd160", "rna_Btla", "rna_Lilrb4a", "rna_Klrd1")


#pdf(paste0(fig_path, "/Co-Stim_co-inhibitoryGeneModule_allCellTypes_res0.8_Chihara_2018.pdf"), width = 10, height = 15)
grid::grid.newpage()
grid::grid.text("Co-stimulatory genes and ADT:\nChihara N, et al. Nature. 2018;558(7710):454-459\n")
VlnPlot(all_integrated_rnm_VlnPlot, features = coStim_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("")
VlnPlot(all_integrated_rnm_VlnPlot, features = coStim_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") 

grid::grid.newpage()
grid::grid.text("Co-inhibitory genes and ADT:\nChihara N, et al. Nature. 2018;558(7710):454-459\n")
VlnPlot(all_integrated_rnm_VlnPlot, features = coInhib_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("")
VlnPlot(all_integrated_rnm_VlnPlot, features = coInhib_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") #"Lilrb4b" not present in this dataset

##dev.off()

#pdf(paste0(fig_path, "/Co-Stim_co-inhibitoryGeneModule_TcellsSubsets_res0.8_Chihara_2018.pdf"), width = 12, height = 10)

grid::grid.newpage()
grid::grid.text("Co-stimulatory genes and ADT:\nChihara N, et al. Nature. 2018;558(7710):454-459\n")
VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coStim_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("")

VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coStim_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("")


grid::grid.newpage()
grid::grid.text("Co-inhibitory genes and ADT:\nChihara N, et al. Nature. 2018;558(7710):454-459\n")

VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coInhib_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("")

VlnPlot(subset(all_integrated_rnm_VlnPlot, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coInhib_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") #"Lilrb4b" not present in this dataset

##dev.off()



# Module score
tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

#Renaming for Seurat's module score analysis without the "rna_" tag
coStim_genes <- c( "Tnfrsf18", "Cd2", "Cd27", "Icos", "Tnfrsf9", "Tnfrsf4", "Slamf6", "Cd226", "Tnfrsf14", "Thy1", "Havcr1")

coInhib_genes <- c("Pdcd1", "Havcr2", "Lag3", "Tigit", "Ctla4", "Cd160", "Btla", "Lilrb4a", "Klrd1")

#pdf(paste0(fig_path, "ModuleScore_Tcell_CoStimulatory_Co-InhibitoryGenes_Chihara_2018.pdf"), width = 12, height = 8)
#DefaultAssay(tmp) <- "ADT"
#tmp <- AddModuleScore(tmp, features = coStim_Inhib_ADT, name = "coStim_Inhib_ADT", assay = "ADT") #does nt analyze for ADT data

DefaultAssay(tmp) <- "RNA"
tmp <- AddModuleScore(tmp, features = list(coStim_genes), name = "coStimulatory_RNA")

tmp <- AddModuleScore(tmp, features = list(coInhib_genes), name = "coInhibitory_RNA")

grid::grid.newpage()
grid::grid.text("FeaturePlot of Co-stimulatory gene module:\nChihara N, et al. Nature. 2018;558(7710):454-459\n")
FeaturePlot(tmp, features = "coStimulatory_RNA1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))


grid::grid.newpage()
grid::grid.text("FeaturePlot of Co-inhibitory gene module:\nChihara N, et al. Nature. 2018;558(7710):454-459\n")
FeaturePlot(tmp, features = "coInhibitory_RNA1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))
##dev.off()




#VlnPlot(all_integrated_rnm_VlnPlot, "adt-CD279", split.by = "group", pt.size = 0.1) + stat_compare_means(method = "wilcox", aes(label=..p.adj..))                        ) #https://github.com/kassambara/ggpubr/issues/65; http://www.sthda.com/english/articles/24-ggpubr-publication-ready-plots/76-add-p-values-and-significance-levels-to-ggplots/


#Module: ADT
tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"

coStim_ADT <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD90.2", "adt-CD365") #"adt-CD365" is added from Chen and Flies

coInhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94")
DefaultAssay(tmp) <- "ADT"

#pdf(paste0(fig_path, "ModuleScore_Tcell_CoStim_Co-Inhib_ADT_ctrl5Module_Chihara_2018.pdf"), width = 12, height = 8)
#DefaultAssay(tmp) <- "ADT"
#tmp <- AddModuleScore(tmp, features = coStim_Inhib_ADT, name = "coStim_Inhib_ADT", assay = "ADT") #does nt analyze for ADT data

DefaultAssay(tmp) <- "ADT"
#Setting ctrl value = 5 as the number of ADT features to plot is less (https://satijalab.org/seurat/reference/addmodulescore)
tmp <- AddModuleScore(tmp, features = list(coStim_ADT), name = "coStimulatory_ADT", assay = "ADT", ctrl = 5)

tmp <- AddModuleScore(tmp, features = list(coInhib_ADT), name = "coInhibitory_ADT", assay = "ADT", ctrl = 5)

grid::grid.newpage()
grid::grid.text("FeaturePlot of Co-stimulatory ADT module:\nChihara N, et al. Nature. 2018;558(7710):454-459\nctrl = 5 AddModuleScore control features\n")
FeaturePlot(tmp, features = "coStimulatory_ADT1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))


grid::grid.newpage()
grid::grid.text("FeaturePlot of Co-inhibitory ADT module:\nChihara N, et al. Nature. 2018;558(7710):454-459\nctrl = 5 AddModuleScore control features\n")
FeaturePlot(tmp, features = "coInhibitory_ADT1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(1.0,0.2))
##dev.off()

#======================================
#MA plot of co-stimulatory and co-inhibitory genes

#pdf(paste0(fig_path, "MAplot_Costim_coInhib_Genes_Chihara_2018_markers.pdf"), height = 8, width = 12)
for(i in unique(differential_genes_all_integrated_rnm$cell_type)){
  #subset differentially expressed gene list to DE genes from each cell type
  subset_deg <- subset(differential_genes_all_integrated_rnm, cell_type %in% i & gene %in% c(coStim_genes, coInhib_genes))
  subset_deg$status <- ifelse(subset_deg$gene %in% coStim_genes, "Co-stimulatory", "Co-inhibitory")
  
  if(nrow(subset_deg) != 0){
    #Source code for MA plot
    avgExprs_costim_coinhib <- AverageExpression(all_integrated_rnm, assays = "RNA", slot = "data", group.by = c("celltype"), features = c(subset_deg$gene) ) #If slot is set to 'data', this function assumes that the data has been log normalized and therefore feature values are exponentiated prior to averaging so that averaging is done in non-log space. 
    # extract average expression for each cell type
    avgExprs.tmp <- as.data.frame(avgExprs_costim_coinhib$RNA[, i])
    colnames(avgExprs.tmp) <- "avg_expr"
    
    #Add average exp value to subset differential exp results
    if(nrow(avgExprs.tmp) > 1){
      subset_deg$avg_exp <- avgExprs.tmp$avg_expr[match(subset_deg$gene, rownames(avgExprs.tmp))]
      
      subset_deg$p_val_adj_color <- ifelse(subset_deg$p_val_adj < 0.05, "Significant", "Non-significant")
      
      subset_deg$status_color <- ifelse(subset_deg$status == "Co-stimulatory", "darkgoldenrod4", "dodgerblue4")
  
    #Source code:https://github.com/satijalab/seurat/issues/4167; https://bioinformatics.stackexchange.com/questions/18059/findmarkers-from-seurat-returns-p-values-as-0-for-highly-significant-genes
      group.color <- c("Significant" = "red", "Non-significant" = "black")
      status_color <- c("Co-stimulatory" ="darkgoldenrod4", "Co-inhibitory" = "dodgerblue4")
  
    #MA plot
    print(ggplot(subset_deg, aes(fill=factor(p_val_adj_color), color  = status) ) +  aes(x= log1p(avg_exp), y=avg_log2FC) +
    geom_point(size = 2, shape = 21) + geom_text_repel(aes(label = gene), color = subset_deg$status_color) + theme_classic() + geom_hline(yintercept = 0, linetype = "dashed") + scale_fill_manual(values = group.color ) + scale_color_manual(values = status_color) +  guides(fill = guide_legend(title = "Significance"), color = guide_legend(title = "Status")) + ggtitle(paste0(i, ": Chihara et al.[PMC6130914]\nco-stimulatory and co-inhibitory markers\n") ) ) #black = non-significant; red= significant
    }else{
      subset_deg$avg_exp <- avgExprs.tmp$avg_expr
      
      subset_deg$p_val_adj_color <- ifelse(subset_deg$p_val_adj < 0.05, "Significant", "Non-significant")
      
      subset_deg$status_color <- ifelse(subset_deg$status == "Co-stimulatory", "darkgoldenrod4", "dodgerblue4")
  
    #Source code:https://github.com/satijalab/seurat/issues/4167; https://bioinformatics.stackexchange.com/questions/18059/findmarkers-from-seurat-returns-p-values-as-0-for-highly-significant-genes
      group.color <- c("Significant" = "red", "Non-significant" = "black")
      #subset_deg$shape <- ifelse(subset_deg$status == "Co-stimulatory", 16, 17)
      status_color <- c("Co-stimulatory" ="darkgoldenrod4", "Co-inhibitory" = "dodgerblue4")
  
    #MA plot
    print(ggplot(subset_deg, aes(fill=factor(p_val_adj_color), color  = status) ) +  aes(x= log1p(avg_exp), y=avg_log2FC) +
    geom_point(size = 2, shape = 21) + geom_text_repel(aes(label = gene), color = subset_deg$status_color) + theme_classic() + geom_hline(yintercept = 0, linetype = "dashed") + scale_fill_manual(values = group.color ) + scale_color_manual(values = status_color) +  guides(fill = guide_legend(title = "Significance"), color = guide_legend(title = "Status")) + ggtitle(paste0(i, ": Chihara et al.[PMC6130914]\nco-stimulatory and co-inhibitory markers\n") ) ) #black = non-significant; red= significant
    }
    
    }
    
}
##dev.off()

#ggplot(subset_deg, aes(fill=factor(p_val_adj_color), color = status) ) +  aes(x= log1p(avg_exp), y=avg_log2FC) + geom_point(size = 2, shape = 21) + geom_text_repel(aes(label = gene), color = subset_deg$status_color) + theme_classic() + geom_hline(yintercept = 0, linetype = "dashed") + scale_fill_manual(values = group.color ) + scale_color_manual(values = c("darkgoldenrod4", "dodgerblue4")) +  guides(fill = guide_legend(title = "Significance"), color = guide_legend(title = "Status"))


#==================================================================================================
#MA plot of co-stimulatory and co-inhibitory ADT
coStim_ADT <- c( "adt-CD357", "adt-CD2", "adt-CD27", "adt-CD28", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD90.2", "adt-CD365") #"adt-CD365" is added from Chen and Flies

coInhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94")

#pdf(paste0(fig_path, "MAplot_Costim_coInhib_ADT_Chihara_2018_markers.pdf"), height = 8, width = 12)
for(i in unique(differential_ADT_all_integrated_rnm$cell_type)){
  #subset differentially expressed gene list to DE genes from each cell type
  subset_deg <- subset(differential_ADT_all_integrated_rnm, cell_type %in% i & gene %in% c(coStim_ADT, coInhib_ADT))
  subset_deg$status <- ifelse(subset_deg$gene %in% coStim_ADT, "Co-stimulatory", "Co-inhibitory")
  
  if(nrow(subset_deg) != 0){
    #Source code for MA plot
    avgExprs_costim_coinhib <- AverageExpression(all_integrated_rnm, assays = "ADT", slot = "data", group.by = c("celltype"), features = c(subset_deg$gene) ) #If slot is set to 'data', this function assumes that the data has been log normalized and therefore feature values are exponentiated prior to averaging so that averaging is done in non-log space. 
    # extract average expression for each cell type
    avgExprs.tmp <- as.data.frame(avgExprs_costim_coinhib$ADT[, i])
    colnames(avgExprs.tmp) <- "avg_expr"
    
    #Add average exp value to subset differential exp results
    if(nrow(avgExprs.tmp) > 1){
      subset_deg$avg_exp <- avgExprs.tmp$avg_expr[match(subset_deg$gene, rownames(avgExprs.tmp))]
      
      subset_deg$p_val_adj_color <- ifelse(subset_deg$p_val_adj < 0.05, "Significant", "Non-significant")
      
      subset_deg$status_color <- ifelse(subset_deg$status == "Co-stimulatory", "darkgoldenrod4", "dodgerblue4")
  
    #Source code:https://github.com/satijalab/seurat/issues/4167; https://bioinformatics.stackexchange.com/questions/18059/findmarkers-from-seurat-returns-p-values-as-0-for-highly-significant-genes
      group.color <- c("Significant" = "red", "Non-significant" = "black")
      status_color <- c("Co-stimulatory" ="darkgoldenrod4", "Co-inhibitory" = "dodgerblue4")
  
    #MA plot
    print(ggplot(subset_deg, aes(fill=factor(p_val_adj_color), color  = status) ) +  aes(x= log1p(avg_exp), y=avg_log2FC) +
    geom_point(size = 2, shape = 21) + geom_text_repel(aes(label = gene), color = subset_deg$status_color) + theme_classic() + geom_hline(yintercept = 0, linetype = "dashed") + scale_fill_manual(values = group.color ) + scale_color_manual(values = status_color) +  guides(fill = guide_legend(title = "Significance"), color = guide_legend(title = "Status")) + ggtitle(paste0(i, ": Chihara et al.[PMC6130914]\nco-stimulatory and co-inhibitory markers\n") ) ) #black = non-significant; red= significant
    }else{
      subset_deg$avg_exp <- avgExprs.tmp$avg_expr
      
      subset_deg$p_val_adj_color <- ifelse(subset_deg$p_val_adj < 0.05, "Significant", "Non-significant")
      
      subset_deg$status_color <- ifelse(subset_deg$status == "Co-stimulatory", "darkgoldenrod4", "dodgerblue4")
  
    #Source code:https://github.com/satijalab/seurat/issues/4167; https://bioinformatics.stackexchange.com/questions/18059/findmarkers-from-seurat-returns-p-values-as-0-for-highly-significant-genes
      group.color <- c("Significant" = "red", "Non-significant" = "black")
      status_color <- c("Co-stimulatory" ="darkgoldenrod4", "Co-inhibitory" = "dodgerblue4")
  
    #MA plot
    print(ggplot(subset_deg, aes(fill=factor(p_val_adj_color), color  = status) ) +  aes(x= log1p(avg_exp), y=avg_log2FC) +
    geom_point(size = 2, shape = 21) + geom_text_repel(aes(label = gene), color = subset_deg$status_color) + theme_classic() + geom_hline(yintercept = 0, linetype = "dashed") + scale_fill_manual(values = group.color ) + scale_color_manual(values = status_color) +  guides(fill = guide_legend(title = "Significance"), color = guide_legend(title = "Status")) + ggtitle(paste0(i, ": Chihara et al.[PMC6130914]\nco-stimulatory and co-inhibitory markers\n") ) ) #black = non-significant; red= significant
    }
    
    }
    
}
##dev.off()

```



#31Aug2022: MA plot of Gzmb
```{r}
#MA plot of Gzmb

#pdf(paste0(fig_path, "MAplot_GZMB_SW0100_P239_Legionella.pdf"), height = 8, width = 12)
for(i in c("CD27loCD11b+_NK", "CD27hiCD11b+_NK", "CD27loCD11blo_NK", "CD8_TCM", "CD8_TEM", "CD4_TEM", "CD27+_gdT")){
  #subset differentially expressed gene list to DE genes from each cell type
  subset_deg <- subset(differential_genes_all_integrated_rnm, cell_type %in% i & gene %in% c("Gzmb"))
  
  if(nrow(subset_deg) != 0){
    #Source code for MA plot
    avgExprs_Gzmb <- AverageExpression(all_integrated_rnm, assays = "RNA", slot = "data", group.by = c("celltype"), features = "Gzmb" ) #If slot is set to 'data', this function assumes that the data has been log normalized and therefore feature values are exponentiated prior to averaging so that averaging is done in non-log space. 
    # extract average expression for each cell type
    avgExprs.tmp <- as.data.frame(avgExprs_Gzmb$RNA[, i])
    colnames(avgExprs.tmp) <- "avg_expr"
    
    #Add average exp value to subset differential exp results
         subset_deg$avg_exp <- avgExprs.tmp$avg_expr
      
      subset_deg$p_val_adj_color <- ifelse(subset_deg$p_val_adj < 0.05, "Significant", "Non-significant")
      
      
    #Source code:https://github.com/satijalab/seurat/issues/4167; https://bioinformatics.stackexchange.com/questions/18059/findmarkers-from-seurat-returns-p-values-as-0-for-highly-significant-genes
      group.color <- c("Significant" = "red", "Non-significant" = "black")
      
  
    #MA plot
    print(ggplot(subset_deg, aes(fill=factor(p_val_adj_color)) ) +  aes(x= log1p(avg_exp), y=avg_log2FC) +
    geom_point(size = 2, shape = 21) + geom_text_repel(aes(label = gene)) + theme_classic() + geom_hline(yintercept = 0, linetype = "dashed") + scale_fill_manual(values = group.color ) +  guides(fill = guide_legend(title = "Significance")) + ggtitle(paste0(i, ": Gzmb\n") ) ) #black = non-significant; red= significant
   
    
    }
    
}
##dev.off()
```


#Upd 2Feb2023: Figures for manuscript
## UMAP plot and vlnplot of costimulatory and coinhibitory receptors (update 3Mar2023: replacing CD200 with CD2 and CD28, adding CD27 from the Chen and Flies 2013 article)
```{r}
tmp <- all_integrated_rnm
tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )

#pdf(paste0(fig_path, "DimPlot_SW0100_P239_Legionella_Manuscript.pdf"), width = 8, height = 6)
DimPlot(tmp, reduction = "wnn.umap", label = TRUE, pt.size = 0.5, repel = TRUE, label.size = 5, split.by = "group") + NoLegend() + theme(axis.text = element_text(size = rel(1.25)), axis.title = element_text(size = rel(1.25)))

##dev.off()

Idents(tmp) <- "celltype"
Idents(tmp) <- factor(Idents(tmp),levels = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell"))

#pdf(paste0(fig_path, "VlnPlot_CD25_CD69_ADT_NK_Tcell_SW0100_P239_Legionella_Manuscript.pdf"), width = 5, height = 4)
VlnPlot(tmp, c("adt-CD25", "adt-CD69"), split.by = "group", stack = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), assay = "ADT") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))
##dev.off()

#pdf(paste0(fig_path, "VlnPlot_Opt2_CD25_CD69_ADT_NK_Tcell_SW0100_P239_Legionella_Manuscript.pdf"), width = 7, height = 5)
VlnPlot(tmp, c("adt-CD25", "adt-CD69"), split.by = "group", stack = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), assay = "ADT", flip = TRUE) & theme(axis.text.y.left =  element_text(size = 14), axis.title.x = element_blank(), axis.text.y.right = element_text(size = 14), axis.text.x = element_text(size = 14, angle = 90,hjust = 1, vjust = 0.5),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))
##dev.off()


#Violin Plot of costimulatory and coinhibitory receptors
coStim_ADT <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365") #"adt-CD365" is added from Chen and Flies

coInhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94")


coStim_genes <- c( "rna_Tnfrsf18", "rna_Cd2", "rna_Cd28", "rna_Cd27", "rna_Icos", "rna_Tnfrsf9", "rna_Tnfrsf4", "rna_Slamf6", "rna_Cd226", "rna_Tnfrsf14", "rna_Havcr1") #"Havcr1" == Tim1 is added from Chen and Flies

coInhib_genes <- c("rna_Pdcd1", "rna_Havcr2", "rna_Lag3", "rna_Tigit", "rna_Ctla4", "rna_Cd160", "rna_Btla", "rna_Lilrb4a", "rna_Klrd1")


#pdf(paste0(fig_path, "/Co-Stim_co-inhibitoryGeneModule_TcellsSubsets_res0.8_Chihara_2018_Chen_and_Flies_2013_forManuscript.pdf"), width = 12, height = 10)

grid::grid.newpage()
grid::grid.text("Co-stimulatory genes and ADT:\nChihara N, et al. Nature. 2018;558(7710):454-459\nandChen L, Flies DB. Nat Rev Immunol. 2013;13(4):227-42\n")
VlnPlot(subset(tmp, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coStim_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

VlnPlot(subset(tmp, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coStim_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))


grid::grid.newpage()
grid::grid.text("Co-inhibitory genes and ADT:\nChihara N, et al. Nature. 2018;558(7710):454-459\nandChen L, Flies DB. Nat Rev Immunol. 2013;13(4):227-42\n")

VlnPlot(subset(tmp, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coInhib_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

VlnPlot(subset(tmp, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")), features = coInhib_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))#"Lilrb4b" not present in this dataset

##dev.off()



# Module score
tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"
tmp$group <- factor(tmp$group, levels = c("wt", "il27raKO"))

#Renaming for Seurat's module score analysis without the "rna_" tag
coStim_genes <- c( "Tnfrsf18", "Cd2", "Cd27", "Icos", "Tnfrsf9", "Tnfrsf4", "Slamf6", "Cd226", "Tnfrsf14", "Havcr1")

coInhib_genes <- c("Pdcd1", "Havcr2", "Lag3", "Tigit", "Ctla4", "Cd160", "Btla", "Lilrb4a", "Klrd1")

#AddModuleScore list https://github.com/satijalab/seurat/issues/495#issuecomment-474608914
#pdf(paste0(fig_path, "FeaturePlot_Tcell_CoStimulatory_Co-InhibitoryGenes_Chihara_2018_Chen_Flies_2013_forManuscript.pdf"), width = 12, height = 8)
DefaultAssay(tmp) <- "RNA"
tmp <- AddModuleScore(tmp, features = list(coStim_genes), name = "coStimulatory_RNA")

tmp <- AddModuleScore(tmp, features = list(coInhib_genes), name = "coInhibitory_RNA")

grid::grid.newpage()
grid::grid.text("FeaturePlot of Co-stimulatory gene module:\nChihara N, et al. Nature. 2018;558(7710):454-459\nandChen L, Flies DB. Nat Rev Immunol. 2013;13(4):227-42\n")
FeaturePlot(tmp, features = "coStimulatory_RNA1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(0.95,0.2))


grid::grid.newpage()
grid::grid.text("FeaturePlot of Co-inhibitory gene module:\nChihara N, et al. Nature. 2018;558(7710):454-459\nandChen L, Flies DB. Nat Rev Immunol. 2013;13(4):227-42\n")
FeaturePlot(tmp, features = "coInhibitory_RNA1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(0.95,0.2))
##dev.off()



#Module: ADT
coStim_ADT <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365") #"adt-CD365" is added from Chen and Flies

coInhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94")
DefaultAssay(tmp) <- "ADT"

#pdf(paste0(fig_path, "FeaturePlot_Tcell_CoStim_Co-Inhib_ADT_ctrl5Module_Chihara_2018_Chen_Flies_2013_forManuscript.pdf"), width = 12, height = 8)

DefaultAssay(tmp) <- "ADT"
#Setting ctrl value = 5 as the number of ADT features to plot is less (https://satijalab.org/seurat/reference/addmodulescore)
tmp <- AddModuleScore(tmp, features = list(coStim_ADT), name = "coStimulatory_ADT", assay = "ADT", ctrl = 5)

tmp <- AddModuleScore(tmp, features = list(coInhib_ADT), name = "coInhibitory_ADT", assay = "ADT", ctrl = 5)

grid::grid.newpage()
grid::grid.text("FeaturePlot of Co-stimulatory ADT module:\nChihara N, et al. Nature. 2018;558(7710):454-459\nandChen L, Flies DB. Nat Rev Immunol. 2013;13(4):227-42\nctrl = 5 AddModuleScore control features\n")
FeaturePlot(tmp, features = "coStimulatory_ADT1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(0.95,0.2))


grid::grid.newpage()
grid::grid.text("FeaturePlot of Co-inhibitory ADT module:\nChihara N, et al. Nature. 2018;558(7710):454-459\nandChen L, Flies DB. Nat Rev Immunol. 2013;13(4):227-42\nctrl = 5 AddModuleScore control features\n")
FeaturePlot(tmp, features = "coInhibitory_ADT1", label = TRUE, repel = TRUE, split.by = "group", order=TRUE) & scale_color_gradient2(midpoint = 0, low = "navy", mid = "white", high = "red") & theme(legend.position = c(0.95,0.2))
##dev.off()

















#=====================================================================
coStim_ADT <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365") #"adt-CD365" is added from Chen and Flies

coInhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94")
DefaultAssay(tmp) <- "ADT"

#pdf(paste0(fig_path, "ModuleScore_Tcell_CoStim_Co-Inhib_ADT_ctrl5Module_Chihara_2018.pdf"), width = 12, height = 8)
#DefaultAssay(tmp) <- "ADT"
#tmp <- AddModuleScore(tmp, features = coStim_Inhib_ADT, name = "coStim_Inhib_ADT", assay = "ADT") #does nt analyze for ADT data

#Setting ctrl value = 5 as the number of ADT features to plot is less (https://satijalab.org/seurat/reference/addmodulescore)
tmp <- AddModuleScore(tmp, features = list(coStim_ADT), name = "coStimulatory_ADT", assay = "ADT", ctrl = 5)

tmp <- AddModuleScore(tmp, features = list(coInhib_ADT), name = "coInhibitory_ADT", assay = "ADT", ctrl = 5)

#https://github.com/satijalab/seurat/issues/3366#issuecomment-674262907
tmp[["coStimulatory_ADT"]] <- CreateAssayObject(data = t(x= FetchData(object = tmp, vars = "coStimulatory_ADT1")))

tmp[["coInhibitory_ADT"]] <- CreateAssayObject(data = t(x= FetchData(object = tmp, vars = "coInhibitory_ADT1")))


tmp[["coStim_CoInhib_ADT"]] <- CreateAssayObject(data = t(x= FetchData(object = tmp, vars = c("coStimulatory_ADT1", "coInhibitory_ADT1"))))


Idents(tmp) <- "cluster0.8.group"
  cluster.averages_coStim_CoInhib_ADT <- AverageExpression(tmp, slot = "data", assays = "coStim_CoInhib_ADT", return.seurat = TRUE)
levels(cluster.averages_coStim_CoInhib_ADT) <- c(
  "CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO",
"CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
"CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
"CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
"CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO",
"CD8_Nave_wt", "CD8_Nave_il27raKO",
"CD8_TCM_wt", "CD8_TCM_il27raKO",
"CD8_TEM_wt", "CD8_TEM_il27raKO",
"CD4_TEM_wt", "CD4_TEM_il27raKO",
"Treg_wt", "Treg_il27raKO",
"CD27+_gdT_wt", "CD27+_gdT_il27raKO",
"CD27-_gdT_wt",  "CD27-_gdT_il27raKO",
"Monocyte_wt", "Monocyte_il27raKO",
"Alveolar_macrophage_wt", "Alveolar_macrophage_il27raKO",
"cDC_wt", "cDC_il27raKO",
"pDC_wt",  "pDC_il27raKO",
"moDC_wt",  "moDC_il27raKO",
"B_cell_wt", "B_cell_il27raKO")
  cluster.averages_coStim_CoInhib_ADT <- ScaleData(cluster.averages_coStim_CoInhib_ADT)

  Seurat::DoHeatmap(cluster.averages_coStim_CoInhib_ADT, features = c("coStimulatory-ADT1", "coInhibitory-ADT1"), size = 3, draw.lines = FALSE, angle = 90) + scale_fill_gradient2(low = "navy", mid = "white", high = "red", midpoint = 0) +  theme(axis.text.y.left = element_text(size = 6.5, colour = "black", margin = margin(r=-15)))




```


# Dotplot of Cell type markers
-shortlisting a few selected cell type markers for visualization, based on Johannes' comment that it would be less complex to show only a few markers rather than the whole list
-markers selected are genes and ADTs that are able to distinguish between cell types and are canonical, more frequently used cell type markers from literature
- Upd16March2023: add3ed CD44 and CD62L to distinguish Naive, TEM and TCM T cells #https://jitc.biomedcentral.com/articles/10.1186/s40425-017-0235-4 #CD44lowCD62L Naive, CD44HighCD62L+ Central memory, CD44highCD62Lneg Effector-effector memory
```{r}
select_markers_list_RNA <- c("rna_Klrb1c", "rna_Ncr1","rna_Cd27", "rna_Itgam", "rna_Cd3d", "rna_Cd4", "rna_Cd8a", "rna_Cd8b1", "rna_Cd44", "rna_Sell", "rna_Tcrg-C1", "rna_Trgv2", "rna_Trdv4", "rna_Foxp3", "rna_Ctla4", "rna_Il2ra", "rna_Siglec1", "rna_Siglecf", "rna_Marco", "rna_F13a1", "rna_Ly6c2", "rna_Cd14", "rna_Fcgr4", "rna_Fcgr1", "rna_Cd40", "rna_Dpp4", "rna_H2-Ab1", "rna_H2-Eb1", "rna_Flt3", "rna_Xcr1", "rna_Cst3", "rna_Siglech", "rna_Bst2", "rna_Cd19", "rna_Cd79a", "rna_Cd79b") #https://jitc.biomedcentral.com/articles/10.1186/s40425-017-0235-4 #CD44lowCD62L Naive, CD44HighCD62L+ Central memory, CD44highCD62Lneg Effector-effector memory


select_markers_list_ADT <- c("adt-CD335", "adt-NK-1.1", "adt-CD27", "adt-CD11b", "adt-CD3","adt-CD4", "adt-CD8a", "adt-CD8b", "adt-CD44", "adt-CD62L", "adt-TCR-gamma-delta", "adt-TCR-V-gamma-2", "adt-CD11c", "adt-CD169-SIGLEC1", "adt-CD170", "adt-Ly-6C", "adt-CD14", "adt-CD16-32", "adt-CD40", "adt-CD26", "adt-I-A-I-E", "adt-CD135", "adt-XCR1", "adt-CD370", "adt-CD197", "adt-SIGLECH", "adt-CD199", "adt-CD317", "adt-CD19", "adt-CD79b") #https://jitc.biomedcentral.com/articles/10.1186/s40425-017-0235-4 #CD44lowCD62L Naive, CD44HighCD62L+ Central memory, CD44highCD62Lneg Effector-effector memory


#pdf(paste0(fig_path, "Manuscript_Dotplot_select_celltypemarkers_all_integrated_res0.8_Feb2023.pdf"), width = 8, height = 10)
DotPlot(all_integrated_rnm_VlnPlot, features = select_markers_list_RNA, assay = "RNA", dot.scale = 4) + scale_color_gradient2(low = "navy", mid = "white", high = "red", midpoint = 0) + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10)) + coord_flip()

DotPlot(all_integrated_rnm_VlnPlot, features = select_markers_list_ADT, assay = "ADT", dot.scale = 4) + scale_color_gradient2(low = "navy", mid = "white", high = "red", midpoint = 0)  + theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text.x = element_text(angle = 90, size = 12, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 12), legend.key.height= unit(0.25, 'in'), legend.key.width= unit(0.25, 'in'), legend.title = element_text(size=10), legend.text = element_text(size = 10)) + coord_flip()

##dev.off()
```


# Il27 violin plot
```{r}
tmp <- all_integrated_rnm_VlnPlot

tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )

#pdf(paste0(fig_path, "Manuscript_VlnPlot_Il27_Opt1_all_integrated_res0.8_Feb2023.pdf"), width = 6, height = 5)
VlnPlot(tmp, features = "Il27", assay = "RNA", split.by = "group") & theme(axis.text.x = element_text(size = 14, hjust = 1, vjust = 0.5, angle = 90), axis.title.x = element_blank(), axis.text.y = element_text(size = 14), axis.title = element_text(size = 14), legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))
##dev.off()

#pdf(paste0(fig_path, "Manuscript_VlnPlot_Il27_Opt2_all_integrated_res0.8_Feb2023.pdf"), width = 6, height = 5)
VlnPlot(tmp, features = "Il27", assay = "RNA", split.by = "group", pt.size = 0.0) & theme(axis.text.x = element_text(size = 14, hjust = 1, vjust = 0.5, angle = 90), axis.title.x = element_blank(), axis.text.y = element_text(size = 14), axis.title = element_text(size = 14), legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))
##dev.off()
```

#Plot enrichR GO BP for DE genes (unadjusted p<0.05)
```{r}
#add cell type name to the Seurat- enrichR GO BP enrichment object. The ernichment was based on DE genes with an unadjusted p value cutoff of 0.05 for the DE gene selection (#Source code for including names to nested list: https://stackoverflow.com/questions/72508806/how-to-add-a-list-name-as-a-column-at-a-nested-list-data-frame-in-r)

x <- enrichr_gobp_all_integrated_rnm

x <- Map(lapply, x, list(cbind), celltype = names(x))

#Add DE_type "pos" or "neg" to each nested dataframe
for(i in 1:length(x)){
  for(j in 1:length(x[[i]])){
    x[[i]][[j]]$DE_type <- names(x[[i]][j])
  }
}

y <- x
for(i in 1:length(y)){
  y[[i]] <- bind_rows(y[[i]])
  y[[i]]$DE_type <- factor(y[[i]]$DE_type, levels = c("pos", "neg"))
}

#Unlist the nested list (https://purrr.tidyverse.org/reference/flatten.html)
#x <- purrr::flatten_dfr(x) %>% data.frame()

p <- list()

for(i in 1:length(y)){
  z <- y[[i]] %>% as.data.frame() %>% dplyr::group_by(DE_type) %>%  dplyr::slice(1:5)
  
  #drop unused y-axis labels: https://stackoverflow.com/questions/66956127/drop-unused-levels-in-facet-wrap-in-ggplot2; label x-axis in decimal format for consistency across plots: https://scales.r-lib.org/reference/label_number.html #accuracy defines how many decimal places to show
  p[[names(y[i])]] <- ggplot(z, aes(x = -log10(GO_Biological_Process_2021.Adjusted.P.value), y = GO_Biological_Process_2021.Term) )+ geom_bar(stat = "identity", aes(fill = DE_type)) + scale_fill_manual(values = c("pos" = "red", "neg" = "blue")) + facet_wrap(z$DE_type, nrow = 2,ncol = 1, drop = TRUE, scales = "free_y", strip.position = "left", )  + theme_classic() + theme(panel.grid = element_blank(), strip.background = element_blank(), strip.placement = "outside", axis.text = element_text(size = 10, color = "black"), axis.title.x = element_text(size = 10, color = "black"), axis.title.y = element_blank(), legend.text = element_text(size = 10, color = "black"), legend.title = element_text(size = 10, color = "black"), strip.text = element_text(size = 10, color = "black")) + scale_x_continuous(expand = c(0, 0), labels = scales::label_number(accuracy = 0.1)) + xlab("-log10(adjusted P)") + ggtitle(names(y[i])) 


}


#pdf(paste0(fig_path, "Manuscript_SeuratDE_EnrichR_GOBP_Barplot_all_integrated_res0.8_Feb2023.pdf"), width = 20, height = 15)

wrap_plots(p[1:9], nrow = 5, ncol = 2) + plot_layout(guides = 'collect')

wrap_plots(p[10:18], nrow = 5, ncol = 2) + plot_layout(guides = 'collect')

##dev.off()


```


# Il10 violin plot
```{r}
tmp <- all_integrated_rnm_VlnPlot

tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )

#pdf(paste0(fig_path, "Manuscript_VlnPlot_Il10_Opt1_all_integrated_res0.8_Feb2023.pdf"), width = 6, height = 5)
VlnPlot(tmp, features = "Il10", assay = "RNA", split.by = "group") & theme(axis.text.x = element_text(size = 14, hjust = 1, vjust = 0.5, angle = 90), axis.title.x = element_blank(), axis.text.y = element_text(size = 14), axis.title = element_text(size = 14), legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))
##dev.off()


#pdf(paste0(fig_path, "Manuscript_VlnPlot_Il10_Opt2_all_integrated_res0.8_Feb2023.pdf"), width = 6, height = 5)
VlnPlot(tmp, features = "Il10", assay = "RNA", split.by = "group", pt.size = 0.0) & theme(axis.text.x = element_text(size = 14, hjust = 1, vjust = 0.5, angle = 90), axis.title.x = element_blank(), axis.text.y = element_text(size = 14), axis.title = element_text(size = 14), legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))
##dev.off()

```


# Ifng, Stat1, Irf1, Gbp2, Ccl5  violin plot
```{r}
tmp <- all_integrated_rnm_VlnPlot

tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )

#pdf(paste0(fig_path, "/VlnPlot_Ifng_Stat1_Irf1_Gbp2_Ccl5_res0.8_forManuscript.pdf"), width = 8, height = 8)
VlnPlot(tmp, features = c("Ifng", "Stat1", "Irf1", "Gbp2", "Ccl5"), stack = TRUE, same.y.lims = TRUE, split.by = "group", assay = "RNA") & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))
##dev.off()

```


#Upd 2March2023: UMAP and vlnplot of coinhibitory and costimulatory receptors in NK cells; plotting Chihara et al. 2018;Chen and Flies 2013; and NK specific receptors: Rahim MM, et al. Front Immunol. 2014;5:145 (https://pubmed.ncbi.nlm.nih.gov/24765094/); Paul S, Lal G. Front Immunol. 2017;8:1124 (https://pubmed.ncbi.nlm.nih.gov/28955340/)
```{r}
tmp <- all_integrated_rnm
tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )
Idents(tmp) <- "celltype"
Idents(tmp) <- factor(Idents(tmp),levels = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell"))

#Violin Plot of costimulatory and coinhibitory receptors
coStim_ADT <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365", "adt-Ly49D", "adt-Ly49H") #"adt-CD365" is added from Chen and Flies; "adt-Ly49D", "adt-Ly49H" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017

coInhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94", "adt-Ly49A", "adt-Ly49G")# "adt-Ly49A", "adt-Ly49G" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017


coStim_genes <- c( "rna_Tnfrsf18", "rna_Cd2", "rna_Cd27", "rna_Icos", "rna_Tnfrsf9", "rna_Tnfrsf4", "rna_Slamf6", "rna_Cd226", "rna_Tnfrsf14", "rna_Havcr1", "rna_Klra4", "rna_Klra8") #"Havcr1" == Tim1 is added from Chen and Flies; "Klra4" = "Ly49D" and "Klra8" = "Ly49H" added from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017


coInhib_genes <- c("rna_Pdcd1", "rna_Havcr2", "rna_Lag3", "rna_Tigit", "rna_Ctla4", "rna_Cd160", "rna_Btla", "rna_Lilrb4a", "rna_Klrd1", "rna_Klra1", "rna_Klra7") #"Klra1" = "adt-Ly49A", "Klra7" = "adt-Ly49G" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017


#pdf(paste0(fig_path, "/Co-Stim_co-inhibitoryGeneModule_NKcellsSubsets_res0.8_Chihara_2018_Chen_and_Flies_2013_forManuscript.pdf"), width = 12, height = 8)

grid::grid.newpage()
grid::grid.text("Co-stimulatory genes and ADT:\nChihara N, et al. Nature. 2018;558(7710):454-459\nandChen L, Flies DB. Nat Rev Immunol. 2013;13(4):227-42\n")
VlnPlot(subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")), features = coStim_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

VlnPlot(subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")), features = coStim_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))


grid::grid.newpage()
grid::grid.text("Co-inhibitory genes and ADT:\nChihara N, et al. Nature. 2018;558(7710):454-459\nandChen L, Flies DB. Nat Rev Immunol. 2013;13(4):227-42\n")

VlnPlot(subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")), features = coInhib_ADT, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

VlnPlot(subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")), features = coInhib_genes, stack = TRUE, same.y.lims = TRUE, split.by = "group") & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))#"Lilrb4b" not present in this dataset

##dev.off()
```


```{r}
tmp <- all_integrated_rnm
tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )
tmp$celltype_whole <- tmp$celltype
tmp$celltype_whole <- ifelse(tmp$celltype_whole %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), "NK_cell", ifelse(tmp$celltype_whole %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), "T_cell", tmp$celltype_whole))
tmp$celltype_whole_group <- paste0(tmp$celltype_whole, "_", tmp$group)

Idents(tmp) <- tmp$celltype_whole_group

coStim_ADT_T <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365") #"adt-CD365" is added from Chen and Flies

coInhib_ADT_T <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94")

coStim_ADT_NK <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365", "adt-Ly49D", "adt-Ly49H") #"adt-CD365" is added from Chen and Flies; "adt-Ly49D", "adt-Ly49H" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017

coInhib_ADT_NK <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94", "adt-Ly49A", "adt-Ly49G")# "adt-Ly49A", "adt-Ly49G" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017

x_diff_NK_costim <- FindMarkers(tmp, features = c(coStim_ADT_NK), assay = "ADT", ident.1 = "NK_cell_il27raKO", ident.2 = "NK_cell_wt", logfc.threshold = 0)

x_diff_NK_coInhib <- FindMarkers(tmp, features = c(coInhib_ADT_NK), assay = "ADT", ident.1 = "NK_cell_il27raKO", ident.2 = "NK_cell_wt", logfc.threshold = 0)

x_diff_T_costim <- FindMarkers(tmp, features = c(coStim_ADT_T), assay = "ADT", ident.1 = "T_cell_il27raKO", ident.2 = "T_cell_wt", logfc.threshold = 0)

x_diff_T_coInhib <- FindMarkers(tmp, features = c(coInhib_ADT_T), assay = "ADT", ident.1 = "T_cell_il27raKO", ident.2 = "T_cell_wt", logfc.threshold = 0)


```


#Plotting avg heatmap for costimulatory and coinhibitory ADTs in NK and T cells
```{r}
#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm #assign seurat object to a temporary variable
Idents(tmp) <- "celltype"
tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )
tmp$celltype_whole <- tmp$celltype
tmp$celltype_whole <- ifelse(tmp$celltype_whole %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), "NK_cell", ifelse(tmp$celltype_whole %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), "T_cell", tmp$celltype_whole))
tmp$celltype_whole_group <- paste0(tmp$celltype_whole, "_", tmp$group)

coStim_ADT <- c( "adt-CD357", "adt-CD2","adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365", "adt-Ly49D", "adt-Ly49H") #"adt-CD365" is added from Chen and Flies; "adt-Ly49D", "adt-Ly49H" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017


coInhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94", "adt-Ly49A", "adt-Ly49G")# "adt-Ly49A", "adt-Ly49G" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017

#tmp <- AddModuleScore(tmp, features = list(coStim_ADT), name = "coStimulatory_ADT", assay = "ADT", ctrl = 5)
#tmp <- AddModuleScore(tmp, features = list(coInhib_ADT), name = "coInhibitory_ADT", assay = "ADT", ctrl = 5)

tmp_subset <- subset(tmp, celltype_whole %in% c("NK_cell", "T_cell"))
Idents(tmp_subset) <- tmp_subset$celltype_whole_group


cluster.averages <- AverageExpression(tmp_subset, slot = "data", assays = "ADT", return.seurat = TRUE)
  levels(cluster.averages) <- c("NK_cell_wt", "NK_cell_il27raKO", "T_cell_wt", "T_cell_il27raKO")
  cluster.averages <- ScaleData(cluster.averages)
 
  
  #pdf(paste0(fig_path, "/Combined_Heatmap_AvgExpr_Coinh_CoStim_NK_T_cell_Chen_Flies_Chihara.pdf"), width = 4, height = 5)
  Seurat::DoHeatmap(cluster.averages, features = c(coInhib_ADT, coStim_ADT), size = 3, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF", "#05BE78", "#0000FF"), raster = FALSE) + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 8, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10)) #raster= FALSE for DoHeatmap to prevent blurred raster images; Source: https://github.com/satijalab/seurat/issues/5996#issuecomment-1138601399
 ##dev.off()
 
  
  coStim_ADT_T <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365") #"adt-CD365" is added from Chen and Flies

coInhib_ADT_T <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94")

coStim_ADT_NK <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365", "adt-Ly49D", "adt-Ly49H") #"adt-CD365" is added from Chen and Flies; "adt-Ly49D", "adt-Ly49H" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017

coInhib_ADT_NK <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94", "adt-Ly49A", "adt-Ly49G")# "adt-Ly49A", "adt-Ly49G" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017

  
 tmp_subset_NK <- subset(tmp, celltype_whole %in% c("NK_cell"))
Idents(tmp_subset_NK ) <- tmp_subset_NK $celltype_whole_group

cluster.averages_NK  <- AverageExpression(tmp_subset_NK , slot = "data", assays = "ADT", return.seurat = TRUE)
  levels(cluster.averages_NK ) <- c("NK_cell_wt", "NK_cell_il27raKO")
  cluster.averages_NK  <- ScaleData(cluster.averages_NK )
 
 htmap_NK <- Seurat::DoHeatmap(cluster.averages_NK , features = c(coInhib_ADT_NK, coStim_ADT_NK), size = 4, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF"), raster = FALSE) + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 7, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10))
 
 
 tmp_subset_T <- subset(tmp, celltype_whole %in% c("T_cell"))
Idents(tmp_subset_T) <- tmp_subset_T$celltype_whole_group

cluster.averages_T <- AverageExpression(tmp_subset_T, slot = "data", assays = "ADT", return.seurat = TRUE)
  levels(cluster.averages_T) <- c("T_cell_wt", "T_cell_il27raKO")
  cluster.averages_T <- ScaleData(cluster.averages_T)
 htmap_T <- Seurat::DoHeatmap(cluster.averages_T, features = c(coInhib_ADT_T, coStim_ADT_T), size = 4, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF"), raster = FALSE) + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 7, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10))
 
#pdf(paste0(fig_path, "/Heatmap_AvgExpr_Coinh_CoStim_NK_T_cell_Chen_Flies_Chihara.pdf"), width = 6, height = 6)
wrap_plots(htmap_NK, htmap_T)
##dev.off()


```



```{r}
#c("adt-KLRG1", "adt-CD94", "adt-Ly49A", "adt-Ly49G", "adt-CD160", "adt-CD226", "adt-CD27", "adt-Ly108", "adt-Ly49D", "adt-Ly49H", "adt-CD137", "adt-CD134", "adt-CD226")
```



#Checking differences between CD8+ T cell subsets, CD4+ T cell subsets (with Tregs and gdT considered separately due to their different functions, and NK cell subsets)
```{r}

#RNA
# Combine all NK clusters into one and different T cell clusters into one

NK_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

CD8_Tcell_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM"), ]

CD4_Tcell_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD4_TEM"), ]

Treg_cell_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("Treg"), ]

gdT_cell_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27+_gdT", "CD27-_gdT"), ]

vennPlot_DGE_NK_vs_Tsubset_upreg <- venndetail(list(
  NK_up = NK_genes$gene[NK_genes$avg_log2FC > 0 & NK_genes$p_val_adj < 0.05], #subset to upregulated genes (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
  CD8_T_up = CD8_Tcell_genes$gene[CD8_Tcell_genes$avg_log2FC > 0 & CD8_Tcell_genes$p_val_adj < 0.05],
  CD4_T_up = CD4_Tcell_genes$gene[CD4_Tcell_genes$avg_log2FC > 0 & CD4_Tcell_genes$p_val_adj < 0.05],
  Treg_up = Treg_cell_genes$gene[Treg_cell_genes$avg_log2FC > 0 & Treg_cell_genes$p_val_adj < 0.05],
  gdT_up = gdT_cell_genes$gene[gdT_cell_genes$avg_log2FC > 0 & gdT_cell_genes$p_val_adj < 0.05]
))


vennPlot_DGE_NK_vs_Tsubset_downreg <- venndetail(list(
  NK_down = NK_genes$gene[NK_genes$avg_log2FC < 0 & NK_genes$p_val_adj < 0.05], #subset to downregulated genes (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  CD8_T_down = CD8_Tcell_genes$gene[CD8_Tcell_genes$avg_log2FC < 0 & CD8_Tcell_genes$p_val_adj < 0.05],
  CD4_T_down = CD4_Tcell_genes$gene[CD4_Tcell_genes$avg_log2FC < 0 & CD4_Tcell_genes$p_val_adj < 0.05],
  Treg_down = Treg_cell_genes$gene[Treg_cell_genes$avg_log2FC < 0 & Treg_cell_genes$p_val_adj < 0.05],
  gdT_down = gdT_cell_genes$gene[gdT_cell_genes$avg_log2FC < 0 & gdT_cell_genes$p_val_adj < 0.05]

  ))

#ADT
# Combine all NK clusters into one and all T cell clusters into one

NK_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

CD8_Tcell_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM"), ]

CD4_Tcell_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD4_TEM"), ]

Treg_cell_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("Treg"), ]

gdT_cell_ADT <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% c("CD27+_gdT", "CD27-_gdT"), ]

vennPlot_DEP_NK_vs_Tsubset_upreg <- venndetail(list(
  NK_up = NK_ADT$gene[NK_ADT$avg_log2FC > 0 & NK_ADT$p_val_adj < 0.05], #subset to upregulated ADT (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
  CD8_T_up = CD8_Tcell_ADT$gene[CD8_Tcell_ADT$avg_log2FC > 0 & CD8_Tcell_ADT$p_val_adj < 0.05],
  CD4_T_up = CD4_Tcell_ADT$gene[CD4_Tcell_ADT$avg_log2FC > 0 & CD4_Tcell_ADT$p_val_adj < 0.05],
  Treg_up = Treg_cell_ADT$gene[Treg_cell_ADT$avg_log2FC > 0 & Treg_cell_ADT$p_val_adj < 0.05],
  gdT_up = gdT_cell_ADT$gene[gdT_cell_ADT$avg_log2FC > 0 & gdT_cell_ADT$p_val_adj < 0.05]
))


vennPlot_DEP_NK_vs_Tsubset_downreg <- venndetail(list(
  NK_down = NK_ADT$gene[NK_ADT$avg_log2FC < 0 & NK_ADT$p_val_adj < 0.05], #subset to downregulated ADT (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  CD8_T_down = CD8_Tcell_ADT$gene[CD8_Tcell_ADT$avg_log2FC < 0 & CD8_Tcell_ADT$p_val_adj < 0.05],
  CD4_T_down = CD4_Tcell_ADT$gene[CD4_Tcell_ADT$avg_log2FC < 0 & CD4_Tcell_ADT$p_val_adj < 0.05],
  Treg_down = Treg_cell_ADT$gene[Treg_cell_ADT$avg_log2FC < 0 & Treg_cell_ADT$p_val_adj < 0.05],
  gdT_down = gdT_cell_ADT$gene[gdT_cell_ADT$avg_log2FC < 0 & gdT_cell_ADT$p_val_adj < 0.05]

  ))



vennPlot_list_subset <- list(vennPlot_DGE_NK_vs_T_upreg = result(vennPlot_DGE_NK_vs_Tsubset_upreg, wide = TRUE), vennPlot_DGE_NK_vs_T_downreg = result(vennPlot_DGE_NK_vs_Tsubset_downreg, wide = TRUE),
                      vennPlot_DPE_NK_vs_Tsubset_downreg = result(vennPlot_DEP_NK_vs_Tsubset_upreg, wide = TRUE), vennPlot_DPE_NK_vs_T_downreg = result(vennPlot_DEP_NK_vs_Tsubset_downreg, wide = TRUE))

#writexl::write_xlsx(vennPlot_list_subset, paste0(tab_path, "VennPlotdata_NK_vs_T_bySubset.xlsx") )


#Unique and commonly DE in NK and T cells
#Uniquely DE in NK cells
NK_vs_Tsubset_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1 & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1 )& !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1 )  ]

NK_vs_Tsubset_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1 & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1 )]

NK_vs_Tsubset_DEP_up <- vennPlot_DEP_NK_vs_Tsubset_upreg@wide$Detail[vennPlot_DEP_NK_vs_Tsubset_upreg@wide$NK_up == 1 & !(vennPlot_DEP_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1 ) & !(vennPlot_DEP_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1 ) & !(vennPlot_DEP_NK_vs_Tsubset_upreg@wide$Treg_up == 1 )& !(vennPlot_DEP_NK_vs_Tsubset_upreg@wide$gdT_up == 1 )  ]

NK_vs_Tsubset_DEP_down <- vennPlot_DEP_NK_vs_Tsubset_downreg@wide$Detail[vennPlot_DEP_NK_vs_Tsubset_downreg@wide$NK_down == 1 & !(vennPlot_DEP_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1 ) & !(vennPlot_DEP_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DEP_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DEP_NK_vs_Tsubset_downreg@wide$Treg_down == 1 ) & !(vennPlot_DEP_NK_vs_Tsubset_downreg@wide$gdT_down == 1 )]



#Uniquely DE in CD8 T cells
CD8_T_vs_NK_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[! (vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1) & vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1 & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1 )& !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1 )  ]

CD8_T_vs_NK_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[!(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1) & vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1 & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1 )]


CD4_T_vs_NK_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[! (vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1) & vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1 & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1 )& !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1 )  ]

CD4_T_vs_NK_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[!(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1) & vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1 )]


Treg_vs_NK_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[! (vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1) & vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1 & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1 )  ]

Treg_vs_NK_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[!(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1 & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1 )]


gdT_vs_NK_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[! (vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1) & vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1  ]

gdT_vs_NK_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[!(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1) & vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1]



#clusterProfiler enrichment
library(clusterProfiler)
goBP_unique_NK_vs_Tsubset_up_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(NK_vs_Tsubset_DE_up, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )
goBP_unique_NK_vs_Tsubset_down_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(NK_vs_Tsubset_DE_down, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )


goBP_unique_CD8_T_vs_NK_up_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(CD8_T_vs_NK_DE_up, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )


goBP_unique_CD8_T_vs_NK_down_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(CD8_T_vs_NK_DE_down, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )



goBP_unique_CD4_T_vs_NK_up_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(CD4_T_vs_NK_DE_up, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )


goBP_unique_CD4_T_vs_NK_down_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(CD4_T_vs_NK_DE_down, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )
#No genes; no enrichment

goBP_unique_Treg_vs_NK_up_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(Treg_vs_NK_DE_up, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )
#No genes; no enrichment

goBP_unique_Treg_vs_NK_down_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(Treg_vs_NK_DE_down, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )
#No genes; no enrichment

goBP_unique_gdT_vs_NK_up_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(gdT_vs_NK_DE_up, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )
#No genes; no enrichment

goBP_unique_gdT_vs_NK_down_DEG <- clusterProfiler::enrichGO(gene = unique(bitr(gdT_vs_NK_DE_down, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                                     OrgDb = "org.Mm.eg.db",
                                                     readable = TRUE,
                                                     keyType = "ENTREZID",
                                                     ont = "BP",
                                                     pvalueCutoff = 0.05,
                                                     pAdjustMethod = "BH"
                                                     )



#pdf(paste0(fig_path, "ClusterProfiler_GO_BP_NK_vs_TcellSubsets.pdf"), height = 9, width = 7)
barplot(goBP_unique_NK_vs_Tsubset_up_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\nupregulated in NK cells vs T cells\n")

barplot(goBP_unique_NK_vs_Tsubset_down_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\ndownregulated in NK cells vs T cells\n")

barplot(goBP_unique_CD8_T_vs_NK_up_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\nupregulated in CD8+ T cells vs NK and other T cell subsets \n")

barplot(goBP_unique_CD8_T_vs_NK_down_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\ndownregulated in CD8+ T cells vs NK and other T cell subsets\n")

barplot(goBP_unique_CD4_T_vs_NK_up_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\nupregulated in CD4+ T cells vs NK and other T cell subsets \n")

barplot(goBP_unique_gdT_vs_NK_down_DEG, showCategory = 20) + ggtitle("ClusterProfiler GO BP enrichment among genes uniquely\ndownregulated in gdT cells vs NK and other T cell subsets\n")

##dev.off()





#topGO enrichement
library(pcaExplorer)
library(topGO)
library(org.Mm.eg.db)
library(AnnotationDbi)
rm(x, y, p, q)
x <- pcaExplorer::topGOtable(DEgenes =  NK_vs_Tsubset_DE_up, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)

y <- pcaExplorer::topGOtable(DEgenes =  NK_vs_Tsubset_DE_down, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)

p <- pcaExplorer::topGOtable(DEgenes =  CD8_T_vs_NK_DE_up, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)

q <- pcaExplorer::topGOtable(DEgenes =  CD8_T_vs_NK_DE_down, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)


r <- pcaExplorer::topGOtable(DEgenes =  CD4_T_vs_NK_DE_up, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD4_TEM")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)

s <- pcaExplorer::topGOtable(DEgenes =  gdT_vs_NK_DE_down, BGgenes = differential_genes_all_integrated_rnm$gene[differential_genes_all_integrated_rnm$cell_type %in% c("CD27+_gdT", "CD27-_gdT")], ontology = "BP",mapping = "org.Mm.eg.db", geneID = "symbol", topTablerows = 500)


#pdf(paste0(fig_path, "topGO_GO_BP_NK_vs_TcellSubsets.pdf"), height = 9, width = 15)
ggplot(x[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\nupregulated in NK cells vs T cell subsets\n")

ggplot(y[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\ndownregulated in NK cells vs T cells subsets\n")

ggplot(p[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\nupregulated in CD8+ T cells vs NK and other T cell subsets\n")

ggplot(q[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\ndownregulated in CD8+ T cells vs NK and other T cell subsets\n")

ggplot(r[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\nupregulated in CD4+ T cells vs NK and other T cell subsets\n")

ggplot(s[1:20, ], aes(x = Significant, y = fct_reorder(Term, p.value_elim, .desc = TRUE))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.1), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + ggtitle("topGO GO BP enrichment among genes uniquely\ndownregulated in gdT cells vs NK and other T cell subsets\n")
##dev.off()
```


```{r}
z <- bind_rows(tbl_fil_all_RNA)

common_Seu_Musc_DE_genes <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")){
  diff_genes_Seu <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% i, ]
  diff_genes_Musc <- z[z$cluster_id %in% i, ]
  common_Seu_Musc <- merge(diff_genes_Seu, diff_genes_Musc, by.x = "gene", by.y = "gene")
  common_Seu_Musc_DE_genes[[i]] <- common_Seu_Musc
}

rm(common_Seu_Musc, diff_genes_Seu, diff_genes_Musc)
p <- bind_rows(tbl_fil_all_ADT)
common_Seu_Musc_DE_ADT <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")){
  diff_ADT_Seu <- differential_ADT_all_integrated_rnm[differential_ADT_all_integrated_rnm$cell_type %in% i, ]
  diff_ADT_Musc <- p[p$cluster_id %in% i, ]
  common_Seu_Musc <- merge(diff_ADT_Seu, diff_ADT_Musc, by.x = "gene", by.y = "gene")
  common_Seu_Musc_DE_ADT[[i]] <- common_Seu_Musc
}
```

#Average expression levels of co-stimulatory and co-inhibitory genes and ADT (DGE and DPE based on Seurat)
- Average non-log expression values of normalized counts
```{r}

#Source code: https://satijalab.org/seurat/articles/integration_introduction.html
tmp <- all_integrated_rnm_VlnPlot #assign seurat object to a temporary variable


coStim_ADT <- c( "adt-CD357", "adt-CD2", "adt-CD28", "adt-CD27", "adt-CD278", "adt-CD137", "adt-CD134", "adt-Ly108", "adt-CD226", "adt-CD270", "adt-CD365", "adt-Ly49D", "adt-Ly49H") #"adt-CD365" is added from Chen and Flies; "adt-Ly49D", "adt-Ly49H" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017

coInhib_ADT <- c("adt-CD279", "adt-CD366", "adt-TIGIT", "adt-CD152", "adt-CD160", "adt-CD272", "adt-CD85k", "adt-CD94", "adt-Ly49A", "adt-Ly49G")# "adt-Ly49A", "adt-Ly49G" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017


coStim_genes <- c( "Tnfrsf18", "Cd2", "Cd27", "Icos", "Tnfrsf9", "Tnfrsf4", "Slamf6", "Cd226", "Tnfrsf14", "Havcr1", "Klra4", "Klra8") #"Havcr1" == Tim1 is added from Chen and Flies; "Klra4" = "Ly49D" and "Klra8" = "Ly49H" added from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017


coInhib_genes <- c("Pdcd1", "Havcr2", "Lag3", "Tigit", "Ctla4", "Cd160", "Btla", "Lilrb4a", "Klrd1", "Klra1", "Klra7") #"Klra1" = "adt-Ly49A", "Klra7" = "adt-Ly49G" from Rahim MM, et al. Front Immunol. 2014; Paul S and Lal G. Front Immunol. 2017


cluster_Average_exp_RNA_costim_coinhib <- list()

#RNA
for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") ){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, features = c(coStim_genes, coInhib_genes), slot = "data", assays = "RNA")
  cluster.averages <- as.data.frame(cluster.averages$RNA)
  cluster.averages$Celltype <- i
  cluster.averages$gene <- row.names(cluster.averages)
  cluster_Average_exp_RNA_costim_coinhib[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_RNA_costim_coinhib, paste0(tab_path, "cluster_average_Exp_costim_coinhib_RNA_NK_Tcells_il27raKO_vs_wt.xlsx"))


#ADT
cluster_Average_exp_ADT_costim_coinhib <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") ){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "group"
  cluster.averages <- AverageExpression(tmp_clus, features = c(coStim_ADT, coInhib_ADT), slot = "data", assays = "ADT")
  cluster.averages <- as.data.frame(cluster.averages$ADT)
  cluster.averages$Celltype <- i
  cluster.averages$protein <- row.names(cluster.averages)
  cluster_Average_exp_ADT_costim_coinhib[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_ADT_costim_coinhib, paste0(tab_path, "cluster_average_Exp_costim_coinhib_ADT_NK_Tcells_il27raKO_vs_wt.xlsx"))

#By sample
cluster_Average_exp_RNA_costim_coinhib_bySampleID <- list()

#RNA
for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") ){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "sample_id"
  cluster.averages <- AverageExpression(tmp_clus, features = c(coStim_genes, coInhib_genes), slot = "data", assays = "RNA")
  cluster.averages <- as.data.frame(cluster.averages$RNA)
  cluster.averages$Celltype <- i
  cluster.averages$gene <- row.names(cluster.averages)
  cluster_Average_exp_RNA_costim_coinhib_bySampleID[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_RNA_costim_coinhib_bySampleID, paste0(tab_path, "cluster_average_Exp_bySampleID_costim_coinhib_RNA_NK_Tcells_il27raKO_vs_wt.xlsx"))


#ADT
cluster_Average_exp_ADT_costim_coinhib_bySampleID <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") ){
  tmp_clus <- subset(tmp, idents = i)
  Idents(tmp_clus) <- "sample_id"
  cluster.averages <- AverageExpression(tmp_clus, features = c(coStim_ADT, coInhib_ADT), slot = "data", assays = "ADT")
  cluster.averages <- as.data.frame(cluster.averages$ADT)
  cluster.averages$Celltype <- i
  cluster.averages$protein <- row.names(cluster.averages)
  cluster_Average_exp_ADT_costim_coinhib_bySampleID[[i]] <- cluster.averages
 
}

#writexl::write_xlsx(cluster_Average_exp_ADT_costim_coinhib_bySampleID, paste0(tab_path, "cluster_average_Exp_bySampleID_costim_coinhib_ADT_NK_Tcells_il27raKO_vs_wt.xlsx"))


```


```{r}

tmp <- all_integrated_rnm
Idents(tmp) <- "cluster0.8.group"

get_differential_genes_forGSEA <- function(celltype){
  Seurat::FindMarkers(object = tmp, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", logfc.threshold = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_genes_all_integrated_rnm_forGSEA <- map_dfr(tmp_celltype_list, get_differential_genes_forGSEA)


differential_genes_all_integrated_rnm_forGSEA$rank <- -log10(differential_genes_all_integrated_rnm_forGSEA$p_val) * sign(differential_genes_all_integrated_rnm_forGSEA$avg_log2FC) #https://github.com/RGLab/MAST/issues/147#issuecomment-996809717 ; setting pvalue instead of adjusted p value to avoid 0 ranking when adjusted p value = 1

gsea_Seurat_DE_genes <- list()

curated_sets_msigdbr <- msigdbr(species = "Mus musculus", category = c("H") ) %>% dplyr::select(gs_name, gene_symbol) #Source code: https://yulab-smu.top/biomedical-knowledge-mining-book/universal-api.html?q=hallmark#msigdb-analysis


for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell") ){
  diff_for_gsea <- differential_genes_all_integrated_rnm_forGSEA[differential_genes_all_integrated_rnm_forGSEA$cell_type %in% i, ]
  diff_for_gsea_subset <- diff_for_gsea$rank
  names(diff_for_gsea_subset) <- diff_for_gsea$gene
  diff_for_gsea_subset <- na.omit(sort(diff_for_gsea_subset, decreasing = TRUE))
  
  gsea_DE <- clusterProfiler::GSEA(diff_for_gsea_subset, TERM2GENE = curated_sets_msigdbr, by = "fgsea", pAdjustMethod = "BH")
  gsea_Seurat_DE_genes[[i]] <- gsea_DE
  
}

gsea_Seurat_DE_genes_table <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell") ){
  x <- gsea_Seurat_DE_genes[[i]]@result
  
  if(nrow(x) != 0) {
    x$celltype <- i
    gsea_Seurat_DE_genes_table[[i]] <- x
  
  }
}

#writexl::write_xlsx(gsea_Seurat_DE_genes_table, paste0(tab_path, "GSEA_all_integrated_rnm_res0.8.xlsx"))
#writexl::write_xlsx(bind_rows(gsea_Seurat_DE_genes_table), paste0(tab_path, "tmp_gsea_xl.xlsx"))

#saveRDS(gsea_Seurat_DE_genes, file = paste0(tab_path, "gsea_Seurat_DE_genes.rds"))
#gsea_Seurat_DE_genes <- readRDS(paste0(tab_path, "gsea_Seurat_DE_genes.rds"))

gsea_plot_combined <- list()





for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell") ){
  gsea_plot <- gsea_Seurat_DE_genes[[i]]
  
  if(nrow(gsea_plot@result) != 0) {
    
    for(j in 1:nrow(gsea_plot@result)){
      
      plot_gsea <- gseaplot(gsea_plot, geneSetID = j, title = paste0(gsea_plot@result$Description[j]) ) & theme(plot.title = element_text(size = 10))
      gsea_plot_combined[[i]][[j]] <- plot_gsea
    }
  }

}

#pdf(paste0(fig_path, "/GSEA_plots_byCellType_March2023.pdf"), width = 15, height = 25)
for(i in 1:length(gsea_plot_combined)){
  grid.newpage()
  grid.text(paste0(names(gsea_plot_combined[i]), "\nIl27raKO vs WT GSEA by cell type") )
    
  print(wrap_plots(gsea_plot_combined[[i]], ncol = 4, nrow = 5))
}
##dev.off()


```


#Upd 22Mar2023: Based on discussion and inputs from Prof. Bosmann
- 1. Plotting heatmap of genes enriched in GSEA "HALLMARK_INTERFERON_GAMMA_RESPONSE"
```{r}
#extract genes enriched in "HALLMARK_INTERFERON_GAMMA_RESPONSE" across different cell types from gsea_Seurat_DE_genes_table

x <- bind_rows(gsea_Seurat_DE_genes_table)
x <- x[x$celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]
gseaGenes_Ifn_gamma_response <- unique(unlist(strsplit(x$core_enrichment[x$Description %in% c("HALLMARK_INTERFERON_GAMMA_RESPONSE")], split='/', fixed=TRUE) ))

tmp <- all_integrated_rnm_VlnPlot
tmp_subset_NK_T <- subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) 

Idents(tmp_subset_NK_T) <- "cluster0.8.group"

Idents(tmp_subset_NK_T) <-  factor(Idents(tmp_subset_NK_T), levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO", 
                                                           "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                           "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                           "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                           "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO", 
                                                           "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                           "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                           "CD8_TEM_wt", "CD8_TEM_il27raKO", 
                                                           "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                           "Treg_wt", "Treg_il27raKO",
                                                           "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                           "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )



cluster.averages.gseaGenes_Ifn_gamma_response <- AverageExpression(tmp_subset_NK_T, features = gseaGenes_Ifn_gamma_response, slot = "data", assays = "RNA", return.seurat = TRUE)

cluster.averages.gseaGenes_Ifn_gamma_response <- ScaleData(cluster.averages.gseaGenes_Ifn_gamma_response)

cluster.averages.gseaGenes_Ifn_gamma_response$orig.ident <- row.names(cluster.averages.gseaGenes_Ifn_gamma_response@meta.data)

htmap_gseaGenes_Ifn_gamma_response <- Seurat::DoHeatmap(cluster.averages.gseaGenes_Ifn_gamma_response, features = gseaGenes_Ifn_gamma_response, size = 3, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF"), raster = FALSE) + NoLegend() + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 7, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10))
 
#pdf(paste0(fig_path, "/Heatmap_AvgExpr_cluster.averages.gseaGenes_Ifn_gamma_response.pdf"), width = 5, height = 15)
htmap_gseaGenes_Ifn_gamma_response + ggtitle("Heatmap of average gene expression of genes enriched in GSEA\n 'HALLMARK_INTERFERON_GAMMA_RESPONSE'\nDE genes may/may not be statistically significant\n ")
##dev.off()

cluster.averages.gseaGenes_Ifn_gamma_response$orig.ident <- factor(cluster.averages.gseaGenes_Ifn_gamma_response$orig.ident, 
                                                                   levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO",
                                                                              "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                                              "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                                              "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                                              "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO",
                                                                              "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                                              "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                                              "CD8_TEM_wt", "CD8_TEM_il27raKO",
                                                                              "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                                              "Treg_wt", "Treg_il27raKO",
                                                                              "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                                              "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )


#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_cluster.averages.gseaGenes_Ifn_gamma_response.pdf"), width = 10, height = 15)
dittoHeatmap(cluster.averages.gseaGenes_Ifn_gamma_response, genes = gseaGenes_Ifn_gamma_response, annot.by =  "orig.ident")
##dev.off()
```

- 2. Plotting heatmap of genes enriched in GSEA "HALLMARK_TNFA_SIGNALING_VIA_NFKB"
```{r}
#extract genes enriched in "HALLMARK_TNFA_SIGNALING_VIA_NFKB" across different cell types from gsea_Seurat_DE_genes_table

x <- bind_rows(gsea_Seurat_DE_genes_table)
x <- x[x$celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]
gseaGenes_TNF_signaling <- unique(unlist(strsplit(x$core_enrichment[x$Description %in% c("HALLMARK_TNFA_SIGNALING_VIA_NFKB")], split='/', fixed=TRUE) ))

tmp <- all_integrated_rnm_VlnPlot
tmp_subset_NK_T <- subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) 

Idents(tmp_subset_NK_T) <- "cluster0.8.group"

Idents(tmp_subset_NK_T) <-  factor(Idents(tmp_subset_NK_T), levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO", 
                                                           "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                           "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                           "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                           "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO", 
                                                           "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                           "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                           "CD8_TEM_wt", "CD8_TEM_il27raKO", 
                                                           "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                           "Treg_wt", "Treg_il27raKO",
                                                           "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                           "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )



cluster.averages.gseaGenes_TNF_signaling <- AverageExpression(tmp_subset_NK_T, features = gseaGenes_TNF_signaling, slot = "data", assays = "RNA", return.seurat = TRUE)

cluster.averages.gseaGenes_TNF_signaling <- ScaleData(cluster.averages.gseaGenes_TNF_signaling)

cluster.averages.gseaGenes_TNF_signaling$orig.ident <- row.names(cluster.averages.gseaGenes_TNF_signaling@meta.data)

htmap_gseaGenes_TNF_signaling <- Seurat::DoHeatmap(cluster.averages.gseaGenes_TNF_signaling, features = gseaGenes_TNF_signaling, size = 3, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF"), raster = FALSE) + NoLegend() + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 7, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10))
 
#pdf(paste0(fig_path, "/Heatmap_AvgExpr_cluster.averages.gseaGenes_TNF_signaling.pdf"), width = 5, height = 15)
htmap_gseaGenes_TNF_signaling + ggtitle("Heatmap of average gene expression of genes enriched in GSEA\n 'HALLMARK_TNFA_SIGNALING_VIA_NFKB'\nDE genes may/may not be statistically significant\n ")
##dev.off()

cluster.averages.gseaGenes_TNF_signaling$orig.ident <- factor(cluster.averages.gseaGenes_TNF_signaling$orig.ident, 
                                                                   levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO",
                                                                              "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                                              "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                                              "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                                              "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO",
                                                                              "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                                              "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                                              "CD8_TEM_wt", "CD8_TEM_il27raKO",
                                                                              "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                                              "Treg_wt", "Treg_il27raKO",
                                                                              "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                                              "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )


#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_cluster.averages.gseaGenes_TNF_signaling.pdf"), width = 10, height = 15)
dittoHeatmap(cluster.averages.gseaGenes_TNF_signaling, genes = gseaGenes_TNF_signaling, annot.by =  "orig.ident")
##dev.off()
```


- 3. Plotting heatmap of genes enriched in GSEA "HALLMARK_IL2_STAT5_SIGNALING"
```{r}
#extract genes enriched in "HALLMARK_IL2_STAT5_SIGNALING" across different cell types from gsea_Seurat_DE_genes_table

x <- bind_rows(gsea_Seurat_DE_genes_table)
x <- x[x$celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]
gseaGenes_IL2_STAT5_signaling <- unique(unlist(strsplit(x$core_enrichment[x$Description %in% c("HALLMARK_IL2_STAT5_SIGNALING")], split='/', fixed=TRUE) ))

tmp <- all_integrated_rnm_VlnPlot
tmp_subset_NK_T <- subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) 

Idents(tmp_subset_NK_T) <- "cluster0.8.group"

Idents(tmp_subset_NK_T) <-  factor(Idents(tmp_subset_NK_T), levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO", 
                                                           "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                           "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                           "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                           "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO", 
                                                           "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                           "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                           "CD8_TEM_wt", "CD8_TEM_il27raKO", 
                                                           "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                           "Treg_wt", "Treg_il27raKO",
                                                           "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                           "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )



cluster.averages.gseaGenes_IL2_STAT5_signaling <- AverageExpression(tmp_subset_NK_T, features = gseaGenes_IL2_STAT5_signaling, slot = "data", assays = "RNA", return.seurat = TRUE)

cluster.averages.gseaGenes_IL2_STAT5_signaling <- ScaleData(cluster.averages.gseaGenes_IL2_STAT5_signaling)

cluster.averages.gseaGenes_IL2_STAT5_signaling$orig.ident <- row.names(cluster.averages.gseaGenes_IL2_STAT5_signaling@meta.data)

htmap_gseaGenes_IL2_STAT5_signaling <- Seurat::DoHeatmap(cluster.averages.gseaGenes_IL2_STAT5_signaling, features = gseaGenes_IL2_STAT5_signaling, size = 3, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF"), raster = FALSE) + NoLegend() + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 7, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10))
 
#pdf(paste0(fig_path, "/Heatmap_AvgExpr_cluster.averages.gseaGenes_IL2_STAT5_signaling.pdf"), width = 5, height = 6)
htmap_gseaGenes_IL2_STAT5_signaling + ggtitle("Heatmap of average gene expression of genes enriched in GSEA\n 'HALLMARK_IL2_STAT5_SIGNALING'\nDE genes may/may not be statistically significant\n ")
##dev.off()

cluster.averages.gseaGenes_IL2_STAT5_signaling$orig.ident <- factor(cluster.averages.gseaGenes_IL2_STAT5_signaling$orig.ident, 
                                                                   levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO",
                                                                              "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                                              "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                                              "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                                              "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO",
                                                                              "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                                              "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                                              "CD8_TEM_wt", "CD8_TEM_il27raKO",
                                                                              "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                                              "Treg_wt", "Treg_il27raKO",
                                                                              "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                                              "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )


#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_cluster.averages.gseaGenes_IL2_STAT5_signaling.pdf"), width = 10, height = 6)
dittoHeatmap(cluster.averages.gseaGenes_IL2_STAT5_signaling, genes = gseaGenes_IL2_STAT5_signaling, annot.by =  "orig.ident")
##dev.off()
```


- 4. Plotting heatmap of genes enriched in GSEA "HALLMARK_IL6_JAK_STAT3_SIGNALING"
```{r}
#extract genes enriched in "HALLMARK_IL6_JAK_STAT3_SIGNALING" across different cell types from gsea_Seurat_DE_genes_table

x <- bind_rows(gsea_Seurat_DE_genes_table)
x <- x[x$celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]
gseaGenes_Il6_Jak_Stat3_signaling <- unique(unlist(strsplit(x$core_enrichment[x$Description %in% c("HALLMARK_IL6_JAK_STAT3_SIGNALING")], split='/', fixed=TRUE) ))

tmp <- all_integrated_rnm_VlnPlot
tmp_subset_NK_T <- subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) 

Idents(tmp_subset_NK_T) <- "cluster0.8.group"

Idents(tmp_subset_NK_T) <-  factor(Idents(tmp_subset_NK_T), levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO", 
                                                           "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                           "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                           "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                           "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO", 
                                                           "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                           "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                           "CD8_TEM_wt", "CD8_TEM_il27raKO", 
                                                           "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                           "Treg_wt", "Treg_il27raKO",
                                                           "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                           "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )



cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling <- AverageExpression(tmp_subset_NK_T, features = gseaGenes_Il6_Jak_Stat3_signaling, slot = "data", assays = "RNA", return.seurat = TRUE)

cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling <- ScaleData(cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling)

cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling$orig.ident <- row.names(cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling@meta.data)

htmap_gseaGenes_Il6_Jak_Stat3_signaling <- Seurat::DoHeatmap(cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling, features = gseaGenes_Il6_Jak_Stat3_signaling, size = 3, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF"), raster = FALSE) + NoLegend() + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 7, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10))
 
#pdf(paste0(fig_path, "/Heatmap_AvgExpr_cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling.pdf"), width = 5, height = 3)
htmap_gseaGenes_Il6_Jak_Stat3_signaling + ggtitle("Heatmap of average gene expression of genes enriched in GSEA\n 'HALLMARK_IL6_JAK_STAT3_SIGNALING'\nDE genes may/may not be statistically significant\n ")
##dev.off()

cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling$orig.ident <- factor(cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling$orig.ident, 
                                                                   levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO",
                                                                              "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                                              "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                                              "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                                              "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO",
                                                                              "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                                              "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                                              "CD8_TEM_wt", "CD8_TEM_il27raKO",
                                                                              "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                                              "Treg_wt", "Treg_il27raKO",
                                                                              "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                                              "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )


#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling.pdf"), width = 10, height = 3)
dittoHeatmap(cluster.averages.gseaGenes_Il6_Jak_Stat3_signaling, genes = gseaGenes_Il6_Jak_Stat3_signaling, annot.by =  "orig.ident")
##dev.off()
```

- 5. Plotting heatmap of genes enriched in GSEA "HALLMARK_MTORC1_SIGNALING"
```{r}
#extract genes enriched in "HALLMARK_MTORC1_SIGNALING" across different cell types from gsea_Seurat_DE_genes_table

x <- bind_rows(gsea_Seurat_DE_genes_table)
x <- x[x$celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]
gseaGenes_mtorc1_signaling <- unique(unlist(strsplit(x$core_enrichment[x$Description %in% c("HALLMARK_MTORC1_SIGNALING")], split='/', fixed=TRUE) ))

tmp <- all_integrated_rnm_VlnPlot
tmp_subset_NK_T <- subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) 

Idents(tmp_subset_NK_T) <- "cluster0.8.group"

Idents(tmp_subset_NK_T) <-  factor(Idents(tmp_subset_NK_T), levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO", 
                                                           "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                           "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                           "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                           "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO", 
                                                           "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                           "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                           "CD8_TEM_wt", "CD8_TEM_il27raKO", 
                                                           "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                           "Treg_wt", "Treg_il27raKO",
                                                           "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                           "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )



cluster.averages.gseaGenes_mtorc1_signaling <- AverageExpression(tmp_subset_NK_T, features = gseaGenes_mtorc1_signaling, slot = "data", assays = "RNA", return.seurat = TRUE)

cluster.averages.gseaGenes_mtorc1_signaling <- ScaleData(cluster.averages.gseaGenes_mtorc1_signaling)

cluster.averages.gseaGenes_mtorc1_signaling$orig.ident <- row.names(cluster.averages.gseaGenes_mtorc1_signaling@meta.data)

htmap_gseaGenes_mtorc1_signaling <- Seurat::DoHeatmap(cluster.averages.gseaGenes_mtorc1_signaling, features = gseaGenes_mtorc1_signaling, size = 3, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF"), raster = FALSE) + NoLegend() + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 7, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10))
 
#pdf(paste0(fig_path, "/Heatmap_AvgExpr_cluster.averages.gseaGenes_mtorc1_signaling.pdf"), width = 5, height = 15)
htmap_gseaGenes_mtorc1_signaling + ggtitle("Heatmap of average gene expression of genes enriched in GSEA\n 'HALLMARK_MTORC1_SIGNALING'\nDE genes may/may not be statistically significant\n ")
##dev.off()

cluster.averages.gseaGenes_mtorc1_signaling$orig.ident <- factor(cluster.averages.gseaGenes_mtorc1_signaling$orig.ident, 
                                                                   levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO",
                                                                              "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                                              "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                                              "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                                              "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO",
                                                                              "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                                              "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                                              "CD8_TEM_wt", "CD8_TEM_il27raKO",
                                                                              "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                                              "Treg_wt", "Treg_il27raKO",
                                                                              "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                                              "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )


#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_cluster.averages.gseaGenes_mtorc1_signaling.pdf"), width = 10, height = 15)
dittoHeatmap(cluster.averages.gseaGenes_mtorc1_signaling, genes = gseaGenes_mtorc1_signaling, annot.by =  "orig.ident")
##dev.off()
```


- 6. Plotting heatmap of genes enriched in GSEA "HALLMARK_MYC_TARGETS_V1"
```{r}
#extract genes enriched in "HALLMARK_MYC_TARGETS_V1" across different cell types from gsea_Seurat_DE_genes_table

x <- bind_rows(gsea_Seurat_DE_genes_table)
x <- x[x$celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]
gseaGenes_myc_targets_v1 <- unique(unlist(strsplit(x$core_enrichment[x$Description %in% c("HALLMARK_MYC_TARGETS_V1")], split='/', fixed=TRUE) ))

tmp <- all_integrated_rnm_VlnPlot
tmp_subset_NK_T <- subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) 

Idents(tmp_subset_NK_T) <- "cluster0.8.group"

Idents(tmp_subset_NK_T) <-  factor(Idents(tmp_subset_NK_T), levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO", 
                                                           "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                           "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                           "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                           "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO", 
                                                           "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                           "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                           "CD8_TEM_wt", "CD8_TEM_il27raKO", 
                                                           "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                           "Treg_wt", "Treg_il27raKO",
                                                           "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                           "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )



cluster.averages.gseaGenes_myc_targets_v1 <- AverageExpression(tmp_subset_NK_T, features = gseaGenes_myc_targets_v1, slot = "data", assays = "RNA", return.seurat = TRUE)

cluster.averages.gseaGenes_myc_targets_v1 <- ScaleData(cluster.averages.gseaGenes_myc_targets_v1)

cluster.averages.gseaGenes_myc_targets_v1$orig.ident <- row.names(cluster.averages.gseaGenes_myc_targets_v1@meta.data)

htmap_gseaGenes_myc_targets_v1 <- Seurat::DoHeatmap(cluster.averages.gseaGenes_myc_targets_v1, features = gseaGenes_myc_targets_v1, size = 3, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF"), raster = FALSE) + NoLegend() + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 7, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10))
 
#pdf(paste0(fig_path, "/Heatmap_AvgExpr_cluster.averages.gseaGenes_myc_targets_v1.pdf"), width = 5, height = 15)
htmap_gseaGenes_myc_targets_v1 + ggtitle("Heatmap of average gene expression of genes enriched in GSEA\n 'HALLMARK_MYC_TARGETS_V1'\nDE genes may/may not be statistically significant\n ")
##dev.off()

cluster.averages.gseaGenes_myc_targets_v1$orig.ident <- factor(cluster.averages.gseaGenes_myc_targets_v1$orig.ident, 
                                                                   levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO",
                                                                              "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                                              "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                                              "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                                              "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO",
                                                                              "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                                              "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                                              "CD8_TEM_wt", "CD8_TEM_il27raKO",
                                                                              "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                                              "Treg_wt", "Treg_il27raKO",
                                                                              "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                                              "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )


#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_cluster.averages.gseaGenes_myc_targets_v1.pdf"), width = 10, height = 15)
dittoHeatmap(cluster.averages.gseaGenes_myc_targets_v1, genes = gseaGenes_myc_targets_v1, annot.by =  "orig.ident")
##dev.off()
```

#Heatmap of statistically significant DE from each NK and T cell subtype 
```{r}
#extract genes that are significantly DE in each cell type

signifGenes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & differential_genes_all_integrated_rnm$p_val_adj < 0.05, ] %>% dplyr::select(gene, cell_type)

tmp <- all_integrated_rnm_VlnPlot
tmp_subset_NK_T <- subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) 

Idents(tmp_subset_NK_T) <- "cluster0.8.group"

Idents(tmp_subset_NK_T) <-  factor(Idents(tmp_subset_NK_T), levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO", 
                                                           "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                           "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                           "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                           "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO", 
                                                           "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                           "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                           "CD8_TEM_wt", "CD8_TEM_il27raKO", 
                                                           "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                           "Treg_wt", "Treg_il27raKO",
                                                           "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                           "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )



cluster.averages.signifGenes <- AverageExpression(tmp_subset_NK_T, features = signifGenes, slot = "data", assays = "RNA", return.seurat = TRUE)

cluster.averages.signifGenes <- ScaleData(cluster.averages.signifGenes)

cluster.averages.signifGenes$orig.ident <- row.names(cluster.averages.signifGenes@meta.data)


cluster.averages.signifGenes$orig.ident <- factor(cluster.averages.signifGenes$orig.ident, 
                                                                   levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO",
                                                                              "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                                              "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                                              "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                                              "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO",
                                                                              "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                                              "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                                              "CD8_TEM_wt", "CD8_TEM_il27raKO",
                                                                              "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                                              "Treg_wt", "Treg_il27raKO",
                                                                              "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                                              "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )


htmap_signifGenes_seu <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")){
  
  if(i %in% signifGenes$cell_type){
    signifGenes_subset <- signifGenes$gene[signifGenes$cell_type %in% i]
    
    htmap_signifGenes_seu[[i]] <- Seurat::DoHeatmap(cluster.averages.signifGenes, features = signifGenes_subset, size = 3, draw.lines = FALSE, angle = 90, group.colors = c("#05BE78", "#0000FF"), raster = FALSE) + NoLegend() + scale_fill_gradientn(colours = c("navy", "white", "red")) +  theme(axis.text.y.left = element_text(size = 7, colour = "black", margin = margin(r=-15)), plot.title = element_text(size = 10))+ ggtitle(paste0("Heatmap of average gene expression of significant DE genes (adjusted P < 0.05)\n ", "in\n", i) )
   }
}

#pdf(paste0(fig_path, "/Heatmap_AvgExpr_cluster.averages.signifGenes.pdf"), width = 5, height = 8)
htmap_signifGenes_seu
##dev.off()

#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_cluster.averages.signifGenes.pdf"), width = 10, height = 18)
for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")){
  
  if(i %in% signifGenes$cell_type){
    signifGenes_subset <- signifGenes$gene[signifGenes$cell_type %in% i]
    
    x <- dittoHeatmap(cluster.averages.signifGenes, genes = signifGenes_subset, annot.by =  "orig.ident", main = paste0("Heatmap of average gene expression of significant\nDE genes (adjusted P < 0.05)\n ", "in\n", i), complex = TRUE) #complex = TRUE draws heatmap using ComplexHeatmap
    x@matrix_param$height <- unit(3, "mm")*nrow(x@matrix) #adjusts height and width of heatmap
    x@matrix_param$width <- unit(5, "mm")*ncol(x@matrix)
    
    print(x)
}
}
##dev.off()

rm(x, signifGenes_subset)

#DE genes that are not statistically significant differentially expressed genes in other clusters
signifGenes_unique <- signifGenes[!(duplicated(signifGenes$gene  ) | duplicated(signifGenes$gene, fromLast = TRUE)), ] #Source code for excluding all duplicated rows: https://stackoverflow.com/questions/13763216/how-can-i-remove-all-duplicates-so-that-none-are-left-in-a-data-frame

#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_cluster.averages.signifGenes_unique_all_NK_T.pdf"), width = 10, height = 18)
for(i in unique(signifGenes_unique$cell_type)){
  
  signifGenes_subset <- signifGenes_unique$gene[signifGenes_unique$cell_type %in% i]
  
  if(! ( i ==  "CD27loCD11b+_NK" ) ) {
  
  x <- dittoHeatmap(cluster.averages.signifGenes, genes = signifGenes_subset, annot.by =  "orig.ident", main = paste0("Heatmap of average gene expression of significant\nDE genes (adjusted P < 0.05)\n ", "specific to\n", i), complex = TRUE) #complex = TRUE draws heatmap using ComplexHeatmap
  
  x@matrix_param$height <- unit(3, "mm")*nrow(x@matrix) #adjusts height and width of heatmap
  x@matrix_param$width <- unit(5, "mm")*ncol(x@matrix)
  
  print(x)
  } else {
    x <- dittoHeatmap(cluster.averages.signifGenes, genes = c(signifGenes_subset, "Wdfy1"), annot.by =  "orig.ident", main = paste0("Heatmap of average gene expression of significant\nDE genes (adjusted P < 0.05)\n ", "specific to\n", i, "\nWdfy1 gene is not unique, and has been\n added to the plot as a single gene\nheatmap cannot be plotted\n"), complex = TRUE) #complex = TRUE draws heatmap using ComplexHeatmap; ADDing Wdfy1 to DE list for plotting for "CD27loCD11b+_NK", as it only has 1 unique DE gene which throws an error while plotting;  Wdfy1 is present in DE in this NK cell subtype and other cell subsets, so selecting this gene for plotting
  
  x@matrix_param$height <- unit(3, "mm")*nrow(x@matrix) #adjusts height and width of heatmap
  x@matrix_param$width <- unit(5, "mm")*ncol(x@matrix)
  }
}

##dev.off()






```


# Heatmap of genes from venn section between overall NK and T cell DE genes 
```{r}
#RNA
# Combine all NK clusters into one and different T cell clusters into one

NK_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

CD8_Tcell_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM"), ]

CD4_Tcell_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD4_TEM"), ]

Treg_cell_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("Treg"), ]

gdT_cell_genes <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27+_gdT", "CD27-_gdT"), ]

vennPlot_DGE_NK_vs_Tsubset_upreg <- venndetail(list(
  NK_up = NK_genes$gene[NK_genes$avg_log2FC > 0 & NK_genes$p_val_adj < 0.05], #subset to upregulated genes (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
  CD8_T_up = CD8_Tcell_genes$gene[CD8_Tcell_genes$avg_log2FC > 0 & CD8_Tcell_genes$p_val_adj < 0.05],
  CD4_T_up = CD4_Tcell_genes$gene[CD4_Tcell_genes$avg_log2FC > 0 & CD4_Tcell_genes$p_val_adj < 0.05],
  Treg_up = Treg_cell_genes$gene[Treg_cell_genes$avg_log2FC > 0 & Treg_cell_genes$p_val_adj < 0.05],
  gdT_up = gdT_cell_genes$gene[gdT_cell_genes$avg_log2FC > 0 & gdT_cell_genes$p_val_adj < 0.05]
))


vennPlot_DGE_NK_vs_Tsubset_downreg <- venndetail(list(
  NK_down = NK_genes$gene[NK_genes$avg_log2FC < 0 & NK_genes$p_val_adj < 0.05], #subset to downregulated genes (avg_log2FC < 0) that are statistically significant (p_val_adj < 0.05)
  CD8_T_down = CD8_Tcell_genes$gene[CD8_Tcell_genes$avg_log2FC < 0 & CD8_Tcell_genes$p_val_adj < 0.05],
  CD4_T_down = CD4_Tcell_genes$gene[CD4_Tcell_genes$avg_log2FC < 0 & CD4_Tcell_genes$p_val_adj < 0.05],
  Treg_down = Treg_cell_genes$gene[Treg_cell_genes$avg_log2FC < 0 & Treg_cell_genes$p_val_adj < 0.05],
  gdT_down = gdT_cell_genes$gene[gdT_cell_genes$avg_log2FC < 0 & gdT_cell_genes$p_val_adj < 0.05]

  ))



vennPlot_list_subset <- list(vennPlot_DGE_NK_vs_T_upreg = result(vennPlot_DGE_NK_vs_Tsubset_upreg, wide = TRUE), vennPlot_DGE_NK_vs_T_downreg = result(vennPlot_DGE_NK_vs_Tsubset_downreg, wide = TRUE))


#Unique and commonly DE in NK and T cells
#Uniquely DE in NK cells
NK_vs_Tsubset_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1 & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1 )& !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1 )  ]

NK_vs_Tsubset_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1 & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1 )]


#Uniquely DE in CD8 T cells
CD8_T_vs_NK_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[! (vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1) & vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1 & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1 )& !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1 )  ]

CD8_T_vs_NK_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[!(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1) & vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1 & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1 )]


CD4_T_vs_NK_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[! (vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1) & vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1 & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1 )& !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1 )  ]

CD4_T_vs_NK_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[!(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1) & vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1 )]


Treg_vs_NK_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[! (vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1) & vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1 & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1 )  ]

Treg_vs_NK_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[!(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1 & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1 )]


gdT_vs_NK_DE_up <- vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Detail[! (vennPlot_DGE_NK_vs_Tsubset_upreg@wide$NK_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD8_T_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$CD4_T_up == 1) & !(vennPlot_DGE_NK_vs_Tsubset_upreg@wide$Treg_up == 1) & vennPlot_DGE_NK_vs_Tsubset_upreg@wide$gdT_up == 1  ]

gdT_vs_NK_DE_down <- vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Detail[!(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$NK_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD8_T_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$CD4_T_down == 1 ) & !(vennPlot_DGE_NK_vs_Tsubset_downreg@wide$Treg_down == 1) & vennPlot_DGE_NK_vs_Tsubset_downreg@wide$gdT_down == 1]


```

## NK cell specific DE genes
```{r}
#NK cell specific DE genes
NK_specific <- c(NK_vs_Tsubset_DE_up, NK_vs_Tsubset_DE_down)

tmp <- all_integrated_rnm_VlnPlot
tmp_subset_NK_T <- subset(tmp, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) 

Idents(tmp_subset_NK_T) <- "cluster0.8.group"

Idents(tmp_subset_NK_T) <-  factor(Idents(tmp_subset_NK_T), levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO", 
                                                           "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                           "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                           "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                           "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO", 
                                                           "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                           "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                           "CD8_TEM_wt", "CD8_TEM_il27raKO", 
                                                           "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                           "Treg_wt", "Treg_il27raKO",
                                                           "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                           "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )



cluster.averages.NK_specific <- AverageExpression(tmp_subset_NK_T, features = NK_specific, slot = "data", assays = "RNA", return.seurat = TRUE)

cluster.averages.NK_specific <- ScaleData(cluster.averages.NK_specific)

cluster.averages.NK_specific$orig.ident <- row.names(cluster.averages.NK_specific@meta.data)


cluster.averages.NK_specific$orig.ident <- factor(cluster.averages.NK_specific$orig.ident, 
                                                                   levels = c("CD27loCD11b+_NK_wt", "CD27loCD11b+_NK_il27raKO",
                                                                              "CD27loCD11bhi_NK_wt", "CD27loCD11bhi_NK_il27raKO",
                                                                              "CD27hiCD11b+_NK_wt", "CD27hiCD11b+_NK_il27raKO",
                                                                              "CD27loCD11b+_mito_NK_wt", "CD27loCD11b+_mito_NK_il27raKO",
                                                                              "CD27loCD11blo_NK_wt", "CD27loCD11blo_NK_il27raKO",
                                                                              "CD8_Nave_wt", "CD8_Nave_il27raKO",
                                                                              "CD8_TCM_wt", "CD8_TCM_il27raKO",
                                                                              "CD8_TEM_wt", "CD8_TEM_il27raKO",
                                                                              "CD4_TEM_wt","CD4_TEM_il27raKO",
                                                                              "Treg_wt", "Treg_il27raKO",
                                                                              "CD27+_gdT_wt", "CD27+_gdT_il27raKO",
                                                                              "CD27-_gdT_wt", "CD27-_gdT_il27raKO") )


x <- dittoHeatmap(cluster.averages.NK_specific, genes = NK_specific, annot.by =  "orig.ident", main = paste0("Heatmap of average gene expression of significant\nDE genes (adjusted P < 0.05)\n ", "specific to\n", "NK cells\n"), complex = TRUE) #complex = TRUE draws heatmap using ComplexHeatmap
  
x@matrix_param$height <- unit(3, "mm")*nrow(x@matrix) #adjusts height and width of heatmap
x@matrix_param$width <- unit(5, "mm")*ncol(x@matrix)

#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_cluster.averages.NK_specific.pdf"), width = 10, height = 6)
  x
##dev.off()
  
VlnPlot(all_integrated_rnm_VlnPlot, features = NK_specific, assay = "RNA", split.by = "group", stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))  
  

rm(x)
```



```{r}
VlnPlot(all_integrated_rnm_VlnPlot, features = c("Ifit1bl1", "Smchd1", "Il6st", "Arl4c", "Il7r", "Ms4a4c", "Samhd1", "Dtx3l", "Gbp4", "Eea1","Lag3", "Gzmk", "Tgfbr2", "Stom", "Xaf1", "Fcer1g", "Ifit1", "Tspan13", "Slfn1", "Serpinb6b", "Wdr83os", "Ifi208", "Igtp", "Gbp2"), idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), split.by = "group", assay = "RNA", stack = TRUE, same.y.lims = TRUE)
```



```{r}
#Source code: https://satijalab.org/seurat/articles/parsebio_sketch_integration
tmp <- all_integrated_rnm
pseudoBulk_RNA <- AggregateExpression(tmp, return.seurat = TRUE, slot = "counts", assays = "RNA", group.by = c("celltype", "sample_id", "group")  )

pseudoBulk_RNA$celltype <- ifelse(!(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) %in% "BM") &  !(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) %in% "mito"), paste0(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 1), "_", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) ), ifelse(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) %in% "mito", paste0(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 1), "_", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2), "_", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 3) ), sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 1) ) )
                                  
pseudoBulk_RNA$sample_id <- ifelse(!(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) %in% "BM") &  !(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) %in% "mito"), paste0(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 3), "_", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 4), "_", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 5)) , ifelse(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) %in% "mito", paste0(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 4), "_", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 5), "_", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 6) ), paste0(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2), "_", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 3), "_", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 4)) ) )

pseudoBulk_RNA$group <- ifelse(!(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) %in% "BM") &  !(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) %in% "mito"), sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 6), ifelse(sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 2) %in% "mito", sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 7), sapply(strsplit(Cells(pseudoBulk_RNA), split = "_"), "[", 5)) )

pseudoBulk_RNA$group <- factor(x = pseudoBulk_RNA$group, levels = c("wt", "il27raKO"))
#pseudoBulk_RNA@meta.data$orig.ident <- row.names(pseudoBulk_RNA)

pseudoBulk_DE_RNA <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell") ){
  pseudoBulk_cell_subset <- subset(pseudoBulk_RNA, celltype == i)
  Idents(pseudoBulk_cell_subset) <- "group"
  RNA_DE <- FindMarkers(pseudoBulk_cell_subset, ident.1 = "il27raKO", ident.2 = "wt", slot = "counts", test.use = "DESeq2") %>%
    rownames_to_column(var = "gene") %>% cbind(cell_type = i, .)
  pseudoBulk_DE_RNA[[i]] <- RNA_DE
  
  
}

tbl_pseudoBulk_DE_RNA <- bind_rows(pseudoBulk_DE_RNA)

VlnPlot(all_integrated_rnm_VlnPlot, features = c("Gzmk", "Serpina3g", "Prnp", "Ifitm3", "Il7r", "Samhd1", "Gbp6", "Rbl2", "Il18r1", "Iigp1", "Podnl1", "Txnip", "Adgre5", "Zfas1"), assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))


#ADT
tmp <- all_integrated_rnm
pseudoBulk_ADT <- AggregateExpression(tmp, return.seurat = TRUE, slot = "counts", assays = "ADT", group.by = c("celltype", "sample_id", "group")  )

pseudoBulk_ADT$celltype <- ifelse(!(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) %in% "BM") &  !(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) %in% "mito"), paste0(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 1), "_", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) ), ifelse(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) %in% "mito", paste0(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 1), "_", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2), "_", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 3) ), sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 1) ) )
                                  
pseudoBulk_ADT$sample_id <- ifelse(!(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) %in% "BM") &  !(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) %in% "mito"), paste0(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 3), "_", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 4), "_", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 5)) , ifelse(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) %in% "mito", paste0(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 4), "_", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 5), "_", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 6) ), paste0(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2), "_", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 3), "_", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 4)) ) )

pseudoBulk_ADT$group <- ifelse(!(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) %in% "BM") &  !(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) %in% "mito"), sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 6), ifelse(sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 2) %in% "mito", sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 7), sapply(strsplit(Cells(pseudoBulk_ADT), split = "_"), "[", 5)) )

pseudoBulk_ADT$group <- factor(x = pseudoBulk_ADT$group, levels = c("wt", "il27raKO"))


pseudoBulk_DE_ADT <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell") ){
  pseudoBulk_cell_subset <- subset(pseudoBulk_ADT, celltype == i)
  Idents(pseudoBulk_cell_subset) <- "group"
  ADT_DE <- FindMarkers(pseudoBulk_cell_subset, ident.1 = "il27raKO", ident.2 = "wt", slot = "counts", test.use = "DESeq2") %>%
    rownames_to_column(var = "ADT") %>% cbind(cell_type = i, .)
  pseudoBulk_DE_ADT[[i]] <- ADT_DE
  
  
}


tbl_pseudoBulk_DE_ADT <- bind_rows(pseudoBulk_DE_ADT)




```


```{r}
y <- AggregateExpression(tmp, return.seurat = TRUE, slot = "counts", assays = "RNA", group.by = c("celltype", "group")  )

y$celltype <- ifelse(!(sapply(strsplit(Cells(y), split = "_"), "[", 1) %in% c("Monocyte", "cDC", "moDC", "pDC", "Treg")) &  !(sapply(strsplit(Cells(y), split = "_"), "[", 2) %in% "mito"), paste0(sapply(strsplit(Cells(y), split = "_"), "[", 1), "_", sapply(strsplit(Cells(y), split = "_"), "[", 2) ), ifelse(sapply(strsplit(Cells(y), split = "_"), "[", 2) %in% "mito", paste0(sapply(strsplit(Cells(y), split = "_"), "[", 1), "_", sapply(strsplit(Cells(y), split = "_"), "[", 2), "_", sapply(strsplit(Cells(y), split = "_"), "[", 3) ), sapply(strsplit(Cells(y), split = "_"), "[", 1) ) )
                                  
y$group <- ifelse(!(sapply(strsplit(Cells(y), split = "_"), "[", 1) %in% c("Monocyte", "cDC", "moDC", "pDC", "Treg")) &  !(sapply(strsplit(Cells(y), split = "_"), "[", 2) %in% "mito"), sapply(strsplit(Cells(y), split = "_"), "[", 3),        ifelse(sapply(strsplit(Cells(y), split = "_"), "[", 2) %in% "mito", sapply(strsplit(Cells(y), split = "_"), "[", 4), sapply(strsplit(Cells(y), split = "_"), "[", 2)) )

z <- subset(y, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))

z$celltype <- factor(z$celltype, levels = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") )
z$group <- factor(z$group, levels = c("wt", "il27raKO"))


x <- dittoHeatmap(z, genes = c("Gzmk", "Serpina3g", "Prnp", "Ifitm3", "Il7r", "Samhd1", "Gbp6", "Rbl2", "Il18r1", "Iigp1", "Podnl1", "Txnip", "Adgre5", "Zfas1"), annot.by =  c("celltype", "group"), complex = TRUE, order.by = c("celltype", "group"))   #order.by source code:  https://github.com/dtm2451/dittoSeq/issues/70#issuecomment-772803365

x@matrix_param$height <- unit(3, "mm")*nrow(x@matrix); x@matrix_param$width <- unit(5, "mm")*ncol(x@matrix)


#pdf(paste0(fig_path, "/DittoHeatmap_AvgExpr_NK_Tcell_different_genes_fromAggExprDE.pdf"), width = 10, height = 6)
  x
##dev.off()


all_integrated_rnm_VlnPlot$group <- factor(all_integrated_rnm_VlnPlot$group, levels = c("wt", "il27raKO"))
#pdf(paste0(fig_path, "/VlnPlot_AvgExpr_NK_Tcell_different_genes_fromAggExprDE.pdf"), width = 10, height = 6)

VlnPlot(all_integrated_rnm_VlnPlot, features = c("Gzmk", "Serpina3g", "Prnp", "Ifitm3", "Il7r", "Samhd1", "Gbp6", "Rbl2", "Il18r1", "Iigp1", "Podnl1", "Txnip", "Adgre5", "Zfas1"), assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & ylab("") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))
##dev.off()


#Heatmap using avg_log2FC values based on DE analysis
Idents(all_integrated_rnm) <- "cluster0.8.group"
get_differential_genes_NK_T <- function(celltype){
    Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", logfc.threshold = 0) %>%
        rownames_to_column(var = "gene") %>%
        cbind(cell_type = celltype, .)
}
tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")

differential_genes_RNA_all_integrated_rnm_NK_T_lfc0 <- map_dfr(tmp_celltype_list, get_differential_genes_NK_T)

#writexl::write_xlsx(differential_genes_RNA_all_integrated_rnm_NK_T_lfc0 , paste0(tab_path, "differential_genes_RNA_all_integrated_rnm_NK_T_lfc0_res0.8_all_integrated_rnm.xlsx"))

x <- differential_genes_RNA_all_integrated_rnm_NK_T_lfc0[differential_genes_RNA_all_integrated_rnm_NK_T_lfc0$gene %in% c("Gzmk", "Serpina3g", "Prnp", "Ifitm3", "Il7r", "Samhd1", "Gbp6", "Rbl2", "Il18r1", "Iigp1", "Podnl1", "Txnip", "Adgre5", "Zfas1"), ] %>% select(cell_type, gene, avg_log2FC)

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_select_NK_vs_T_genes <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)


#create the actual heatmap object
hmap_mat_select_NK_vs_T_genes <- Heatmap(mat_select_NK_vs_T_genes,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = TRUE,
      show_row_dend = TRUE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = '',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = NULL,
    height = unit(3, "mm")*nrow(mat_select_NK_vs_T_genes), width = unit(5, "mm")*ncol(mat_select_NK_vs_T_genes)
)


#draw the heatmap
#pdf(paste0(fig_path, "logFC_DE_NK_Tcell_different_genes_SeuratDE_LFC0.pdf"), width = 8, height = 5)
hmap_mat_select_NK_vs_T_genes
#dev.off()

rm(x, hmap_mat_select_NK_vs_T_genes, mat_select_NK_vs_T_genes)

#Heatmap based on aggregate expression

x <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$gene %in% c("Gzmk", "Serpina3g", "Prnp", "Ifitm3", "Il7r", "Samhd1", "Gbp6", "Rbl2", "Il18r1", "Iigp1", "Podnl1", "Txnip", "Adgre5", "Zfas1") & tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] %>% select(cell_type, gene, avg_log2FC)

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_select_NK_vs_T_genes <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)


#create the actual heatmap object
hmap_mat_select_NK_vs_T_genes <- Heatmap(mat_select_NK_vs_T_genes,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = TRUE,
      show_row_dend = TRUE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = '',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = NULL,
    height = unit(3, "mm")*nrow(mat_select_NK_vs_T_genes), width = unit(5, "mm")*ncol(mat_select_NK_vs_T_genes)
)





#draw the heatmap
#pdf(paste0(fig_path, "logFC_AvgExpr_NK_Tcell_different_genes_fromAggExprDE.pdf"), width = 8, height = 5)
hmap_mat_select_NK_vs_T_genes
#dev.off()


```

#checking survival genes https://www.informatics.jax.org/vocab/gene_ontology/GO:0043066
```{r}
#checking survival genes https://www.informatics.jax.org/vocab/gene_ontology/GO:0043066
genes_select_1 <- c("Aamdc", "Aars", "Aatf",  "Abl1", "Abl2", "Abraxas2", "Acaa2", "Ackr3", "Acot1", "Actc1", "Acvr1", "Ada", "Adam8", "Adam17", "Adamts20", "Adar", "Adcyap1", "Adnp", "Adora1", "Adora2a", "Agap2", "Ago4", "Agt", "Agtr1a", "Agtr1b", "Agtr2", "Ahi1", "Aif1", "Aipl1", "Akr1a1",   "Akr1b3",   "Akt1",  "Akt2", "Alb", "Aldh2", "Alkbh1", "Alms1", "Alox12","Ambra1", "Amigo2", "Angpt1", "Angpt4", "Angptl4", "Ankle2", "Anp32b", "Apbb2", "Apbb3", "Apc", "Api5", "Apip",  "Apoe",  "Apoh",  "Aqp1", "Ar", "Araf",  "Arel1", "Arf4",  "Arg2",  "Arhgap10", "Arl6ip1", "Armc10", "Arnt2", "Arrb1", "Arrb2", "Ascl1", "Asic2", "Asns",  "Atad5", "Atf5",  "Atg5", "Atg7",  "Atoh1", "Atox1", "Atp7a", "Aurka", "Aurkb", "Aven",  "Avp", "Axl", "Babam2", "Bad", "Bag1",  "Bag3",  "Bag5", "Bag6", "Bak1", "Bard1", "Barhl1", "Bax", "Bcl2", "Bcl2a1a",  "Bcl2a1d", "Bcl2l1", "Bcl2l2", "Bcl2l10", "Bcl2l12", "Bcl3", "Bcl6", "Bcl10", "Bcl11b", "Bdkrb2", "Bdnf", "Becn1", "Bfar", "Bhlhb9", "Bid", "Birc2", "Birc3", "Birc5", "Birc6", "Birc7", "Blm", "Bmf", "Bmi1", "Bmp4", "Bmp7",  "Bnip3", "Bnip3l", "Bok", "Braf", "Brca1", "Btc",   "Btg2",  "C5ar1", "Caap1", "Cabin1",   "Cacna1a",  "Camk1d",   "Capn3", "Card14",   "Casp2", "Casp3", "Cast",  "Cat",   "Cav1",  "Cbl",   "Cblb",  "Cbs",   "Ccar2", "Ccl2",  "Ccl5",  "Ccl12", "Ccl19", "Ccl21a", "Ccl21b", "Ccl21d", "Ccn1",  "Ccnd1", "Ccnd2", "Ccng1", "Ccr5",  "Ccr7",  "Cd27",  "Cd38",  "Cd40lg",   "Cd44", "Cd59a", "Cd59b", "Cd74",  "Cdc73", "Cdh5",  "Cdk1",  "Cdk11b", "Cdkn1a", "Cdkn1b", "Cdkn2d", "Cebpb", "Cep63", "Cerkl", "Cfdp1", "Cflar", "Cftr",  "Chd8",  "Chl1",  "Chst11",   "Ciapin1",  "Cib1",  "Cidea", "Cited1",   "Cited2", "Ckmt1", "Clcf1", "Cldn7", "Clec5a",   "Cln3",  "Cln8",  "Clu",   "Cntf",  "Cntfr", "Col2a1",   "Comp",  "Cops5", "Coro1a",  "Cpeb4", "Creb3", "Creb3l1",  "Crlf1", "Cryaa", "Cryab", "Csf1r", "Csf2",  "Cth",   "Ctnna1",   "Ctnnb1", "Ctns",  "Ctsh",  "Cttn",  "Cx3cl1",   "Cx3cr1",   "Cxcl12", "Cxcr2", "Cxcr3")

genes_select_upd_1 <- genes_select_1[genes_select_1 %in% row.names(tmp[["RNA"]] )]


genes_select_2 <- c("D1Pas1",   "Dab2",  "Dad1",  "Dapk1", "Daxx",  "Ddb1",  "Ddias", "Ddr2",  "Ddrgk1", "Ddx3x", "Dffa",  "Dhcr24",   "Dhrs2", "Dipk2a", "Dkk1",  "Dll1", "Dlx1",  "Dnaja1",   "Dnaja3",   "Dnajb6", "Dnajc3", "Dnajc5", "Dnmt1", "Dock8", "Dpep1", "Draxin",   "Drd3",  "Dspp", "Dstyk", "Dusp1", "Dyrk3", "Edn1",  "Ednra", "Ednrb", "Eef2k", "Efna1", "Egfr",  "Egr3",  "Eif2ak2",  "Eif2ak3", "Eif5a", "Ell3",  "En1",   "En2",   "Eno1",  "Eno1b", "Ep300", "Epcam", "Epo",   "Epor",  "Erbb2", "Erbb3", "Erbb4", "Ercc5", "Evi2b", "Eya1",  "Eya2",  "Eya3",  "Eya4",  "F2r",   "Fabp1", "Fadd",  "Faim",  "Faim2", "Faiml", "Fank1", "Fas",   "Fate1", "Fcer1g",   "Fcgr2b",   "Fcmr",  "Ffar4", "Fga",   "Fgb",   "Fgf4",  "Fgf8", "Fgf10", "Fgf20", "Fgf21", "Fgfr2", "Fgg",  "Fhl2",  "Fignl1",   "Fkbp8", "Flna",  "Flt3l", "Flt4",  "Fmn2", "Fmr1",  "Fn1",   "Fnip1", "Fnta",  "Foxb1", "Foxc1", "Foxc2", "Foxe3", "Foxo1", "Foxp1", "Foxq1", "Fstl1", "Fxn",   "Fyn",   "Fzd3",  "Fzd9",  "G2e3",  "Gapdh", "Gas1",  "Gas6",  "Gata1", "Gata2", "Gata3", "Gata4", "Gata6", "Gba",   "Gbe1",  "Gcg",   "Gclc",  "Gclm",  "Gcm2",  "Gdf5",  "Gdnf",  "Gfer",  "Gfral", "Ghitm", "Ghrh",  "Ghrl",  "Ghsr",  "Gimap3",   "Gimap5",   "Gli2",  "Gli3",  "Glo1",  "Glp1r", "Gm20594",  "Gnai2", "Gnai3", "Gnaq",  "Gnrh1", "Golph3",  "Gpam",  "Gpi1",  "Gpx1",  "Grem1", "Gria4", "Grik2", "Grin1", "Grina", "Grk1", "Grk2",  "Grk5",  "Grm7",  "Grn",   "Gsk3b", "Gstp1", "Gstp2", "Gstp3", "Hand2", "Hax1",  "Hck",   "Hcls1", "Hdac1", "Hdac2", "Hdgf",  "Hells", "Herpud1",  "Hey2",  "Hgf",   "Hhip",  "Hif1a", "Higd1a",   "Higd2a",   "Hip1r", "Hipk2", "Hipk3", "Hk1",   "Hmga2", "Hmgb2", "Hmgcr", "Hmgn5", "Hmox1", "Hnf1a", "Hnf1b", "Hnrnpk",   "Hpn", "Hras",  "Hsf1",  "Hsh2d", "Hsp90ab1", "Hspa1b",   "Hspa5", "Hspb1", "Hspb6", "Hspd1", "Hsph1", "Htr2b", "Htt", "Hyou1", "Hypk",  "Icam1", "Id1",   "Ido1",  "Ier3",  "Ifit3", "Ifit3b", "Igbp1", "Igf1",  "Igf1r", "Ihh", "Ikbkg", "Il1b",  "Il1rn", "Il2",   "Il2rb", "Il3",   "Il4",   "Il6",   "Il6st", "Il7",   "Il7r",  "Il9", "Il9r",  "Il10",  "Il13",  "Il18",  "Il19",  "Il24",  "Il27ra",   "Ilk",  "Ing2", "Insl3", "Insl6", "Ints1", "Irs2",  "Isl1",  "Itch",  "Itga5", "Itga6", "Itgav", "Itgb1", "Itgb3", "Itpkb", "Itprip",   "Itsn1", "Ivns1abp", "Jak2",  "Jak3",  "Jun")

genes_select_upd_2 <- genes_select_2[genes_select_2 %in% row.names(tmp[["RNA"]] )]


genes_select_3 <- c("Kcnh8",  "Kcnj1",  "Kdm1a",  "Kdm2a",  "Kdm2b",  "Kdm6a",  "Kdr",  "Khdc3",  "Kif14",  "Kifap3", "Kitl", "Klf4", "Klhl20", "Kras", "Krit1",  "Krt18",  "Lamp3",  "Lamtor5",  "Lcn2", "Lef1", "Lep",  "Lgals3", "Lgmn", "Lhx3", "Lhx4", "Lifr", "Lig4", "Lims1",  "Lims2", "Lmna", "Lrp1", "Lrp2", "Lrp6", "Lrrk2",  "Ltk",  "Lypd3",  "Mad2l1", "Madd", "Maea", "Mael", "Mag",  "Map2k1", "Map2k4", "Map2k5", "Map3k7", "Map3k12",  "Map4k4", "Mapk7",  "Mapk8",  "Mapk8ip1", "Mapk8ip2", "Mapk8ip3", "Mapkap1",  "Marchf7",  "Maz",  "Mcl1", "Mdk",  "Mdm2", "Mdm4", "Mecp2", "Med1", "Mef2c",  "Meis3",  "Mertk",  "Mfn2", "Mgmt", "Mical1", "Mien1",  "Mif",  "Mir17", "Mir18",  "Mir18b", "Mir19b-1", "Mir19b-2", "Mir20a", "Mir20b", "Mir25",  "Mir92-1", "Mir92-2", "Mir93", "Mir106a", "Mir106b",  "Mir290a",  "Mir292", "Mir293", "Mir294", "Mir295", "Mir363", "Mir494", "Mitf", "Mlst8",  "Mmp9", "Mnat1",  "Mnt",  "Mpl",  "Mpv17l", "Mpz", "Mre11a", "Mrtfa",  "Msh2", "Mst1", "Msx1", "Msx2", "Mt1",  "Mt3",  "Mtdh", "Mtnr1b", "Mtor", "Muc1", "Muc4", "Myc",  "Mydgf",  "Myo18a", "Myocd",  "Naa15",  "Naa16",  "Naa35",  "Naa38",  "Naca", "Naip1", "Naip2",  "Naip5",  "Naip6",  "Naip7",  "Nanos3", "Nckap1l",  "Ncl", "Ncoa3",  "Ndnf", "Ndufa13", "Ndufaf4",  "Ndufs3", "Nefl", "Nes",  "Neurod1",  "Nfatc4", "Nfe2l2", "Nfix", "Nfkb1",  "Ngb", "Ngf",  "Ngfr", "Niban2", "Nkx2-5", "Nkx2-6", "Nkx3-2", "Nle1", "Nme5", "Nmnat1", "Nnt", "Noc2l",  "Nod2", "Nog",  "Nol3", "Nono", "Nos1", "Notch1", "Npc1", "Npm1", "Nppc", "Npy5r",  "Nqo1", "Nr1h4",  "Nr2e1",  "Nr3c1",  "Nr4a1",  "Nr4a2",  "Nr4a3",  "Nrbp2",  "Nrg1",  "Nrp1", "Ntf3", "Ntf5", "Ntrk1",  "Ntrk2",  "Ntsr1",  "Nuak2",  "Nuggc",  "Nup62",  "Nupr1", "Ogg1", "Oog2", "Oog3", "Opa1", "Opn3", "Optn", "Ormdl3", "Osr1", "Oxr1", "Pa2g4", "Pafah2", "Pak2", "Pak4", "Pak5", "Palb2",  "Pam16",  "Park7",  "Parl", "Pax2", "Pax4", "Pax8", "Pcdhgc3",  "Pcdhgc4",  "Pcdhgc5", "Pcgf2",  "Pcid2",  "Pcmt1",  "Pcnt", "Pcp4", "Pdcd1", "Pdcd4",  "Pdcd10", "Pde3a",  "Pdgfrb", "Pdk4", "Pdpk1",  "Pdpn", "Pdx1", "Pdxk", "Pea15a",  "Peli3",  "Pgr",  "Phb",  "Phb2", "Phip", "Pias1",  "Pidd1",  "Pih1d1", "Pik3ca", "Pik3cb",  "Pik3cg", "Pik3r1", "Pim1", "Pim2", "Pim3", "Pin1", "Pin1rt1",  "Pink1",  "Pip",  "Pkhd1", "Pla2g3", "Plac8",  "Plaur",  "Plk1", "Plk2", "Plk3", "Plxnd1", "Pnp",  "Por",  "Pou3f3", "Pou3f4", "Pou4f1", "Ppara",  "Ppard",  "Ppargc1a", "Ppia", "Ppif", "Ppp1r10",  "Ppp5c",  "Ppt1", "Pramel1",  "Pramel7",  "Prap1",  "Prdm9",  "Prdx2",  "Prdx3",  "Prdx5",  "Prelid1",  "Prkaa1", "Prkaa2", "Prkca",  "Prkcd",  "Prkce",  "Prkcg",  "Prkch",  "Prkci",  "Prkcq",  "Prkcz",  "Prkdc",  "Prkn", "Prlr", "Prnp", "Proc", "Prok2",  "Prokr1", "Prop1",  "Prr5", "Psen1",  "Psen2",  "Psmd10", "Psme3",  "Psmg2",  "Ptcra",  "Pten", "Ptgfr",  "Ptgs2",  "Pth",  "Ptk2", "Ptk2b",  "Ptma", "Ptms", "Ptpn1",  "Ptprz1", "Ptrh2",  "Pttg1ip",  "Qars", "Qki")

genes_select_upd_3 <- genes_select_3[genes_select_3 %in% row.names(tmp[["RNA"]] )]



genes_select_4 <- c("Raf1", "Rag1", "Ramp2", "Rapgef3",  "Rara", "Rarb", "Rarg", "Rasa1",  "Rb1",  "Rb1cc1", "Reg3b",  "Rela", "Retreg1", "Rffl", "Rgl2", "Rgn",  "Rhbdd1", "Rhoa", "Rictor", "Ripk1",  "Rln1", "Rnf7", "Rnf34", "Rnf144b",  "Rnf157", "Rnu12",  "Rock1",  "Rorc", "Rpl10",  "Rps3a1", "Rps6", "Rps6ka1",  "Rps6ka3", "Rps6kb1",  "Rrm2b",  "Rrn3", "Rtkn2",  "Rxfp2",  "Scg2", "Sct",  "Scx",  "Selenos",  "Sema3e",  "Sema5a", "Serpinb2", "Serpinb9", "Serpinb9b",  "Serpinb9c",  "Serpinb9d",  "Serpinb9e",  "Serpinb9f",  "Serpinb9g",  "Serpinb9h", "Serpinb13",  "Serpine1", "Set",  "Setx", "Sfn",  "Sfrp1",  "Sfrp2",  "Sgk1", "Sgk3", "Sgms1", "Sh3glb1",  "Sh3rf1", "Sh3rf2", "Shc1", "Shh",  "Siah2",  "Sin3a",  "Sirt1",  "Sirt2",  "Sirt4", "Sirt5",  "Six1", "Six4", "Slc1a1", "Slc2a3", "Slc7a5", "Slc9a1", "Slc25a4",  "Slc25a5",  "Slc25a27", "Slc25a31", "Slc35f6",  "Slc39a10", "Slc40a1",  "Slc46a2",  "Smad3",  "Smad5",  "Smad6",  "Smarca4",  "Smo", "Snai1",  "Snai2",  "Snca", "Sncb", "Snx6", "Sod1", "Sod2", "Son",  "Sox8", "Sox9", "Sox10",  "Sphk1",  "Spp1", "Spry2",  "Src",  "Srsf6",  "St3gal1",  "St6gal1",  "Stambp", "Star", "Stat5a", "Stat5b", "Stil", "Stk40",  "Stradb", "Stxbp1", "Supv3l1",  "Sycp2",  "Syngap1",  "Syvn1", "Taf9", "Taf9b",  "Tax1bp1",  "Tbx1", "Tbx3", "Tcf7", "Tcf7l2", "Tcim", "Tcl1", "Tek", "Tent5b", "Tert", "Tex11",  "Tfap2a", "Tfap2b", "Tfap2d", "Tfrc", "Tgfa", "Tgfb2",  "Tgfb3", "Tgfbr1", "Tgfbr3", "Thap11", "Thbs1",  "Thoc6",  "Timp1",  "Tjp1", "Tle1", "Tmbim1", "Tmbim4", "Tmbim6", "Tmem14a",  "Tmem161a", "Tmf1", "Tmigd1", "Tnf",  "Tnfaip3",  "Tnfrsf1a", "Tnfrsf4",  "Tnfrsf18", "Tnfrsf22", "Tnfrsf23", "Tnfsf4", "Tnip2",  "Tox3", "Tpt1", "Traf2",  "Trap1",  "Trem2",  "Triap1", "Trim32", "Trip10", "Trp53",  "Trp63",  "Trp73",  "Tsc22d1",  "Tsc22d3",  "Tslp", "Ttpa", "Twist1", "Twist2", "Txndc12",  "Tyro3",  "Ube2b",  "Ube2v2", "Ube2z",  "Ucn", "Ucp2", "Ufl1", "Ufm1", "Unc5b",  "Ung",  "Uri1", "Usp47",  "Vdac1",  "Vdac2",  "Vegfa",  "Vegfb",  "Vhl",  "Vip", "Vps54",  "Vstm2l", "Vtcn1",  "Wdr73",  "Wfs1", "Wnk3", "Wnt1", "Wnt4", "Wnt5a",  "Wnt7a", "Wnt9a",  "Wt1", "Xbp1", "Xiap", "Xrcc2",  "Xrcc4",  "Yap1", "Ybx3", "Yme1l1", "Ywhah", "Zc3h12a",  "Zc3hc1", "Zfand6", "Zfp385a",  "Zfp830", "Zmynd11",  "Zpr1")

genes_select_upd_4 <- genes_select_4[genes_select_4 %in% row.names(tmp[["RNA"]] )]



genes_select <- c(genes_select_1, genes_select_2, genes_select_3, genes_select_4)

genes_select <- genes_select[genes_select %in% differential_genes_all_integrated_rnm$gene]



#pdf(paste0(fig_path, "Htmap_negative regulation of apoptotic process_GO_0043066_fromAggExprDE.pdf"), width = 10, height = 30)
x <- dittoHeatmap(z, genes = genes_select[1:200], annot.by =  c("celltype", "group"), complex = TRUE, order.by = c("celltype", "group")) 
x@matrix_param$height <- unit(3, "mm")*nrow(x@matrix) #adjusts height and width of heatmap
x@matrix_param$width <- unit(5, "mm")*ncol(x@matrix)
x

y <- dittoHeatmap(z, genes = genes_select[201:407], annot.by =  c("celltype", "group"), complex = TRUE, order.by = c("celltype", "group")) 
y@matrix_param$height <- unit(3, "mm")*nrow(y@matrix) #adjusts height and width of heatmap
y@matrix_param$width <- unit(5, "mm")*ncol(y@matrix)
y
#dev.off()






#Heatmap based on aggregate expression

x <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$gene %in% genes_select[1:200] & tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] %>% select(cell_type, gene, avg_log2FC)

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]

y <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$gene %in% genes_select[201:407] & tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] %>% select(cell_type, gene, avg_log2FC)

y <- spread(y, cell_type, avg_log2FC)
row.names(y) <- y$gene
y <- y[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_select_NK_vs_T_genes_1 <- as.matrix(x)
mat_select_NK_vs_T_genes_2 <- as.matrix(y)


#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)


#create the actual heatmap object
hmap_mat_select_NK_vs_T_genes_1 <- ComplexHeatmap::Heatmap(mat_select_NK_vs_T_genes_1,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = TRUE,
      show_row_dend = TRUE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = '',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = NULL,
    height = unit(3, "mm")*nrow(mat_select_NK_vs_T_genes_1), width = unit(5, "mm")*ncol(mat_select_NK_vs_T_genes_1)
)


hmap_mat_select_NK_vs_T_genes_2 <- ComplexHeatmap::Heatmap(mat_select_NK_vs_T_genes_2,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = TRUE,
      show_row_dend = TRUE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = '',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = NULL,
    height = unit(3, "mm")*nrow(mat_select_NK_vs_T_genes_2), width = unit(5, "mm")*ncol(mat_select_NK_vs_T_genes_2)
)



#draw the heatmap
#pdf(paste0(fig_path, "logFC_negative regulation of apoptotic process_GO_0043066_fromAggExprDE.pdf"), width = 8, height = 30)
hmap_mat_select_NK_vs_T_genes_1
hmap_mat_select_NK_vs_T_genes_2
#dev.off()


```

#muscatDS NK, T cell heatmap with logFC based on unique genes in either NK or T cell
```{r}
x <- bind_rows(tbl_fil_all_RNA)
y <- x[x$p_adj.loc < 0.05, ]
y_NK <- y[y$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]
y_T <- y[y$cluster_id %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]
y_NK <- unique(y_NK$gene)
y_T <- unique(y_T$gene)

unique_NK_T <- c(y_NK[!(y_NK %in% y_T)], y_T[!(y_T %in% y_NK)])

mat_NK_T_muscatDS <- x[x$gene %in% unique_NK_T, ] %>% select(cluster_id, gene, logFC)
mat_NK_T_muscatDS <- spread(mat_NK_T_muscatDS, cluster_id, logFC)

row.names(mat_NK_T_muscatDS) <- mat_NK_T_muscatDS$gene
mat_NK_T_muscatDS <- mat_NK_T_muscatDS[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]

mat_NK_T_muscatDS <- as.matrix(mat_NK_T_muscatDS)

#create the actual heatmap object
hmap_mat_NK_T_muscatDS <- ComplexHeatmap::Heatmap(mat_NK_T_muscatDS,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = TRUE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = '',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = NULL,
    height = unit(3, "mm")*nrow(mat_NK_T_muscatDS), width = unit(5, "mm")*ncol(mat_NK_T_muscatDS)
) #cluster_rows=FALSE because of NA values; source: https://stackoverflow.com/questions/67524158/why-dont-my-na-values-work-inside-of-complexheatmap-in-r

#draw the heatmap
#pdf(paste0(fig_path, "logFC_hmap_NK_T_p.loc0.05_muscatDS.pdf"), width = 8, height = 30)
hmap_mat_NK_T_muscatDS
#dev.off()



```



#muscatDS NK, T cell heatmap with logFC
```{r}
x <- bind_rows(tbl_fil_all_RNA)
y <- x[x$p_adj.loc < 0.05, ]
y <- y[y$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]



mat_NK_T_both_muscatDS <- x[x$gene %in% y$gene, ] %>% select(cluster_id, gene, logFC)
mat_NK_T_both_muscatDS <- spread(mat_NK_T_both_muscatDS, cluster_id, logFC)

row.names(mat_NK_T_both_muscatDS) <- mat_NK_T_both_muscatDS$gene
mat_NK_T_both_muscatDS <- mat_NK_T_both_muscatDS[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK",  "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT")] #excluding "CD27loCD11blo_NK" and "CD27-_gdT" as they did not have statistically significant DE as per muscatDS

mat_NK_T_both_muscatDS <- as.matrix(mat_NK_T_both_muscatDS)

#create the actual heatmap object
hmap_mat_NK_T_both_muscatDS <- ComplexHeatmap::Heatmap(mat_NK_T_both_muscatDS,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = TRUE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = '',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = NULL,
    height = unit(3, "mm")*nrow(mat_NK_T_both_muscatDS), width = unit(5, "mm")*ncol(mat_NK_T_both_muscatDS)
) #cluster_rows=FALSE because of NA values; source: https://stackoverflow.com/questions/67524158/why-dont-my-na-values-work-inside-of-complexheatmap-in-r

#draw the heatmap
#pdf(paste0(fig_path, "logFC_hmap_NK_T_both_p.loc0.05_muscatDS.pdf"), width = 8, height = 30)
hmap_mat_NK_T_both_muscatDS
#dev.off()


```


#Seurat DE NK, T cell heatmap with avg_log2FC
```{r}
Idents(all_integrated_rnm) <- "cluster0.8.group"
get_differential_genes_NK_T <- function(celltype){
    Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", logfc.threshold = 0) %>%
        rownames_to_column(var = "gene") %>%
        cbind(cell_type = celltype, .)
}
tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")

differential_genes_RNA_all_integrated_rnm_NK_T_lfc0 <- map_dfr(tmp_celltype_list, get_differential_genes_NK_T)



x <- differential_genes_RNA_all_integrated_rnm_NK_T_lfc0[differential_genes_RNA_all_integrated_rnm_NK_T_lfc0$p_val_adj < 0.05,]

mat_NK_T_both_SeuratDE <- x %>% select(cell_type, gene, avg_log2FC)
mat_NK_T_both_SeuratDE <- spread(mat_NK_T_both_SeuratDE, cell_type, avg_log2FC)

row.names(mat_NK_T_both_SeuratDE) <- mat_NK_T_both_SeuratDE$gene

mat_NK_T_both_SeuratDE <- mat_NK_T_both_SeuratDE[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT")] #excluding "CD27loCD11blo_NK" and "CD27-_gdT" which do not have statistically significant DE

mat_NK_T_both_SeuratDE <- as.matrix(mat_NK_T_both_SeuratDE)

#create the actual heatmap object
hmap_mat_NK_T_both_SeuratDE <- ComplexHeatmap::Heatmap(mat_NK_T_both_SeuratDE,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = TRUE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = '',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = NULL,
    height = unit(3, "mm")*nrow(mat_NK_T_both_SeuratDE), width = unit(5, "mm")*ncol(mat_NK_T_both_SeuratDE)
) #cluster_rows=FALSE because of NA values; source: https://stackoverflow.com/questions/67524158/why-dont-my-na-values-work-inside-of-complexheatmap-in-r

#draw the heatmap
#pdf(paste0(fig_path, "logFC_hmap_NK_T_both_p_adj0.05_SeuratDE.pdf"), width = 8, height = 40)
hmap_mat_NK_T_both_SeuratDE
#dev.off()


```




```{r}

tmp <- all_integrated_rnm
Idents(tmp) <- "celltype"
DefaultAssay(tmp) <- "RNA"

enrichGOBP_Seurat_DE_genes <- list()


for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell") ){#excluding "CD27loCD11blo_NK", "CD27-_gdT", "B_cell" as there are no statistically significant DE genes based on adjusted p value
  diff_for_goBP <- subset(differential_genes_all_integrated_rnm, subset = cell_type %in% i & p_val_adj < 0.05)$gene
  #clust <- tmp[, tmp$celltype %in% i] #subset to a particular cell type
  #expr_clust <- clust[Matrix::rowSums(clust[["RNA"]]@counts > 0) > 0, ] #selecting background as genes which are at least expressed in a particular cluster/celltype
  #background.genes <- rownames(expr_clust) #NOT including background as the number of significant DE genes is low, and including background genes results in no enrichment
  
  goBP_DE <- clusterProfiler::enrichGO(gene = unique(bitr(diff_for_goBP, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                       #universe = background.genes,
                                       OrgDb = "org.Mm.eg.db",
                                       readable = TRUE,
                                       keyType = "ENTREZID",
                                       ont = "BP",
                                       pvalueCutoff = 0.05,
                                       pAdjustMethod = "BH"
                                                     )
  enrichGOBP_Seurat_DE_genes[[i]] <- goBP_DE
  
}




enrichGOBP_Seurat_DE_genes_table <- list()

for(i in c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "moDC") ){ #excluding "CD27loCD11b+_mito_NK", "pDC", as they was not enriched
  x <- enrichGOBP_Seurat_DE_genes[[i]]@result
  
  if(nrow(x) != 0) {
    x$celltype <- i
    enrichGOBP_Seurat_DE_genes_table[[i]] <- x
  
  }
}

#writexl::write_xlsx(bind_rows(enrichGOBP_Seurat_DE_genes_table), paste0(tab_path, "tmp_enrichgoBP_xl.xlsx"))

#saveRDS(enrichGOBP_Seurat_DE_genes, file = paste0(tab_path, "enrichGOBP_Seurat_DE_genes.rds"))
#enrichGOBP_Seurat_DE_genes <- readRDS(paste0(tab_path, "enrichGOBP_Seurat_DE_genes.rds"))

```

#Plotting genes inversely correlated between NK and T cells (NK up, T down: Seurat DE, lfc threshold default)
- Notes:
- Checking only NK up and T down as T down indicates the group of genes that are upregulated in NK cells and downregulated in all T clls in global IL27RA KO deficient mice.
- Looking at T cells down DE gives an idea of the genes that are perhaps upregulated in WT T cells.
- However NK down, T up is not considered, as T up are DE genes specific to the global deficient IL27RA KO mice but not the NK cell specific IL27RA KO mice.
```{r}
x <- dittoHeatmap(z, genes = unique(c(vennPlot_list_inverse_correlated$vennPlot_DGE_NK_up_vs_T_down$Detail)), annot.by =  c("celltype", "group"), complex = TRUE, order.by = c("celltype", "group"), main = "Genes that are upregulated in NK cell\n subsets and downregulated in \n T cell subets")  #order.by source code:  https://github.com/dtm2451/dittoSeq/issues/70#issuecomment-772803365

x@matrix_param$height <- unit(3, "mm")*nrow(x@matrix); x@matrix_param$width <- unit(5, "mm")*ncol(x@matrix)


#pdf(paste0(fig_path, "/DittoHeatmap_NK_T_SeuratDE_inverselyCorrelated_fromAggExprDE.pdf"), width = 10, height = 15)
  x
##dev.off()

VlnPlot(all_integrated_rnm_VlnPlot, features = c("Cybb", "Iigp1", "Gm4951", "St8sia1", "Ly6c1", "Ifit1bl1", "Ifit3", "Slfn1", "Il7r", "Pdcd1", "Ckb", "Podnl1", "Tnfrsf25"), assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) 


```


```{r}

suppressPackageStartupMessages({
library(ggpubr)
library(VennDetail)
library(VennDiagram)
})

#RNA
# Combine all NK clusters into one and all T cell clusters into one

NK_genes_pB <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

Tcell_genes_pB <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM", "Treg", "CD27+_gdT", "CD27-_gdT"), ]


#genes upregulated in NK cells and downregulated in T cells
vennPlot_DGE_NK_up_vs_T_down_pB <- venndetail(list(
    NK_pB_up = NK_genes_pB$gene[NK_genes_pB$avg_log2FC > 0 & NK_genes_pB$p_val_adj < 0.05], #subset to upregulated genes (avg_log2FC > 0) that are statistically significant (p_val_adj < 0.05)
    T_pB_down = Tcell_genes_pB$gene[Tcell_genes_pB$avg_log2FC < 0 & Tcell_genes_pB$p_val_adj < 0.05],
    NK_pB_down = NK_genes_pB$gene[NK_genes_pB$avg_log2FC < 0], 
    T_pB_up = Tcell_genes_pB$gene[Tcell_genes_pB$avg_log2FC > 0]
))


```


```{r}
#Checking correlation between cell types based on aggregated RNA counts
y <- AggregateExpression(tmp, return.seurat = FALSE, slot = "counts", assays = "RNA", group.by = c("celltype", "group")  )

cor_celltype_group <- y$RNA %>% cor(method = "spearman") #Source code:  https://github.com/maxkarlsson/HPA-SingleCellType/blob/b7b70d7e4b164956fe674531d415f9c0068b4397/scripts/main_HPA21.Rmd (lines 226-231)

cor_celltype_group %>% pheatmap::pheatmap()


z <- y$RNA[, c( c("CD27loCD11b+_NK_il27raKO", "CD27loCD11bhi_NK_il27raKO", "CD27hiCD11b+_NK_il27raKO", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11blo_NK_il27raKO", "CD8_Nave_il27raKO", "CD8_TCM_il27raKO","CD8_TEM_il27raKO", "CD4_TEM_il27raKO", "Treg_il27raKO", "CD27+_gdT_il27raKO", "CD27-_gdT_il27raKO"))]

cor_celltype_il27raKO <- z %>% cor(method = "spearman") #Source code:  https://github.com/maxkarlsson/HPA-SingleCellType/blob/b7b70d7e4b164956fe674531d415f9c0068b4397/scripts/main_HPA21.Rmd (lines 226-231)

cor_celltype_il27raKO %>% pheatmap::pheatmap()

#==========================================================================

#Checking DE between NK_cell_il27raKO vs NK_cell_wt, T_cell_wt and T_cell_il27raKO to identify genes specifically deregulated in NK_cell_il27raKO
tmp <- all_integrated_rnm #assign seurat object to a temporary variable
Idents(tmp) <- "celltype"
tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )
tmp$celltype_whole <- tmp$celltype
tmp$celltype_whole <- ifelse(tmp$celltype_whole %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), "NK_cell", ifelse(tmp$celltype_whole %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), "T_cell", tmp$celltype_whole)) #assign all NK cell subsets to NK and T cell subsets to T cell
tmp$celltype_whole_group <- paste0(tmp$celltype_whole, "_", tmp$group)
Idents(tmp) <- "celltype_whole_group"

q <- FindMarkers(tmp, assay = "RNA", ident.1 = "NK_cell_il27raKO", ident.2 = c("T_cell_wt", "T_cell_il27raKO", "NK_cell_wt")  )
q$gene <- row.names(q)

qp <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$gene %in% q$gene[q$avg_log2FC > 0], ] #Check which genes upregulated in the DE between NK_cell_il27raKO vs NK_cell_wt, T_cell_wt and T_cell_il27raKO is also present in the differential expression analysis table

qp_genes <- qp[qp$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & qp$p_val_adj < 0.05, ] #subset to statistically significant DE genes enriched in NK and T cell subsets

VlnPlot(all_integrated_rnm_VlnPlot, features = c("Wdfy1", "Ccl5", "Nabp1", "Serpinb6b", "Gzmb", "P2ry14", "Gpr171", "Lilrb4a", "Serpina3g", "Serpinb9"), assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))

dittoHeatmap(z, genes = unique(qp_genes$gene), annot.by =  c("celltype", "group"), complex = TRUE, order.by = c("celltype", "group"))


VlnPlot(all_integrated_rnm_VlnPlot, features = unique(qp_genes$gene)[1:16], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))
VlnPlot(all_integrated_rnm_VlnPlot, features = unique(qp_genes$gene)[17:33], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))



p <- FindMarkers(tmp, assay = "RNA", ident.1 = "NK_cell_il27raKO", ident.2 = "T_cell_wt")
p$gene <- row.names(p)
dittoHeatmap(z, genes = p$gene, annot.by =  c("celltype", "group"), complex = TRUE, order.by = c("celltype", "group"))



VlnPlot(all_integrated_rnm_VlnPlot, features = c(q$gene[q$pct.1 > 0.5 & q$pct.2 < 0.5][1:15]), assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))

#==============================================================================
#Checking DE between NK_cell_il27raKO vs NK_cell_wt, T_cell_wt and T_cell_il27raKO to identify genes specifically deregulated in NK_cell_il27raKO
tmp <- all_integrated_rnm #assign seurat object to a temporary variable
Idents(tmp) <- "celltype"
tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )
Idents(tmp) <- "cluster0.8.group"

nk_1 <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_NK_il27raKO", ident.2 = c("CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt", "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

nk_1$gene <- row.names(nk_1)
nk_1$contrast <- paste0( "CD27loCD11b+_NK_il27raKO", "_vs_", "CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")



nk_2 <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11bhi_NK_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt", "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

nk_2$gene <- row.names(nk_2)
nk_2$contrast <- paste0( "CD27loCD11bhi_NK_il27raKO" , "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")


nk_3 <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27hiCD11b+_NK_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt", "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

nk_3$gene <- row.names(nk_3)
nk_3$contrast <- paste0( "CD27hiCD11b+_NK_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")


nk_4 <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_mito_NK_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt", "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

nk_4$gene <- row.names(nk_4)
nk_4$contrast <- paste0( "CD27loCD11b+_mito_NK_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")


nk_5 <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11blo_NK_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt", "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

nk_5$gene <- row.names(nk_5)
nk_5$contrast <- paste0( "CD27loCD11blo_NK_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt,  CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")


t_1 <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD8_Nave_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_wt", "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

t_1$gene <- row.names(t_1)
t_1$contrast <- paste0( "CD8_Nave_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")



t_2 <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD8_TCM_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt",  "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

t_2$gene <- row.names(t_2)
t_2$contrast <- paste0("CD8_TCM_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")


t_3 <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD8_TEM_il27raKO" , ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt",  "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

t_3$gene <- row.names(t_3)
t_3$contrast <- paste0("CD8_TEM_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")


t_4 <- FindMarkers(tmp, assay = "RNA", ident.1 =  "CD4_TEM_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt",  "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

t_4$gene <- row.names(t_4)
t_4$contrast <- paste0("CD4_TEM_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")


t_5 <- FindMarkers(tmp, assay = "RNA", ident.1 =  "Treg_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt",  "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

t_5$gene <- row.names(t_5)
t_5$contrast <- paste0("Treg_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")


t_6 <- FindMarkers(tmp, assay = "RNA", ident.1 =  "CD27+_gdT_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt",  "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_wt", "CD27-_gdT_il27raKO", "CD27-_gdT_wt") ) 

t_6$gene <- row.names(t_6)
t_6$contrast <- paste0("CD27+_gdT_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_wt, CD27-_gdT_il27raKO, CD27-_gdT_wt")


t_7 <- FindMarkers(tmp, assay = "RNA", ident.1 =  "CD27-_gdT_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_il27raKO", "CD27loCD11blo_NK_wt", "CD8_Nave_il27raKO", "CD8_Nave_wt",  "CD8_TCM_il27raKO", "CD8_TCM_wt", "CD8_TEM_il27raKO", "CD8_TEM_wt", "CD4_TEM_il27raKO", "CD4_TEM_wt", "Treg_il27raKO", "Treg_wt", "CD27+_gdT_il27raKO", "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

t_7$gene <- row.names(t_7)
t_7$contrast <- paste0("CD27-_gdT_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt, CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_wt")


# #writexl::write_xlsx(list("CD27loCD11b+_NK_il27raKO_vs_NK_T" = nk_1,
#                          "CD27loCD11bhi_NK_il27raKO" = nk_2, 
#                          "CD27hiCD11b+_NK_il27raKO" = nk_3, 
#                          "CD27loCD11b+_mito_NK_il27raKO" = nk_4, 
#                          "CD27loCD11blo_NK_il27raKO" = nk_5, 
#                          "CD8_Nave_il27raKO" = t_1, 
#                          "CD8_TCM_il27raKO" = t_2, 
#                          "CD8_TEM_il27raKO" = t_3, 
#                          "CD4_TEM_il27raKO" = t_4, 
#                          "Treg_il27raKO" = t_5, 
#                          "CD27+_gdT_il27raKO" = t_6, 
#                          "CD27-_gdT_il27raKO" = t_7), paste0(tab_path, "NK_T_il27rako_vs_Other_NK_T.xlsx"))


all_integrated_rnm_VlnPlot$group <- factor(all_integrated_rnm_VlnPlot$group, levels = c("wt", "il27raKO"))

#pdf(paste0(fig_path, "/VlnPlot_DE_NK_T_subsets_vs_otherNK_T_subsets.pdf"), width = 12, height = 8)
 
VlnPlot(all_integrated_rnm_VlnPlot, features = nk_1$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD27loCD11b+_NK_il27raKO  vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_il27raKO,\nCD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt,\nCD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt,  CD27loCD11blo_NK_wt, CD8_Nave_il27raKO,\nCD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt, CD8_TEM_il27raKO,\nCD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt, Treg_il27raKO,\nTreg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt, CD27-_gdT_il27raKO,\nCD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD27loCD11b+_NK_il27raKO and\n<50% of cells in the other NK and T cell subsets\n") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

VlnPlot(all_integrated_rnm_VlnPlot, features = nk_2$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle( "CD27loCD11bhi_NK_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO,\n CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\nCD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD27loCD11bhi_NK_il27raKO and\n<50% of cells in the other NK and T cell subsets\n") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))


VlnPlot(all_integrated_rnm_VlnPlot, features = nk_3$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD27hiCD11b+_NK_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_il27raKO,\n CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\nCD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD27hiCD11b+_NK_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

VlnPlot(all_integrated_rnm_VlnPlot, features = nk_4$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD27loCD11b+_mito_NK_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, \n CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\nCD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD27loCD11b+_mito_NK_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

VlnPlot(all_integrated_rnm_VlnPlot, features = nk_5$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD27loCD11blo_NK_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, \n CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_wt,\nCD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD27loCD11blo_NK_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))


VlnPlot(all_integrated_rnm_VlnPlot, features = t_1$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD8_Nave_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, \n CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\n CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD8_Nave_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

VlnPlot(all_integrated_rnm_VlnPlot, features = t_2$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD8_TCM_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, \n CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\n CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_wt,\n CD8_TEM_il27raKO, CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD8_TCM_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))


VlnPlot(all_integrated_rnm_VlnPlot, features = t_3$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD8_TEM_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, \n CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\n CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n  CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD8_TEM_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))


VlnPlot(all_integrated_rnm_VlnPlot, features = t_4$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD4_TEM_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, \n CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\n CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n CD8_TEM_il27raKO,  CD8_TEM_wt, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD4_TEM_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

VlnPlot(all_integrated_rnm_VlnPlot, features = t_5$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("Treg_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, \n CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\n CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n CD8_TEM_il27raKO,  CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\n Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in Treg_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))


VlnPlot(all_integrated_rnm_VlnPlot, features = t_6$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD27+_gdT_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, \n CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\n CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n CD8_TEM_il27raKO,  CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_wt,\nCD27-_gdT_il27raKO, CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD27+_gdT_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))


VlnPlot(all_integrated_rnm_VlnPlot, features = t_7$gene[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) + ggtitle("CD27-_gdT_il27raKO vs CD27loCD11b+_NK_il27raKO, CD27loCD11b+_NK_wt,\n CD27loCD11bhi_NK_il27raKO, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_il27raKO, CD27hiCD11b+_NK_wt, \n CD27loCD11b+_mito_NK_il27raKO, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_il27raKO, CD27loCD11blo_NK_wt,\n CD8_Nave_il27raKO, CD8_Nave_wt, CD8_TCM_il27raKO, CD8_TCM_wt,\n CD8_TEM_il27raKO,  CD8_TEM_wt, CD4_TEM_il27raKO, CD4_TEM_wt,\nTreg_il27raKO, Treg_wt, CD27+_gdT_il27raKO, CD27+_gdT_wt,\n CD27-_gdT_wt;\n plotting top 15 DE genes that are expressed in atleast 50% of cells in CD27-_gdT_il27raKO and\n<50% of cells in the other NK and T cell subsets") & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))

##dev.off()

```

#Comparing each NK and T cell subset with other subsets within the same group
```{r}
tmp <- all_integrated_rnm #assign seurat object to a temporary variable
Idents(tmp) <- "celltype"
tmp$group <- factor(tmp$group, levels = c( "wt", "il27raKO") )
Idents(tmp) <- "cluster0.8.group"

#il27raKO
nk_1_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_NK_il27raKO", ident.2 = c("CD27loCD11bhi_NK_il27raKO", "CD27hiCD11b+_NK_il27raKO", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11blo_NK_il27raKO") ) 

nk_1_il27raKO$gene <- row.names(nk_1_il27raKO)
nk_1_il27raKO$contrast <- paste0( "CD27loCD11b+_NK_il27raKO", "_vs_", "CD27loCD11bhi_NK_il27raKO, CD27hiCD11b+_NK_il27raKO,  CD27loCD11b+_mito_NK_il27raKO, CD27loCD11blo_NK_il27raKO")



nk_2_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11bhi_NK_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27hiCD11b+_NK_il27raKO","CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11blo_NK_il27raKO") ) 

nk_2_il27raKO$gene <- row.names(nk_2_il27raKO)
nk_2_il27raKO$contrast <- paste0( "CD27loCD11bhi_NK_il27raKO" , "_vs_", "CD27loCD11b+_NK_il27raKO, CD27hiCD11b+_NK_il27raKO, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11blo_NK_il27raKO")


nk_3_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27hiCD11b+_NK_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11bhi_NK_il27raKO", "CD27loCD11b+_mito_NK_il27raKO", "CD27loCD11blo_NK_il27raKO") ) 

nk_3_il27raKO$gene <- row.names(nk_3_il27raKO)
nk_3_il27raKO$contrast <- paste0( "CD27hiCD11b+_NK_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11bhi_NK_il27raKO, CD27loCD11b+_mito_NK_il27raKO, CD27loCD11blo_NK_il27raKO")


nk_4_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_mito_NK_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11bhi_NK_il27raKO", "CD27hiCD11b+_NK_il27raKO", "CD27loCD11blo_NK_il27raKO") ) 

nk_4_il27raKO$gene <- row.names(nk_4_il27raKO)
nk_4_il27raKO$contrast <- paste0( "CD27loCD11b+_mito_NK_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11bhi_NK_il27raKO, CD27hiCD11b+_NK_il27raKO, CD27loCD11blo_NK_il27raKO")


nk_5_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11blo_NK_il27raKO", ident.2 = c("CD27loCD11b+_NK_il27raKO", "CD27loCD11bhi_NK_il27raKO", "CD27hiCD11b+_NK_il27raKO", "CD27loCD11b+_mito_NK_il27raKO") ) 

nk_5_il27raKO$gene <- row.names(nk_5_il27raKO)
nk_5_il27raKO$contrast <- paste0( "CD27loCD11blo_NK_il27raKO", "_vs_", "CD27loCD11b+_NK_il27raKO, CD27loCD11bhi_NK_il27raKO, CD27hiCD11b+_NK_il27raKO, CD27loCD11b+_mito_NK_il27raKO")


t_1_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD8_Nave_il27raKO", ident.2 = c("CD8_TCM_il27raKO", "CD8_TEM_il27raKO", "CD4_TEM_il27raKO", "Treg_il27raKO", "CD27+_gdT_il27raKO", "CD27-_gdT_il27raKO") ) 

t_1_il27raKO$gene <- row.names(t_1_il27raKO)
t_1_il27raKO$contrast <- paste0( "CD8_Nave_il27raKO", "_vs_", "CD8_TCM_il27raKO, CD8_TEM_il27raKO, CD4_TEM_il27raKO,Treg_il27raKO,  CD27+_gdT_il27raKO, CD27-_gdT_il27raKO")



t_2_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD8_TCM_il27raKO", ident.2 = c("CD8_Nave_il27raKO", "CD8_TEM_il27raKO", "CD4_TEM_il27raKO", "Treg_il27raKO", "CD27+_gdT_il27raKO", "CD27-_gdT_il27raKO") ) 

t_2_il27raKO$gene <- row.names(t_2_il27raKO)
t_2_il27raKO$contrast <- paste0("CD8_TCM_il27raKO", "_vs_", "CD8_Nave_il27raKO, CD8_TEM_il27raKO, CD4_TEM_il27raKO, Treg_il27raKO,  CD27+_gdT_il27raKO, CD27-_gdT_il27raKO")


t_3_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD8_TEM_il27raKO" , ident.2 = c("CD8_Nave_il27raKO", "CD8_TCM_il27raKO", "CD4_TEM_il27raKO", "Treg_il27raKO", "CD27+_gdT_il27raKO", "CD27-_gdT_il27raKO") ) 

t_3_il27raKO$gene <- row.names(t_3_il27raKO)
t_3_il27raKO$contrast <- paste0("CD8_TEM_il27raKO", "_vs_", "CD8_Nave_il27raKO, CD8_TCM_il27raKO,CD4_TEM_il27raKO, Treg_il27raKO,  CD27+_gdT_il27raKO, CD27-_gdT_il27raKO")


t_4_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 =  "CD4_TEM_il27raKO", ident.2 = c("CD8_Nave_il27raKO", "CD8_TCM_il27raKO", "CD8_TEM_il27raKO", "Treg_il27raKO", "CD27+_gdT_il27raKO", "CD27-_gdT_il27raKO") ) 

t_4_il27raKO$gene <- row.names(t_4_il27raKO)
t_4_il27raKO$contrast <- paste0("CD4_TEM_il27raKO", "_vs_", "CD8_Nave_il27raKO, CD8_TCM_il27raKO, CD8_TEM_il27raKO, Treg_il27raKO, CD27+_gdT_il27raKO, CD27-_gdT_il27raKO")


t_5_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 =  "Treg_il27raKO", ident.2 = c("CD8_Nave_il27raKO", "CD8_TCM_il27raKO", "CD8_TEM_il27raKO", "CD4_TEM_il27raKO", "CD27+_gdT_il27raKO", "CD27-_gdT_il27raKO") ) 

t_5_il27raKO$gene <- row.names(t_5_il27raKO)
t_5_il27raKO$contrast <- paste0("Treg_il27raKO", "_vs_", "CD8_Nave_il27raKO, CD8_TCM_il27raKO, CD8_TEM_il27raKO, CD4_TEM_il27raKO, CD27+_gdT_il27raKO, CD27-_gdT_il27raKO")


t_6_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 =  "CD27+_gdT_il27raKO", ident.2 = c("CD8_Nave_il27raKO", "CD8_TCM_il27raKO", "CD8_TEM_il27raKO", "CD4_TEM_il27raKO", "Treg_il27raKO", "CD27-_gdT_il27raKO") ) 

t_6_il27raKO$gene <- row.names(t_6_il27raKO)
t_6_il27raKO$contrast <- paste0("CD27+_gdT_il27raKO", "_vs_", "CD8_Nave_il27raKO, CD8_TCM_il27raKO,CD8_TEM_il27raKO, CD4_TEM_il27raKO, Treg_il27raKO, CD27-_gdT_il27raKO")


t_7_il27raKO <- FindMarkers(tmp, assay = "RNA", ident.1 =  "CD27-_gdT_il27raKO", ident.2 = c("CD8_Nave_il27raKO",  "CD8_TCM_il27raKO", "CD8_TEM_il27raKO", "CD4_TEM_il27raKO", "Treg_il27raKO", "CD27+_gdT_il27raKO", "CD27-_gdT_wt") ) 

t_7_il27raKO$gene <- row.names(t_7_il27raKO)
t_7_il27raKO$contrast <- paste0("CD27-_gdT_il27raKO", "_vs_", "CD8_Nave_il27raKO, CD8_TCM_il27raKO, CD8_TEM_il27raKO, CD4_TEM_il27raKO, Treg_il27raKO, CD27+_gdT_il27raKO")


#wt
nk_1_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_NK_wt", ident.2 = c("CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_wt") ) 

nk_1_wt$gene <- row.names(nk_1_wt)
nk_1_wt$contrast <- paste0( "CD27loCD11b+_NK_wt", "_vs_", "CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_wt,  CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_wt")


nk_2_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11bhi_NK_wt", ident.2 = c("CD27loCD11b+_NK_wt", "CD27hiCD11b+_NK_wt","CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_wt") ) 

nk_2_wt$gene <- row.names(nk_2_wt)
nk_2_wt$contrast <- paste0( "CD27loCD11bhi_NK_wt" , "_vs_", "CD27loCD11b+_NK_wt, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_wt")


nk_3_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27hiCD11b+_NK_wt", ident.2 = c("CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_wt", "CD27loCD11b+_mito_NK_wt", "CD27loCD11blo_NK_wt") ) 

nk_3_wt$gene <- row.names(nk_3_wt)
nk_3_wt$contrast <- paste0( "CD27hiCD11b+_NK_wt", "_vs_", "CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_wt, CD27loCD11b+_mito_NK_wt, CD27loCD11blo_NK_wt")


nk_4_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_mito_NK_wt", ident.2 = c("CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_wt", "CD27loCD11blo_NK_wt") ) 

nk_4_wt$gene <- row.names(nk_4_wt)
nk_4_wt$contrast <- paste0( "CD27loCD11b+_mito_NK_wt", "_vs_", "CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_wt, CD27loCD11blo_NK_wt")


nk_5_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11blo_NK_wt", ident.2 = c("CD27loCD11b+_NK_wt", "CD27loCD11bhi_NK_wt", "CD27hiCD11b+_NK_wt", "CD27loCD11b+_mito_NK_wt") ) 

nk_5_wt$gene <- row.names(nk_5_wt)
nk_5_wt$contrast <- paste0( "CD27loCD11blo_NK_wt", "_vs_", "CD27loCD11b+_NK_wt, CD27loCD11bhi_NK_wt, CD27hiCD11b+_NK_wt, CD27loCD11b+_mito_NK_wt")


t_1_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD8_Nave_wt", ident.2 = c("CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt", "Treg_wt", "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

t_1_wt$gene <- row.names(t_1_wt)
t_1_wt$contrast <- paste0( "CD8_Nave_wt", "_vs_", "CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt,Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")



t_2_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD8_TCM_wt", ident.2 = c("CD8_Nave_wt", "CD8_TEM_wt", "CD4_TEM_wt", "Treg_wt", "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

t_2_wt$gene <- row.names(t_2_wt)
t_2_wt$contrast <- paste0("CD8_TCM_wt", "_vs_", "CD8_Nave_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")


t_3_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD8_TEM_wt" , ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD4_TEM_wt", "Treg_wt", "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

t_3_wt$gene <- row.names(t_3_wt)
t_3_wt$contrast <- paste0("CD8_TEM_wt", "_vs_", "CD8_Nave_wt, CD8_TCM_wt,CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")


t_4_wt <- FindMarkers(tmp, assay = "RNA", ident.1 =  "CD4_TEM_wt", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "Treg_wt", "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

t_4_wt$gene <- row.names(t_4_wt)
t_4_wt$contrast <- paste0("CD4_TEM_wt", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, Treg_wt, CD27+_gdT_wt, CD27-_gdT_wt")


t_5_wt <- FindMarkers(tmp, assay = "RNA", ident.1 =  "Treg_wt", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt", "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

t_5_wt$gene <- row.names(t_5_wt)
t_5_wt$contrast <- paste0("Treg_wt", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, CD27+_gdT_wt, CD27-_gdT_wt")


t_6_wt <- FindMarkers(tmp, assay = "RNA", ident.1 =  "CD27+_gdT_wt", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt", "Treg_wt", "CD27-_gdT_wt") ) 

t_6_wt$gene <- row.names(t_6_wt)
t_6_wt$contrast <- paste0("CD27+_gdT_wt", "_vs_", "CD8_Nave_wt, CD8_TCM_wt,CD8_TEM_wt, CD4_TEM_wt, Treg_wt, CD27-_gdT_wt")


t_7_wt <- FindMarkers(tmp, assay = "RNA", ident.1 =  "CD27-_gdT_wt", ident.2 = c("CD8_Nave_wt",  "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt", "Treg_wt", "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

t_7_wt$gene <- row.names(t_7_wt)
t_7_wt$contrast <- paste0("CD27-_gdT_wt", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt, CD27+_gdT_wt")


#nk_il27raKO vs t_wt
#il27raKO
nk_1_il27raKO_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_NK_il27raKO", ident.2 = c( "CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_1_il27raKO_vs_t_wt$gene <- row.names(nk_1_il27raKO_vs_t_wt)
nk_1_il27raKO_vs_t_wt$contrast <- paste0( "CD27loCD11b+_NK_il27raKO", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")



nk_2_il27raKO_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11bhi_NK_il27raKO", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_2_il27raKO_vs_t_wt$gene <- row.names(nk_2_il27raKO_vs_t_wt)
nk_2_il27raKO_vs_t_wt$contrast <- paste0( "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")


nk_3_il27raKO_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27hiCD11b+_NK_il27raKO", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_3_il27raKO_vs_t_wt$gene <- row.names(nk_3_il27raKO_vs_t_wt)
nk_3_il27raKO_vs_t_wt$contrast <- paste0( "CD27hiCD11b+_NK_il27raKO", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")


nk_4_il27raKO_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_mito_NK_il27raKO", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_4_il27raKO_vs_t_wt$gene <- row.names(nk_4_il27raKO_vs_t_wt)
nk_4_il27raKO_vs_t_wt$contrast <- paste0( "CD27loCD11b+_mito_NK_il27raKO", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")


nk_5_il27raKO_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11blo_NK_il27raKO", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_5_il27raKO_vs_t_wt$gene <- row.names(nk_5_il27raKO_vs_t_wt)
nk_5_il27raKO_vs_t_wt$contrast <- paste0( "CD27loCD11blo_NK_il27raKO", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")


#nk_wt vs t_wt
#wt
nk_1_wt_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_NK_wt", ident.2 = c( "CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_1_wt_vs_t_wt$gene <- row.names(nk_1_wt_vs_t_wt)
nk_1_wt_vs_t_wt$contrast <- paste0( "CD27loCD11b+_NK_wt", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")



nk_2_wt_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11bhi_NK_wt", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_2_wt_vs_t_wt$gene <- row.names(nk_2_wt_vs_t_wt)
nk_2_wt_vs_t_wt$contrast <- paste0( "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")


nk_3_wt_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27hiCD11b+_NK_wt", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_3_wt_vs_t_wt$gene <- row.names(nk_3_wt_vs_t_wt)
nk_3_wt_vs_t_wt$contrast <- paste0( "CD27hiCD11b+_NK_wt", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")


nk_4_wt_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11b+_mito_NK_wt", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_4_wt_vs_t_wt$gene <- row.names(nk_4_wt_vs_t_wt)
nk_4_wt_vs_t_wt$contrast <- paste0( "CD27loCD11b+_mito_NK_wt", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")


nk_5_wt_vs_t_wt <- FindMarkers(tmp, assay = "RNA", ident.1 = "CD27loCD11blo_NK_wt", ident.2 = c("CD8_Nave_wt", "CD8_TCM_wt", "CD8_TEM_wt", "CD4_TEM_wt","Treg_wt",  "CD27+_gdT_wt", "CD27-_gdT_wt") ) 

nk_5_wt_vs_t_wt$gene <- row.names(nk_5_wt_vs_t_wt)
nk_5_wt_vs_t_wt$contrast <- paste0( "CD27loCD11blo_NK_wt", "_vs_", "CD8_Nave_wt, CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,  CD27+_gdT_wt, CD27-_gdT_wt")






#pdf(paste0(fig_path, "/VlnPlot_DE_NK_il27raKO_vs_T_wt_subsets.pdf"), width = 12, height = 8) 


x <- nk_1_il27raKO_vs_t_wt[!(nk_1_il27raKO_vs_t_wt$gene %in% nk_1_wt_vs_t_wt$gene), ]
VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 > 0.5 & x$pct.2 < 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("CD27loCD11b+_NK_il27raKO vs CD8_Nave_wt,\n CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,\nCD27+_gdT_wt, CD27-_gdT_wt\n; pct.1 > 0.5 & pct.2 < 0.5\n")

VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 < 0.5 & x$pct.2 > 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("CD27loCD11b+_NK_il27raKO vs CD8_Nave_wt,\n CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,\nCD27+_gdT_wt, CD27-_gdT_wt\n; pct.1 < 0.5 & pct.2 > 0.5\n")




x <- nk_2_il27raKO_vs_t_wt[!(nk_2_il27raKO_vs_t_wt$gene %in% nk_2_wt_vs_t_wt$gene), ]
VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 > 0.5 & x$pct.2 < 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("CD27loCD11bhi_NK_il27raKO vs CD8_Nave_wt,\n CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,\nCD27+_gdT_wt, CD27-_gdT_wt\n; pct.1 > 0.5 & pct.2 < 0.5\n")

VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 < 0.5 & x$pct.2 > 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("CD27loCD11bhi_NK_il27raKO vs CD8_Nave_wt,\n CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,\nCD27+_gdT_wt, CD27-_gdT_wt\n; pct.1 < 0.5 & pct.2 > 0.5\n")


x <- nk_3_il27raKO_vs_t_wt[!(nk_3_il27raKO_vs_t_wt$gene %in% nk_3_wt_vs_t_wt$gene), ]
VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 > 0.5 & x$pct.2 < 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("CD27hiCD11b+_NK_il27raKO vs CD8_Nave_wt,\n CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,\nCD27+_gdT_wt, CD27-_gdT_wt\n; pct.1 > 0.5 & pct.2 < 0.5\n")

#VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 < 0.5 & x$pct.2 > 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) ## No genes match this criteria



x <- nk_4_il27raKO_vs_t_wt[!(nk_4_il27raKO_vs_t_wt$gene %in% nk_4_wt_vs_t_wt$gene), ]
#VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 > 0.5 & x$pct.2 < 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF"))  ## No genes match this criteria
VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 < 0.5 & x$pct.2 > 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("CD27loCD11b+_mito_NK vs CD8_Nave_wt,\n CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,\nCD27+_gdT_wt, CD27-_gdT_wt\n; pct.1 < 0.5 & pct.2 > 0.5\n")


x <- nk_5_il27raKO_vs_t_wt[!(nk_5_il27raKO_vs_t_wt$gene %in% nk_5_wt_vs_t_wt$gene), ]
#VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 > 0.5 & x$pct.2 < 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) ## No genes match this criteria
VlnPlot(all_integrated_rnm_VlnPlot, features = x$gene[x$pct.1 < 0.5 & x$pct.2 > 0.5][1:20], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("CD27loCD11blo_NK vs CD8_Nave_wt,\n CD8_TCM_wt, CD8_TEM_wt, CD4_TEM_wt, Treg_wt,\nCD27+_gdT_wt, CD27-_gdT_wt\n; pct.1 < 0.5 & pct.2 > 0.5\n")


##dev.off()
```


#Heatmap of significant NK and T cell DE genes from muscat DS RNA analysis
```{r}
muscatDS_DE <- bind_rows(tbl_fil_all_RNA)

muscatDS_NK_T_cell_signif <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]

muscatDS_NK_T_cell_signif <- muscatDS_NK_T_cell_signif[muscatDS_NK_T_cell_signif$p_adj.loc < 0.05, ]

muscatDS_NK_T_cell_signif_DEgenes <- unique(muscatDS_NK_T_cell_signif$gene)

htmap_muscatDS <- dittoHeatmap(z, genes = muscatDS_NK_T_cell_signif_DEgenes, annot.by =  c("celltype", "group"), complex = TRUE, order.by = c("celltype", "group"))   

htmap_muscatDS@matrix_param$height <- unit(3, "mm")*nrow(htmap_muscatDS@matrix); htmap_muscatDS@matrix_param$width <- unit(5, "mm")*ncol(htmap_muscatDS@matrix)


#pdf(paste0(fig_path, "/DittoHeatmap_signifDEp_adj.loc0.05_muscatDS_NK_Tcell_different_genes_fromAggExprDE.pdf"), width = 10, height = 25)
  htmap_muscatDS
##dev.off()
  
  VlnPlot(all_integrated_rnm_VlnPlot, features = muscatDS_NK_T_cell_signif_DEgenes[1:15], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))
  
  
  #VlnPlot(all_integrated_rnm_VlnPlot, features = c("Il7r", "Iigp1", "Rsad2", "Gbp6", "Tgtp2", "Ifit3", "Ifit1", "Ifi208", "Podnl1", "Casp1", "Ptms", "Serpina3g", "P2ry14", "Ndrg1","Stom", "Sult2b1", "Xcl1", "Timp2", "Wdfy1", "Ppp1r3b", "Ifng", "Ccl5", "Casp3", "Tspan13", "Metrnl", "Dusp5", "Uhrf2", "Nt5c3", "Pycard", "Plaat3", "Gbp3", "Zfas1", "Gadd45gip1", "Prdx2", "Wdr83os", "Rad23a", "Emc3", "Asna1", "Asrgl1", "Yipf4", "Il2ra", "Furin", "Dnajc15", "Satb1", "Usp18", "Jun", "Gpr171", "Snx10", "Prnp", "Adgre5", "Fosl2", "Rbl2", "Il18r1", "Odc1", "Igtp", "Gbp2", "Ipcef1", "Txnip", "Nmi", "Irf1", "Phf11a", "Phf11b", "Tor3a", "Crbn", "Samhd1", "Stat2", "Irgm2", "Ifit2", "Gbp9", "Stat1", "Uba7", "Hmgb2", "Gzmk", "Ifitm3", "Cxcr6", "Gzmb", "Gzmc","Gzmk"), assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))
  
muscatDS_signif_NK_to_plot <-  c("P2ry14", "Ndrg1","Stom", "Sult2b1", "Timp2", "Ppp1r3b", "Metrnl", "Dusp5", "Pycard", "Wdr83os", "Asna1", "Usp18", "Gpr171", "Snx10", "Odc1") 
  
  
muscatDS_signif_NK_and_T_to_plot <- c("Ifi208", "Podnl1", "Wdfy1", "Ifng", "Gadd45gip1", "Prdx2","Rad23a", "Il2ra","Satb1", "Fosl2", "Igtp", "Gbp2", "Irf1","Stat1")
                                      
muscatDS_signif_T_to_plot <- c("Il7r", "Iigp1", "Rsad2", "Gbp6", "Tgtp2", "Ifit3", "Ifit1", "Casp1", "Ptms", "Serpina3g", "Xcl1", "Casp3", "Tspan13", "Uhrf2", "Nt5c3", "Plaat3", "Gbp3", "Zfas1", "Emc3", "Asrgl1", "Yipf4", "Furin", "Dnajc15", "Jun", "Prnp", "Adgre5", "Rbl2", "Il18r1", "Ipcef1", "Txnip", "Nmi", "Phf11a", "Phf11b", "Tor3a", "Crbn", "Samhd1", "Stat2", "Irgm2", "Ifit2", "Gbp9", "Uba7", "Hmgb2", "Gzmk", "Ifitm3", "Cxcr6")    

all_integrated_rnm_VlnPlot$group <- factor(all_integrated_rnm_VlnPlot$group, levels = c("wt", "il27raKO"))

#pdf(paste0(fig_path, "/VlnPlot_muscatDS_signifc_NK_T_cells_10Apr2023.pdf"), width = 12, height = 8) 


VlnPlot(all_integrated_rnm_VlnPlot, features = muscatDS_signif_NK_to_plot, assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("DE genes (muscatDS pseudo-bulk) significantly (adj.p.loc < 0.05)\n differentially expressed in NK cells\n")



VlnPlot(all_integrated_rnm_VlnPlot, features = muscatDS_signif_NK_and_T_to_plot, assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("DE genes (muscatDS pseudo-bulk) significantly (adj.p.loc < 0.05)\n differentially expressed in NK and T cells\n")


VlnPlot(all_integrated_rnm_VlnPlot, features = muscatDS_signif_T_to_plot[1:12], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("DE genes (muscatDS pseudo-bulk) significantly (adj.p.loc < 0.05)\n differentially expressed in T cells\n(1/4)\n")

VlnPlot(all_integrated_rnm_VlnPlot, features = muscatDS_signif_T_to_plot[13:23], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("DE genes (muscatDS pseudo-bulk) significantly (adj.p.loc < 0.05)\n differentially expressed in T cells\n(2/4)\n")


VlnPlot(all_integrated_rnm_VlnPlot, features = muscatDS_signif_T_to_plot[24:34], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("DE genes (muscatDS pseudo-bulk) significantly (adj.p.loc < 0.05)\n differentially expressed in T cells\n(3/4)\n")

VlnPlot(all_integrated_rnm_VlnPlot, features = muscatDS_signif_T_to_plot[35:45], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("DE genes (muscatDS pseudo-bulk) significantly (adj.p.loc < 0.05)\n differentially expressed in T cells\n(4/4)\n")

##dev.off()

#qp <- VlnPlot(all_integrated_rnm_VlnPlot, features = muscatDS_signif_T_to_plot[35:45], assay = "RNA", split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("DE genes (muscatDS pseudo-bulk) significantly (adj.p.loc < 0.05)\n differentially expressed in T cells\n(4/4)\n")
#qp1 <-ggpubr::add_summary(qp, "mean", group = qp$data$split)
# qp1$layers[[2]]$aes_params$shape <-  "|"
# qp1$layers[[2]]$aes_params$size <- 2
#qp1$layers[[2]]$aes_params$colour <- "black"


x <- dittoHeatmap(z, genes = c(muscatDS_signif_NK_to_plot, muscatDS_signif_NK_and_T_to_plot, muscatDS_signif_T_to_plot), annot.by =  c("celltype", "group"), complex = TRUE, order.by = c("celltype", "group"), cluster_rows = FALSE)   #order.by source code:  https://github.com/dtm2451/dittoSeq/issues/70#issuecomment-772803365

x@matrix_param$height <- unit(3, "mm")*nrow(x@matrix); x@matrix_param$width <- unit(5, "mm")*ncol(x@matrix)


#pdf(paste0(fig_path, "/DittoHeatmap_Signif_MuscatDS_DE_NK_Tcell_different_genes_fromAggExprDE_11Apr2023.pdf"), width = 10, height = 11)
  x
##dev.off()




#Heatmap using avg_log2FC values based on DE analysis
x <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"),]

x <- x[x$gene %in% c(muscatDS_signif_NK_to_plot, muscatDS_signif_NK_and_T_to_plot, muscatDS_signif_T_to_plot), ] %>% dplyr::select(cluster_id, gene, logFC)

x <- spread(x, cluster_id, logFC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_muscatDS_NK_vs_T_genes <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_muscatDS_NK_vs_T_genes) )
row.names(annotation_col) <- colnames(mat_muscatDS_NK_vs_T_genes)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_muscatDS_NK_vs_T_genes <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

row_split = rep("NK", 74); row_split[16:29] = "NK_and_T"; row_split[30:74] = "T" # splitting rows based on gene list c(muscatDS_signif_NK_to_plot, muscatDS_signif_NK_and_T_to_plot, muscatDS_signif_T_to_plot) (https://github.com/jokergoo/ComplexHeatmap/issues/884#issuecomment-1068901666)

mat_muscatDS_NK_vs_T_genes <- mat_muscatDS_NK_vs_T_genes[c(muscatDS_signif_NK_to_plot, muscatDS_signif_NK_and_T_to_plot, muscatDS_signif_T_to_plot), ] #reorder matrix rows based on order of the gene list genes


#create the actual heatmap object
hmap_mat_muscatDS_NK_vs_T_genes <- ComplexHeatmap::Heatmap(mat_muscatDS_NK_vs_T_genes,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
    row_order = row.names(mat_muscatDS_NK_vs_T_genes),
    row_split = row_split,
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = '',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_muscatDS_NK_vs_T_genes,
    height = unit(3, "mm")*nrow(mat_muscatDS_NK_vs_T_genes), width = unit(5, "mm")*ncol(mat_muscatDS_NK_vs_T_genes)
)

#relevel the heatmap legend
hmap_mat_muscatDS_NK_vs_T_genes@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")

#draw the heatmap
#pdf(paste0(fig_path, "logFC_signif_muscatDS_DE_NK_Tcell.pdf"), width = 8, height = 11)
hmap_mat_muscatDS_NK_vs_T_genes
#dev.off()



#enrichment of genes from each gene set c(muscatDS_signif_NK_to_plot, muscatDS_signif_NK_and_T_to_plot, muscatDS_signif_T_to_plot)

enrichgoBP_muscatDS_signif_NK <- clusterProfiler::enrichGO(gene = unique(bitr(muscatDS_signif_NK_to_plot, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                       #universe = background.genes,
                                       OrgDb = "org.Mm.eg.db",
                                       readable = TRUE,
                                       keyType = "ENTREZID",
                                       ont = "BP",
                                       pvalueCutoff = 0.05,
                                       pAdjustMethod = "BH"
                                                     )


enrichgoBP_muscatDS_signif_NK_and_T <- clusterProfiler::enrichGO(gene = unique(bitr(muscatDS_signif_NK_and_T_to_plot, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                       #universe = background.genes,
                                       OrgDb = "org.Mm.eg.db",
                                       readable = TRUE,
                                       keyType = "ENTREZID",
                                       ont = "BP",
                                       pvalueCutoff = 0.05,
                                       pAdjustMethod = "BH"
                                                     )

enrichgoBP_muscatDS_signif_T <- clusterProfiler::enrichGO(gene = unique(bitr(muscatDS_signif_T_to_plot, fromType ="SYMBOL", toType = c("ENTREZID"), OrgDb = "org.Mm.eg.db"))$ENTREZID,
                                       #universe = background.genes,
                                       OrgDb = "org.Mm.eg.db",
                                       readable = TRUE,
                                       keyType = "ENTREZID",
                                       ont = "BP",
                                       pvalueCutoff = 0.05,
                                       pAdjustMethod = "BH"
                                                     )

#Setting background genes
tmp_goBP <- all_integrated_rnm; DefaultAssay(tmp_goBP) <- "RNA"

clust <- tmp_goBP [, tmp_goBP$celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") ] #subset to a particular cell type
expr_clust <- clust[Matrix::rowSums(clust[["RNA"]]@counts > 0) > 0, ] #selecting background as genes/proteins which are atleast expressed in a particular cluster/celltype
background.genes <- rownames(expr_clust)

topgoDE_muscatDS_signif_NK <-
  pcaExplorer::topGOtable(DEgenes =  muscatDS_signif_NK_to_plot,
                          BGgenes = background.genes,
                          ontology = "BP",
                          mapping = "org.Mm.eg.db",
                          geneID = "symbol",
                          topTablerows = 500)

rm(clust, expr_clust, background.genes)


clust <- tmp_goBP [, tmp_goBP$celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") ] #subset to a particular cell type
expr_clust <- clust[Matrix::rowSums(clust[["RNA"]]@counts > 0) > 0, ] #selecting background as genes/proteins which are atleast expressed in a particular cluster/celltype
background.genes <- rownames(expr_clust)
topgoDE_muscatDS_signif_NK_and_T <-
  pcaExplorer::topGOtable(DEgenes =  muscatDS_signif_NK_and_T_to_plot,
                          BGgenes = background.genes,
                          ontology = "BP",
                          mapping = "org.Mm.eg.db",
                          geneID = "symbol",
                          topTablerows = 500)

rm(clust, expr_clust, background.genes)


clust <- tmp_goBP [, tmp_goBP$celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") ] #subset to a particular cell type
expr_clust <- clust[Matrix::rowSums(clust[["RNA"]]@counts > 0) > 0, ] #selecting background as genes/proteins which are atleast expressed in a particular cluster/celltype
background.genes <- rownames(expr_clust)
topgoDE_muscatDS_signif_T <-
  pcaExplorer::topGOtable(DEgenes =  muscatDS_signif_T_to_plot,
                          BGgenes = background.genes,
                          ontology = "BP",
                          mapping = "org.Mm.eg.db",
                          geneID = "symbol",
                          topTablerows = 500)
rm(clust, expr_clust, background.genes)



#Subset topGO results to retain GO BP enriched in at least 5 genes and which have a statistical significance (p.value_elim) < 0.05
topgoDE_muscatDS_signif_NK_filt <- topgoDE_muscatDS_signif_NK[topgoDE_muscatDS_signif_NK$Significant >= 5 & topgoDE_muscatDS_signif_NK$p.value_elim < 0.05, ]

topgoDE_muscatDS_signif_NK_and_T_filt <- topgoDE_muscatDS_signif_NK_and_T[topgoDE_muscatDS_signif_NK_and_T$Significant >= 5 & topgoDE_muscatDS_signif_NK_and_T$p.value_elim < 0.05, ]

topgoDE_muscatDS_signif_T_filt <- topgoDE_muscatDS_signif_T[topgoDE_muscatDS_signif_T$Significant >= 5 & topgoDE_muscatDS_signif_T$p.value_elim < 0.05, ]



ggplot(topgoDE_muscatDS_signif_NK_filt, aes(x = Significant, y = fct_reorder(Term,  p.value_elim, .desc = TRUE ))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.05), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10))


ggplot(topgoDE_muscatDS_signif_NK_and_T_filt, aes(x = Significant, y = fct_reorder(Term,  p.value_elim, .desc = TRUE ))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.05), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + scale_x_continuous(labels = scales::label_number(accuracy = 1)  )

ggplot(topgoDE_muscatDS_signif_T_filt, aes(x = Significant, y = fct_reorder(Term,  p.value_elim, .desc = TRUE ))) + geom_bar(stat = "identity", aes(fill = p.value_elim)) + xlab("Count") + ylab("") + scale_fill_gradient(limits = c(0, 0.05), low="red", high = " blue") + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10))



x_NK_up <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") & muscatDS_DE$logFC > 0 & muscatDS_DE$p_adj.loc < 0.05,]
x_NK_down <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") & muscatDS_DE$logFC < 0 & muscatDS_DE$p_adj.loc < 0.05,]
x_T_up <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & muscatDS_DE$logFC > 0 & muscatDS_DE$p_adj.loc < 0.05,]
x_T_down <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")& muscatDS_DE$logFC < 0 & muscatDS_DE$p_adj.loc < 0.05,]

count_topgoDE_muscatDS_signif_NK_filt <- list()


for(i in 1:nrow(topgoDE_muscatDS_signif_NK_filt)){
  split_genes <- strsplit(topgoDE_muscatDS_signif_NK_filt$genes[i], "," , fixed=TRUE)
  split_genes <- unlist(split_genes)
  up_count <- 0; down_count <- 0
  for(j in 1:length(split_genes)){
    if(split_genes[j] %in%  x_NK_up$gene){
      up_count <- up_count + 1
    } else {
      down_count <- down_count + 1
    }
  }
  
  df <- data.frame(data.frame(Term = topgoDE_muscatDS_signif_NK_filt$Term[i], type = "Upregulated", Count = up_count) )
  df[2, ] <- data.frame(Term = topgoDE_muscatDS_signif_NK_filt$Term[i], type = "Downregulated", Count = down_count)
  
  count_topgoDE_muscatDS_signif_NK_filt[topgoDE_muscatDS_signif_NK_filt$Term[i]] <- list(df)
}

count_topgoDE_muscatDS_signif_NK_filt <- bind_rows(count_topgoDE_muscatDS_signif_NK_filt)
count_topgoDE_muscatDS_signif_NK_filt$p.value_elim <- topgoDE_muscatDS_signif_NK_filt$p.value_elim[match(count_topgoDE_muscatDS_signif_NK_filt$Term, topgoDE_muscatDS_signif_NK_filt$Term) ] #add p.value_elim data for reordering the plot


#ggplot(count_topgoDE_muscatDS_signif_NK_filt, aes(x = Count,  y = fct_reorder(Term, p.value_elim, .desc = TRUE )))+ geom_col(aes(fill = type)) + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + scale_fill_manual(values = c("Upregulated" = "#F8766D", "Downregulated" = "#00BFC4") ) + ylab("") + scale_x_continuous(labels = scales::label_number(accuracy = 1) , expand = c(0,0.05))
#ggplot stacked barplot source code(#https://www.datanovia.com/en/blog/how-to-create-a-ggplot-stacked-bar-chart-2/)



count_topgoDE_muscatDS_signif_NK_and_T_filt <- list()


for(i in 1:nrow(topgoDE_muscatDS_signif_NK_and_T_filt)){
  split_genes <- strsplit(topgoDE_muscatDS_signif_NK_and_T_filt$genes[i], "," , fixed=TRUE)
  split_genes <- unlist(split_genes)
  up_count <- 0; down_count <- 0
  for(j in 1:length(split_genes)){
    if(split_genes[j] %in%  c(x_NK_up$gene, x_T_up$gene )  ){
      up_count <- up_count + 1
    } else {
      down_count <- down_count + 1
    }
  }
  
  df <- data.frame(data.frame(Term = topgoDE_muscatDS_signif_NK_and_T_filt$Term[i], type = "Upregulated", Count = up_count) )
  df[2, ] <- data.frame(Term = topgoDE_muscatDS_signif_NK_and_T_filt$Term[i], type = "Downregulated", Count = down_count)
  
  count_topgoDE_muscatDS_signif_NK_and_T_filt[topgoDE_muscatDS_signif_NK_and_T_filt$Term[i]] <- list(df)
}

count_topgoDE_muscatDS_signif_NK_and_T_filt <- bind_rows(count_topgoDE_muscatDS_signif_NK_and_T_filt)
count_topgoDE_muscatDS_signif_NK_and_T_filt$p.value_elim <- topgoDE_muscatDS_signif_NK_and_T_filt$p.value_elim[match(count_topgoDE_muscatDS_signif_NK_and_T_filt$Term, topgoDE_muscatDS_signif_NK_and_T_filt$Term) ] #add p.value_elim data for reordering the plot


#ggplot(count_topgoDE_muscatDS_signif_NK_and_T_filt, aes(x = Count,  y = fct_reorder(Term, p.value_elim, .desc = TRUE )))+ geom_col(aes(fill = type)) + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + scale_fill_manual(values = c("Upregulated" = "#F8766D", "Downregulated" = "#00BFC4") ) + ylab("") + scale_x_continuous(labels = scales::label_number(accuracy = 1) , expand = c(0,0.05))
#ggplot stacked barplot source code(#https://www.datanovia.com/en/blog/how-to-create-a-ggplot-stacked-bar-chart-2/)


count_topgoDE_muscatDS_signif_T_filt <- list()


for(i in 1:nrow(topgoDE_muscatDS_signif_T_filt)){
  split_genes <- strsplit(topgoDE_muscatDS_signif_T_filt$genes[i], "," , fixed=TRUE)
  split_genes <- unlist(split_genes)
  up_count <- 0; down_count <- 0
  for(j in 1:length(split_genes)){
    if(split_genes[j] %in%  c(x_NK_up$gene, x_T_up$gene )  ){
      up_count <- up_count + 1
    } else {
      down_count <- down_count + 1
    }
  }
  
  df <- data.frame(data.frame(Term = topgoDE_muscatDS_signif_T_filt$Term[i], type = "Upregulated", Count = up_count) )
  df[2, ] <- data.frame(Term = topgoDE_muscatDS_signif_T_filt$Term[i], type = "Downregulated", Count = down_count)
  
  count_topgoDE_muscatDS_signif_T_filt[topgoDE_muscatDS_signif_T_filt$Term[i]] <- list(df)
}

count_topgoDE_muscatDS_signif_T_filt <- bind_rows(count_topgoDE_muscatDS_signif_T_filt)

count_topgoDE_muscatDS_signif_T_filt$p.value_elim <- topgoDE_muscatDS_signif_T_filt$p.value_elim[match(count_topgoDE_muscatDS_signif_T_filt$Term, topgoDE_muscatDS_signif_T_filt$Term) ] #add p.value_elim data for reordering the plot

#ggplot(count_topgoDE_muscatDS_signif_T_filt, aes(x = Count,  y = fct_reorder(Term, p.value_elim, .desc = TRUE )))+ geom_col(aes(fill = type)) + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + scale_fill_manual(values = c("Upregulated" = "#F8766D", "Downregulated" = "#00BFC4") ) + ylab("") + scale_x_continuous(expand = c(0,0.05))
#ggplot stacked barplot source code(#https://www.datanovia.com/en/blog/how-to-create-a-ggplot-stacked-bar-chart-2/)




#pdf(paste0(fig_path, "topGO_enrichment_signif_muscatDS_DE_NK.pdf"), width = 8, height = 6)

ggplot(count_topgoDE_muscatDS_signif_NK_filt, aes(x = Count,  y = fct_reorder(Term, p.value_elim, .desc = TRUE )))+ geom_col(aes(fill = type)) + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + scale_fill_manual(values = c("Upregulated" = "#F8766D", "Downregulated" = "#00BFC4") ) + ylab("") + scale_x_continuous(labels = scales::label_number(accuracy = 1) , expand = c(0,0.05)) + ggtitle("Barplot showing signficant topGO BP\n enrichment for muscatDS DE genes significant in NK cells\n")
#ggplot stacked barplot source code(#https://www.datanovia.com/en/blog/how-to-create-a-ggplot-stacked-bar-chart-2/)
#dev.off()

#pdf(paste0(fig_path, "topGO_enrichment_signif_muscatDS_DE_NK_Tcell.pdf"), width = 8, height = 6)

ggplot(count_topgoDE_muscatDS_signif_NK_and_T_filt, aes(x = Count,  y = fct_reorder(Term, p.value_elim, .desc = TRUE )))+ geom_col(aes(fill = type)) + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + scale_fill_manual(values = c("Upregulated" = "#F8766D", "Downregulated" = "#00BFC4") ) + ylab("") + scale_x_continuous(labels = scales::label_number(accuracy = 1) , expand = c(0,0.05)) + ggtitle("Barplot showing signficant topGO BP\n enrichment for muscatDS DE genes significant\n in NK and T cells\n")
#ggplot stacked barplot source code(#https://www.datanovia.com/en/blog/how-to-create-a-ggplot-stacked-bar-chart-2/)
#dev.off()

#pdf(paste0(fig_path, "topGO_enrichment_signif_muscatDS_DE_Tcell.pdf"), width = 9, height = 12)

ggplot(count_topgoDE_muscatDS_signif_T_filt, aes(x = Count,  y = fct_reorder(Term, p.value_elim, .desc = TRUE )))+ geom_col(aes(fill = type)) + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + scale_fill_manual(values = c("Upregulated" = "#F8766D", "Downregulated" = "#00BFC4") ) + ylab("") + scale_x_continuous(expand = c(0,0.05)) + ggtitle("Barplot showing signficant topGO BP\n enrichment for muscatDS DE genes significant in T cells\n")
#ggplot stacked barplot source code(#https://www.datanovia.com/en/blog/how-to-create-a-ggplot-stacked-bar-chart-2/)


#dev.off()


x_temp_Seu <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] %>% dplyr::select(gene, cell_type, avg_log2FC)

x_temp_muscTDS <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] %>% dplyr::select(gene, cluster_id, logFC)

x_temp_Seu_muscatDS <- merge(x= x_temp_Seu, y = x_temp_muscTDS, by.x = c("gene", "cell_type" ), by.y = c("gene", "cluster_id")  )
              
x_temp_Seu_muscatDS$DE_Status <- ifelse(sign(x_temp_Seu_muscatDS$avg_log2FC) == sign(x_temp_Seu_muscatDS$logFC), "same", "different" )
                      
```


#Upd 12Apr2023: Based on discussion with Prof. Bosmann, checking the IFNGamma response in alveolar macrophages from muscatDS results
```{r}
#Setting background genes
tmp_goBP <- all_integrated_rnm; DefaultAssay(tmp_goBP) <- "RNA"

clust <- tmp_goBP [, tmp_goBP$celltype %in% c("Alveolar_macrophage") ] #subset to a particular cell type
expr_clust <- clust[Matrix::rowSums(clust[["RNA"]]@counts > 0) > 0, ] #selecting background as genes/proteins which are atleast expressed in a particular cluster/celltype
background.genes <- rownames(expr_clust)


muscatDS_signif_AM <- muscatDS_DE$gene[muscatDS_DE$cluster_id %in% c("Alveolar_macrophage") & muscatDS_DE$p_adj.loc < 0.05]

x_AM_up <-  muscatDS_DE[muscatDS_DE$cluster_id %in% c("Alveolar_macrophage") & muscatDS_DE$p_adj.loc < 0.05 & muscatDS_DE$logFC > 0, ]

x_AM_down <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("Alveolar_macrophage") & muscatDS_DE$p_adj.loc < 0.05 & muscatDS_DE$logFC < 0, ]

topgoDE_muscatDS_signif_AM <-
  pcaExplorer::topGOtable(DEgenes =  muscatDS_signif_AM,
                          BGgenes = background.genes,
                          ontology = "BP",
                          mapping = "org.Mm.eg.db",
                          geneID = "symbol",
                          topTablerows = 500)

rm(clust, expr_clust, background.genes)


#Subset topGO results to retain GO BP enriched in at least 2 genes and which have a statistical significance (p.value_elim) < 0.05
topgoDE_muscatDS_signif_AM_filt <- topgoDE_muscatDS_signif_AM[topgoDE_muscatDS_signif_AM$Significant >= 2 & topgoDE_muscatDS_signif_AM$p.value_elim < 0.05, ]


count_topgoDE_muscatDS_signif_AM_filt <- list()


for(i in 1:nrow(topgoDE_muscatDS_signif_AM_filt)){
  split_genes <- strsplit(topgoDE_muscatDS_signif_AM_filt$genes[i], "," , fixed=TRUE)
  split_genes <- unlist(split_genes)
  up_count <- 0; down_count <- 0
  for(j in 1:length(split_genes)){
    if(split_genes[j] %in%  c(x_AM_up$gene )  ){
      up_count <- up_count + 1
    } else {
      down_count <- down_count + 1
    }
  }
  
  df <- data.frame(data.frame(Term = topgoDE_muscatDS_signif_AM_filt$Term[i], type = "Upregulated", Count = up_count) )
  df[2, ] <- data.frame(Term = topgoDE_muscatDS_signif_AM_filt$Term[i], type = "Downregulated", Count = down_count)
  
  count_topgoDE_muscatDS_signif_AM_filt[topgoDE_muscatDS_signif_AM_filt$Term[i]] <- list(df)
}

count_topgoDE_muscatDS_signif_AM_filt <- bind_rows(count_topgoDE_muscatDS_signif_AM_filt)

count_topgoDE_muscatDS_signif_AM_filt$p.value_elim <- topgoDE_muscatDS_signif_AM_filt$p.value_elim[match(count_topgoDE_muscatDS_signif_AM_filt$Term, topgoDE_muscatDS_signif_AM_filt$Term) ] #add p.value_elim data for reordering the plot




#pdf(paste0(fig_path, "topGO_enrichment_signif_muscatDS_DE_AM.pdf"), width = 9, height = 15)

ggplot(count_topgoDE_muscatDS_signif_AM_filt, aes(x = Count,  y = fct_reorder(Term, p.value_elim, .desc = TRUE )))+ geom_col(aes(fill = type)) + theme_bw() + theme(axis.text = element_text(color = "black", size = 10), panel.grid = element_blank(), axis.title = element_text(color = "black", size = 10), legend.text = element_text(color = "black", size = 10), legend.title = element_text(color = "black", size = 10)) + scale_fill_manual(values = c("Upregulated" = "#F8766D", "Downregulated" = "#00BFC4") ) + ylab("") + scale_x_continuous(labels = scales::label_number(accuracy = 1) , expand = c(0,0.05)) + scale_y_discrete(labels = scales::label_wrap(50)) + ggtitle("Barplot showing signficant topGO BP\n enrichment for muscatDS DE genes significant in AM\nEnrichment terms with at least 2 genes enriched are depicted\n")
#ggplot stacked barplot source code(#https://www.datanovia.com/en/blog/how-to-create-a-ggplot-stacked-bar-chart-2/)
#dev.off()

#Heatmap using avg_log2FC values based on DE analysis
x <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT", "Alveolar_macrophage"),]

x <- x[x$gene %in% c("Casp1", "Gbp3", "Gbp6", "Gbp9", "Ifitm3", "Irgm2", "Nmi", "Xcl1", "Gbp2", "Ifng", "Igtp", "Irf1", "Stat1"), ] %>% dplyr::select(cluster_id, gene, logFC) #subset to genes enriched in "cellular response to interferon-gamma"/ "response to interferon-gamma" in muscatDS DE enrichment for T cells and combined NK and T cells

x <- spread(x, cluster_id, logFC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT", "Alveolar_macrophage")]


mat_muscatDS_AM_IFNg_response_genes <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_muscatDS_AM_IFNg_response_genes) )
row.names(annotation_col) <- colnames(mat_muscatDS_AM_IFNg_response_genes)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF", "Alveolar_macrophage" =  "#005685FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_muscatDS_AM_IFNg_response_genes <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_muscatDS_AM_IFNg_response_genes <- ComplexHeatmap::Heatmap(mat_muscatDS_AM_IFNg_response_genes,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of IFNgamma response genes from NK and T cell\nmuscatDS pseudo-bulk RNA DE enrichment\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_muscatDS_AM_IFNg_response_genes,
    height = unit(3, "mm")*nrow(mat_muscatDS_AM_IFNg_response_genes), width = unit(5, "mm")*ncol(mat_muscatDS_AM_IFNg_response_genes)
)

#draw the heatmap
#pdf(paste0(fig_path, "logFC_signif_muscatDS_DE_AM_IFNGammaResponse.pdf"), width = 8, height = 5)
hmap_mat_muscatDS_AM_IFNg_response_genes
#dev.off()


```

# Heatmap of topGO BP enriched in NK and T cells from muscat DS DE genes
```{r}
#Heatmap using avg_log2FC values based on DE analysis
x <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT", "Alveolar_macrophage"),]

genes_t_plot_topgoDE_muscatDS_signif_NK_T_T_filt <- unique(c( unique(unlist(strsplit( topgoDE_muscatDS_signif_T_filt$genes[topgoDE_muscatDS_signif_T_filt$Term %in% c("cellular response to interferon-beta", "cellular response to interferon-gamma", "innate immune response", "immune response", "inflammatory response", "response to bacterium", "response to interferon-gamma")], "," )) ),  unique(unlist(strsplit(topgoDE_muscatDS_signif_NK_and_T_filt$genes[topgoDE_muscatDS_signif_NK_and_T_filt$Term %in% c("cellular response to interferon-beta", "cellular response to interferon-gamma")], "," ))) ) )

x <- x[x$gene %in% genes_t_plot_topgoDE_muscatDS_signif_NK_T_T_filt, ] %>% dplyr::select(cluster_id, gene, logFC) #subset to genes enriched in "cellular response to interferon-gamma"/ "response to interferon-gamma" in muscatDS DE enrichment for T cells and combined NK and T cells

x <- spread(x, cluster_id, logFC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT", "Alveolar_macrophage")]


mat_muscatDS_IFN_Immun_Inflamm_genes <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_muscatDS_IFN_Immun_Inflamm_genes) )
row.names(annotation_col) <- colnames(mat_muscatDS_IFN_Immun_Inflamm_genes)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF", "Alveolar_macrophage" =  "#005685FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_muscatDS_IFN_Immun_Inflamm_genes <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_muscatDS_IFN_Immun_Inflamm_genes <- ComplexHeatmap::Heatmap(mat_muscatDS_IFN_Immun_Inflamm_genes,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of genes enriched in NK and T and T cell specific\nmuscatDS pseudo-bulk RNA DE enrichment\nin topGO BP: cellular response to interferon-beta [GO:0035458]\n, cellular response to interferon-gamma [GO:0071346]\n, response to interferon-gamma [GO:0034341]\n, innate immune response [GO:0045087]\n, immune response [GO:0006955]\n, inflammatory response [GO:0006954]\n, response to bacterium [GO:0009617]\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_muscatDS_IFN_Immun_Inflamm_genes,
    height = unit(3, "mm")*nrow(mat_muscatDS_IFN_Immun_Inflamm_genes), width = unit(5, "mm")*ncol(mat_muscatDS_IFN_Immun_Inflamm_genes)
)

#draw the heatmap
#pdf(paste0(fig_path, "logFC_signif_muscatDS_DE_topGO_NK_and_T_T_Ifn_Immun_Innfla_Resp.pdf"), width = 8, height = 7)
hmap_mat_muscatDS_IFN_Immun_Inflamm_genes
#dev.off()
```


```{r}
#NK up and T down
muscatDS_NK_up_signifGenes <- unique(muscatDS_DE$gene[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") & muscatDS_DE$p_adj.loc < 0.05 & muscatDS_DE$logFC > 0] )

muscatDS_T_down_signifGenes <-unique( muscatDS_DE$gene[muscatDS_DE$cluster_id %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & muscatDS_DE$p_adj.loc < 0.05 & muscatDS_DE$logFC < 0] )

muscatDS_NK_up_T_down_signif <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & muscatDS_DE$gene %in% c(muscatDS_NK_up_signifGenes, muscatDS_T_down_signifGenes), ] %>% dplyr::select(cluster_id, gene, logFC)


x <- spread(muscatDS_NK_up_T_down_signif, cluster_id, logFC)
row.names(x) <- x$gene


x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_muscatDS_NK_up_T_down_signif <- as.matrix(x)
mat_muscatDS_NK_up_T_down_signif <- mat_muscatDS_NK_up_T_down_signif[c(muscatDS_NK_up_signifGenes, muscatDS_T_down_signifGenes), ] #reorder matrix rows based on order of the gene list genes

row_split = rep("NK", 134); row_split[53:134] = "T" # splitting rows based on gene list c(muscatDS_NK_up_signifGenes, muscatDS_T_down_signifGenes) (https://github.com/jokergoo/ComplexHeatmap/issues/884#issuecomment-1068901666)



#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_muscatDS_NK_up_T_down_signif) )
row.names(annotation_col) <- colnames(mat_muscatDS_NK_up_T_down_signif)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_muscatDS_NK_up_T_down_signif <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_muscatDS_NK_up_T_down_signif <- ComplexHeatmap::Heatmap(mat_muscatDS_NK_up_T_down_signif,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
    row_order = row.names(mat_muscatDS_NK_up_T_down_signif),
    row_split = row_split,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of genes muscatDS pseudo-bulk RNA DE genes that are\nsignificantly (p_adj.loc < 0.05) upregulated in NK cells or\ndownregulated in T cells\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_muscatDS_NK_up_T_down_signif,
    height = unit(3, "mm")*nrow(mat_muscatDS_NK_up_T_down_signif), width = unit(5, "mm")*ncol(mat_muscatDS_NK_up_T_down_signif)
)

#draw the heatmap
#pdf(paste0(fig_path, "logFC_signif_muscatDS_DE_NK_up_T_down.pdf"), width = 8, height = 20)
hmap_mat_muscatDS_NK_up_T_down_signif
#dev.off()

#============================================

# NK up and CD4 TEM down
muscatDS_NK_up_signifGenes <- unique(muscatDS_DE$gene[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK") & muscatDS_DE$p_adj.loc < 0.05 & muscatDS_DE$logFC > 0] )

muscatDS_CD4T_down_signifGenes <-unique( muscatDS_DE$gene[muscatDS_DE$cluster_id %in% c("CD4_TEM") & muscatDS_DE$p_adj.loc < 0.05 & muscatDS_DE$logFC < 0] )

muscatDS_NK_up_CD4T_down_signif <- muscatDS_DE[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & muscatDS_DE$gene %in% c(muscatDS_NK_up_signifGenes, muscatDS_CD4T_down_signifGenes), ] %>% dplyr::select(cluster_id, gene, logFC)


x <- spread(muscatDS_NK_up_CD4T_down_signif, cluster_id, logFC)
row.names(x) <- x$gene


x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_muscatDS_NK_up_CD4T_down_signif <- as.matrix(x)
mat_muscatDS_NK_up_CD4T_down_signif <- mat_muscatDS_NK_up_CD4T_down_signif[c(muscatDS_NK_up_signifGenes, muscatDS_CD4T_down_signifGenes), ] #reorder matrix rows based on order of the gene list genes

row_split = rep("NK", 75); row_split[53:75] = "CD4 T" # splitting rows based on gene list c(muscatDS_NK_up_signifGenes, muscatDS_T_down_signifGenes) (https://github.com/jokergoo/ComplexHeatmap/issues/884#issuecomment-1068901666)



#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_muscatDS_NK_up_CD4T_down_signif) )
row.names(annotation_col) <- colnames(mat_muscatDS_NK_up_CD4T_down_signif)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_muscatDS_NK_up_CD4T_down_signif <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_muscatDS_NK_up_CD4T_down_signif <- ComplexHeatmap::Heatmap(mat_muscatDS_NK_up_CD4T_down_signif,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
    row_order = row.names(mat_muscatDS_NK_up_CD4T_down_signif),
    row_split = row_split,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of genes muscatDS pseudo-bulk RNA DE genes that are\nsignificantly (p_adj.loc < 0.05) upregulated in NK cells or\ndownregulated in CD4 TEM cells\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_muscatDS_NK_up_CD4T_down_signif,
    height = unit(3, "mm")*nrow(mat_muscatDS_NK_up_CD4T_down_signif), width = unit(5, "mm")*ncol(mat_muscatDS_NK_up_CD4T_down_signif)
)

#draw the heatmap
#pdf(paste0(fig_path, "logFC_signif_muscatDS_DE_NK_up_CD4T_down.pdf"), width = 8, height = 15)
hmap_mat_muscatDS_NK_up_CD4T_down_signif
#dev.off()


```














```{r}
#Pseudobulk analysis using Libra (https://github.com/neurorestore/Libra)
library(Libra)
tmp <- all_integrated_rnm
DefaultAssay(tmp) <- "RNA"
tmp$group <- factor(tmp$group, levels = c("il27raKO", "wt")) #the second factor is used as reference level, and DE is calculated as factor A vs factor B (Source: https://github.com/neurorestore/Libra/issues/37#issuecomment-1458899877)
DE_RNA = run_de(tmp, meta = tmp@meta.data, cell_type_col = "celltype", label_col = "group", replicate_col = "sample_id", de_family = "pseudobulk", de_method = "DESeq2", de_type = "Wald")

DefaultAssay(tmp) <- "ADT"
tmp$group <- factor(tmp$group, levels = c("il27raKO", "wt")) #the second factor is used as reference level, and DE is calculated as factor A vs factor B (Source: https://github.com/neurorestore/Libra/issues/37#issuecomment-1458899877)
DE_ADT = run_de(tmp, meta = tmp@meta.data, cell_type_col = "celltype", label_col = "group", replicate_col = "sample_id", de_family = "pseudobulk", de_method = "DESeq2", de_type = "Wald")
```




```{r}
#c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")
tmp_NK <- subset(all_integrated_rnm, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"))
DefaultAssay(tmp_NK) <- "RNA"
#https://www.informatics.jax.org/vocab/gene_ontology/GO:0002729: positive regulation of natural killer cell cytokine production
NK_cytokine_posReg <- list("Cd160", "Cd226", "Clnk", "Gimap3", "Gimap5", "H2-Bl", "H2-D1", "H2-K1", "H2-L", "H2-M1", "H2-M2", "H2-M5", "H2-M9", "H2-M10.1", "H2-M10.2", "H2-M10.3", "H2-M10.4", "H2-M10.5", "H2-M10.6", "H2-M11", "H2-Q1", "H2-Q2", "H2-Q4", "H2-Q6", "H2-Q7", "H2-Q10", "H2-T3", "H2-T10", "H2-T22", "H2-T23", "H2-T-ps", "Il21")

#Filter NK_cytokine_posReg to only include genes that are DE in NK cells in muscatDS RNA DE analysis
NK_cytokine_posReg_filt <- NK_cytokine_posReg[NK_cytokine_posReg %in% muscatDS_DE$gene[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")]]


tmp_NK <- AddModuleScore(tmp_NK, features = list(NK_cytokine_posReg_filt), name = "NK_cytokine")

qp1 <- VlnPlot(tmp_NK, features = c("NK_cytokine1"), split.by = "group", pt.size = 0.0); ggpubr::add_summary(qp1, "mean", group = qp1$data$split)

#https://www.informatics.jax.org/go/term/GO:0045954: positive regulation of natural killer cell mediated cytotoxicity

NK_cytotox_posReg <- list("Ap1g1", "Cadm1", "Cd160", "Cd226", "Crtam", "Gimap3", "Gimap5", "H2-Bl", "H2-D1", "H2-K1", "H2-L", "H2-M1", "H2-M2", "H2-M3", "H2-M5", "H2-M9", "H2-M10.1", "H2-M10.2", "H2-M10.3", "H2-M10.4", "H2-M10.5", "H2-M10.6", "H2-M11", "H2-Q1", "H2-Q2", "H2-Q4", "H2-Q6", "H2-Q7", "H2-Q10", "H2-T3", "H2-T10", "H2-T22", "H2-T23", "H2-T-ps", "H60a", "Il12a", "Il12b", "Il18rap", "Il21", "Klrb1c", "Klrc1", "Klrc2", "Klrc3", "Klrd1", "Klre1", "Klri1", "Klri2", "Klrk1", "Lag3", "Lamp1", "Ncr3-ps", "Nectin2", "Pvr", "Raet1b", "Rasgrp1", "Sh2d1a", "Sh2d1b1", "Sh2d1b2", "Slamf6", "Stat5a", "Stat5b", "Ulbp1", "Vav1")

#Filter NK_cytotox_posReg to only include genes that are DE in NK cells in muscatDS RNA DE analysis
NK_cytotox_posReg_filt <- NK_cytotox_posReg[NK_cytotox_posReg %in% muscatDS_DE$gene[muscatDS_DE$cluster_id %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK")]]

tmp_NK <- AddModuleScore(tmp_NK, features = list(NK_cytotox_posReg_filt), name = "NK_cytotoxic")

qp2 <- VlnPlot(tmp_NK, features = c("NK_cytotoxic1"), split.by = "group", pt.size = 0.0); ggpubr::add_summary(qp2, "mean", group = qp2$data$split)


#https://www.informatics.jax.org/go/term/GO:0002726: positive regulation of T cell cytokine production

#c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")
tmp_T <- subset(all_integrated_rnm, celltype %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))
DefaultAssay(tmp_T) <- "RNA"
#https://www.informatics.jax.org/vocab/gene_ontology/GO:0002729: positive regulation of natural killer cell cytokine production
T_cytokine_posReg <- list("Arid5a", "B2m", "Cd55", "Cd55b", "Cd81", "Dennd1b", "Fzd5", "Gata3", "Il1b", "Il1r1", "Il4", "Il6", "Il18", "Il18r1", "Malt1", "Map3k7", "Nlrp3", "Prkcz", "Rsad2", "Sash3", "Slamf1", "Tbx21", "Tnfsf4", "Traf2", "Traf6", "Xcl1")

#Filter T_cytokine_posReg to only include genes that are DE in NK cells in muscatDS RNA DE analysis
T_cytokine_posReg_filt <- T_cytokine_posReg[T_cytokine_posReg %in% muscatDS_DE$gene[muscatDS_DE$cluster_id %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]]


tmp_T <- AddModuleScore(tmp_T, features = list(T_cytokine_posReg_filt), name = "T_cytokine")

qp3 <- VlnPlot(tmp_T, features = c("T_cytokine1"), split.by = "group", pt.size = 0.0); ggpubr::add_summary(qp3, "mean", group = qp3$data$split)


#https://www.informatics.jax.org/vocab/gene_ontology/GO:0001916: positive regulation of T cell mediated cytotoxicity
T_cytotox_posReg <- list("2410137M14Rik", "B2m", "Cd1d1", "Cd1d2", "Cyrib", "Fadd", "Gm7030", "Gm8909", "Gm11127", "H2-Bl", "H2-D1", "H2-Ea", "H2-K1", "H2-L", "H2-M1", "H2-M2", "H2-M3", "H2-M5", "H2-M9", "H2-M10.1", "H2-M10.2", "H2-M10.3", "H2-M10.4", "H2-M10.5", "H2-M10.6", "H2-M11", "H2-Q1", "H2-Q2", "H2-Q4", "H2-Q6", "H2-Q7", "H2-Q8", "H2-Q9", "H2-Q10", "H2-T3", "H2-T10", "H2-T22", "H2-T23", "H2-T24", "H2-T-ps", "H60b", "H60c", "Hfe", "Il12a", "Il12b", "Il23a", "Mill1", "Mill2", "Mr1", "Nectin2", "P2rx7", "Pnp", "Ptprc", "Pvr", "Raet1a", "Raet1b", "Raet1c", "Raet1d", "Raet1e", "Slc22a13", "Stx7", "Tap2", "Ulbp1")

#Filter T_cytotox_posReg to only include genes that are DE in NK cells in muscatDS RNA DE analysis
T_cytotox_posReg_filt <- T_cytotox_posReg[T_cytotox_posReg %in% muscatDS_DE$gene[muscatDS_DE$cluster_id %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]]

tmp_T <- AddModuleScore(tmp_T, features = list(T_cytotox_posReg_filt), name = "T_cytotoxic")

qp4 <- VlnPlot(tmp_T, features = c("T_cytotoxic1"), split.by = "group", pt.size = 0.0); ggpubr::add_summary(qp4, "mean", group = qp4$data$split)

tmp_NK_T <- subset(all_integrated_rnm, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))
DefaultAssay(tmp_NK_T) <- "RNA"
tmp_NK_T <- AddModuleScore(tmp_NK_T, features = list(NK_cytokine_posReg_filt), name = "NK_cytokine")
tmp_NK_T <- AddModuleScore(tmp_NK_T, features = list(NK_cytotox_posReg_filt), name = "NK_cytotoxic")
tmp_NK_T <- AddModuleScore(tmp_NK_T, features = list(T_cytokine_posReg_filt), name = "T_cytokine")
tmp_NK_T <- AddModuleScore(tmp_NK_T, features = list(T_cytotox_posReg_filt), name = "T_cytotoxic")

qp5 <- VlnPlot(tmp_NK_T, features = c("NK_cytokine1", "NK_cytotoxic1", "T_cytokine1", "T_cytotoxic1"), split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE); qp5 <- ggpubr::add_summary(qp5, "mean", group = qp5$data$split)
# qp5$layers[[2]]$aes_params$shape <-  "|"
# qp5$layers[[2]]$aes_params$size <- 2

```


```{r}

#pdf(paste0(fig_path, "VlnPlot_IL2_signaling.pdf"), width = 8, height = 6)

VlnPlot(all_integrated_rnm_VlnPlot, features = c("rna_Prnp", "rna_Ifitm3", "rna_Cd44", "rna_Il2ra", "adt-CD25"), split.by = "group", pt.size = 0.0, stack = TRUE, same.y.lims = TRUE, idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")) & theme(axis.text.x.bottom = element_text(size = 14), axis.title.y = element_blank(), axis.text.x.top = element_text(size = 14, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 14),  legend.title = element_text(size=rel(1)), legend.text = element_text(size = 14), plot.title = element_text(size = 8)) & scale_fill_manual(values = c("wt" = "#05BE78", "il27raKO" = "#0000FF")) & ggtitle("Violin plot of IL2-IL2R signaling genes\n")

#dev.off()
```

#Heatmap of GO BP: immune response (GO:0006955) enriched in CD4_TEM Seurat DE
```{r}
genes_to_plot_topgoDE_SeuratDE_ImmunResp <- unique(unlist(strsplit( topgoDE_cluster$CD4_TEM$genes[topgoDE_cluster$CD4_TEM$Term %in% c("immune response")], "," )) )

x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] #subset to NK and T cell clusters

x <- x[x$gene %in% genes_to_plot_topgoDE_SeuratDE_ImmunResp, ] %>% dplyr::select(cell_type, gene, avg_log2FC) #retrieve log2FC for genes enriched in "immune response" (GO:0006955) in CD4_TEM cluster in Seurat DE

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_SeuratDE_ImmunResp <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_SeuratDE_ImmunResp) )
row.names(annotation_col) <- colnames(mat_SeuratDE_ImmunResp)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF", "Alveolar_macrophage" =  "#005685FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_SeuratDE_ImmunResp <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_SeuratDE_ImmunResp <- ComplexHeatmap::Heatmap(mat_SeuratDE_ImmunResp,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of log2FoldChange in NK and T and T cell of genes enriched in\n"immune response" (GO:0006955) in CD4_TEM cluster\n inSeurat DE\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_SeuratDE_ImmunResp,
    height = unit(3, "mm")*nrow(mat_SeuratDE_ImmunResp), width = unit(5, "mm")*ncol(mat_SeuratDE_ImmunResp)
)

hmap_mat_SeuratDE_ImmunResp@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") #reorder color legend

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SeuratDE_topGO_NK_and_T_T_ImmuneResponse.pdf"), width = 8, height = 7)
hmap_mat_SeuratDE_ImmunResp
#dev.off()

```

#Heatmap of GO BP: immune system process (GO:0002376) enriched in CD4_TEM Seurat DE
```{r}
genes_to_plot_topgoDE_SeuratDE_ImmunSystProc <- unique(unlist(strsplit( topgoDE_cluster$CD4_TEM$genes[topgoDE_cluster$CD4_TEM$Term %in% c("immune system process")], "," )) )

x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] #subset to NK and T cell clusters

x <- x[x$gene %in% genes_to_plot_topgoDE_SeuratDE_ImmunSystProc, ] %>% dplyr::select(cell_type, gene, avg_log2FC) #retrieve log2FC for genes enriched in immune system process (GO:0002376) in CD4_TEM cluster in Seurat DE

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_SeuratDE_ImmunSystProc <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_SeuratDE_ImmunSystProc) )
row.names(annotation_col) <- colnames(mat_SeuratDE_ImmunSystProc)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF", "Alveolar_macrophage" =  "#005685FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_SeuratDE_ImmunSystProc <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_SeuratDE_ImmunSystProc <- ComplexHeatmap::Heatmap(mat_SeuratDE_ImmunSystProc,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of log2FoldChange in NK and T and T cell of genes enriched in\n"immune system process" (GO:0002376) in CD4_TEM cluster\n inSeurat DE\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_SeuratDE_ImmunSystProc,
    height = unit(3, "mm")*nrow(mat_SeuratDE_ImmunSystProc), width = unit(5, "mm")*ncol(mat_SeuratDE_ImmunSystProc)
)

hmap_mat_SeuratDE_ImmunSystProc@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") #reorder color legend

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SeuratDE_topGO_NK_and_T_T_ImmuneSystemProcess.pdf"), width = 8, height = 7)
hmap_mat_SeuratDE_ImmunSystProc
#dev.off()
```


#Heatmap of GO BP: lymphocyte proliferation (GO:0046651) enriched in CD4_TEM Seurat DE
```{r}
genes_to_plot_topgoDE_SeuratDE_lymphoProlif <- unique(unlist(strsplit( topgoDE_cluster$CD4_TEM$genes[topgoDE_cluster$CD4_TEM$Term %in% c("lymphocyte proliferation")], "," )) )

x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] #subset to NK and T cell clusters

x <- x[x$gene %in% genes_to_plot_topgoDE_SeuratDE_lymphoProlif, ] %>% dplyr::select(cell_type, gene, avg_log2FC) #retrieve log2FC for genes enriched in lymphocyte proliferation (GO:0046651) in CD4_TEM cluster in Seurat DE

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_SeuratDE_lymphoProlif <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_SeuratDE_lymphoProlif) )
row.names(annotation_col) <- colnames(mat_SeuratDE_lymphoProlif)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF", "Alveolar_macrophage" =  "#005685FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_SeuratDE_lymphoProlif <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_SeuratDE_lymphoProlif <- ComplexHeatmap::Heatmap(mat_SeuratDE_lymphoProlif,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of log2FoldChange in NK and T and T cell of genes enriched in\n"lymphocyte proliferation" (GO:0046651) in CD4_TEM cluster\n inSeurat DE\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_SeuratDE_lymphoProlif,
    height = unit(3, "mm")*nrow(mat_SeuratDE_lymphoProlif), width = unit(5, "mm")*ncol(mat_SeuratDE_lymphoProlif)
)

hmap_mat_SeuratDE_lymphoProlif@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") #reorder color legend

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SeuratDE_topGO_NK_and_T_T_LymphocyteProliferation.pdf"), width = 8, height = 7)
hmap_mat_SeuratDE_lymphoProlif
#dev.off()
```


#Heatmap of GO BP: response to interferon-gamma (GO:0034341) enriched in CD4_TEM Seurat DE
```{r}
genes_to_plot_topgoDE_SeuratDE_IFNgResp <- unique(unlist(strsplit( topgoDE_cluster$CD4_TEM$genes[topgoDE_cluster$CD4_TEM$Term %in% c("response to interferon-gamma")], "," )) )

genes_to_plot_topgoDE_SeuratDE_IFNgResp <- c(genes_to_plot_topgoDE_SeuratDE_IFNgResp, "Ifng") #adding Ifng to the plot

x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] #subset to NK and T cell clusters

x <- x[x$gene %in% genes_to_plot_topgoDE_SeuratDE_IFNgResp, ] %>% dplyr::select(cell_type, gene, avg_log2FC) #retrieve log2FC for genes enriched in response to interferon-gamma (GO:0034341)  in CD4_TEM cluster in Seurat DE

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK",  "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")] #Excluding "CD27loCD11bhi_NK" and "CD8_Nave", as they do not express any of the Ifngamma response genes


mat_SeuratDE_IFNgResp <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_SeuratDE_IFNgResp) )
row.names(annotation_col) <- colnames(mat_SeuratDE_IFNgResp)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_SeuratDE_IFNgResp <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_SeuratDE_IFNgResp <- ComplexHeatmap::Heatmap(mat_SeuratDE_IFNgResp,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of log2FoldChange in NK and T and T cell of genes enriched in\n"response to interferon-gamma" (GO:0034341)  in CD4_TEM cluster\n inSeurat DE\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_SeuratDE_IFNgResp,
    height = unit(3, "mm")*nrow(mat_SeuratDE_IFNgResp), width = unit(5, "mm")*ncol(mat_SeuratDE_IFNgResp)
)

hmap_mat_SeuratDE_IFNgResp@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK",  "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") #reorder color legend

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SeuratDE_topGO_NK_and_T_T_IFNGammaResp.pdf"), width = 8, height = 7)
hmap_mat_SeuratDE_IFNgResp
#dev.off()
```


#Heatmap of GO BP: regulation of defense response (GO:0031347) enriched in CD4_TEM Seurat DE
```{r}
genes_to_plot_topgoDE_SeuratDE_RegDefenseResp <- unique(unlist(strsplit( topgoDE_cluster$CD4_TEM$genes[topgoDE_cluster$CD4_TEM$Term %in% c("regulation of defense response")], "," )) )

x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] #subset to NK and T cell clusters

x <- x[x$gene %in% genes_to_plot_topgoDE_SeuratDE_RegDefenseResp, ] %>% dplyr::select(cell_type, gene, avg_log2FC) #retrieve log2FC for genes enriched in regulation of defense response (GO:0031347) in CD4_TEM cluster in Seurat DE

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")] #Excluding "CD27hiCD11b+_NK" that does not have any of any of enriched DE genes


mat_SeuratDE_RegDefenseResp <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_SeuratDE_RegDefenseResp) )
row.names(annotation_col) <- colnames(mat_SeuratDE_RegDefenseResp)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_SeuratDE_RegDefenseResp <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_SeuratDE_RegDefenseResp <- ComplexHeatmap::Heatmap(mat_SeuratDE_RegDefenseResp,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of log2FoldChange in NK and T and T cell of genes enriched in\n"regulation of defense response" (GO:0031347) in CD4_TEM cluster\n inSeurat DE\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_SeuratDE_RegDefenseResp,
    height = unit(3, "mm")*nrow(mat_SeuratDE_RegDefenseResp), width = unit(5, "mm")*ncol(mat_SeuratDE_RegDefenseResp)
)

hmap_mat_SeuratDE_RegDefenseResp@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") #reorder color legend

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SeuratDE_topGO_NK_and_T_T_RegDefenseResp.pdf"), width = 8, height = 7)
hmap_mat_SeuratDE_RegDefenseResp
#dev.off()
```


#Heatmap of GO BP:  T cell proliferation (GO:0042098) enriched in CD4_TEM Seurat DE
```{r}
genes_to_plot_topgoDE_SeuratDE_TcellProlif <- unique(unlist(strsplit( topgoDE_cluster$CD4_TEM$genes[topgoDE_cluster$CD4_TEM$Term %in% c("T cell proliferation")], "," )) )

x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] #subset to NK and T cell clusters

x <- x[x$gene %in% genes_to_plot_topgoDE_SeuratDE_TcellProlif, ] %>% dplyr::select(cell_type, gene, avg_log2FC) #retrieve log2FC for genes enriched in T cell proliferation (GO:0042098)  in CD4_TEM cluster in Seurat DE

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_SeuratDE_TcellProlif <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_SeuratDE_TcellProlif) )
row.names(annotation_col) <- colnames(mat_SeuratDE_TcellProlif)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF", "Alveolar_macrophage" =  "#005685FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_SeuratDE_TcellProlif <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_SeuratDE_TcellProlif <- ComplexHeatmap::Heatmap(mat_SeuratDE_TcellProlif,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of log2FoldChange in NK and T and T cell of genes enriched in\n"T cell proliferation" (GO:0042098)  in CD4_TEM cluster\n inSeurat DE\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_SeuratDE_TcellProlif,
    height = unit(3, "mm")*nrow(mat_SeuratDE_TcellProlif), width = unit(5, "mm")*ncol(mat_SeuratDE_TcellProlif)
)

hmap_mat_SeuratDE_TcellProlif@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") #reorder color legend

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SeuratDE_topGO_NK_and_T_T_TcellProliferation.pdf"), width = 8, height = 7)
hmap_mat_SeuratDE_TcellProlif
#dev.off()
```

#Heatmap of GO BP: inflammatory response (GO:0006954) enriched in CD4_TEM Seurat DE
```{r}
genes_to_plot_topgoDE_SeuratDE_inflammResp <- unique(unlist(strsplit( topgoDE_cluster$CD4_TEM$genes[topgoDE_cluster$CD4_TEM$Term %in% c("inflammatory response")], "," )) )

x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] #subset to NK and T cell clusters

x <- x[x$gene %in% genes_to_plot_topgoDE_SeuratDE_inflammResp, ] %>% dplyr::select(cell_type, gene, avg_log2FC) #retrieve log2FC for genes enriched in inflammatory response (GO:0006954) in CD4_TEM cluster in Seurat DE

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")] #Excluding "CD27loCD11b+_mito_NK",  which does not DE any of the enriched genes


mat_SeuratDE_inflammResp <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_SeuratDE_inflammResp) )
row.names(annotation_col) <- colnames(mat_SeuratDE_inflammResp)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF", "Alveolar_macrophage" =  "#005685FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_SeuratDE_inflammResp <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_SeuratDE_inflammResp <- ComplexHeatmap::Heatmap(mat_SeuratDE_inflammResp,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of log2FoldChange in NK and T and T cell of genes enriched in\n"inflammatory response" (GO:0006954) in CD4_TEM cluster\n inSeurat DE\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_SeuratDE_inflammResp,
    height = unit(3, "mm")*nrow(mat_SeuratDE_inflammResp), width = unit(5, "mm")*ncol(mat_SeuratDE_inflammResp)
)

hmap_mat_SeuratDE_inflammResp@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") #reorder color legend

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SeuratDE_topGO_NK_and_T_T_InflammResp.pdf"), width = 8, height = 7)
hmap_mat_SeuratDE_inflammResp
#dev.off()
```



#Heatmap of core enrichment genes from Hallmark Interferon Gamma response from CD4_TEM
```{r}

genes_to_plot_topgoDE_SeuratDE_Hallm_IfngammaResp <- c("Xcl1", "Gch1", "Cd38", "Eif2ak2", "Nfkbia", "Trim26", "B2m", "Stat3", "Irf9", "H2-M3", "Ifit3", "Rsad2", "Rnf213", "Nampt", "Lgals3bp", "Parp14", "Ogfr", "Pfkp", "Psmb2", "Ptpn1", "Stat2", "Cd74", "Xaf1", "Gbp8", "Ripk1", "Tapbp", "Zbp1", "Cmpk2", "Icam1", "Ube2l6", "Irf8", "Gbp3", "Socs3", "Pnpt1", "Casp3", "Sp110", "Jak2", "Cmtr1", "Cxcl10", "Tap1", "Trafd1", "Dhx58", "Samhd1", "Gbp4", "Arid5b", "Casp1", "H2-Q10", "H2-T23", "Trim21", "Nmi", "Psmb9", "Gbp10", "H2-DMa", "H2-Q7", "Gbp9", "Psmb10", "Irf1", "Psme1", "Stat1")

DefaultAssay(object = all_integrated_rnm) <- "RNA"
Idents(all_integrated_rnm) <- "cluster0.8.group"

#Finding differentially expressed genes across clusters (Source code adapted from: https://github.com/hbctraining/scRNA-seq_online/blob/master/lessons/09_merged_SC_marker_identification.md)

get_differential_genes_lfc0 <- function(celltype){
  Seurat::FindMarkers(object = all_integrated_rnm, ident.1 = paste0(celltype, "_il27raKO"), ident.2 = paste0(celltype, "_wt"), verbose = FALSE, assay = "RNA", logfc.threshold = 0) %>%
    rownames_to_column(var = "gene") %>%
    cbind(cell_type = celltype, .)
}

tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

differential_genes_all_integrated_rnm_lfc0 <- map_dfr(tmp_celltype_list, get_differential_genes_lfc0)


x <- differential_genes_all_integrated_rnm_lfc0[differential_genes_all_integrated_rnm_lfc0$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] #subset to NK and T cell clusters

x <- x[x$gene %in% genes_to_plot_topgoDE_SeuratDE_Hallm_IfngammaResp, ] %>% dplyr::select(cell_type, gene, avg_log2FC) #retrieve log2FC for genes enriched in GSEA "Hallmark_Interferon_Gamma_Response" in CD4_TEM cluster in Seurat DE

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_SeuratDE_Hallm_IfngammaResp <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_SeuratDE_Hallm_IfngammaResp) )
row.names(annotation_col) <- colnames(mat_SeuratDE_Hallm_IfngammaResp)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF", "Alveolar_macrophage" =  "#005685FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_SeuratDE_Hallm_IfngammaResp <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

#create the actual heatmap object
hmap_mat_SeuratDE_Hallm_IfngammaResp <- ComplexHeatmap::Heatmap(mat_SeuratDE_Hallm_IfngammaResp,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of log2FoldChange in NK and T and T cell of genes enriched in\n GSEA "Hallmark_Interferon_Gamma_Response" in CD4_TEM cluster\n inSeurat DE\nBased on Seurat DE with lfc.threshold=0\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_SeuratDE_Hallm_IfngammaResp,
    height = unit(3, "mm")*nrow(mat_SeuratDE_Hallm_IfngammaResp), width = unit(5, "mm")*ncol(mat_SeuratDE_Hallm_IfngammaResp)
)

hmap_mat_SeuratDE_Hallm_IfngammaResp@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") #reorder color legend

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SeuratDE_topGO_NK_and_T_T_Hallmark_IfngammaResp_LFC0.pdf"), width = 8, height = 7)
hmap_mat_SeuratDE_Hallm_IfngammaResp
#dev.off()



```

## Upd 10May2023
#Heatmap of DE genes from Seurat, specific to NK or T cells or common to both
```{r}
seu_pseudoBulk_DE_RNA_NK_T <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & tbl_pseudoBulk_DE_RNA$p_val_adj < 0.05, ] %>% na.omit()



#Heatmap using avg_log2FC values based on DE analysis
x <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"),]

x <- x[x$gene %in% unique(seu_pseudoBulk_DE_RNA_NK_T$gene), ] %>% dplyr::select(cell_type, gene, avg_log2FC)

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_seu_pseudoBulk_DE_RNA_NK_T_genes <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_seu_pseudoBulk_DE_RNA_NK_T_genes) )
row.names(annotation_col) <- colnames(mat_seu_pseudoBulk_DE_RNA_NK_T_genes)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_seu_pseudoBulk_DE_RNA_NK_T_genes <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

row_split = rep("NK", 74); row_split[16:29] = "NK_and_T"; row_split[30:74] = "T" # splitting rows based on gene list c(muscatDS_signif_NK_to_plot, muscatDS_signif_NK_and_T_to_plot, muscatDS_signif_T_to_plot) (https://github.com/jokergoo/ComplexHeatmap/issues/884#issuecomment-1068901666)

mat_seu_pseudoBulk_DE_RNA_NK_T_genes <- mat_seu_pseudoBulk_DE_RNA_NK_T_genes[c(muscatDS_signif_NK_to_plot, muscatDS_signif_NK_and_T_to_plot, muscatDS_signif_T_to_plot), ] #reorder matrix rows based on order of the gene list genes


#create the actual heatmap object
hmap_mat_seu_pseudoBulk_DE_RNA_NK_T_genes <- ComplexHeatmap::Heatmap(mat_seu_pseudoBulk_DE_RNA_NK_T_genes,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
    row_order = row.names(mat_seu_pseudoBulk_DE_RNA_NK_T_genes),
    #row_split = row_split,
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = '',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_seu_pseudoBulk_DE_RNA_NK_T_genes,
    height = unit(3, "mm")*nrow(mat_seu_pseudoBulk_DE_RNA_NK_T_genes), width = unit(5, "mm")*ncol(mat_seu_pseudoBulk_DE_RNA_NK_T_genes)
)

#relevel the heatmap legend
hmap_mat_seu_pseudoBulk_DE_RNA_NK_T_genes@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")

#draw the heatmap
#pdf(paste0(fig_path, "logFC_signif_Seurat_Pseudobulk_DE_NK_Tcell.pdf"), width = 8, height = 11)
hmap_mat_seu_pseudoBulk_DE_RNA_NK_T_genes
#dev.off()


```


# Upd 11May2023: Single cell pathway analysis (SCPA)
Notes:  (Reference: https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383; https://jackbibby1.github.io/SCPA/articles/comparing_two_populations.html#some-conclusions-about-scpa)
- "If doing a 2 sample comparison there will be a fold change column in the output from compare_pathways, which will tells the directionality."<br>
- "The fold change is calculated from population1 - population2, so a negative value will be higher in population2." <br> 
- "SCPA measures changes in the multivariate distribution of a pathway (reflecting multiple parameters including enrichment) as its primary outcome, so there will be pathways that show large qvals, but no fold change. These pathways are still importantly different. qval should be used as the primary statistic when judging biological relevance, and the pathway fold change/enirchment should only be a secondary informative value." <br>
- The authors/developers of SCPA "recommend using the qval as the measure of differential regulation of a pathway, rather than fold change. And just use the fold change to get an idea if there is enrichment of that pathway or not." <br>
```{r}
# Performing SCPA analysis (Source vignette: https://jackbibby1.github.io/SCPA/articles/disease_comparison.html; https://jackbibby1.github.io/SCPA/articles/systematic_tissue_comparison.html; https://igordot.github.io/msigdbr/articles/msigdbr-intro.html; https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383; https://jackbibby1.github.io/SCPA/articles/comparing_two_populations.html#some-conclusions-about-scpa)
#remotes::install_github("jackbibby1/SCPA@v1.1.0")

#Hallmark pathways based on FC

# temprary vector of cell types
tmp_celltype_list <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT",  "Monocyte", "Alveolar_macrophage", "cDC", "pDC", "moDC", "B_cell")

# vector comprising pathways
library(msigdbr)
library(SCPA)

#pws <- c("bp", "cc", "mf", "kegg", "reactome")
#pathways <- msigdbr("Mus musculus") %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()

pws <- c("bp")
pathways_Hallmark <- msigdbr("Mus musculus", category = c("H")) %>% filter(grepl(paste(pws, collapse = "|"), gs_subcat, ignore.case = T) |  grepl("HALLMARK", x = gs_name, ignore.case = T)) %>% format_pathways()



scpa_out_il27rako_wt <- list()

for (i in tmp_celltype_list) {
  
  
  il27raKO <- seurat_extract(subset(all_integrated_rnm_VlnPlot, group %in% "il27raKO"), 
                          meta1 = "celltype", value_meta1 = i)
  
  wt <- seurat_extract(subset(all_integrated_rnm_VlnPlot, group %in% "wt"), 
                            meta1 = "celltype", value_meta1 = i)
  
  print(paste("comparing", i))
  scpa_out_il27rako_wt[[i]] <- compare_pathways(samples = list(il27raKO, wt), pathways = pathways_Hallmark) %>% 
    dplyr::select(Pathway, qval, FC) %>%
    set_colnames(c("Pathway", paste(i, "qval", sep = "_"), paste(i, "FC", sep = "_")))
  #fold change in pathways calculated from population 1 - population 2 (https://github.com/jackbibby1/SCPA/issues/8#issuecomment-1069478383)

}
  
scpa_out_il27rako_wt_fig_filt_FC <- scpa_out_il27rako_wt %>% 
  purrr::reduce(full_join, by = "Pathway") %>%  
  set_colnames(gsub(colnames(.), pattern = " ", replacement = "_")) %>%
  dplyr::select(c("Pathway", grep("_FC", colnames(.)))) %>%
  column_to_rownames("Pathway") #specify purrr::reduce to prevent error (Source code: https://stackoverflow.com/questions/66776517/how-do-i-solve-the-problem-with-reduce-function-in-r)



#c("HALLMARK_INTERFERON_GAMMA_RESPONSE", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_COAGULATION", "HALLMARK_APOPTOSIS", "HALLMARK_CHOLESTEROL_HOMEOSTASIS", "HALLMARK_COMPLEMENT", "HALLMARK_DNA_REPAIR", "HALLMARK_E2F_TARGETS", "HALLMARK_FATTY_ACID_METABOLISM", "HALLMARK_G2M_CHECKPOINT", "HALLMARK_GLYCOLYSIS", "HALLMARK_HEDGEHOG_SIGNALING", "HALLMARK_HYPOXIA", "HALLMARK_HEME_METABOLISM", "HALLMARK_HYPOXIA", "HALLMARK_IL2_STAT5_SIGNALING", "HALLMARK_IL6_JAK_STAT3_SIGNALING", "HALLMARK_INFLAMMATORY_RESPONSE", "HALLMARK_INTERFERON_ALPHA_RESPONSE", "HALLMARK_MTORC1_SIGNALING", "HALLMARK_MYC_TARGETS_V1", "HALLMARK_MYC_TARGETS_V2", "HALLMARK_NOTCH_SIGNALING", "HALLMARK_OXIDATIVE_PHOSPHORYLATION", "HALLMARK_P53_PATHWAY", "HALLMARK_PEROXISOME", "HALLMARK_PI3K_AKT_MTOR_SIGNALING", "HALLMARK_PROTEIN_SECRETION", "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY", "HALLMARK_TGF_BETA_SIGNALING", "HALLMARK_UNFOLDED_PROTEIN_RESPONSE", "HALLMARK_WNT_BETA_CATENIN_SIGNALING")

scpa_out_il27rako_wt_fig_filt_NK_T_FC <-  scpa_out_il27rako_wt_fig_filt_FC[, c(1:12)] #Filter to only keep columns that pertain to NK and T cells

#Filter to retain only relevant immune related pathways
relevant_pathways <- c("HALLMARK_INTERFERON_GAMMA_RESPONSE", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_COAGULATION", "HALLMARK_APOPTOSIS", "HALLMARK_CHOLESTEROL_HOMEOSTASIS", "HALLMARK_COMPLEMENT", "HALLMARK_DNA_REPAIR", "HALLMARK_E2F_TARGETS", "HALLMARK_FATTY_ACID_METABOLISM", "HALLMARK_G2M_CHECKPOINT", "HALLMARK_GLYCOLYSIS", "HALLMARK_HEDGEHOG_SIGNALING", "HALLMARK_HYPOXIA", "HALLMARK_HEME_METABOLISM", "HALLMARK_HYPOXIA", "HALLMARK_IL2_STAT5_SIGNALING", "HALLMARK_IL6_JAK_STAT3_SIGNALING", "HALLMARK_INFLAMMATORY_RESPONSE", "HALLMARK_INTERFERON_ALPHA_RESPONSE", "HALLMARK_MTORC1_SIGNALING", "HALLMARK_MYC_TARGETS_V1", "HALLMARK_MYC_TARGETS_V2", "HALLMARK_NOTCH_SIGNALING", "HALLMARK_OXIDATIVE_PHOSPHORYLATION", "HALLMARK_P53_PATHWAY", "HALLMARK_PEROXISOME", "HALLMARK_PI3K_AKT_MTOR_SIGNALING", "HALLMARK_PROTEIN_SECRETION", "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY", "HALLMARK_TGF_BETA_SIGNALING", "HALLMARK_UNFOLDED_PROTEIN_RESPONSE", "HALLMARK_WNT_BETA_CATENIN_SIGNALING")

scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune <- scpa_out_il27rako_wt_fig_filt_NK_T_FC[row.names(scpa_out_il27rako_wt_fig_filt_NK_T_FC) %in% relevant_pathways, ]


library(ComplexHeatmap)
library(circlize)

col_hm_FC <- colorRamp2(colors = c("blue", "white", "red"), breaks = c(-2, 0, 2))

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(gsub("_FC", "", colnames(scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune)) )
row.names(annotation_col) <- colnames(gsub("_FC", "", colnames(scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune)))
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

colAnn_scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune <- as.matrix(scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune)

htmap_scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune <- ComplexHeatmap::Heatmap(scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune ,
                        col = col_hm_FC,
                        name = "Fold-change",
                        show_row_names = TRUE,
                        column_names_gp = gpar(fontsize = 8),row_names_gp = gpar(fontsize = 8),
                        column_title = 'SCPA pathway heatmap of relevant Hallmark\nGO BP based on fold change\nin IL27RAKO vs WT\n',
                        border = TRUE,
                        cluster_columns = FALSE,
                        column_labels = gsub("_FC", "", colnames(scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune)),
                        top_annotation = colAnn_scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune,
                        height = unit(3, "mm")*nrow(scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune), width = unit(5, "mm")*ncol(scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune)  )

calc_ht_size = function(ht, unit = "inch") {
    #pdf(NULL)
    ht = draw(ht)
    w = ComplexHeatmap:::width(ht)
    w = convertX(w, unit, valueOnly = TRUE)
    h = ComplexHeatmap:::height(ht)
    h = convertY(h, unit, valueOnly = TRUE)
    #dev.off()

    c(w, h)
}


size = calc_ht_size(htmap_scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune)


#relevel the heatmap legend
htmap_scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")


#pdf(paste0(fig_path, "Htmap_SCPA_relevant_GOBP_Il27raKO_vs_wt_FoldChange_NK_Tcell.pdf"), width = size[1] + 2, height = size[2])

draw(htmap_scpa_out_il27rako_wt_fig_filt_NK_T_FC_immune,
    heatmap_legend_side = 'right',
    annotation_legend_side = 'right',
    row_sub_title_side = 'left',
    row_title_side = "right")

##dev.off()


scpa_out_il27rako_wt_fig_filt_qval <- scpa_out_il27rako_wt %>% 
  purrr::reduce(full_join, by = "Pathway") %>%  
  set_colnames(gsub(colnames(.), pattern = " ", replacement = "_")) %>%
  select(c("Pathway", grep("_qval", colnames(.)))) %>%
  column_to_rownames("Pathway") #specify purrr::reduce to prevent error (Source code: https://stackoverflow.com/questions/66776517/how-do-i-solve-the-problem-with-reduce-function-in-r)


scpa_out_il27rako_wt_fig_filt_NK_T_qval <-  scpa_out_il27rako_wt_fig_filt_qval[, c(1:12)] #Filter to only keep columns that pertain to NK and T cells

#Filter to retain only relevant immune related pathways
relevant_pathways <- c("HALLMARK_INTERFERON_GAMMA_RESPONSE", "HALLMARK_TNFA_SIGNALING_VIA_NFKB", "HALLMARK_COAGULATION", "HALLMARK_APOPTOSIS", "HALLMARK_CHOLESTEROL_HOMEOSTASIS", "HALLMARK_COMPLEMENT", "HALLMARK_DNA_REPAIR", "HALLMARK_E2F_TARGETS", "HALLMARK_FATTY_ACID_METABOLISM", "HALLMARK_G2M_CHECKPOINT", "HALLMARK_GLYCOLYSIS", "HALLMARK_HEDGEHOG_SIGNALING", "HALLMARK_HYPOXIA", "HALLMARK_HEME_METABOLISM", "HALLMARK_HYPOXIA", "HALLMARK_IL2_STAT5_SIGNALING", "HALLMARK_IL6_JAK_STAT3_SIGNALING", "HALLMARK_INFLAMMATORY_RESPONSE", "HALLMARK_INTERFERON_ALPHA_RESPONSE", "HALLMARK_MTORC1_SIGNALING", "HALLMARK_MYC_TARGETS_V1", "HALLMARK_MYC_TARGETS_V2", "HALLMARK_NOTCH_SIGNALING", "HALLMARK_OXIDATIVE_PHOSPHORYLATION", "HALLMARK_P53_PATHWAY", "HALLMARK_PEROXISOME", "HALLMARK_PI3K_AKT_MTOR_SIGNALING", "HALLMARK_PROTEIN_SECRETION", "HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY", "HALLMARK_TGF_BETA_SIGNALING", "HALLMARK_UNFOLDED_PROTEIN_RESPONSE", "HALLMARK_WNT_BETA_CATENIN_SIGNALING")

scpa_out_il27rako_wt_fig_filt_NK_T_qval_immune <- scpa_out_il27rako_wt_fig_filt_NK_T_qval[row.names(scpa_out_il27rako_wt_fig_filt_NK_T_qval) %in% relevant_pathways, ]

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(gsub("_qval", "", colnames(scpa_out_il27rako_wt_fig_filt_NK_T_qval_immune)) )
row.names(annotation_col) <- colnames(gsub("_qval", "", colnames(scpa_out_il27rako_wt_fig_filt_NK_T_qval_immune)))
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

colAnn_scpa_out_il27rako_wt_fig_filt_NK_T_qval_immune <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )

col_hm_Qval <- colorRamp2(colors = c("blue", "white", "red"), breaks = c(0, 1, 2))


htmap_scpa_out_il27rako_wt_fig_filt_NK_T_Qval_immune  <- ComplexHeatmap::Heatmap(scpa_out_il27rako_wt_fig_filt_NK_T_qval_immune  ,
                        #col = col_hm,
                        name = "Qval",
                        show_row_names = TRUE,
                        column_names_gp = gpar(fontsize = 8),row_names_gp = gpar(fontsize = 8),
                        column_title = 'SCPA pathway heatmap of relevant Hallmark\nGO BP based on Q value\nin IL27RAKO vs WT\n',
                        border = TRUE,
                        cluster_columns = FALSE,
                        column_labels = gsub("_qval", "", colnames(scpa_out_il27rako_wt_fig_filt_NK_T_qval_immune)),
                        top_annotation = colAnn_scpa_out_il27rako_wt_fig_filt_NK_T_qval_immune,
                        height = unit(3, "mm")*nrow(scpa_out_il27rako_wt_fig_filt_NK_T_qval_immune ), width = unit(5, "mm")*ncol(scpa_out_il27rako_wt_fig_filt_NK_T_qval_immune )  )

size = calc_ht_size(htmap_scpa_out_il27rako_wt_fig_filt_NK_T_Qval_immune )

#pdf(paste0(fig_path, "Htmap_SCPA_relevant_GOBP_Il27raKO_vs_wt_Qvalue_NK_Tcell.pdf"), width = size[1] + 2, height = size[2])

draw(htmap_scpa_out_il27rako_wt_fig_filt_NK_T_Qval_immune ,
    heatmap_legend_side = 'right',
    annotation_legend_side = 'right',
    row_sub_title_side = 'left',
    row_title_side = "right")

##dev.off()


scpa_out_il27rako_wt_dotplot <- scpa_out_il27rako_wt

for (i in 1:length(scpa_out_il27rako_wt_dotplot)){ 
  scpa_out_il27rako_wt_dotplot[[i]]$Celltype <- names(scpa_out_il27rako_wt_dotplot[i])
  colnames(scpa_out_il27rako_wt_dotplot[[i]]) <- c("Pathway", "qval",  "FC",  "Celltype")  
  }

scpa_out_il27rako_wt_dotplot <- bind_rows(scpa_out_il27rako_wt_dotplot)

scpa_out_il27rako_wt_NK_T_filt_dotplot <- scpa_out_il27rako_wt_dotplot[scpa_out_il27rako_wt_dotplot$Celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ]

scpa_out_il27rako_wt_filt_NK_T_filt_immune_dotplot <- scpa_out_il27rako_wt_NK_T_filt_dotplot[scpa_out_il27rako_wt_NK_T_filt_dotplot$Pathway %in% relevant_pathways, ]

scpa_out_il27rako_wt_filt_NK_T_filt_immune_dotplot$Celltype <- factor(scpa_out_il27rako_wt_filt_NK_T_filt_immune_dotplot$Celltype, levels = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))

#pdf(paste0(fig_path, "Dotplot_SCPA_relevant_GOBP_Il27raKO_vs_wt_NK_Tcell.pdf"), width = 10, height = 8)

ggplot(scpa_out_il27rako_wt_filt_NK_T_filt_immune_dotplot, aes(x = Celltype, y = Pathway, fill = FC)) + 
    geom_point(aes(size = qval), alpha = 0.75, shape = 21) + scale_fill_gradient2(low= "blue", mid= "white", high = "red", midpoint = 0, limits = c(-2,2), breaks = c(-2, -1, 0, 1, 2), oob = scales::squish) + theme_bw() + theme(axis.text = element_text(size = 10, color = "black"), axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5), axis.title.x = element_blank(), axis.title.y = element_text(size = 10, color = "black"), panel.grid = element_blank()) + ggtitle('SCPA pathway dotplot of relevant Hallmark GO BP\nin IL27RAKO vs WT\n')
##dev.off()



#=====================================================================================================
#Retrieve the genes from the Msigdbr for the relevant pathways (#Source code: https://github.com/jackbibby1/SCPA/issues/38#issuecomment-1451126708)
relevant_pathway_genes <- msigdbr("Mus musculus", "H") %>%
  filter(gs_name %in% relevant_pathways) %>%
  select(gs_name, gene_symbol) %>% # select whatever columns you want here
  group_split(gs_name) 

y <- bind_rows(relevant_pathway_genes)

#HALLMARK_INTERFERON_ALPHA_RESPONSE
ifnaResp_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_INTERFERON_ALPHA_RESPONSE")] %>% unlist()


seu_pseudoBulk_DE_RNA_NK_T <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & tbl_pseudoBulk_DE_RNA$p_val_adj < 0.05, ] %>% na.omit()


ifnaResp_genes_NK_T_seu_pseudo <- seu_pseudoBulk_DE_RNA_NK_T[seu_pseudoBulk_DE_RNA_NK_T$gene %in% ifnaResp_genes, ]

VlnPlot(all_integrated_rnm_VlnPlot, features = ifnaResp_genes_NK_T_seu_pseudo$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)

#HALLMARK_INTERFERON_GAMMA_RESPONSE
ifngResp_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_INTERFERON_GAMMA_RESPONSE")] %>% unlist()

ifngResp_genes_NK_T_seu_pseudo <- seu_pseudoBulk_DE_RNA_NK_T[seu_pseudoBulk_DE_RNA_NK_T$gene %in% ifngResp_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = ifngResp_genes_NK_T_seu_pseudo$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)


#HALLMARK_TNFA_SIGNALING_VIA_NFKB
TnfaSignal_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_TNFA_SIGNALING_VIA_NFKB")] %>% unlist()

TnfaSignal_genes_NK_T_seu_pseudo <- seu_pseudoBulk_DE_RNA_NK_T[seu_pseudoBulk_DE_RNA_NK_T$gene %in% TnfaSignal_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = TnfaSignal_genes_NK_T_seu_pseudo$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)


#HALLMARK_IL2_STAT5_SIGNALING
IL2_STAT5_Signal_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_IL2_STAT5_SIGNALING")] %>% unlist()

IL2_STAT5_Signal_genes_NK_T_seu_pseudo <- seu_pseudoBulk_DE_RNA_NK_T[seu_pseudoBulk_DE_RNA_NK_T$gene %in% IL2_STAT5_Signal_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = IL2_STAT5_Signal_genes_NK_T_seu_pseudo$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)



#HALLMARK_IL6_JAK_STAT3_SIGNALING
IL6_JAK_STAT3_Signal_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_IL6_JAK_STAT3_SIGNALING")] %>% unlist()

IL6_JAK_STAT3_Signal_genes_NK_T_seu_pseudo <- seu_pseudoBulk_DE_RNA_NK_T[seu_pseudoBulk_DE_RNA_NK_T$gene %in% IL6_JAK_STAT3_Signal_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = IL6_JAK_STAT3_Signal_genes_NK_T_seu_pseudo$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)



#HALLMARK_INFLAMMATORY_RESPONSE
inflammResp_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_INFLAMMATORY_RESPONSE")] %>% unlist()

inflammResp_genes_NK_T_seu_pseudo <- seu_pseudoBulk_DE_RNA_NK_T[seu_pseudoBulk_DE_RNA_NK_T$gene %in% inflammResp_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = inflammResp_genes_NK_T_seu_pseudo$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)


#HALLMARK_P53_PATHWAY
p53Pathway_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_P53_PATHWAY")] %>% unlist()

p53Pathway_genes_NK_T_seu_pseudo <- seu_pseudoBulk_DE_RNA_NK_T[seu_pseudoBulk_DE_RNA_NK_T$gene %in% p53Pathway_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = p53Pathway_genes_NK_T_seu_pseudo$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)



#HALLMARK_COAGULATION
coagulation_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_COAGULATION")] %>% unlist()

coagulation_genes_NK_T_seu_pseudo <- seu_pseudoBulk_DE_RNA_NK_T[seu_pseudoBulk_DE_RNA_NK_T$gene %in% coagulation_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = coagulation_genes_NK_T_seu_pseudo$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))


#Heatmap of genes significantly DE (Seurat pseudobulk) in any of the NK and T cells in the relevant pathway 


#Heatmap using avg_log2FC values based on DE analysis
x <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"),]

x <- x[x$gene %in% unique(c(ifnaResp_genes_NK_T_seu_pseudo$gene, ifngResp_genes_NK_T_seu_pseudo$gene, TnfaSignal_genes_NK_T_seu_pseudo$gene, IL2_STAT5_Signal_genes_NK_T_seu_pseudo$gene, IL6_JAK_STAT3_Signal_genes_NK_T_seu_pseudo$gene, inflammResp_genes_NK_T_seu_pseudo$gene, p53Pathway_genes_NK_T_seu_pseudo$gene, coagulation_genes_NK_T_seu_pseudo$gene)), ] %>% dplyr::select(cell_type, gene, avg_log2FC)

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA) )
row.names(annotation_col) <- colnames(mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )


#create the actual heatmap object
hmap_mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA <- ComplexHeatmap::Heatmap(mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
    row_order = row.names(mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA),
    #row_split = row_split,
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of significantly Seurat pseudobulk DE genes\nin NK and T cells enriched in\nselected SCPA Hallmark GO BP\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA,
    height = unit(3, "mm")*nrow(mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA), width = unit(5, "mm")*ncol(mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA)
)

#relevel the heatmap legend
hmap_mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SCPA__SelectRelevantGOBP_signif_Seurat_Pseudobulk_DE_NK_Tcell.pdf"), width = 8, height = 6)
hmap_mat_seu_pseudoBulk_DE_RNA_NK_T_genes_SCPA
#dev.off()

rm(x)

#=====================================================================================================
#Heatmap of genes significantly DE in any of the NK and T cells in the relevant pathway 


#Retrieve the genes from the Msigdbr for the relevant pathways (#Source code: https://github.com/jackbibby1/SCPA/issues/38#issuecomment-1451126708)
relevant_pathway_genes <- msigdbr("Mus musculus", "H") %>%
  filter(gs_name %in% relevant_pathways) %>%
  select(gs_name, gene_symbol) %>% # select whatever columns you want here
  group_split(gs_name) 

y <- bind_rows(relevant_pathway_genes)

#HALLMARK_INTERFERON_ALPHA_RESPONSE
ifnaResp_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_INTERFERON_ALPHA_RESPONSE")] %>% unlist()


seu_DE_RNA_NK_T <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & differential_genes_all_integrated_rnm$p_val_adj < 0.05, ] %>% na.omit()


ifnaResp_genes_NK_T <- seu_DE_RNA_NK_T[seu_DE_RNA_NK_T$gene %in% ifnaResp_genes, ]

VlnPlot(all_integrated_rnm_VlnPlot, features = ifnaResp_genes_NK_T$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)

#HALLMARK_INTERFERON_GAMMA_RESPONSE
ifngResp_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_INTERFERON_GAMMA_RESPONSE")] %>% unlist()

ifngResp_genes_NK_T <- seu_DE_RNA_NK_T[seu_DE_RNA_NK_T$gene %in% ifngResp_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = ifngResp_genes_NK_T$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)


#HALLMARK_TNFA_SIGNALING_VIA_NFKB
TnfaSignal_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_TNFA_SIGNALING_VIA_NFKB")] %>% unlist()

TnfaSignal_genes_NK_T <- seu_DE_RNA_NK_T[seu_DE_RNA_NK_T$gene %in% TnfaSignal_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = TnfaSignal_genes_NK_T$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)


#HALLMARK_IL2_STAT5_SIGNALING
IL2_STAT5_Signal_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_IL2_STAT5_SIGNALING")] %>% unlist()

IL2_STAT5_Signal_genes_NK_T <- seu_DE_RNA_NK_T[seu_DE_RNA_NK_T$gene %in% IL2_STAT5_Signal_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = IL2_STAT5_Signal_genes_NK_T$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)



#HALLMARK_IL6_JAK_STAT3_SIGNALING
IL6_JAK_STAT3_Signal_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_IL6_JAK_STAT3_SIGNALING")] %>% unlist()

IL6_JAK_STAT3_Signal_genes_NK_T <- seu_DE_RNA_NK_T[seu_DE_RNA_NK_T$gene %in% IL6_JAK_STAT3_Signal_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = IL6_JAK_STAT3_Signal_genes_NK_T$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)



#HALLMARK_INFLAMMATORY_RESPONSE
inflammResp_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_INFLAMMATORY_RESPONSE")] %>% unlist()

inflammResp_genes_NK_T <- seu_DE_RNA_NK_T[seu_DE_RNA_NK_T$gene %in% inflammResp_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = inflammResp_genes_NK_T$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)


#HALLMARK_P53_PATHWAY
p53Pathway_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_P53_PATHWAY")] %>% unlist()

p53Pathway_genes_NK_T <- seu_pseudoBulk_DE_RNA_NK_T[seu_pseudoBulk_DE_RNA_NK_T$gene %in% p53Pathway_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = p53Pathway_genes_NK_T$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), stack = TRUE)



#HALLMARK_COAGULATION
coagulation_genes <- y$gene_symbol[y$gs_name %in% c("HALLMARK_COAGULATION")] %>% unlist()

coagulation_genes_NK_T <- seu_DE_RNA_NK_T[seu_DE_RNA_NK_T$gene %in% coagulation_genes, ]


VlnPlot(all_integrated_rnm_VlnPlot, features = coagulation_genes_NK_T$gene, assay = "RNA",split.by = "group", idents = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))


#Heatmap using avg_log2FC values based on DE analysis
x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"),]

x <- x[x$gene %in% unique(c(ifnaResp_genes_NK_T$gene, ifngResp_genes_NK_T$gene, TnfaSignal_genes_NK_T$gene, IL2_STAT5_Signal_genes_NK_T$gene, IL6_JAK_STAT3_Signal_genes_NK_T$gene, inflammResp_genes_NK_T$gene, p53Pathway_genes_NK_T$gene, coagulation_genes_NK_T$gene)), ] %>% dplyr::select(cell_type, gene, avg_log2FC)

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_seu_DE_RNA_NK_T_genes_SCPA <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_seu_DE_RNA_NK_T_genes_SCPA) )
row.names(annotation_col) <- colnames(mat_seu_DE_RNA_NK_T_genes_SCPA)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_seu_DE_RNA_NK_T_genes_SCPA <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )


#create the actual heatmap object
hmap_mat_seu_DE_RNA_NK_T_genes_SCPA <- ComplexHeatmap::Heatmap(mat_seu_DE_RNA_NK_T_genes_SCPA,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
    row_order = row.names(mat_seu_DE_RNA_NK_T_genes_SCPA),
    #row_split = row_split,
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of significantly Seurat DE genes\nin NK and T cells enriched in\nselected SCPA Hallmark GO BP\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_seu_DE_RNA_NK_T_genes_SCPA,
    height = unit(3, "mm")*nrow(mat_seu_DE_RNA_NK_T_genes_SCPA), width = unit(5, "mm")*ncol(mat_seu_DE_RNA_NK_T_genes_SCPA)
)

#relevel the heatmap legend
hmap_mat_seu_DE_RNA_NK_T_genes_SCPA@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")

#draw the heatmap
#pdf(paste0(fig_path, "logFC_SCPA__SelectRelevantGOBP_signif_Seurat_DE_NK_Tcell.pdf"), width = 8, height = 6)
hmap_mat_seu_DE_RNA_NK_T_genes_SCPA
#dev.off()

rm(x)
```


#Correlation analysis: NK vs T cells (Seurat Pseudobulk)
```{r}
#RNA
#subset the NK_genes and Tcell_genes objects to only include significant genes (adjusted p_val < 0.05) 

# Combine all NK clusters into one and all T cell clusters into one

NK_genes_seu_pseudobulk <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK"), ]

Tcell_genes_seu_pseudobulk <- tbl_pseudoBulk_DE_RNA[tbl_pseudoBulk_DE_RNA$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM", "Treg", "CD27+_gdT", "CD27-_gdT"), ]

NK_genes_seu_pseudobulk_subset <- NK_genes_seu_pseudobulk[NK_genes_seu_pseudobulk$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))
Tcell_genes_seu_pseudobulk_subset <- Tcell_genes_seu_pseudobulk[Tcell_genes_seu_pseudobulk$p_val_adj < 0.05, ] %>% group_by(gene) %>%  summarize(avg_log2FC = mean(avg_log2FC, na.rm = TRUE))

df_NK_T_seu_pseudobulk_corr <- Reduce(function(...) merge(..., by = "gene", all = TRUE), list(
                data.frame(gene = NK_genes_seu_pseudobulk_subset$gene,
                          NKcell_log2FC = round(as.numeric(NK_genes_seu_pseudobulk_subset$avg_log2FC), 2),
                           stringsAsFactors = FALSE),
                   data.frame(gene = Tcell_genes_seu_pseudobulk_subset$gene,
                              Tcell_log2FC = round(as.numeric(Tcell_genes_seu_pseudobulk_subset$avg_log2FC), 2),
                              stringsAsFactors = FALSE)
                                     ))


df_NK_T_seu_pseudobulk_corr <- df_NK_T_seu_pseudobulk_corr[!is.na(df_NK_T_seu_pseudobulk_corr$gene), ]

row.names(df_NK_T_seu_pseudobulk_corr) <- unique(df_NK_T_seu_pseudobulk_corr$gene)
df_NK_T_seu_pseudobulk_corr$gene <- NULL
df_NK_T_seu_pseudobulk_corr$onecol <- rep("", nrow(df_NK_T_seu_pseudobulk_corr))


#pdf(paste0(fig_path, "Correlation_plot_NK_Tcells_SeuratPseudoBulkDGE.pdf"))
#Plotting correlation based on avg_log2FC
sp_logFC_NK_T_seu_pseudobulk <- ggscatter(df_NK_T_seu_pseudobulk_corr, x = "NKcell_log2FC", y = "Tcell_log2FC",
                add = "reg.line",  # Add regression line
                add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
                conf.int = TRUE, # Add confidence interval
                label = row.names(df_NK_T_seu_pseudobulk_corr),
                repel = TRUE, 
                font.label = c(14, "plain"), size = 0.8, cor.coef.size = 14
)
# Add correlation coefficient
sp_logFC_NK_T_seu_pseudobulk + stat_cor(method = "pearson") + xlab("NK cells (avg_log2FC)") + ylab("T cells (avg_log2FC)") + theme(axis.text = element_text(size = rel(1.05)), axis.title = element_text(size = rel(1.15)))

#dev.off()

```



#Heatmap of correlated and non-correlated genes from Seurat DE
```{r}
genes_to_plot_corr_nonCorr <- row.names(df_NK_T_corr)

#Heatmap using avg_log2FC values based on DE analysis
x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"),]

x <- x[x$gene %in% unique(c(genes_to_plot_corr_nonCorr)), ] %>% dplyr::select(cell_type, gene, avg_log2FC)

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_seu_DE_RNA_NK_T_genes_corr_nonCorr <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr) )
row.names(annotation_col) <- colnames(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )


#create the actual heatmap object
hmap_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr <- ComplexHeatmap::Heatmap(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
    #row_order = row.names(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr),
    #row_split = row_split,
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of significant Seurat DE genes\nin NK and T cells\nthat show correlated and non-correlated expression\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr,
    height = unit(3, "mm")*nrow(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr), width = unit(5, "mm")*ncol(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr)
)

#relevel the heatmap legend
hmap_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")

#draw the heatmap
#pdf(paste0(fig_path, "logFC_corr_nonCorr__SelectRelevantGOBP_signif_Seurat_DE_NK_Tcell.pdf"), width = 8, height = 6)
hmap_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr
#dev.off()

rm(x)

#==========================================================================================================

#Plot NK specific non-correlated genes, T specific correlated genes, and NK and T correlated genes
genes_corr_nonCorr_NK_T <- as.data.frame(df_NK_T_corr)
genes_corr_nonCorr_NK_T$genes <- row.names(genes_corr_nonCorr_NK_T)

genes_nonCorr_NKspecific <- genes_corr_nonCorr_NK_T$genes[is.na(genes_corr_nonCorr_NK_T$Tcell_log2FC) ]


genes_nonCorr_Tspecific <- genes_corr_nonCorr_NK_T$genes[is.na(genes_corr_nonCorr_NK_T$NKcell_log2FC) ]

pq <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") & differential_genes_all_integrated_rnm$p_val_adj < 0.05 & differential_genes_all_integrated_rnm$gene %in% genes_corr_nonCorr_NK_T_Tspecific,  ]

#Filter the T cell specific genes to only include genes that are DE in at least 2 T cell subtypes
gene_T_freq <- as.data.frame(table(pq$gene)) #Source code: https://stackoverflow.com/questions/36761644/count-the-number-of-duplicate-for-a-column

genes_nonCorr_Tspecific_filt <- pq$gene[pq$gene %in% gene_T_freq$Var1[gene_T_freq$Freq >=2]]


genes_corr_NK_T <- genes_corr_nonCorr_NK_T$genes[!is.na(genes_corr_nonCorr_NK_T$NKcell_log2FC)   & !is.na(genes_corr_nonCorr_NK_T$Tcell_log2FC) ]


#Heatmap using avg_log2FC values based on DE analysis
x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"),]

x <- x[x$gene %in% unique(c(genes_nonCorr_NKspecific, genes_nonCorr_Tspecific_filt, genes_corr_NK_T)), ] %>% dplyr::select(cell_type, gene, avg_log2FC)

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_seu_DE_RNA_NK_T_genes_corr_nonCorr <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr) )
row.names(annotation_col) <- colnames(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )


#create the actual heatmap object
hmap_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr <- ComplexHeatmap::Heatmap(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
    #row_order = row.names(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr),
    #row_split = row_split,
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of significant Seurat DE genes\nin NK and T cells\nthat show correlated and non-correlated expression\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr,
    height = unit(3, "mm")*nrow(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr), width = unit(5, "mm")*ncol(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr)
)

#relevel the heatmap legend
hmap_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")

#draw the heatmap
#pdf(paste0(fig_path, "logFC_corr_nonCorr__SelectRelevantGOBP_signif_Seurat_DE_NK_Tcell.pdf"), width = 8, height = 6)
hmap_mat_seu_DE_RNA_NK_T_genes_corr_nonCorr
#dev.off()


```



#Dotplot of top 20 (or less) upregulated genes in each cell type in il27raKO vs WT
```{r}
#retrieve list of top 20 DEGs from each cluster
top20_seurat_DGE_NK_T <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$p_val_adj < 0.05 & differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"), ] %>% dplyr::group_by(cell_type) %>% dplyr::slice_head(n = 20)


#Heatmap using avg_log2FC values based on DE analysis
x <- differential_genes_all_integrated_rnm[differential_genes_all_integrated_rnm$cell_type %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"),]

x <- x[x$gene %in% unique(c(top20_seurat_DGE_NK_T$gene)), ] %>% dplyr::select(cell_type, gene, avg_log2FC)

x <- spread(x, cell_type, avg_log2FC)
row.names(x) <- x$gene
x <- x[, c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")]


mat_seu_DE_RNA_NK_T_genes_top20 <- as.matrix(x)
#setting up color and breaks for all heatmaps
#blue-grey90-red (down-noChange-upregulated)
myCol1 <- colorRampPalette(c('navy', 'grey90', 'red'))(100)
myBreaks1 <- seq(-1, 1, length.out = 100)

#create data.frame for annotation of column clusters
annotation_col <- as.data.frame(colnames(mat_seu_DE_RNA_NK_T_genes_top20) )
row.names(annotation_col) <- colnames(mat_seu_DE_RNA_NK_T_genes_top20)
colnames(annotation_col) <- "Cell_type"

annotation_colors <- list(Cell_type = c("CD27loCD11b+_NK" =  "#E69F00FF",  "CD27loCD11bhi_NK" = "#56B4E9FF", "CD27hiCD11b+_NK" = "#009E73FF", "CD27loCD11b+_mito_NK"  = "#F0E442FF", "CD27loCD11blo_NK" = "#0072B2FF", "CD8_Nave" =  "#D55E00FF", "CD8_TCM" = "#CC79A7FF", "CD8_TEM"  =   "#666666FF", "CD4_TEM"  = "#AD7700FF", "Treg" = "#1C91D4FF", "CD27+_gdT" = "#007756FF", "CD27-_gdT" =  "#D5C711FF") ) #colors from dittoHeatmap colors (https://github.com/dtm2451/dittoSeq/blob/devel/R/dittoColors.R)

 
colAnn_mat_seu_DE_RNA_NK_T_genes_top20 <- ComplexHeatmap::HeatmapAnnotation(
    df = annotation_col,
    which = 'col',
    na_col = 'white',
    col = annotation_colors
    )


#create the actual heatmap object
hmap_mat_seu_DE_RNA_NK_T_genes_top20 <- ComplexHeatmap::Heatmap(mat_seu_DE_RNA_NK_T_genes_top20,

    name = 'Log2\nfold change',
    
    rect_gp = gpar(col = "white", lwd = 1),

    col = colorRamp2(myBreaks1, myCol1),

    # parameters for the colour-bar that represents gradient of expression
      heatmap_legend_param = list(
        color_bar = 'continuous',
        legend_direction = 'vertical',
        legend_width = unit(8, 'cm'),
        legend_height = unit(5.0, 'cm'),
        title_position = 'topcenter',
        title_gp=gpar(fontsize = 8, fontface = 'plain'),
        labels_gp=gpar(fontsize = 8, fontface = 'plain')),

    # row (gene) parameters
      cluster_rows = FALSE,
      show_row_dend = FALSE,
      row_title_side = 'left',
      row_title_gp = gpar(fontsize = 8,  fontface = 'plain'),
      row_title_rot = 90,
      show_row_names = TRUE,
      row_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      row_names_side = 'right',
    #row_order = row.names(mat_seu_DE_RNA_NK_T_genes_corr_nonCorr),
    #row_split = row_split,
      row_dend_width = unit(10,'mm'),

    # column (sample) parameters
      cluster_columns = FALSE,
      show_column_dend = FALSE,
      column_title = 'Heatmap of top 20 (or less) significant Seurat DE genes\nin NK and T cells\n',
      column_title_side = 'top',
      column_title_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_title_rot = 0,
      show_column_names = TRUE,
      column_names_gp = gpar(fontsize = 8, fontface = 'plain'),
      column_names_max_height = unit(10, 'cm'),
      column_dend_height = unit(25,'mm'),

    # cluster methods for rows and columns
      clustering_distance_columns = NA,
      clustering_distance_rows = "euclidean",
      
    # specify top and bottom annotations
      top_annotation = colAnn_mat_seu_DE_RNA_NK_T_genes_top20,
    height = unit(3, "mm")*nrow(mat_seu_DE_RNA_NK_T_genes_top20), width = unit(5, "mm")*ncol(mat_seu_DE_RNA_NK_T_genes_top20)
)

#relevel the heatmap legend
hmap_mat_seu_DE_RNA_NK_T_genes_top20@top_annotation@anno_list$Cell_type@color_mapping@levels <- c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT")

#draw the heatmap
#pdf(paste0(fig_path, "logFC_top20_signif_Seurat_DE_NK_Tcell.pdf"), width = 8, height = 6)
hmap_mat_seu_DE_RNA_NK_T_genes_top20
#dev.off()



#Heatmap based on aggregate expression
y <- AggregateExpression(tmp, return.seurat = TRUE, slot = "counts", assays = "RNA", group.by = c("celltype", "group")  )

y$celltype <- ifelse(!(sapply(strsplit(Cells(y), split = "_"), "[", 1) %in% c("Monocyte", "cDC", "moDC", "pDC", "Treg")) &  !(sapply(strsplit(Cells(y), split = "_"), "[", 2) %in% "mito"), paste0(sapply(strsplit(Cells(y), split = "_"), "[", 1), "_", sapply(strsplit(Cells(y), split = "_"), "[", 2) ), ifelse(sapply(strsplit(Cells(y), split = "_"), "[", 2) %in% "mito", paste0(sapply(strsplit(Cells(y), split = "_"), "[", 1), "_", sapply(strsplit(Cells(y), split = "_"), "[", 2), "_", sapply(strsplit(Cells(y), split = "_"), "[", 3) ), sapply(strsplit(Cells(y), split = "_"), "[", 1) ) )
                                  
y$group <- ifelse(!(sapply(strsplit(Cells(y), split = "_"), "[", 1) %in% c("Monocyte", "cDC", "moDC", "pDC", "Treg")) &  !(sapply(strsplit(Cells(y), split = "_"), "[", 2) %in% "mito"), sapply(strsplit(Cells(y), split = "_"), "[", 3),        ifelse(sapply(strsplit(Cells(y), split = "_"), "[", 2) %in% "mito", sapply(strsplit(Cells(y), split = "_"), "[", 4), sapply(strsplit(Cells(y), split = "_"), "[", 2)) )

z <- subset(y, celltype %in% c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT"))

z$celltype <- factor(z$celltype, levels = c("CD27loCD11b+_NK", "CD27loCD11bhi_NK", "CD27hiCD11b+_NK", "CD27loCD11b+_mito_NK", "CD27loCD11blo_NK", "CD8_Nave", "CD8_TCM", "CD8_TEM", "CD4_TEM","Treg", "CD27+_gdT", "CD27-_gdT") )
z$group <- factor(z$group, levels = c("wt", "il27raKO"))


x <- dittoHeatmap(z, genes = unique(top20_seurat_DGE_NK_T$gene), annot.by =  c("celltype", "group"), complex = TRUE, order.by = c("celltype", "group"), main = "Heatmap of average expression of top 20 (or less)\nsignificant (adjP < 0.05) Seurat DE genes\nin NK and T cells\n")  #order.by source code:  https://github.com/dtm2451/dittoSeq/issues/70#issuecomment-772803365

x@matrix_param$height <- unit(3, "mm")*nrow(x@matrix); x@matrix_param$width <- unit(5, "mm")*ncol(x@matrix)

#pdf(paste0(fig_path, "/DittoHtmap_AvgExpr_NK_Tcell_top20DE_fromSeuDE_SW0100_P239_May2023.pdf"), width = 10, height = 15)
  x 
##dev.off()


```



#sessionInfo
```{r}
#sink("sessionInfo_SW0100_P239_TotalSeq_IL27RAKO_Johannes_v1.2.txt")
sessionInfo()
#sink()
```




